<!DOCTYPE html>
<html lang="en"><head>
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <title>Documentation: MultiQC</title>
    <meta name="description" content="">
    <meta name="author" content="">
    <link rel="apple-touch-icon" sizes="57x57" href="https://multiqc.info/apple-icon-57x57.png">
    <link rel="apple-touch-icon" sizes="60x60" href="https://multiqc.info/apple-icon-60x60.png">
    <link rel="apple-touch-icon" sizes="72x72" href="https://multiqc.info/apple-icon-72x72.png">
    <link rel="apple-touch-icon" sizes="76x76" href="https://multiqc.info/apple-icon-76x76.png">
    <link rel="apple-touch-icon" sizes="114x114" href="https://multiqc.info/apple-icon-114x114.png">
    <link rel="apple-touch-icon" sizes="120x120" href="https://multiqc.info/apple-icon-120x120.png">
    <link rel="apple-touch-icon" sizes="144x144" href="https://multiqc.info/apple-icon-144x144.png">
    <link rel="apple-touch-icon" sizes="152x152" href="https://multiqc.info/apple-icon-152x152.png">
    <link rel="apple-touch-icon" sizes="180x180" href="https://multiqc.info/apple-icon-180x180.png">
    <link rel="icon" type="image/png" sizes="192x192" href="https://multiqc.info/android-icon-192x192.png">
    <link rel="manifest" href="https://multiqc.info/manifest.json">
    <meta name="msapplication-TileColor" content="#ffffff">
    <meta name="msapplication-TileImage" content="../ms-icon-144x144.png">
    <meta name="theme-color" content="#ffffff">

    <link href="MultiQC%20installation_files/bootstrap.css" rel="stylesheet">
    <link rel="stylesheet" href="MultiQC%20installation_files/font-awesome.css">
    <link rel="stylesheet" href="MultiQC%20installation_files/font-mfizz.css">
    <link rel="stylesheet" href="MultiQC%20installation_files/github.css">
    <link rel="stylesheet" href="MultiQC%20installation_files/styles.css">

    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
    <!--[if lt IE 9]>
      <script src="https://oss.maxcdn.com/html5shiv/3.7.2/html5shiv.min.js"></script>
      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
    <![endif]-->
  </head>

  <body id="mqc_docs" data-spy="scroll" data-target=".toc" data-offset="100">

    <div class="header">

      <!-- Static navbar -->
      <nav class="navbar navbar-default navbar-inverse navbar-static-top">
        <div class="container">
          <div class="navbar-header">
            <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false" aria-controls="navbar">
              <span class="sr-only">Toggle navigation</span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
              <span class="icon-bar"></span>
            </button>
            <a class="navbar-brand visible-xs" href="#">MultiQC</a>
          </div>
          <div id="navbar" class="navbar-collapse collapse">
            <ul class="nav navbar-nav navbar-right">
              <li><a href="https://multiqc.info/">Home</a></li>
              <li class="active"><a href="#">Docs</a></li>
              <li><a href="https://multiqc.info/plugins.html">Plugins</a></li>
              <li><a href="https://multiqc.info/logos">Logo</a></li>
              <li class="dropdown">
                <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-haspopup="true" aria-expanded="false">Example Reports <span class="caret"></span></a>
                <ul class="dropdown-menu">
                  <li class="dropdown-header">Analysis Types</li>
                  <li><a href="https://multiqc.info/examples/rna-seq/multiqc_report.html">RNA-Seq</a></li>
                  <li><a href="https://multiqc.info/examples/wgs/multiqc_report.html">Whole Genome</a></li>
                  <li><a href="https://multiqc.info/examples/bs-seq/multiqc_report.html">Bisulfite</a></li>
                  <li><a href="https://multiqc.info/examples/hi-c/multiqc_report.html">Hi-C</a></li>
                  <li class="dropdown-header">MultiQC Customisation</li>
                  <li><a href="https://multiqc.info/examples/ngi-rna/Test_NGI_Project_multiqc_report.html">MultiQC_NGI Plugin</a></li>
                </ul>
              </li>
            </ul>
          </div><!--/.nav-collapse -->
        </div>
      </nav>
      <div class="container">
        <div class="row">
          <div class="col-sm-6">
            <object type="image/svg+xml" title="MultiQC" data="MultiQC%20installation_files/Docs_logo.svg">
              <img src="MultiQC%20installation_files/Docs_logo.png" title="MultiQC">
            </object>
          </div>
          <div class="col-sm-6" style="margin-top:40px;">
            <p class="lead">Welcome to the MultiQC docs.</p>
            <p>These docs are bundled with the MultiQC download for your convenience,
               so you can also read in your installation or on <a href="https://github.com/ewels/MultiQC/tree/master/docs">Github</a>.
            </p>
          </div>
        </div>
      </div>
    </div>

    <div class="container docs-container">
      <div class="row">
        <div class="col-sm-3 col-md-push-9" id="toc_column">
          <h3 id="toc_header">Table of Contents</h3>
          <div class="toc affix" data-spy="affix" data-offset-top="254">
            
<ul class="nav nav-stacked">
<li class="active"><a href="#using-multiqc">Using MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#introduction">Introduction</a></li>
<li class="active"><a href="#installing-multiqc">Installing MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#system-python">System Python</a></li>
<li><a href="#installing-python">Installing Python</a></li>
<li><a href="#installing-with-conda">Installing with conda</a></li>
<li><a href="#installation-with-pip">Installation with pip</a></li>
<li><a href="#manual-installation">Manual installation</a></li>
<li><a href="#updating-multiqc">Updating MultiQC</a></li>
<li><a href="#using-a-specific-python-interpreter">Using a specific python interpreter</a></li>
<li><a href="#using-with-a-python-script">Using with a Python script</a></li>
<li><a href="#installing-on-windows">Installing on Windows</a></li>
<li class="active"><a href="#using-the-docker-container">Using the Docker container</a>
<ul class="nav nav-stacked">
<li><a href="#docker-bash-alias">Docker bash alias</a></li>
</ul>
</li>
<li><a href="#using-singularity">Using Singularity</a>
<ul class="nav nav-stacked">
<li><a href="#import-errors-with-singularity">Import errors with Singularity</a></li>
</ul>
</li>
<li><a href="#python-2">Python 2</a></li>
<li><a href="#using-multiqc-through-galaxy">Using MultiQC through Galaxy</a>
<ul class="nav nav-stacked">
<li><a href="#on-the-main-galaxy-instance">On the main Galaxy instance</a></li>
<li><a href="#on-your-instance">On your instance</a></li>
</ul>
</li>
<li><a href="#installing-from-freebsd">Installing from FreeBSD</a></li>
<li><a href="#installing-as-an-environment-module">Installing as an environment module</a></li>
</ul>
</li>
<li><a href="#running-multiqc">Running MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#choosing-where-to-scan">Choosing where to scan</a></li>
<li><a href="#renaming-reports">Renaming reports</a></li>
<li><a href="#overwriting-existing-reports">Overwriting existing reports</a></li>
<li><a href="#sample-names-prefixed-with-directories">Sample names prefixed with directories</a></li>
<li><a href="#using-different-templates">Using different templates</a></li>
<li><a href="#pdf-reports">PDF Reports</a></li>
<li><a href="#printing-to-stdout">Printing to stdout</a></li>
<li><a href="#parsed-data-directory">Parsed data directory</a></li>
<li><a href="#exporting-plots">Exporting Plots</a></li>
<li><a href="#choosing-which-modules-to-run">Choosing which modules to run</a></li>
</ul>
</li>
<li><a href="#using-multiqc-reports">Using MultiQC Reports</a>
<ul class="nav nav-stacked">
<li><a href="#browser-compatibility">Browser compatibility</a></li>
<li><a href="#report-layout">Report layout</a></li>
<li><a href="#general-statistics-table">General Statistics table</a></li>
<li><a href="#plots">Plots</a>
<ul class="nav nav-stacked">
<li><a href="#interactive-plots">Interactive plots</a></li>
<li><a href="#flat-plots">Flat plots</a></li>
<li><a href="#exporting-plots-1">Exporting plots</a></li>
<li><a href="#dynamic-plots">Dynamic plots</a></li>
</ul>
</li>
<li><a href="#toolbox">Toolbox</a>
<ul class="nav nav-stacked">
<li><a href="#highlight-samples">Highlight Samples</a></li>
<li><a href="#renaming-samples">Renaming Samples</a></li>
<li><a href="#hiding-samples">Hiding Samples</a></li>
<li><a href="#export">Export</a></li>
<li><a href="#save-settings">Save Settings</a></li>
</ul>
</li>

</ul>
</li>
<li><a href="#configuring-multiqc">Configuring MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#sample-name-cleaning">Sample name cleaning</a>
<ul class="nav nav-stacked">
<li><a href="#other-search-types">Other search types</a></li>
<li><a href="#clashing-sample-names">Clashing sample names</a></li>
</ul>
</li>
<li><a href="#module-search-patterns">Module search patterns</a></li>
<li><a href="#using-log-filenames-as-sample-names">Using log filenames as sample names</a></li>
<li><a href="#ignoring-files">Ignoring Files</a></li>
<li><a href="#ignoring-samples">Ignoring samples</a></li>
<li><a href="#large-sample-numbers">Large sample numbers</a>
<ul class="nav nav-stacked">
<li><a href="#disabling-on-load-plotting">Disabling on-load plotting</a></li>
<li><a href="#flat--interactive-plots">Flat / interactive plots</a></li>
<li><a href="#tables--beeswarm-plots">Tables / Beeswarm plots</a></li>
</ul>
</li>
<li><a href="#coloured-log-output">Coloured log output</a></li>
<li><a href="#checks-for-new-versions">Checks for new versions</a></li>
<li><a href="#command-line-config">Command-line config</a></li>
<li><a href="#optimising-run-time">Optimising run-time</a>
<ul class="nav nav-stacked">
<li><a href="#profile-your-multiqc-run-time">Profile your MultiQC run time</a></li>
<li><a href="#be-picky-with-which-modules-are-run">Be picky with which modules are run</a></li>
<li><a href="#optimise-file-search-patterns">Optimise file search patterns</a></li>
<li><a href="#force-interactive-plots">Force interactive plots</a></li>
<li><a href="#skip-the-report-if-you-dont-need-it">Skip the report if you don't need it</a></li>
</ul>
</li>
<li><a href="#custom-css-files">Custom CSS files</a></li>
</ul>
</li>
<li><a href="#customising-reports">Customising Reports</a>
<ul class="nav nav-stacked">
<li><a href="#titles-and-introductory-text">Titles and introductory text</a>
<ul class="nav nav-stacked">
<li><a href="#report-time-and-analysis-paths">Report time and analysis paths</a></li>
</ul>
</li>
<li><a href="#report-logo">Report Logo</a></li>
<li><a href="#project-level-information">Project level information</a></li>
<li><a href="#sample-name-replacement">Sample name replacement</a></li>
<li><a href="#bulk-sample-renaming-in-reports">Bulk sample renaming in reports</a></li>
<li><a href="#show--hide-samples-buttons">Show / Hide samples buttons</a></li>
<li><a href="#module-and-section-comments">Module and section comments</a></li>
<li><a href="#removing-modules-or-sections">Removing modules or sections</a></li>
<li><a href="#order-of-modules">Order of modules</a>
<ul class="nav nav-stacked">
<li><a href="#order-of-module-and-module-subsection-output">Order of module and module subsection output</a></li>
</ul>
</li>
<li><a href="#customising-plots">Customising plots</a></li>
<li><a href="#customising-tables">Customising tables</a>
<ul class="nav nav-stacked">
<li><a href="#hiding-columns">Hiding columns</a></li>
<li><a href="#column-order">Column order</a></li>
<li><a href="#column-titles">Column titles</a></li>
<li><a href="#conditional-formatting">Conditional formatting</a></li>
</ul>
</li>
<li><a href="#number-base-multiplier">Number base (multiplier)</a></li>
<li><a href="#number-formatting">Number formatting</a></li>
<li><a href="#troubleshooting">Troubleshooting</a></li>
</ul>
</li>
<li><a href="#using-multiqc-in-pipelines">Using MultiQC in pipelines</a>
<ul class="nav nav-stacked">
<li><a href="#nextflow">Nextflow</a>
<ul class="nav nav-stacked">
<li><a href="#clashing-input-filenames">Clashing input filenames</a></li>
<li><a href="#custom-run-name">Custom run name</a></li>
<li><a href="#multiqc-config-file">MultiQC config file</a></li>
</ul>
</li>
<li><a href="#snakemake">Snakemake</a></li>
</ul>
</li>
<li><a href="#downstream-analysis">Downstream analysis</a>
<ul class="nav nav-stacked">
<li><a href="#tidymultiqc">TidyMultiqc</a></li>
<li><a href="#megaqc">MegaQC</a></li>
<li><a href="#chronqc">ChronQC</a></li>
</ul>
</li>
<li><a href="#troubleshooting-1">Troubleshooting</a>
<ul class="nav nav-stacked">
<li><a href="#not-enough-samples-found">Not enough samples found</a>
<ul class="nav nav-stacked">
<li><a href="#big-log-files">Big log files</a></li>
</ul>
</li>
<li><a href="#no-logs-found-for-a-tool">No logs found for a tool</a></li>
<li><a href="#error-messages-about-mkl-trial-mode--licences">Error messages about mkl trial mode / licences</a></li>
<li><a href="#locale-error-messages">Locale Error Messages</a></li>
<li><a href="#no-space-left-on-device">No space left on device</a></li>
</ul>
</li>

</ul>
</li>
<li><a href="#multiqc-modules">MultiQC Modules</a>
<ul class="nav nav-stacked">
<li><a href="#pre-alignment">Pre-alignment</a>
<ul class="nav nav-stacked">
<li><a href="#adapter-removal">Adapter Removal</a></li>
<li><a href="#afterqc">AfterQC</a></li>
<li><a href="#bcl2fastq">bcl2fastq</a></li>
<li><a href="#bclconvert">bclconvert</a></li>
<li><a href="#biobloom-tools">BioBloom Tools</a></li>
<li><a href="#ccs">CCS</a></li>
<li><a href="#checkqc">CheckQC</a></li>
<li><a href="#clipandmerge">ClipAndMerge</a></li>
<li><a href="#cluster-flow">Cluster Flow</a></li>
<li><a href="#cutadapt">Cutadapt</a></li>
<li><a href="#fastp">Fastp</a></li>
<li><a href="#fastq-screen">FastQ Screen</a></li>
<li><a href="#fastqc">FastQC</a>
<ul class="nav nav-stacked">
<li><a href="#theoretical-gc-content">Theoretical GC Content</a></li>
<li><a href="#changing-the-order-of-sections">Changing the order of sections</a></li>
</ul>
</li>
<li><a href="#flash">FLASh</a></li>
<li><a href="#flexbar">Flexbar</a></li>
<li><a href="#interop">InterOp</a></li>
<li><a href="#ivar">iVar</a></li>
<li><a href="#jellyfish">Jellyfish</a></li>
<li><a href="#kat">KAT</a></li>
<li><a href="#leehom">leeHom</a></li>
<li><a href="#lima">Lima</a></li>
<li><a href="#minionqc">minionqc</a></li>
<li><a href="#pbmarkdup">pbmarkdup</a></li>
<li><a href="#pychopper">Pychopper</a></li>
<li><a href="#pycoqc">pycoQC</a></li>
<li><a href="#seqyclean">SeqyClean</a></li>
<li><a href="#sickle">Sickle</a></li>
<li><a href="#skewer">Skewer</a></li>
<li><a href="#sortmerna">SortMeRNA</a></li>
<li><a href="#trimmomatic">Trimmomatic</a></li>
</ul>
</li>
<li><a href="#aligners">Aligners</a>
<ul class="nav nav-stacked">
<li><a href="#bbmap">BBMap</a></li>
<li><a href="#biscuit">BISCUIT</a>
<ul class="nav nav-stacked">
<li><a href="#insert-size-distribution">Insert Size Distribution</a></li>
</ul>
</li>
<li><a href="#bismark">Bismark</a></li>
<li><a href="#bowtie-1">Bowtie 1</a></li>
<li><a href="#bowtie-2">Bowtie 2</a></li>
<li><a href="#dragen">DRAGEN</a></li>
<li><a href="#hic-pro">HiC-Pro</a></li>
<li><a href="#hicup">HiCUP</a></li>
<li><a href="#hisat2">HISAT2</a></li>
<li><a href="#kallisto">Kallisto</a></li>
<li><a href="#longranger">Longranger</a></li>
<li><a href="#malt">MALT</a></li>
<li><a href="#salmon">Salmon</a></li>
<li><a href="#star">STAR</a></li>
<li><a href="#tophat">TopHat</a></li>
</ul>
</li>
<li><a href="#post-alignment">Post-alignment</a>
<ul class="nav nav-stacked">
<li><a href="#bamtools">Bamtools</a></li>
<li><a href="#bcftools">Bcftools</a></li>
<li><a href="#biobambam2">biobambam2</a></li>
<li><a href="#busco">BUSCO</a></li>
<li><a href="#bustools">bustools</a></li>
<li><a href="#conpair">Conpair</a></li>
<li><a href="#damageprofiler">DamageProfiler</a></li>
<li><a href="#dedup">DeDup</a></li>
<li><a href="#deeptools">deepTools</a></li>
<li><a href="#disambiguate">Disambiguate</a></li>
<li><a href="#eigenstratdatabasetools">EigenStratDatabaseTools</a></li>
<li><a href="#featurecounts">featureCounts</a></li>
<li><a href="#fgbio">Fgbio</a></li>
<li><a href="#gatk">GATK</a>
<ul class="nav nav-stacked">
<li><a href="#baserecalibrator">BaseRecalibrator</a></li>
<li><a href="#varianteval">VariantEval</a></li>
</ul>
</li>
<li><a href="#gffcompare">GffCompare</a></li>
<li><a href="#goleft-indexcov">goleft indexcov</a></li>
<li><a href="#hap.py">Hap.py</a></li>
<li><a href="#hicexplorer">HiCExplorer</a></li>
<li><a href="#homer">HOMER</a></li>
<li><a href="#hops">HOPS</a></li>
<li><a href="#htseq">HTSeq</a></li>
<li><a href="#jcvi-genome-annotation">JCVI Genome Annotation</a></li>
<li><a href="#kaiju">Kaiju</a></li>
<li><a href="#kraken">Kraken</a></li>
<li><a href="#macs2">MACS2</a></li>
<li><a href="#methylqa">methylQA</a></li>
<li><a href="#mirtop">mirtop</a></li>
<li><a href="#mirtrace">miRTrace</a></li>
<li><a href="#mosdepth">mosdepth</a></li>
<li><a href="#mtnucratio">MTNucRatio</a></li>
<li><a href="#multivcfanalyzer">MultiVCFAnalyzer</a></li>
<li><a href="#nanostat">NanoStat</a></li>
<li><a href="#ngsderive">ngsderive</a></li>
<li><a href="#odgi">odgi</a></li>
<li><a href="#optitype">OptiType</a></li>
<li><a href="#pangolin">Pangolin</a></li>
<li><a href="#peddy">Peddy</a></li>
<li><a href="#phantompeakqualtools">phantompeakqualtools</a></li>
<li><a href="#picard">Picard</a></li>
<li><a href="#preseq">Preseq</a></li>
<li><a href="#prokka">Prokka</a></li>
<li><a href="#purple">PURPLE</a></li>
<li><a href="#qc3c">qc3C</a></li>
<li><a href="#qorts">QoRTs</a></li>
<li><a href="#qualimap">Qualimap</a></li>
<li><a href="#quast">QUAST</a></li>
<li><a href="#rna-seqc">RNA-SeQC</a></li>
<li><a href="#rockhopper">Rockhopper</a></li>
<li><a href="#rsem">RSEM</a></li>
<li><a href="#rseqc">RSeQC</a></li>
<li><a href="#sambamba">Sambamba</a>
<ul class="nav nav-stacked">
<li><a href="#markdup">markdup</a></li>
</ul>
</li>
<li><a href="#samblaster">Samblaster</a></li>
<li><a href="#samtools">Samtools</a>
<ul class="nav nav-stacked">
<li><a href="#idxstats">idxstats</a></li>
</ul>
</li>
<li><a href="#sargasso">Sargasso</a></li>
<li><a href="#sentieon">Sentieon</a></li>
<li><a href="#sex.deterrmine">Sex.DetErrMine</a></li>
<li><a href="#slamdunk">Slamdunk</a></li>
<li><a href="#snippy">Snippy</a></li>
<li><a href="#snpeff">SnpEff</a></li>
<li><a href="#snpsplit">SNPsplit</a></li>
<li><a href="#somalier">Somalier</a></li>
<li><a href="#stacks">Stacks</a></li>
<li><a href="#supernova">Supernova</a></li>
<li><a href="#theta2">THeTA2</a></li>
<li><a href="#varscan2">VarScan2</a></li>
<li><a href="#vcftools">VCFTools</a>
<ul class="nav nav-stacked">
<li><a href="#important-general-note">Important General Note</a></li>
<li><a href="#implemented">Implemented</a></li>
<li><a href="#to-do">To do</a></li>
</ul>
</li>
<li><a href="#vep">VEP</a></li>
<li><a href="#verifybamid">VerifyBAMID</a></li>
<li><a href="#whatshap">WhatsHap</a></li>
</ul>
</li>

</ul>
</li>
<li><a href="#custom-content">Custom Content</a>
<ul class="nav nav-stacked">
<li><a href="#introduction-1">Introduction</a>
<ul class="nav nav-stacked">
<li><a href="#data-from-a-released-tool">Data from a released tool</a></li>
<li><a href="#images">Images</a></li>
<li><a href="#multiqc-specific-data-file">MultiQC-specific data file</a></li>
<li><a href="#data-as-part-of-multiqc-config">Data as part of MultiQC config</a></li>
<li><a href="#separate-configuration-and-data-files">Separate configuration and data files</a></li>
</ul>
</li>
<li><a href="#configuration-1">Configuration</a>
<ul class="nav nav-stacked">
<li><a href="#grouping-sections-and-subsections">Grouping sections and subsections</a></li>
<li><a href="#order-of-sections">Order of sections</a></li>
<li><a href="#section-configuration">Section configuration</a></li>
<li><a href="#plot-configuration">Plot configuration</a></li>
<li><a href="#tricky-extras">Tricky extras</a>
<ul class="nav nav-stacked">
<li><a href="#differences-between-tables-and-general-stats">Differences between Tables and General Stats</a></li>
<li><a href="#first-columns-in-tables-are-special">First columns in tables are special</a></li>
</ul>
</li>
<li><a href="#linting">Linting</a></li>
</ul>
</li>
<li><a href="#examples">Examples</a></li>
</ul>
</li>
<li><a href="#coding-with-multiqc">Coding with MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#writing-new-modules">Writing New Modules</a>
<ul class="nav nav-stacked">
<li><a href="#introduction-2">Introduction</a></li>
<li><a href="#core-modules--plugins">Core modules / plugins</a></li>
<li><a href="#multiqc-lint-tests">MultiQC Lint Tests</a></li>
<li><a href="#code-formatting">Code formatting</a>
<ul class="nav nav-stacked">
<li><a href="#use-version-control">Use version control</a></li>
<li><a href="#black">Black</a></li>
<li><a href="#prettier">Prettier</a></li>
<li><a href="#markdownlint">Markdownlint</a></li>
</ul>
</li>
<li><a href="#initial-setup">Initial setup</a>
<ul class="nav nav-stacked">
<li><a href="#multiqc-file-structure">MultiQC file structure</a></li>
<li><a href="#submodule">Submodule</a></li>
<li><a href="#entry-points">Entry points</a></li>
<li><a href="#multiqc-config">MultiQC config</a></li>
<li><a href="#documentation">Documentation</a></li>
<li><a href="#changelog">Changelog</a></li>
<li><a href="#multiqcmodule-class">MultiqcModule Class</a></li>
<li><a href="#logging">Logging</a></li>
</ul>
</li>
<li><a href="#step-1-find-log-files">Step 1 - Find log files</a></li>
<li><a href="#step-2-parse-data-from-the-input-files">Step 2 - Parse data from the input files</a>
<ul class="nav nav-stacked">
<li><a href="#filtering-by-parsed-sample-names">Filtering by parsed sample names</a></li>
<li><a href="#no-files-found">No files found</a></li>
<li><a href="#custom-sample-names">Custom sample names</a></li>
<li><a href="#identical-sample-names">Identical sample names</a></li>
<li><a href="#printing-to-the-sources-file">Printing to the sources file</a></li>
</ul>
</li>
<li><a href="#step-3-adding-to-the-general-statistics-table">Step 3 - Adding to the general statistics table</a>
<ul class="nav nav-stacked">
<li><a href="#table-colour-scales">Table colour scales</a></li>
</ul>
</li>
<li><a href="#step-4-writing-data-to-a-file">Step 4 - Writing data to a file</a></li>
<li><a href="#step-5-create-report-sections">Step 5 - Create report sections</a></li>
<li><a href="#step-6-plot-some-data">Step 6 - Plot some data</a></li>
<li><a href="#appendices">Appendices</a>
<ul class="nav nav-stacked">
<li><a href="#user-configuration">User configuration</a></li>
<li><a href="#profiling-performance">Profiling Performance</a></li>
<li><a href="#adding-custom-css--javascript">Adding Custom CSS / Javascript</a></li>
</ul>
</li>

</ul>
</li>
<li><a href="#plotting-functions">Plotting Functions</a>
<ul class="nav nav-stacked">
<li><a href="#common-options">Common options</a></li>
<li><a href="#bar-graphs">Bar graphs</a>
<ul class="nav nav-stacked">
<li><a href="#switching-datasets">Switching datasets</a></li>
<li><a href="#interactive--flat-image-plots">Interactive / Flat image plots</a></li>
</ul>
</li>
<li><a href="#line-graphs">Line graphs</a>
<ul class="nav nav-stacked">
<li><a href="#switching-datasets-1">Switching datasets</a></li>
<li><a href="#additional-data-series">Additional data series</a></li>
</ul>
</li>
<li><a href="#scatter-plots">Scatter Plots</a></li>
<li><a href="#creating-a-table">Creating a table</a>
<ul class="nav nav-stacked">
<li><a href="#table-decimal-places">Table decimal places</a></li>
<li><a href="#table-colour-scales-1">Table colour scales</a></li>
<li><a href="#custom-cell-background-colours">Custom cell background colours</a></li>
<li><a href="#zero-centrepoints">Zero centrepoints</a></li>
<li><a href="#conditional-formatting-of-data-values">Conditional formatting of data values</a></li>
</ul>
</li>
<li><a href="#beeswarm-plots-dot-plots">Beeswarm plots (dot plots)</a></li>
<li><a href="#heatmaps">Heatmaps</a></li>
<li><a href="#javascript-functions">Javascript Functions</a>
<ul class="nav nav-stacked">
<li><a href="#plotting-line-graphs">Plotting line graphs</a></li>
<li><a href="#plotting-bar-graphs">Plotting bar graphs</a></li>
<li><a href="#switching-counts-and-percentages">Switching counts and percentages</a></li>
<li><a href="#switching-plot-datasets">Switching plot datasets</a></li>
<li><a href="#custom-event-triggers">Custom event triggers</a></li>
</ul>
</li>

</ul>
</li>
<li><a href="#multiqc-plugins">MultiQC Plugins</a>
<ul class="nav nav-stacked">
<li><a href="#entry-points-1">Entry Points</a></li>
<li><a href="#modules">Modules</a></li>
<li><a href="#templates">Templates</a></li>
<li><a href="#command-line-options">Command line options</a></li>
<li><a href="#hooks">Hooks</a></li>
</ul>
</li>
<li><a href="#writing-new-templates">Writing New Templates</a>
<ul class="nav nav-stacked">
<li><a href="#core-or-plugin">Core or plugin</a></li>
<li><a href="#creating-a-template-skeleton">Creating a template skeleton</a></li>
<li><a href="#child-templates">Child templates</a></li>
<li><a href="#extra-init-variables">Extra init variables</a></li>
<li><a href="#jinja-template-variables">Jinja template variables</a></li>
<li><a href="#appendices-1">Appendices</a>
<ul class="nav nav-stacked">
<li><a href="#custom-plotting-functions">Custom plotting functions</a></li>
</ul>
</li>

</ul>
</li>
<li><a href="#updating-for-compatibility">Updating for compatibility</a>
<ul class="nav nav-stacked">
<li><a href="#v1.0-updates">v1.0 Updates</a></li></ul></li></ul></li></ul>            <p class="backtotop"><a href="http://gitter.im/ewels/MultiQC">Discuss on Gitter</a></p>
            <p class="backtotop"><a href="#">Back to top</a></p>
          </div>
        </div>
        <div class="col-sm-9 col-md-pull-3">
          <div class="docs_section">
<h1 class="section-header" id="using-multiqc"><a href="#using-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using MultiQC</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#introduction">Introduction</a></li>
<li class="active"><a href="#installing-multiqc">Installing MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#system-python">System Python</a></li>
<li><a href="#installing-python">Installing Python</a></li>
<li><a href="#installing-with-conda">Installing with conda</a></li>
<li><a href="#installation-with-pip">Installation with pip</a></li>
<li><a href="#manual-installation">Manual installation</a></li>
<li><a href="#updating-multiqc">Updating MultiQC</a></li>
<li><a href="#using-a-specific-python-interpreter">Using a specific python interpreter</a></li>
<li><a href="#using-with-a-python-script">Using with a Python script</a></li>
<li><a href="#installing-on-windows">Installing on Windows</a></li>
<li class="active"><a href="#using-the-docker-container">Using the Docker container</a>
<ul class="nav nav-stacked">
<li><a href="#docker-bash-alias">Docker bash alias</a></li>
</ul>
</li>
<li><a href="#using-singularity">Using Singularity</a>
<ul class="nav nav-stacked">
<li><a href="#import-errors-with-singularity">Import errors with Singularity</a></li>
</ul>
</li>
<li><a href="#python-2">Python 2</a></li>
<li><a href="#using-multiqc-through-galaxy">Using MultiQC through Galaxy</a>
<ul class="nav nav-stacked">
<li><a href="#on-the-main-galaxy-instance">On the main Galaxy instance</a></li>
<li><a href="#on-your-instance">On your instance</a></li>
</ul>
</li>
<li><a href="#installing-from-freebsd">Installing from FreeBSD</a></li>
<li><a href="#installing-as-an-environment-module">Installing as an environment module</a></li>
</ul>
</li>
<li><a href="#running-multiqc">Running MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#choosing-where-to-scan">Choosing where to scan</a></li>
<li><a href="#renaming-reports">Renaming reports</a></li>
<li><a href="#overwriting-existing-reports">Overwriting existing reports</a></li>
<li><a href="#sample-names-prefixed-with-directories">Sample names prefixed with directories</a></li>
<li><a href="#using-different-templates">Using different templates</a></li>
<li><a href="#pdf-reports">PDF Reports</a></li>
<li><a href="#printing-to-stdout">Printing to stdout</a></li>
<li><a href="#parsed-data-directory">Parsed data directory</a></li>
<li><a href="#exporting-plots">Exporting Plots</a></li>
<li><a href="#choosing-which-modules-to-run">Choosing which modules to run</a></li>
</ul>
</li>
<li><a href="#using-multiqc-reports">Using MultiQC Reports</a>
<ul class="nav nav-stacked">
<li><a href="#browser-compatibility">Browser compatibility</a></li>
<li><a href="#report-layout">Report layout</a></li>
<li><a href="#general-statistics-table">General Statistics table</a></li>
<li><a href="#plots">Plots</a>
<ul class="nav nav-stacked">
<li><a href="#interactive-plots">Interactive plots</a></li>
<li><a href="#flat-plots">Flat plots</a></li>
<li><a href="#exporting-plots-1">Exporting plots</a></li>
<li><a href="#dynamic-plots">Dynamic plots</a></li>
</ul>
</li>
<li><a href="#toolbox">Toolbox</a>
<ul class="nav nav-stacked">
<li><a href="#highlight-samples">Highlight Samples</a></li>
<li><a href="#renaming-samples">Renaming Samples</a></li>
<li><a href="#hiding-samples">Hiding Samples</a></li>
<li><a href="#export">Export</a></li>
<li><a href="#save-settings">Save Settings</a></li>
</ul>
</li>

</ul>
</li>
<li><a href="#configuring-multiqc">Configuring MultiQC</a>
<ul class="nav nav-stacked">
<li><a href="#sample-name-cleaning">Sample name cleaning</a>
<ul class="nav nav-stacked">
<li><a href="#other-search-types">Other search types</a></li>
<li><a href="#clashing-sample-names">Clashing sample names</a></li>
</ul>
</li>
<li><a href="#module-search-patterns">Module search patterns</a></li>
<li><a href="#using-log-filenames-as-sample-names">Using log filenames as sample names</a></li>
<li><a href="#ignoring-files">Ignoring Files</a></li>
<li><a href="#ignoring-samples">Ignoring samples</a></li>
<li><a href="#large-sample-numbers">Large sample numbers</a>
<ul class="nav nav-stacked">
<li><a href="#disabling-on-load-plotting">Disabling on-load plotting</a></li>
<li><a href="#flat--interactive-plots">Flat / interactive plots</a></li>
<li><a href="#tables--beeswarm-plots">Tables / Beeswarm plots</a></li>
</ul>
</li>
<li><a href="#coloured-log-output">Coloured log output</a></li>
<li><a href="#checks-for-new-versions">Checks for new versions</a></li>
<li><a href="#command-line-config">Command-line config</a></li>
<li><a href="#optimising-run-time">Optimising run-time</a>
<ul class="nav nav-stacked">
<li><a href="#profile-your-multiqc-run-time">Profile your MultiQC run time</a></li>
<li><a href="#be-picky-with-which-modules-are-run">Be picky with which modules are run</a></li>
<li><a href="#optimise-file-search-patterns">Optimise file search patterns</a></li>
<li><a href="#force-interactive-plots">Force interactive plots</a></li>
<li><a href="#skip-the-report-if-you-dont-need-it">Skip the report if you don't need it</a></li>
</ul>
</li>
<li><a href="#custom-css-files">Custom CSS files</a></li>
</ul>
</li>
<li><a href="#customising-reports">Customising Reports</a>
<ul class="nav nav-stacked">
<li><a href="#titles-and-introductory-text">Titles and introductory text</a>
<ul class="nav nav-stacked">
<li><a href="#report-time-and-analysis-paths">Report time and analysis paths</a></li>
</ul>
</li>
<li><a href="#report-logo">Report Logo</a></li>
<li><a href="#project-level-information">Project level information</a></li>
<li><a href="#sample-name-replacement">Sample name replacement</a></li>
<li><a href="#bulk-sample-renaming-in-reports">Bulk sample renaming in reports</a></li>
<li><a href="#show--hide-samples-buttons">Show / Hide samples buttons</a></li>
<li><a href="#module-and-section-comments">Module and section comments</a></li>
<li><a href="#removing-modules-or-sections">Removing modules or sections</a></li>
<li><a href="#order-of-modules">Order of modules</a>
<ul class="nav nav-stacked">
<li><a href="#order-of-module-and-module-subsection-output">Order of module and module subsection output</a></li>
</ul>
</li>
<li><a href="#customising-plots">Customising plots</a></li>
<li><a href="#customising-tables">Customising tables</a>
<ul class="nav nav-stacked">
<li><a href="#hiding-columns">Hiding columns</a></li>
<li><a href="#column-order">Column order</a></li>
<li><a href="#column-titles">Column titles</a></li>
<li><a href="#conditional-formatting">Conditional formatting</a></li>
</ul>
</li>
<li><a href="#number-base-multiplier">Number base (multiplier)</a></li>
<li><a href="#number-formatting">Number formatting</a></li>
<li><a href="#troubleshooting">Troubleshooting</a></li>
</ul>
</li>
<li><a href="#using-multiqc-in-pipelines">Using MultiQC in pipelines</a>
<ul class="nav nav-stacked">
<li><a href="#nextflow">Nextflow</a>
<ul class="nav nav-stacked">
<li><a href="#clashing-input-filenames">Clashing input filenames</a></li>
<li><a href="#custom-run-name">Custom run name</a></li>
<li><a href="#multiqc-config-file">MultiQC config file</a></li>
</ul>
</li>
<li><a href="#snakemake">Snakemake</a></li>
</ul>
</li>
<li><a href="#downstream-analysis">Downstream analysis</a>
<ul class="nav nav-stacked">
<li><a href="#tidymultiqc">TidyMultiqc</a></li>
<li><a href="#megaqc">MegaQC</a></li>
<li><a href="#chronqc">ChronQC</a></li>
</ul>
</li>
<li><a href="#troubleshooting-1">Troubleshooting</a>
<ul class="nav nav-stacked">
<li><a href="#not-enough-samples-found">Not enough samples found</a>
<ul class="nav nav-stacked">
<li><a href="#big-log-files">Big log files</a></li>
</ul>
</li>
<li><a href="#no-logs-found-for-a-tool">No logs found for a tool</a></li>
<li><a href="#error-messages-about-mkl-trial-mode--licences">Error messages about mkl trial mode / licences</a></li>
<li><a href="#locale-error-messages">Locale Error Messages</a></li>
<li><a href="#no-space-left-on-device">No space left on device</a></li>
</ul>
</li>

</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<div class="docs_block" id="installation.md"><h1 id="introduction"><a href="#introduction" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Introduction</h1>
<p>MultiQC is a reporting tool that parses summary statistics from results and log files
generated by other bioinformatics tools. MultiQC doesn't run other tools for you -
it's designed to be placed at the end of analysis pipelines or to be run manually
when you've finished running your tools.</p>
<p>When you launch MultiQC, it recursively searches through any provided file paths
and finds files that it recognises. It parses relevant information from these and
generates a single stand-alone HTML report file. It also saves a directory of data
files with all parsed data for further downstream use.</p>
<h1 id="installing-multiqc"><a href="#installing-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installing MultiQC</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#system-python">System Python</a></li>
<li><a href="#installing-python">Installing Python</a></li>
<li><a href="#installing-with-conda">Installing with conda</a></li>
<li><a href="#installation-with-pip">Installation with pip</a></li>
<li><a href="#manual-installation">Manual installation</a></li>
<li><a href="#updating-multiqc">Updating MultiQC</a></li>
<li><a href="#using-a-specific-python-interpreter">Using a specific python interpreter</a></li>
<li><a href="#using-with-a-python-script">Using with a Python script</a></li>
<li><a href="#installing-on-windows">Installing on Windows</a></li>
<li class="active"><a href="#using-the-docker-container">Using the Docker container</a>
<ul class="nav nav-stacked">
<li><a href="#docker-bash-alias">Docker bash alias</a></li>
</ul>
</li>
<li><a href="#using-singularity">Using Singularity</a>
<ul class="nav nav-stacked">
<li><a href="#import-errors-with-singularity">Import errors with Singularity</a></li>
</ul>
</li>
<li><a href="#python-2">Python 2</a></li>
<li><a href="#using-multiqc-through-galaxy">Using MultiQC through Galaxy</a>
<ul class="nav nav-stacked">
<li><a href="#on-the-main-galaxy-instance">On the main Galaxy instance</a></li>
<li><a href="#on-your-instance">On your instance</a></li>
</ul>
</li>
<li><a href="#installing-from-freebsd">Installing from FreeBSD</a></li>
<li><a href="#installing-as-an-environment-module">Installing as an environment module</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<h2 id="system-python"><a href="#system-python" class="header-link"><span class="glyphicon glyphicon-link"></span></a>System Python</h2>
<p>Before we start - a quick note that using the system-wide installation of Python
is <em>not</em> recommended. This often causes problems and it's a little risky to mess
with it. If you find yourself prepending <code>sudo</code> to any MultiQC commands, take a
step back and think about Python virtual environments / conda instead (see below).</p>
<h2 id="installing-python"><a href="#installing-python" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installing Python</h2>
<p>To see if you have python installed, run <code>python --version</code> on the command line.
MultiQC needs Python version 2.7+, 3.4+ or 3.5+.</p>
<p>We recommend using virtual environments to manage your Python installation.
Our favourite is <em>conda</em>, a cross-platform tool to manage Python environments.
You can installation instructions for Miniconda
<a href="https://docs.conda.io/en/latest/miniconda.html">here</a>.</p>
<p>Once conda is installed, you can create a Python environment with the following commands:</p>
<pre><code class="language-bash hljs">conda create --name py3.7 python=3.7
conda activate py3.7</code></pre>
<p>You'll want to add the <code>conda activate py3.7</code> line to your <code>.bashrc</code> file so
that the environment is loaded every time you load the terminal.</p>
<h2 id="installing-with-conda"><a href="#installing-with-conda" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installing with conda</h2>
<p>If you're using <code>conda</code> as described above, you can install MultiQC from the <code>bioconda</code>
channel as follows:</p>
<pre><code class="language-bash hljs">conda install -c bioconda -c conda-forge multiqc</code></pre>
<p>Please see the <a href="https://bioconda.github.io/user/install.html">Bioconda documentation</a> for more details.</p>
<h2 id="installation-with-pip"><a href="#installation-with-pip" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installation with pip</h2>
<p>This is the easiest way to install MultiQC. <code>pip</code> is the package manager for
the Python Package Manager. It comes bundled with recent versions of Python,
otherwise you can find installation instructions <a href="http://pip.readthedocs.org/en/stable/installing/">here</a>.</p>
<p>You can now install MultiQC from
<a href="https://pypi.python.org/pypi/multiqc">PyPI</a> as follows:</p>
<pre><code class="language-bash hljs">pip install multiqc</code></pre>
<p>If you would like the development version, the command is:</p>
<pre><code class="language-bash hljs">pip install git+https://github.com/ewels/MultiQC.git</code></pre>
<p>Note that if you have problems with read-only directories, you can install to
your home directory with the <code>--user</code> parameter (though it's probably better
to use virtual environments, as described above).</p>
<pre><code class="language-bash hljs">pip install --user multiqc</code></pre>
<h2 id="manual-installation"><a href="#manual-installation" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Manual installation</h2>
<p>If you'd rather not use either of these tools, you can clone the code and install the code yourself:</p>
<pre><code class="language-bash hljs">git <span class="hljs-built_in">clone</span> https://github.com/ewels/MultiQC.git
<span class="hljs-built_in">cd</span> MultiQC
pip install .</code></pre>
<p><code>git</code> not installed? No problem - just download the flat files:</p>
<pre><code class="language-bash hljs">curl -LOk https://github.com/ewels/MultiQC/archive/master.zip
unzip master.zip
<span class="hljs-built_in">cd</span> MultiQC-master
pip install .</code></pre>
<p>Note that it is <em>not</em> recommended to use the command <code>python setup.py install</code>
as this has been superseded by <code>pip</code> and does not correctly handle some package
management, such as pre-releases.</p>
<h2 id="updating-multiqc"><a href="#updating-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Updating MultiQC</h2>
<p>You can update MultiQC from <a href="https://pypi.python.org/pypi/multiqc">PyPI</a>
at any time by running the following command:</p>
<pre><code class="language-bash hljs">pip install --upgrade multiqc</code></pre>
<p>To update the development version, use:</p>
<pre><code class="language-bash hljs">pip install --force git+https://github.com/ewels/MultiQC.git</code></pre>
<p>If you cloned the <code>git</code> repo, just pull the latest changes and install:</p>
<pre><code class="language-bash hljs"><span class="hljs-built_in">cd</span> MultiQC
git pull
pip install .</code></pre>
<p>If you downloaded the flat files, just repeat the installation procedure.</p>
<h2 id="using-a-specific-python-interpreter"><a href="#using-a-specific-python-interpreter" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using a specific python interpreter</h2>
<p>If you prefer, you can also run MultiQC with a specific python interpreter.
The command line usage and flags are then exactly the same as if you ran just <code>multiqc</code>.</p>
<p>For example:</p>
<pre><code class="language-bash hljs">python -m multiqc .
python3 -m multiqc .
~/my_env/bin/python -m multiqc .</code></pre>
<h2 id="using-with-a-python-script"><a href="#using-with-a-python-script" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using with a Python script</h2>
<p>You can import and run MultiQC from within a Python script, using
the <code>multiqc.run()</code> function as follows:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">import</span> multiqc
multiqc.run(<span class="hljs-string">"/path/to/dir"</span>)</code></pre>
<h2 id="installing-on-windows"><a href="#installing-on-windows" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installing on Windows</h2>
<p>MultiQC is has primarily been designed for us on Unix systems (Linux, Mac OSX).
However, it <em>should</em> work on Windows too. Indeed, automated
<a href="https://github.com/ewels/MultiQC/actions">continuous integration tests</a>
run using GitHub Actions to check compatibility (see test config
<a href="https://github.com/ewels/MultiQC/blob/master/.github/workflows/multiqc_windows.yml">here</a>).</p>
<p>Note that support for using the base <code>multiqc</code> command was improved in MultiQC version 1.8.</p>
<h2 id="using-the-docker-container"><a href="#using-the-docker-container" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using the Docker container</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#docker-bash-alias">Docker bash alias</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>A Docker container is provided on Docker Hub called ewels/MultiQC.
It's based on <code>czentye/matplotlib-minimal</code> to give the smallest size I could manage (~80MB).</p>
<p>To use, call the <code>docker run</code> with your current working directory mounted as a volume and working directory:</p>
<pre><code class="language-bash hljs">docker run -t -v `<span class="hljs-built_in">pwd</span>`:`<span class="hljs-built_in">pwd</span>` -w `<span class="hljs-built_in">pwd</span>` ewels/multiqc</code></pre>
<p>By default, docker will use the <code>:latest</code> tag. For MultiQC, this is set to be the most recent release.
To use the most recent development code, use <code>ewels/multiqc::dev</code>.
You can also specify specific versions, eg: <code>ewels/multiqc:1.9</code>.</p>
<p>You can also specify additional MultiQC parameters as normal:</p>
<pre><code class="language-bash hljs">docker run -t -v `<span class="hljs-built_in">pwd</span>`:`<span class="hljs-built_in">pwd</span>` -w `<span class="hljs-built_in">pwd</span>` ewels/multiqc . --title <span class="hljs-string">"My amazing report"</span> -b <span class="hljs-string">"This was made with docker"</span></code></pre>
<p>Note that all files on the command line (eg. config files) must also be mounted in the docker container to be accessible.
For more help, look into <a href="https://docs.docker.com/engine/reference/commandline/run/">the Docker documentation</a></p>
<h3 id="docker-bash-alias"><a href="#docker-bash-alias" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Docker bash alias</h3>
<p>The above base command is a little verbose, so if you are using this a
 lot it may be worth adding the following bash alias to your <code>~/.bashrc</code> file:</p>
<pre><code class="language-bash hljs"><span class="hljs-built_in">alias</span> multiqc=<span class="hljs-string">"docker run -tv `pwd`:`pwd` -w `pwd` ewels/multiqc"</span></code></pre>
<p>Once applied (you may need to reload your shell if added to your <code>.bashrc</code>) you can then just use the <code>multiqc</code> instead:</p>
<pre><code class="language-bash hljs">multiqc .</code></pre>
<h2 id="using-singularity"><a href="#using-singularity" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using Singularity</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#import-errors-with-singularity">Import errors with Singularity</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Although there is no dedicated Singularity image available for MultiQC, you can use the above Docker container.</p>
<p>First, build a singularity container image from the docker image (where <code>1.9</code> is the MultiQC version):</p>
<pre><code class="language-bash hljs">singularity build multiqc-1.9.sif docker://ewels/multiqc:1.9</code></pre>
<p>Then, use <code>singularity run</code> to run the image with the normal MultiQC arguments:</p>
<pre><code class="language-bash hljs">singularity run multiqc-1.9.sif my_results/ --title <span class="hljs-string">"Report made using Singularity"</span></code></pre>
<h3 id="import-errors-with-singularity"><a href="#import-errors-with-singularity" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Import errors with Singularity</h3>
<p>Sometimes, Singularity can be over-ambitious with sharing file paths which can result in the Python
environment in your local system interacting with Python inside the image.
This can give rise to <code>ImportError</code> errors for <code>numpy</code> and other packages.</p>
<p>The giveaway for when this is the problem is that traceback will list python package paths which
are on your system and look different that of MultiQC inside the container (eg. <code>/usr/lib/python3.8/site-packages/multiqc/</code>).</p>
<p>To fix this, run the command <code>export PYTHONNOUSERSITE=1</code> before running MultiQC.
This variable <a href="https://docs.python.org/3/using/cmdline.html#envvar-PYTHONNOUSERSITE">tells Python</a>
not to add site-packages to the system path when loading, which should avoid the conflicts.</p>
<h2 id="python-2"><a href="#python-2" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Python 2</h2>
<p>As of MultiQC version 1.9, <strong>Python 2 is no longer officially supported</strong>.
Automatic CI tests will no longer run with Python 2 and Python 2 specific workarounds
are no longer guaranteed.</p>
<p>Whilst it may be possible to continue using MultiQC with Python 2 for a short time by
pinning dependencies, MultiQC compatibility for Python 2 will now slowly drift and start
to break. If you haven't already, <strong>you need to switch to Python 3 now</strong>.</p>
<p>Python 2 had its <a href="https://www.python.org/doc/sunset-python-2/">official sunset date</a>
on January 1st 2020, meaning that it will no longer be developed by the Python community.
Part of the <a href="https://www.python.org/doc/sunset-python-2/">python.org statement</a> reads:</p>
<blockquote>
<p>That means that we will not improve it anymore after that day,
even if someone finds a security problem in it.
You should upgrade to Python 3 as soon as you can.</p>
</blockquote>
<p><a href="https://python3statement.org/">Very many Python packages no longer support Python 2</a>
and it whilst the MultiQC code is currently compatible with both Python 2 and Python 3,
it is increasingly difficult to maintain compatibility with the dependency packages it
uses, such as MatPlotLib, numpy and more.</p>
<h2 id="using-multiqc-through-galaxy"><a href="#using-multiqc-through-galaxy" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using MultiQC through Galaxy</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#on-the-main-galaxy-instance">On the main Galaxy instance</a></li>
<li><a href="#on-your-instance">On your instance</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<h3 id="on-the-main-galaxy-instance"><a href="#on-the-main-galaxy-instance" class="header-link"><span class="glyphicon glyphicon-link"></span></a>On the main Galaxy instance</h3>
<p>The easiest and fast manner to use MutliQC is to use the <a href="https://usegalaxy.org/">usegalaxy.org</a> main Galaxy instance where you will find <a href="https://usegalaxy.org/?tool_id=toolshed.g2.bx.psu.edu%2Frepos%2Fengineson%2Fmultiqc%2Fmultiqc%2F1.0.0.0&amp;version=1.0.0.0&amp;__identifer=2sjdq8d9r3l">MultiQC Galaxy tool</a> under the <em>NGS: QC and manipualtion</em> tool panel section.</p>
<h3 id="on-your-instance"><a href="#on-your-instance" class="header-link"><span class="glyphicon glyphicon-link"></span></a>On your instance</h3>
<p>You can install MultiQC on your own Galaxy instance through your Galaxy admin space, searching on the <a href="https://toolshed.g2.bx.psu.edu/">main Toolshed</a> for the <a href="https://toolshed.g2.bx.psu.edu/view/iuc/multiqc/3bad335ccea9">MultiQC repository</a> available under the <em>visualization</em>, <em>statistics</em> and <em>Fastq Manipulation</em> sections.</p>
<h2 id="installing-from-freebsd"><a href="#installing-from-freebsd" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installing from FreeBSD</h2>
<p>If you're using FreeBSD you can install MultiQC via the FreeBSD ports system:</p>
<pre><code class="language-bash hljs">pkg install py36-multiqc</code></pre>
<p><em>(or <code>py27-multiqc</code>, <code>py37-multiqc</code>, or any other currently mainstream python version).</em></p>
<p>This will install a prebuilt binary using only highly-portable
optimizations, much like <code>apt</code>, <code>yum</code>, etc.</p>
<p>FreeBSD ports can also be built and installed from source:</p>
<pre><code class="language-bash hljs"><span class="hljs-built_in">cd</span> /usr/ports/biology/py-multiqc
make install</code></pre>
<p>To report issues with a FreeBSD port, please submit a PR on the
<a href="https://www.freebsd.org/support/bugreports.html">FreeBSD bug reports page</a>.
For more information, visit <a href="https://www.freebsd.org/ports/index.html"></a><a href="https://www.freebsd.org/ports/">https://www.freebsd.org/ports/</a></p>
<h2 id="installing-as-an-environment-module"><a href="#installing-as-an-environment-module" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Installing as an environment module</h2>
<p>Many people using MultiQC will be working on a HPC environment.
Every server / cluster is different, and you're probably best off asking
your friendly sysadmin to install MultiQC for you. However, with that
in mind, here are a few general tips for installing MultiQC into an
environment module system:</p>
<p>MultiQC comes in two parts - the <code>multiqc</code> python package and the
<code>multiqc</code> executable script. The former must be available in <code>$PYTHONPATH</code>
and the script must be available on the <code>$PATH</code>.</p>
<p>A typical installation procedure with an environment module Python install
might look like this: <em>(Note that <code>$PYTHONPATH</code> must be defined before <code>pip</code> installation.)</em></p>
<pre><code class="language-bash hljs">VERSION=0.7
INST=/path/to/software/multiqc/<span class="hljs-variable">$VERSION</span>
module load python/2.7.6
mkdir <span class="hljs-variable">$INST</span>
<span class="hljs-built_in">export</span> PYTHONPATH=<span class="hljs-variable">$INST</span>/lib/python2.7/site-packages
pip install --install-option=<span class="hljs-string">"--prefix=<span class="hljs-variable">$INST</span>"</span> multiqc</code></pre>
<p>Once installed, you'll need to create an environment module file.
Again, these vary between systems a lot, but here's an example:</p>
<pre><code class="language-bash hljs"><span class="hljs-comment">#%Module1.0#####################################################################</span>
<span class="hljs-comment">##</span>
<span class="hljs-comment">## MultiQC</span>
<span class="hljs-comment">##</span>

<span class="hljs-built_in">set</span> components [ file split [ module-info name ] ]
<span class="hljs-built_in">set</span> version [ lindex <span class="hljs-variable">$components</span> 1 ]
<span class="hljs-built_in">set</span> modroot /path/to/software/multiqc/<span class="hljs-variable">$version</span>

proc ModulesHelp { } {
    global version modroot
    puts stderr <span class="hljs-string">"\tMultiQC - use MultiQC <span class="hljs-variable">$version</span>"</span>
    puts stderr <span class="hljs-string">"\n\tVersion <span class="hljs-variable">$version</span>\n"</span>
}
module-whatis   <span class="hljs-string">"Loads MultiQC environment."</span>

<span class="hljs-comment"># load required modules</span>
module load python/2.7.6

<span class="hljs-comment"># only one version at a time</span>
conflict multiqc

<span class="hljs-comment"># Make the directories available</span>
prepend-path    PATH        <span class="hljs-variable">$modroot</span>/bin
prepend-path    PYTHONPATH  <span class="hljs-variable">$modroot</span>/lib/python2.7/site-packages</code></pre></div><div class="docs_block" id="usage.md"><h1 id="running-multiqc"><a href="#running-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Running MultiQC</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#choosing-where-to-scan">Choosing where to scan</a></li>
<li><a href="#renaming-reports">Renaming reports</a></li>
<li><a href="#overwriting-existing-reports">Overwriting existing reports</a></li>
<li><a href="#sample-names-prefixed-with-directories">Sample names prefixed with directories</a></li>
<li><a href="#using-different-templates">Using different templates</a></li>
<li><a href="#pdf-reports">PDF Reports</a></li>
<li><a href="#printing-to-stdout">Printing to stdout</a></li>
<li><a href="#parsed-data-directory">Parsed data directory</a></li>
<li><a href="#exporting-plots">Exporting Plots</a></li>
<li><a href="#choosing-which-modules-to-run">Choosing which modules to run</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Once installed, just go to your analysis directory and run <code>multiqc</code>, followed
by a list of directories to search. At it's simplest, this can just be <code>.</code>
(the current working directory):</p>
<pre><code class="language-bash hljs">multiqc .</code></pre>
<p>That's it! MultiQC will scan the specified directories and produce a report
based on details found in any log files that it recognises.</p>
<p>See <a href="http://multiqc.info/docs/#using-multiqc-reports">Using MultiQC Reports</a> for more information about how
to use the generated report.</p>
<p>For a description of all command line parameters, run <code>multiqc --help</code>.</p>
<h2 id="choosing-where-to-scan"><a href="#choosing-where-to-scan" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Choosing where to scan</h2>
<p>You can supply MultiQC with as many directories or files as you like. Above,
we supply <code>.</code> - just the current directory, but all of these would work too:</p>
<pre><code class="language-bash hljs">multiqc data/
multiqc data/ ../proj_one/analysis/ /tmp/results
multiqc data/*_fastqc.zip
multiqc data/sample_1*</code></pre>
<p>If the <code>--ignore-symlinks</code> flag is set, MultiQC will ignore symlinked directories and files.</p>
<p>You can also ignore files or directories using the <code>-x</code>/<code>--ignore</code> option.
This can be specified multiple times and accepts glob patterns (eg. using the <code>*</code> and <code>?</code> wildcards).</p>
<blockquote>
<p>Note that glob patterns should be enclosed in quotes to prevent them being expanded by bash</p>
</blockquote>
<p>The argument can match filenames, directory names and entire paths. For example:</p>
<pre><code class="language-bash hljs">multiqc . --ignore <span class="hljs-string">"file"</span>
multiqc . --ignore <span class="hljs-string">"fileA"</span> --ignore <span class="hljs-string">"fileB"</span>
multiqc . --ignore <span class="hljs-string">"_R?.zip"</span>
multiqc . --ignore <span class="hljs-string">"run_two/*"</span>
multiqc . --ignore <span class="hljs-string">"*/run_three/*/fastqc/*_R2.zip"</span></code></pre>
<p>Some modules get sample names from the contents of the file and not the filename
(for example, <code>stdout</code> logs can contain multiple samples). In this case, you can
skip samples by name instead:</p>
<pre><code class="language-bash hljs">multiqc . --ignore-samples <span class="hljs-string">"sample_3*"</span></code></pre>
<p>These strings are matched using glob logic (<code>*</code> and <code>?</code> are wildcards).</p>
<p>All of these settings can be saved in a MultiQC config file so that you don't have
to type them on the command line for every run.</p>
<p>Finally, you can supply a file containing a list of file paths, one per row.
MultiQC only search the listed files.</p>
<pre><code class="language-bash hljs">multiqc --file-list my_file_list.txt</code></pre>
<h2 id="renaming-reports"><a href="#renaming-reports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Renaming reports</h2>
<p>The report is called <code>multiqc_report.html</code> by default. Tab-delimited data files
are created in <code>multiqc_data/</code>, containing additional information.
You can use a custom name for the report with the <code>-n</code>/<code>--filename</code> parameter, or instruct
MultiQC to create them in a subdirectory using the <code>-o</code>/<code>-outdir</code> parameter.</p>
<p>Note that different MultiQC templates may have different defaults.</p>
<h2 id="overwriting-existing-reports"><a href="#overwriting-existing-reports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Overwriting existing reports</h2>
<p>It's quite common to repeatedly create new reports as new analysis results
are generated. Instead of manually deleting old reports, you can just specify
the <code>-f</code> parameter and MultiQC will overwrite any conflicting report filenames.</p>
<h2 id="sample-names-prefixed-with-directories"><a href="#sample-names-prefixed-with-directories" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Sample names prefixed with directories</h2>
<p>Sometimes, the same samples may be processed in different ways. If MultiQC
finds log files with the same sample name, the previous data will be overwritten
(this can be inspected by running MultiQC with <code>-v</code>/<code>--verbose</code>).</p>
<p>To avoid this, run MultiQC with the <code>-d</code>/<code>--dirs</code> parameter. This will prefix every
sample name with the directory path for that log file. As such, sample names should
now be unique, and not overwrite one-another.</p>
<p>By default, <code>--dirs</code> will prepend the entire path to each sample name. You can choose
which directories are added with the <code>-dd</code>/<code>--dirs-depth</code> parameter. Set to a positive
integer to use that many directories at the end of the path. A negative integer takes
directories from the start of the path.</p>
<p>For example:</p>
<pre><code class="language-console">$ multiqc -d .
# analysis_1 | results | type | sample_1 | file.log
# analysis_2 | results | type | sample_2 | file.log
# analysis_3 | results | type | sample_3 | file.log

$ multiqc -d -dd 1 .
# sample_1 | file.log
# sample_2 | file.log
# sample_3 | file.log

$ multiqc -d -dd -1 .
# analysis_1 | file.log
# analysis_2 | file.log
# analysis_3 | file.log</code></pre>
<h2 id="using-different-templates"><a href="#using-different-templates" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using different templates</h2>
<p>MultiQC is built around a templating system. You can produce reports with
different styling by using the <code>-t</code>/<code>--template</code> option. The available templates
are listed with <code>multiqc --help</code>.</p>
<p>If you're interested in creating your own custom template, see the
<a href="http://multiqc.info/docs/#writing-new-templates">writing new templates</a> section.</p>
<h2 id="pdf-reports"><a href="#pdf-reports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>PDF Reports</h2>
<p>Whilst HTML is definitely the format of choice for MultiQC reports due to
the interactive features that it can offer, PDF files are an integral part
of some people's workflows. To try to accommodate this, MultiQC has a
<code>--pdf</code> command line flag which will try to create a PDF report for you.</p>
<blockquote>
<p>PDF export support for MultiQC can be difficult to use and disables
many core MultiQC features and even some plots. It should only be used
as a last resort.</p>
</blockquote>
<p>To generate PDFs, MultiQC uses the <code>simple</code> template. This uses flat plots,
has no navigation or toolbar and strips out all JavaScript. The resulting
HTML report is pretty basic, but this simplicity is helpful when generating
PDFs.</p>
<p>Once the report is generated MultiQC attempts to call <a href="http://pandoc.org/">Pandoc</a>,
a command line tool able to convert documents between different file formats.
<strong>You must have Pandoc already installed for this to work</strong>. If you don't have
Pandoc installed, you will get an error message that looks like this:</p>
<pre><code class="language-txt">Error creating PDF - pandoc not found. Is it installed? http://pandoc.org/</code></pre>
<p>Please note that Pandoc is a complex tool and has a number of its own dependencies
for PDF generation. Notably, it uses LaTeX / XeLaTeX which you must also have installed.
Please make sure that you have the latest version of Pandoc and
that it can successfully convert basic HTML files to PDF before reporting
and errors.</p>
<p>Error messages from Pandoc are piped through to the MultiQC log,
for example if the xelatex dependency is not installed you will see the following:</p>
<pre><code class="language-txt">xelatex not found. Please select a different --pdf-engine or install xelatex</code></pre>
<p>Note that not all plots have flat image equivalents, so
some will be missing (at time of writing: FastQC sequence content plot,
beeswarm dot plots, heatmaps).</p>
<h2 id="printing-to-stdout"><a href="#printing-to-stdout" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Printing to stdout</h2>
<p>If you would like to generate MultiQC reports on the fly, you can print the
output to standard out by specifying <code>-n stdout</code>. Note that the data directory
will not be generated and the template used must create stand-alone HTML reports.</p>
<h2 id="parsed-data-directory"><a href="#parsed-data-directory" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Parsed data directory</h2>
<p>By default, MultiQC creates a directory alongside the report containing
tab-delimited files with the parsed data. This is useful for downstream
processing, especially if you're running MultiQC with very large numbers
of samples.</p>
<p>Typically, these files are tab-delimited tables. However, you can get <code>JSON</code>
or <code>YAML</code> output for easier downstream parsing by specifying <code>-k</code>/<code>--data-format</code>
on the command line or <code>data_format</code> in your configuration file.</p>
<p>You can also choose whether to produce the data by specifying either the
<code>--data-dir</code> or <code>--no-data-dir</code> command line flags or the <code>make_data_dir</code>
variable in your configuration file. Note that the data directory
is never produced when printing the MultiQC report to <code>stdout</code>.</p>
<p>To zip the data directory, use the <code>-z</code>/<code>--zip-data-dir</code> flag.</p>
<h2 id="exporting-plots"><a href="#exporting-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Exporting Plots</h2>
<p>In addition to the HTML report, it's also possible to get MultiQC to save
plots as stand alone files. You can do this with the <code>-p</code>/<code>--export</code> command
line flag. By default, plots will be saved in a directory called <code>multiqc_plots</code>
as <code>.png</code>, <code>.svg</code> and <code>.pdf</code> files. Raw data for the plots are also saved to files.</p>
<p>You can instruct MultiQC to always do this by setting the <code>export_plots</code> config
option to <code>true</code>, though note that this will add a few seconds on to execution time.
The <code>plots_dir_name</code> changes the default directory name for plots and the
<code>export_plot_formats</code> specifies what file formats should be created (must be
supported by MatPlotLib).</p>
<p>Note that not all plot types are yet supported, so you may find some plots are
missing.</p>
<blockquote>
<p>Note: You can always save static image versions of plots from within
MultiQC reports, using the <a href="http://multiqc.info/docs/#export">Export toolbox</a> in the side bar.</p>
</blockquote>
<h2 id="choosing-which-modules-to-run"><a href="#choosing-which-modules-to-run" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Choosing which modules to run</h2>
<p>Sometimes, it's desirable to choose which MultiQC modules run. This could be
because you're only interested in one type of output and want to keep the
reports small. Or perhaps the output from one module is misleading in your
situation.</p>
<p>You can do this by using <code>-m</code>/<code>--modules</code> to explicitly define which modules
you want to run. Alternatively, use <code>-e</code>/<code>--exclude</code> to run all modules
except those listed.</p>
<p>You can get a group of modules by using <code>--tag</code> followed by a tag e.g. RNA or DNA.</p></div><div class="docs_block" id="reports.md"><h1 id="using-multiqc-reports"><a href="#using-multiqc-reports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using MultiQC Reports</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#browser-compatibility">Browser compatibility</a></li>
<li><a href="#report-layout">Report layout</a></li>
<li><a href="#general-statistics-table">General Statistics table</a></li>
<li><a href="#plots">Plots</a>
<ul class="nav nav-stacked">
<li><a href="#interactive-plots">Interactive plots</a></li>
<li><a href="#flat-plots">Flat plots</a></li>
<li><a href="#exporting-plots-1">Exporting plots</a></li>
<li><a href="#dynamic-plots">Dynamic plots</a></li>
</ul>
</li>
<li><a href="#toolbox">Toolbox</a>
<ul class="nav nav-stacked">
<li><a href="#highlight-samples">Highlight Samples</a></li>
<li><a href="#renaming-samples">Renaming Samples</a></li>
<li><a href="#hiding-samples">Hiding Samples</a></li>
<li><a href="#export">Export</a></li>
<li><a href="#save-settings">Save Settings</a></li>
</ul>
</li>

</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Once MultiQC has finished, you should have a HTML report file called
<code>multiqc_report.html</code> (or something similar, depending on how you ran MultiQC).
You can launch this report with <code>open multiqc_report.html</code> on the command
line, or double clicking the file in a file browser.</p>
<h2 id="browser-compatibility"><a href="#browser-compatibility" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Browser compatibility</h2>
<p>MultiQC reports should work in any modern browser. They have been tested
using OSX Chrome, Firefox and Safari. If you find any report bugs, please
report them as a <a href="https://github.com/ewels/MultiQC/issues">GitHub issue</a>.</p>
<h2 id="report-layout"><a href="#report-layout" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Report layout</h2>
<p>MultiQC reports have three main page sections:</p>
<ul>
<li>The navigation menu (left side)
<ul>
<li>Links to the different module sections in the report</li>
<li>Click the logo to go to the top of the page</li>
</ul></li>
<li>The toolbox (right side)
<ul>
<li>Contains various tools to modify the report data (see below)</li>
</ul></li>
<li>The report (middle)
<ul>
<li>This is what you came here for, the data!</li>
</ul></li>
</ul>
<p>Note that if you're viewing the report on a mobile device / small window,
the content will be reformatted to fit the screen.</p>
<h2 id="general-statistics-table"><a href="#general-statistics-table" class="header-link"><span class="glyphicon glyphicon-link"></span></a>General Statistics table</h2>
<p>At the top of every MultiQC report is the 'General Statistics' table.
This shows an overview of key values, taken from all modules. The aim
of the table is to bring together stats for each sample from across the
analysis so that you can see it in one place.</p>
<p>Hovering over column headers will show a longer description, including which
module produced the data. Clicking a header will sort the table by that value.
Clicking it again will change the sort direction. You can shift-click multiple
headers to sort by multiple columns.</p>
<p><img src="MultiQC%20installation_files/genstats_sort.png" alt="sort column"></p>
<p>Above the table there is a button called 'Configure Columns'. Clicking this
will launch a modal window with more detailed information about each column,
plus options to show/hide and change the order of columns.</p>
<p><img src="MultiQC%20installation_files/genstats_config_cols.png" alt="configure columns"></p>
<h2 id="plots"><a href="#plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Plots</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#interactive-plots">Interactive plots</a></li>
<li><a href="#flat-plots">Flat plots</a></li>
<li><a href="#exporting-plots-1">Exporting plots</a></li>
<li><a href="#dynamic-plots">Dynamic plots</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>MultiQC modules can take plot more extensive data in the sections below
the general statistics table.</p>
<h3 id="interactive-plots"><a href="#interactive-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Interactive plots</h3>
<p>Plots in MultiQC reports are usually interactive, using the
<a href="http://www.highcharts.com/">HighCharts</a> JavaScript library.</p>
<p>You can hover the mouse over data to see a tooltip with more information
about that dataset. Clicking and dragging on line graphs will zoom into that area.</p>
<p><img src="MultiQC%20installation_files/plot_zoom.png" alt="plot zoom"></p>
<p>To reset the zoom, use the button in the top right:</p>
<p><img src="MultiQC%20installation_files/plot_reset_zoom.png" alt="reset zoom"></p>
<p>Plots have a grey bar along their base; clicking and dragging this will
resize the plot's height:</p>
<p><img src="MultiQC%20installation_files/plot_height.png" alt="plot zoom"></p>
<p>You can force reports to use interactive plots instead of flat by specifying
the <code>--interactive</code> command line option (see below).</p>
<h3 id="flat-plots"><a href="#flat-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Flat plots</h3>
<p>Reports with large numbers of samples may contain flat plots. These are
rendered when the MultiQC report is generated using
<a href="http://matplotlib.org/">MatPlotLib</a> and are non-interactive (flat)
images within the report. The reason for generating these is that large
sample numbers can make MultiQC reports very data-intensive and unresponsive
(crashing people's browsers in extreme cases). Plotting data in flat images
is scalable to any number of samples, however.</p>
<p>Flat plots in MultiQC have been designed to look as similar to their interactive
versions as possible. They are also copied to <code>multiqc_data/multiqc_plots</code></p>
<p>You can force reports to use flat plots with the <code>--flat</code> command line option.</p>
<p>See the <a href="http://multiqc.info/docs/#large-sample-numbers">Large sample numbers</a> section of the
<em>Configuring MultiQC</em> docs for more on how to customise the flat / interactive
plot behaviour.</p>
<h3 id="exporting-plots-1"><a href="#exporting-plots-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Exporting plots</h3>
<p>If you want to use the plot elsewhere (<em>eg.</em> in a presentation or paper),
you can export it in a range of formats. Just click the menu button in
the top right of the plot:</p>
<p><img src="MultiQC%20installation_files/plot_export.png" alt="plot zoom"></p>
<p>This opens the MultiQC Toolbox <em>Export Plots</em> panel with the current plot
selected. You have a range of export options here. When deciding on output
format bear in mind that SVG is a vector format, so can be edited in tools
such as <a href="http://www.adobe.com/products/illustrator.html">Adobe Illustrator</a>
or the free tool <a href="https://inkscape.org/en/">Inkscape</a>. This makes it ideal
for use in publications and manual customisation / annotation. The
<em>Plot scaling</em> option changes how large the labels are relative to the plot.</p>
<h3 id="dynamic-plots"><a href="#dynamic-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Dynamic plots</h3>
<p>Some plots have buttons above them which allow you to change the data
that they show or their axis. For example, many bar plots have the option
to show the data as percentages instead of counts:</p>
<p><img src="MultiQC%20installation_files/plot_percentage_button.png" alt="percentage button"></p>
<h2 id="toolbox"><a href="#toolbox" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Toolbox</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#highlight-samples">Highlight Samples</a></li>
<li><a href="#renaming-samples">Renaming Samples</a></li>
<li><a href="#hiding-samples">Hiding Samples</a></li>
<li><a href="#export">Export</a></li>
<li><a href="#save-settings">Save Settings</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>MultiQC reports come with a 'toolbox', accessible by clicking the buttons
on the right hand side of the report:</p>
<p><img src="MultiQC%20installation_files/toolbox_buttons.png" alt="toolbox buttons"></p>
<p>Active toolbox panels have their button highlighted with a blue outline.
You can hide the toolbox by clicking the open panel button a second time,
or pressing <code>Escape</code> on your keyboard.</p>
<h3 id="highlight-samples"><a href="#highlight-samples" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Highlight Samples</h3>
<p>If you run MultiQC plots with a lot of samples, plots can become very
data-heavy. This makes it difficult to find specific samples, or subsets
of samples.</p>
<p>To help with this, you can use the Highlight Samples tool to colour datasets
of interest. Simply enter some text which will match the samples you want to
highlight and press enter (or click the add button). If you like, you can also
customise the highlight colour.</p>
<p><img src="MultiQC%20installation_files/toolbox_highlight.png" alt="toolbox highlight"></p>
<p>To make it easier to match groups of samples, you can use a regular expressions
by turning on 'Regex mode'. You can test regexes using a nice tool at
<a href="https://regex101.com/">regex101.com</a>. See a nice introduction to regexes
<a href="http://www.regular-expressions.info/quickstart.html">here</a>. Note that pattern
delimiters are not needed (use <code>pattern</code>, not <code>/pattern/</code>).</p>
<p>Here, we highlight any sample names that end in <code>_1</code>:</p>
<p><img src="MultiQC%20installation_files/toolbox_highlight_regex.png" alt="highligh regex"></p>
<p>Note that a new button appears above the General Statistics table when samples
are highlighted, allowing you to sort the table according to highlights.</p>
<p>Search patterns can be changed after creation, just click to edit. To remove,
click the grey cross on the right hand side.</p>
<p>Searching for an empty string will match all samples.</p>
<h3 id="renaming-samples"><a href="#renaming-samples" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Renaming Samples</h3>
<p>Sample names are typically generated based on processed file names. These
file names are not always informative. To help with this, you can do a search
and replace within sample names. Here, we remove the <code>SRR1067</code> and <code>_1</code> parts
of the sample names, which are the same for all samples:</p>
<p><img src="MultiQC%20installation_files/toolbox_rename.png" alt="rename samples"></p>
<p>Again, regular expressions can be used. See above for details. Note that
regex groups can be used - define a group match with parentheses and
use the matching value with <code>$1</code>, <code>$2</code> etc. For example - a search string
<code>SRR283(\d{3})</code> and replace string <code>$1_SRR283</code> would move the final three
digits of matching sample names to the start of the name.</p>
<p>Often, you may have a spreadsheet with filenames and informative sample
names. To avoid having to manually enter each name, you can paste from a
spreadsheet using the 'bulk import' tool:</p>
<p><img src="MultiQC%20installation_files/toolbox_bulk_rename.png" alt="bulk rename"></p>
<h3 id="hiding-samples"><a href="#hiding-samples" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Hiding Samples</h3>
<p>Sometimes, you want to focus on a subset of samples. To temporarily
hide samples from the report, enter a search string as described above into
the 'Hide Samples' toolbox panel.</p>
<p>Here, we hide all samples with <code>_trimmed</code> in their sample name:
<em>(Note that plots will tell you how many samples have been hidden)</em></p>
<p><img src="MultiQC%20installation_files/toolbox_hide_samples.png" alt="hide samples"></p>
<h3 id="export"><a href="#export" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Export</h3>
<p>This panel allows you to download MultiQC plots as images or as raw data.
You can configure the size and characteristics of exported plot images:
Width and Height set the output size of the images, scale sets
how "zoomed-in" they should look (typically you want the plot to be more
zoomed for printing). The tick boxes below these settings allow you to
download multiple plots in one go.</p>
<p>Plots with multiple tabs will all be exported as files when using the
Data tab. For plots with multiple tags, the currently visible plot
will be exported.</p>
<blockquote>
<p>Note: You can also save static plot images when you run MultiQC.
See <a href="http://multiqc.info/docs/#exporting-plots">Exporting Plots</a>
for more information.</p>
</blockquote>
<h3 id="save-settings"><a href="#save-settings" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Save Settings</h3>
<p>To avoid having to re-enter the same toolbox setup repeatedly, you can
save your settings using the 'Save Settings' panel. Just pick a name
and click save. To load, choose your set of settings and press load
(or delete). Loaded settings are applied on top of current settings.
All configs are saved in browser local storage - they do not travel
with the report and may not work in older browsers.</p></div><div class="docs_block" id="config.md"><h1 id="configuring-multiqc"><a href="#configuring-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Configuring MultiQC</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#sample-name-cleaning">Sample name cleaning</a>
<ul class="nav nav-stacked">
<li><a href="#other-search-types">Other search types</a></li>
<li><a href="#clashing-sample-names">Clashing sample names</a></li>
</ul>
</li>
<li><a href="#module-search-patterns">Module search patterns</a></li>
<li><a href="#using-log-filenames-as-sample-names">Using log filenames as sample names</a></li>
<li><a href="#ignoring-files">Ignoring Files</a></li>
<li><a href="#ignoring-samples">Ignoring samples</a></li>
<li><a href="#large-sample-numbers">Large sample numbers</a>
<ul class="nav nav-stacked">
<li><a href="#disabling-on-load-plotting">Disabling on-load plotting</a></li>
<li><a href="#flat--interactive-plots">Flat / interactive plots</a></li>
<li><a href="#tables--beeswarm-plots">Tables / Beeswarm plots</a></li>
</ul>
</li>
<li><a href="#coloured-log-output">Coloured log output</a></li>
<li><a href="#checks-for-new-versions">Checks for new versions</a></li>
<li><a href="#command-line-config">Command-line config</a></li>
<li><a href="#optimising-run-time">Optimising run-time</a>
<ul class="nav nav-stacked">
<li><a href="#profile-your-multiqc-run-time">Profile your MultiQC run time</a></li>
<li><a href="#be-picky-with-which-modules-are-run">Be picky with which modules are run</a></li>
<li><a href="#optimise-file-search-patterns">Optimise file search patterns</a></li>
<li><a href="#force-interactive-plots">Force interactive plots</a></li>
<li><a href="#skip-the-report-if-you-dont-need-it">Skip the report if you don't need it</a></li>
</ul>
</li>
<li><a href="#custom-css-files">Custom CSS files</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Whilst most MultiQC settings can be specified on the command line, MultiQC
is also able to parse system-wide and personal config files. At run time,
it collects the configuration settings from the following places in this order
(overwriting at each step if a conflicting config variable is found):</p>
<ol>
<li>Hardcoded defaults in MultiQC code</li>
<li>System-wide config in <code>&lt;installation_dir&gt;/multiqc_config.yaml</code>
<ul>
<li>Manual installations only, not <code>pip</code> or <code>conda</code></li>
</ul></li>
<li>User config in <code>~/.multiqc_config.yaml</code></li>
<li>File path set in environment variable <code>MULTIQC_CONFIG_PATH</code>
<ul>
<li>For example, define this in your <code>~/.bashrc</code> file and keep the file anywhere you like</li>
</ul></li>
<li>Config file in the current working directory: <code>multiqc_config.yaml</code></li>
<li>Config file paths specified in the command with <code>--config</code> / <code>-c</code>
<ul>
<li>You can specify multiple files like this, they can have any filename.</li>
</ul></li>
<li>Command line config (<code>--cl_config</code>)</li>
<li>Specific command line options (<em>e.g.</em> <code>--force</code>)</li>
</ol>
<h2 id="sample-name-cleaning"><a href="#sample-name-cleaning" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Sample name cleaning</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#other-search-types">Other search types</a></li>
<li><a href="#clashing-sample-names">Clashing sample names</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>MultiQC typically generates sample names by taking the input or log file name,
and 'cleaning' it. To do this, it uses the <code>fn_clean_exts</code> settings and looks
for any matches. If it finds any matches, everything to the right is removed.
For example, consider the following config:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">fn_clean_exts:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">".gz"</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">".fastq"</span></code></pre>
<p>This would make the following sample names:</p>
<pre><code class="language-txt">mysample.fastq.gz  -&gt;  mysample
secondsample.fastq.gz_trimming_log.txt  -&gt;  secondsample
thirdsample.fastq_aligned.sam.gz  -&gt;  thirdsample</code></pre>
<p>There is also a config list called <code>fn_clean_trim</code> which just removes
strings if they are present at the start or end of the sample name.</p>
<p>Usually you don't want to overwrite the defaults (though you can).
Instead, add to the special variable names <code>extra_fn_clean_exts</code>
and <code>extra_fn_clean_trim</code>:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">extra_fn_clean_exts:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">".myformat"</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">"_processedFile"</span>
<span class="hljs-attr">extra_fn_clean_trim:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">"#"</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">".myext"</span></code></pre>
<h3 id="other-search-types"><a href="#other-search-types" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Other search types</h3>
<p>File name cleaning can also take strings to remove (instead of removing with truncation).
Also regex strings can be supplied to match patterns and remove or keep matching substrings.</p>
<h4><code>truncate</code> (default)</h4>
<p>If you just supply a string, the default behavior is similar to 
"trim". The filename will be truncated beginning with the matching 
string.</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">extra_fn_clean_exts:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">".fastq"</span></code></pre>
<p>The above is equivalent to the more explicit:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">extra_fn_clean_exts:</span>
<span class="hljs-attr">  - type:</span> <span class="hljs-string">"truncate"</span>
<span class="hljs-attr">    pattern:</span> <span class="hljs-string">".fastq"</span></code></pre>
<p>This rule would produce the following sample names:</p>
<pre><code class="language-txt">mysample.fastq.gz  -&gt;  mysample
thirdsample.fastq_aligned.sam.gz  -&gt;  thirdsample</code></pre>
<h4><code>remove</code> (formerly <code>replace</code>)</h4>
<p>The <code>remove</code> type allows you to remove the exact match from the filename.</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">extra_fn_clean_exts:</span>
<span class="hljs-attr">  - type:</span> <span class="hljs-string">remove</span>
<span class="hljs-attr">    pattern:</span> <span class="hljs-string">.sorted</span></code></pre>
<p>This rule would produce the following sample names:</p>
<pre><code class="language-txt">secondsample.sorted.deduplicated  -&gt;  secondsample.deduplicated</code></pre>
<h4><code>regex</code></h4>
<p>You can also remove a substring with a regular expression. Here's a <a href="https://regex101.com/">good resource</a> to interactively try it out.</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">extra_fn_clean_exts:</span>
<span class="hljs-attr">  - type:</span> <span class="hljs-string">regex</span>
<span class="hljs-attr">    pattern:</span> <span class="hljs-string">"^processed."</span></code></pre>
<p>This rule would produce the following sample names:</p>
<pre><code class="language-txt">processed.thirdsample.processed  -&gt;  thirdsample.processed</code></pre>
<h4><code>regex_keep</code></h4>
<p>If you'd rather like to <em>keep</em> the match of a regular expression you can use the <code>regex_keep</code> type. This simplifies things if you can e.g. directly target samples names.</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">extra_fn_clean_exts:</span>
<span class="hljs-attr">  - type:</span> <span class="hljs-string">regex_keep</span>
<span class="hljs-attr">    pattern:</span> <span class="hljs-string">"[A-Z]{3}[1-9]{2}"</span></code></pre>
<p>This rule would produce the following sample names:</p>
<pre><code class="language-txt">merged.recalibrated.XZY97.alignment.bam  -&gt;  XZY97</code></pre>
<h4><code>module</code></h4>
<p>This key will tell MultiQC to only apply the pattern to a specific MultiQC module.
This should be a string that matches the module's <code>anchor</code> - the <code>#module</code> bit when you click the main module heading in the sidebar (remove the <code>#</code>).</p>
<p>For example, to truncate all sample names to 5 characters for just Kallisto:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">extra_fn_clean_exts:</span>
<span class="hljs-attr">  - type:</span> <span class="hljs-string">regex_keep</span>
<span class="hljs-attr">    pattern:</span> <span class="hljs-string">"^.{5}"</span>
<span class="hljs-attr">    module:</span> <span class="hljs-string">kallisto</span></code></pre>
<p>You can also supply a list of multiple module anchors if you wish:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">extra_fn_clean_exts:</span>
<span class="hljs-attr">  - type:</span> <span class="hljs-string">regex_keep</span>
<span class="hljs-attr">    pattern:</span> <span class="hljs-string">"^.{5}"</span>
<span class="hljs-attr">    module:</span>
<span class="hljs-bullet">      -</span> <span class="hljs-string">kallisto</span>
<span class="hljs-bullet">      -</span> <span class="hljs-string">cutadapt</span></code></pre>
<h3 id="clashing-sample-names"><a href="#clashing-sample-names" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Clashing sample names</h3>
<p>This process of cleaning sample names can sometimes result in exact duplicates.
A duplicate sample name will overwrite previous results. Warnings showing these events
can be seen with verbose logging using the <code>--verbose</code>/<code>-v</code> flag, or in <code>multiqc_data/multiqc.log</code>.</p>
<p>Problems caused by this will typically be discovered be fewer results than expected. If you're ever
unsure about where the data from results within MultiQC reports come from, have a look at
<code>multiqc_data/multiqc_sources.txt</code>, which lists the path to the file used for every section
of the report.</p>
<h4 id="directory-names"><a href="#directory-names" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Directory names</h4>
<p>One scenario where clashing names can occur is when the same file is processed in different directories.
For example, if <code>sample_1.fastq</code> is processed with four sets of parameters in four different
directories, they will all have the same name - <code>sample_1</code>. Only the last will be shown.
If the directories are different, this can be avoided with the <code>--dirs</code>/<code>-d</code> flag.</p>
<p>For example, given the following files:</p>
<pre><code class="language-txt"> analysis_1
    sample_1.fastq.gz.aligned.log
 analysis_2
    sample_1.fastq.gz.aligned.log
 analysis_3
     sample_1.fastq.gz.aligned.log</code></pre>
<p>Running <code>multiqc -d .</code> will give the following sample names:</p>
<pre><code class="language-txt">analysis_1 | sample_1
analysis_2 | sample_1
analysis_3 | sample_1</code></pre>
<h4 id="filename-truncation"><a href="#filename-truncation" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Filename truncation</h4>
<p>If the problem is with filename truncation, you can also use the <code>--fullnames</code>/<code>-s</code> flag,
which disables all sample name cleaning. For example:</p>
<pre><code class="language-txt"> sample_1.fastq.gz.aligned.log
 sample_1.fastq.gz.subsampled.fastq.gz.aligned.log</code></pre>
<p>Running <code>multiqc -s .</code> will give the following sample names:</p>
<pre><code class="language-txt">sample_1.fastq.gz.aligned.log
sample_1.fastq.gz.subsampled.fastq.gz.aligned.log</code></pre>
<p>You can turn off sample name cleaning permanently by setting
<code>fn_clean_sample_names</code> to <code>false</code> in your config file.</p>
<h2 id="module-search-patterns"><a href="#module-search-patterns" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Module search patterns</h2>
<p>Many bioinformatics tools have standard output formats, filenames and other
signatures. MultiQC uses these to find output; for example, the FastQC module
looks for files that end in <code>_fastqc.zip</code>.</p>
<p>This works well most of the time, until someone has an automated processing
pipeline that renames things. For this reason, as of version v0.3.2 of MultiQC,
the file search patterns are loaded as part of the main config. This means that
they can be overwritten in <code>&lt;installation_dir&gt;/multiqc_config.yaml</code> or
<code>~/.multiqc_config.yaml</code>. So if you always rename your <code>_fastqc.zip</code> files to
<code>_qccheck.zip</code>, MultiQC can still work.</p>
<p>To see the default search patterns, check a given module in the MultiQC documentation.
Each module has its search patterns listed beneath any free-text docs.
Alternatively, see the <a href="https://github.com/ewels/MultiQC/blob/master/multiqc/utils/search_patterns.yaml"><code>search_patterns.yaml</code></a>
file in the MultiQC source code. Copy the section for the program that you want to modify and paste this
into your config file. Make sure you make it part of a dictionary called <code>sp</code>
as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sp:</span>
<span class="hljs-attr">  mqc_module:</span>
<span class="hljs-attr">    fn:</span> <span class="hljs-string">_mysearch.txt</span></code></pre>
<p>Search patterns can specify a filename match (<code>fn</code>) or a file contents
match (<code>contents</code>), as well as a number of additional search keys.
See <a href="#step-1-find-log-files">below</a> for the full reference.</p>
<h2 id="using-log-filenames-as-sample-names"><a href="#using-log-filenames-as-sample-names" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using log filenames as sample names</h2>
<p>A substantial number of MultiQC modules take the sample name identifiers that you
see in the report from the file contents - typically the filename of the input file.
This is because log files can often be called things like <code>mytool.log</code> or even concatenated.
Using the input filename used by the tool is typically safer and more consistent across modules.</p>
<p>However, sometimes this does not work well. For example, if the input filename is not
relevant (eg. using a temporary file or FIFO, process substitution or stdin etc.).
In these cases your log files may have useful filenames but MultiQC will not be using them.</p>
<p>To force MultiQC to use the log filename as the sample identifier, you can use the
<code>--fn_as_s_name</code> command line flag or set the <code>use_filename_as_sample_name</code>:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">use_filename_as_sample_name:</span> <span class="hljs-literal">true</span></code></pre>
<p>This affects all modules and all search patterns. If you want to limit this to just
one or more specific search patterns, you can do by giving a list:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">use_filename_as_sample_name:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">cutadapt</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">picard/gcbias</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">picard/markdups</span></code></pre>
<p>Note that this should be the search pattern key (see above) and not just the module name.
This is because some modules search for multiple files.</p>
<p>The log filename will still be cleaned. To use the raw log filenames,
combine with the <code>--fullnames</code>/<code>-s</code> flag or <code>fn_clean_sample_names</code> config option described above.</p>
<h2 id="ignoring-files"><a href="#ignoring-files" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Ignoring Files</h2>
<p>MultiQC begins by indexing all of the files that you specified and building a list
of the ones it will use. You can specify files and directories to skip on the command
line using <code>-x</code>/<code>--ignore</code>, or for more permanent memory, with the following config file
options: <code>fn_ignore_files</code>, <code>fn_ignore_dirs</code> and <code>fn_ignore_paths</code> (the command line
option simply adds to all of these).</p>
<p>For example, given the following files:</p>
<pre><code class="language-txt"> analysis_1
    sample_1.fastq.gz.aligned.log
 analysis_2
    sample_1.fastq.gz.aligned.log
 analysis_3
     sample_1.fastq.gz.aligned.log</code></pre>
<p>You could specify the following relevant config options:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">fn_ignore_files:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">"*.log"</span>
<span class="hljs-attr">fn_ignore_dirs:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">"analysis_1"</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">"analysis_2"</span>
<span class="hljs-attr">fn_ignore_paths:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">"*/analysis_*/sample_1*"</span></code></pre>
<p>Note that the searched file paths will usually be relative to the working
directory and can be highly variable, so you'll typically want to start patterns
with a <code>*</code> to match any preceding directory structure.</p>
<h2 id="ignoring-samples"><a href="#ignoring-samples" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Ignoring samples</h2>
<p>Some modules get sample names from the contents of the file and not the filename
(for example, <code>stdout</code> logs can contain multiple samples). You can skip samples
by their resolved sample names (after cleaning) with two config options:
<code>sample_names_ignore</code> and <code>sample_names_ignore_re</code>. The first takes a list of
strings to be used for glob pattern matching (same behaviour as the command line
option <code>--ignore-samples</code>), the latter takes a list of regex patterns. For example:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sample_names_ignore:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">"SRR*"</span>
<span class="hljs-attr">sample_names_ignore_re:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">'^SR{2}\d{7}_1$'</span></code></pre>
<h2 id="large-sample-numbers"><a href="#large-sample-numbers" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Large sample numbers</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#disabling-on-load-plotting">Disabling on-load plotting</a></li>
<li><a href="#flat--interactive-plots">Flat / interactive plots</a></li>
<li><a href="#tables--beeswarm-plots">Tables / Beeswarm plots</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>MultiQC has been written with the intention of being used for any number of samples.
This means that it <em>should</em> work well with 6 samples or 6000. Very large sample numbers
are becoming increasingly common, for example with single cell data.</p>
<p>Producing reports with data from many hundreds or thousands of samples provides some
challenges, both technically and also in terms of data visualisation and report usability.</p>
<h3 id="disabling-on-load-plotting"><a href="#disabling-on-load-plotting" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Disabling on-load plotting</h3>
<p>One problem with large reports is that the browser can hang when the report is first loaded.
This is because it loading and processing the data for all plots at once. To mitigate this,
large reports may show plots as grey boxes with a <em>"Show Plot"</em> button. Clicking this will
render the plot as normal and prevents the browser from trying to do everything at once.</p>
<p>By default this behaviour kicks in when a plot has 50 samples or more. This can be customised
by changing the <code>num_datasets_plot_limit</code> config option.</p>
<h3 id="flat--interactive-plots"><a href="#flat--interactive-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Flat / interactive plots</h3>
<p>Reports with many samples start to need a lot of data for plots. This results in inconvenient
report file sizes (can be 100s of megabytes) and worse, web browser crashes. To allow MultiQC
to scale to these sample numbers, most plot types have two plotting functions in the code base -
interactive (using HighCharts) and flat (rendered with MatPlotLib). Flat plots take up the
same disk space irrespective of sample number and do not consume excessive resources to display.</p>
<p>By default, MultiQC generates flat plots when there are 100 or more samples. This cutoff
can be changed by changing the <code>plots_flat_numseries</code> config option. This behaviour can also
be changed by running MultiQC with the <code>--flat</code> / <code>--interactive</code> command line options or by
setting the <code>plots_force_flat</code> / <code>plots_force_interactive</code> config options to <code>True</code>.</p>
<h3 id="tables--beeswarm-plots"><a href="#tables--beeswarm-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Tables / Beeswarm plots</h3>
<p>Report tables with thousands of samples (table rows) can quickly become impossible to use.
To avoid this, tables with large numbers of rows are instead plotted as a Beeswarm plot
(aka. a strip chart / jitter plot). These plots have fixed dimensions with any number
of samples. Hovering on a dot will highlight the same sample in other rows.</p>
<p>By default, MultiQC starts using beeswarm plots when a table has 500 rows or more. This
can be changed by setting the <code>max_table_rows</code> config option.</p>
<h2 id="coloured-log-output"><a href="#coloured-log-output" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Coloured log output</h2>
<p>As of MultiQC version 1.8, log output is coloured using the <a href="https://pypi.org/project/coloredlogs/">coloredlogs</a>
Python package. The code attempts to detect if the logs on the terminal are being redirected to a file
or piped to another tool and will disable colours if so. If the colours annoy you or you're ending
up with weird characters in your MultiQC output, you can disable this feature with the command line
flag <code>--no-ansi</code>. Sadly it's not possible to set this in a config file, as the logger is initilised
before configs are loaded.</p>
<h2 id="checks-for-new-versions"><a href="#checks-for-new-versions" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Checks for new versions</h2>
<p>When MultiQC runs it automatically checks to see if there is a new version available to download.
If so, a log message is printed at the top of the run saying where to download it
(<em>MultiQC Version v0.6 now available!</em>).
This helps people stay up to date and reduces the number of bug reports that are
due to outdated MultiQC versions.</p>
<p>The timeout for the version check is set to 5 seconds, so if you're running offline it should
fail silently and add negligable run time.
However, if you prefer you can explicitly disable the version check by adding
<code>no_version_check: true</code> to your MultiQC config.</p>
<p>The check is done with the main MultiQC website (<a href="https://multiqc.info/">https://multiqc.info</a>).
The only statistics that are collected are the number of checks and the version of MultiQC
that is being run. No identifiable information (eg. IP address etc) is stored.</p>
<h2 id="command-line-config"><a href="#command-line-config" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Command-line config</h2>
<p>Sometimes it's useful to specify a single small config option just once, where creating
a config file for the occasion may be overkill. In these cases you can use the
<code>--cl_config</code> option to supply additional config values on the command line.</p>
<p>Config variables should be given as a YAML string. You will usually need to enclose
this in quotes. If MultiQC is unable to understand your config you will get an error message
saying <code>Could not parse command line config</code>.</p>
<p>As an example, the following command configures the coverage levels to use for the
Qualimap module: <em>(as <a href="http://multiqc.info/docs/#qualimap">described in the docs</a>)</em></p>
<pre><code class="language-bash hljs">multiqc ./datadir --cl_config <span class="hljs-string">"qualimap_config: { general_stats_coverage: [20,40,200] }"</span></code></pre>
<h2 id="optimising-run-time"><a href="#optimising-run-time" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Optimising run-time</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#profile-your-multiqc-run-time">Profile your MultiQC run time</a></li>
<li><a href="#be-picky-with-which-modules-are-run">Be picky with which modules are run</a></li>
<li><a href="#optimise-file-search-patterns">Optimise file search patterns</a></li>
<li><a href="#force-interactive-plots">Force interactive plots</a></li>
<li><a href="#skip-the-report-if-you-dont-need-it">Skip the report if you don't need it</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Usually, MultiQC run time is fairly insignificant - in the order of seconds.
Unless you are running MultiQC on many thousands of analysis files, the optimisations
described below will have limited practical benefit.</p>
<p>In other words, if you're running with 15 RNAseq samples, you may as well save yourself
some time and stick with the defaults.</p>
<h3 id="profile-your-multiqc-run-time"><a href="#profile-your-multiqc-run-time" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Profile your MultiQC run time</h3>
<p>As of version 1.9, MultiQC has a command line option to profile what it spends its time
doing: <code>--profile-runtime</code> (<code>config.profile_runtime</code>). Whilst you're working with writing
your pipeline / setting up your analysis, you can specify and MultiQC will add a section
to the bottom of your report describing how much time it spent searching files and
what it did with those files. You'll also get a breakdown in the command-line log
of how long the different steps of MultiQC execution took:</p>
<pre><code class="language-txt">[INFO   ]         multiqc : MultiQC complete
[INFO   ]         multiqc : Run took 35.28 seconds
[INFO   ]         multiqc :  - 31.01s: Searching files
[INFO   ]         multiqc :  - 1.75s: Running modules
[INFO   ]         multiqc :  - 0.96s: Compressing report data
[INFO   ]         multiqc : For more information, see the 'Run Time' section in multiqc_report.html</code></pre>
<p>If MultiQC is finishing in a few seconds or minutes, you probably don't need to do anything.
If you are working with huge numbers of files then it may be worth looking into these
results to see if you can speed up MultiQC. The documentation below explains how to do this.</p>
<h3 id="be-picky-with-which-modules-are-run"><a href="#be-picky-with-which-modules-are-run" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Be picky with which modules are run</h3>
<p>Probably the easiest way to speed up MultiQC is to only use the modules that you
know you have files for. MultiQC supports a <em>lot</em> of different tools and searches
for matching files for all of them every time you run it.</p>
<p>You can do this with the <code>-m</code> / <code>--module</code> flag (can be repeated) or in a MultiQC
config file by using <code>config.module_order</code>. See <a href="#order-of-modules">Order of modules</a>.</p>
<h3 id="optimise-file-search-patterns"><a href="#optimise-file-search-patterns" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Optimise file search patterns</h3>
<p>Secondly, think about customising the search patterns of the slowest searches.</p>
<p>As an example, logs from Picard are published to <code>STDOUT</code> and so can have any file name.
Some people concatenate logs, so the contents can be anywhere in the file and the files
must also be searched by subsequent tools in case they contain multiple outputs.
If you know that all of your Picard MarkDuplicate log files have the filename
<code>mysamplename_markduplicates.log</code> then you can safely customise that search pattern
with the following MultiQC config:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sp:</span>
  <span class="hljs-string">picard/markdups:</span>
<span class="hljs-attr">    fn:</span> <span class="hljs-string">"*_markduplicates.log"</span></code></pre>
<p>If you know that this is the only type of Picard output that you're interested in,
you can also change all of the other Picard search patterns to use <code>skip: True</code>:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sp:</span>
  <span class="hljs-string">picard/markdups:</span>
<span class="hljs-attr">    fn:</span> <span class="hljs-string">"*_markduplicates.log"</span>
  <span class="hljs-string">picard/alignment_metrics:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/basedistributionbycycle:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/gcbias:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/hsmetrics:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/insertsize:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/oxogmetrics:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/pcr_metrics:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/quality_by_cycle:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/quality_score_distribution:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/quality_yield_metrics:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/rnaseqmetrics:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/rrbs_metrics:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/sam_file_validation:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/variant_calling_metrics:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span>
  <span class="hljs-string">picard/wgs_metrics:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span></code></pre>
<p>This can speed up execution a bit if you really want to squeeze that running time.
The <a href="#multiqc-modules">MultiQC Modules documentation</a> shows the search patterns for every module.</p>
<blockquote>
<p>Note that it's only worth using <code>skip: true</code> on search patterns if you want to use one from a module that has several.
Usually it's better to just <a href="#be-picky-with-which-modules-are-run">specify which modules you want to run</a> instead.</p>
</blockquote>
<h3 id="force-interactive-plots"><a href="#force-interactive-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Force interactive plots</h3>
<p>One step that can take some time is running MatPlotLib to generate static-image plots
(see <a href="#flat--interactive-plots">Flat / interactive plots</a>).
You can force MultiQC to skip this and only use interactive plots by using the <code>--interactive</code>
command line option (<code>config.plots_force_interactive</code>).</p>
<p>This approach is <strong>not recommended if you have a very large number of samples</strong>, as this can
produce a huge report file with all of the embedded plot data and crash your browser when opening it.
If you are running MultiQC for the <code>multiqc_data</code> folder and never intend to look at the report, it
speed things up though.</p>
<h3 id="skip-the-report-if-you-dont-need-it"><a href="#skip-the-report-if-you-dont-need-it" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Skip the report if you don't need it</h3>
<p>If you're running MultiQC just to get parsed data / exported plots (<code>multiqc_data</code>) or the output for MegaQC
and don't actually need the report, you can skip it with <code>--no-report</code>.
This prevents any HTML report from being generated, including the data compression step that precedes it.
This can cut a few seconds off the MultiQC execution time.</p>
<h2 id="custom-css-files"><a href="#custom-css-files" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Custom CSS files</h2>
<p>MultiQC generates HTML reports. You can include custom CSS in your final report if you wish.
Simply add CSS files to the <code>custom_css_files</code> config option:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_css_files:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">myfile.css</span></code></pre>
<p>Or pass <code>--custom-css-file</code> (can be specified multiple times) and MultiQC will include
them in the final report HTML.</p></div><div class="docs_block" id="customisation.md"><h1 id="customising-reports"><a href="#customising-reports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Customising Reports</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#titles-and-introductory-text">Titles and introductory text</a>
<ul class="nav nav-stacked">
<li><a href="#report-time-and-analysis-paths">Report time and analysis paths</a></li>
</ul>
</li>
<li><a href="#report-logo">Report Logo</a></li>
<li><a href="#project-level-information">Project level information</a></li>
<li><a href="#sample-name-replacement">Sample name replacement</a></li>
<li><a href="#bulk-sample-renaming-in-reports">Bulk sample renaming in reports</a></li>
<li><a href="#show--hide-samples-buttons">Show / Hide samples buttons</a></li>
<li><a href="#module-and-section-comments">Module and section comments</a></li>
<li><a href="#removing-modules-or-sections">Removing modules or sections</a></li>
<li><a href="#order-of-modules">Order of modules</a>
<ul class="nav nav-stacked">
<li><a href="#order-of-module-and-module-subsection-output">Order of module and module subsection output</a></li>
</ul>
</li>
<li><a href="#customising-plots">Customising plots</a></li>
<li><a href="#customising-tables">Customising tables</a>
<ul class="nav nav-stacked">
<li><a href="#hiding-columns">Hiding columns</a></li>
<li><a href="#column-order">Column order</a></li>
<li><a href="#column-titles">Column titles</a></li>
<li><a href="#conditional-formatting">Conditional formatting</a></li>
</ul>
</li>
<li><a href="#number-base-multiplier">Number base (multiplier)</a></li>
<li><a href="#number-formatting">Number formatting</a></li>
<li><a href="#troubleshooting">Troubleshooting</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>MultiQC offers a few ways to customise reports to easily add your own
branding and some additional report-level information. These features
are primarily designed for core genomics facilities.</p>
<p>Note that much more extensive customisation of reports is possible using
<a href="http://multiqc.info/docs/#writing-new-templates">custom templates</a>.</p>
<h2 id="titles-and-introductory-text"><a href="#titles-and-introductory-text" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Titles and introductory text</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#report-time-and-analysis-paths">Report time and analysis paths</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>You can specify a custom title for the report using the <code>-i</code>/<code>--title</code>
command line option. The <code>-b</code>/<code>--comment</code> option can be used to add a
longer comment to the top of the report at run time.</p>
<p>You can also specify the title and comment, as well as
a subtitle and the introductory text in your config file:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">title:</span> <span class="hljs-string">"My Title"</span>
<span class="hljs-attr">subtitle:</span> <span class="hljs-string">"A subtitle to go underneath in grey"</span>
<span class="hljs-attr">intro_text:</span> <span class="hljs-string">"MultiQC reports summarise analysis results."</span>
<span class="hljs-attr">report_comment:</span> <span class="hljs-string">"This is a comment about this report."</span></code></pre>
<p>Note that if <code>intro_text</code> is <code>None</code> the template will display the default
introduction sentence. Set this to <code>False</code> to hide this, or set it to a
string to use your own text.</p>
<h3 id="report-time-and-analysis-paths"><a href="#report-time-and-analysis-paths" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Report time and analysis paths</h3>
<p>It's not always appropriate to include the file paths that MultiQC was run with
in a report, for example if sharing reports with others outside your organisation.</p>
<p>If you wish, you can disable the analysis paths and/or time in the report header
with the following config parameters:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">show_analysis_paths:</span> <span class="hljs-literal">False</span>
<span class="hljs-attr">show_analysis_time:</span> <span class="hljs-literal">False</span></code></pre>
<h2 id="report-logo"><a href="#report-logo" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Report Logo</h2>
<p>To add your own custom logo to reports, you can add the following
three lines to your MultiQC configuration file:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_logo:</span> <span class="hljs-string">"/abs/path/to/logo.png"</span>
<span class="hljs-attr">custom_logo_url:</span> <span class="hljs-string">"https://www.example.com"</span>
<span class="hljs-attr">custom_logo_title:</span> <span class="hljs-string">"Our Institute Name"</span></code></pre>
<p>Only <code>custom_logo</code> is needed. The URL will make the logo open up
a new web browser tab with your address and the title sets the mouse
hover title text.</p>
<h2 id="project-level-information"><a href="#project-level-information" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Project level information</h2>
<p>You can add custom information at the top of reports by adding key:value
pairs to the config option <code>report_header_info</code>. Note that if you have
a file called <code>multiqc_config.yaml</code> in the working directory, this will
automatically be parsed and added to the config. For example, if you have
the following saved:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">report_header_info:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">Contact</span> <span class="hljs-attr">E-mail:</span> <span class="hljs-string">"phil.ewels@scilifelab.se"</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">Application</span> <span class="hljs-attr">Type:</span> <span class="hljs-string">"RNA-seq"</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">Project</span> <span class="hljs-attr">Type:</span> <span class="hljs-string">"Application"</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">Sequencing</span> <span class="hljs-attr">Platform:</span> <span class="hljs-string">"HiSeq 2500 High Output V4"</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">Sequencing</span> <span class="hljs-attr">Setup:</span> <span class="hljs-string">"2x125"</span></code></pre>
<p>Then this will be displayed at the top of reports:</p>
<p><img src="MultiQC%20installation_files/report_proj_info.png" alt="report project info"></p>
<p>Note that you can also specify a path to a config file using <code>-c</code>.</p>
<h2 id="sample-name-replacement"><a href="#sample-name-replacement" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Sample name replacement</h2>
<p>Occasionally, when you run MultiQC you may know that you want to change the resulting
sample names at run time. You can do this using the <code>--replace-names</code> option, which
allows you to change sample names during report creation.</p>
<p>Unlike <code>--sample-names</code> below, the original names never make it through to the report.
This can be useful if you know that you have a range of outputs that result in varying
sample names but want to create a consistent report - especially if you want
samples to line up properly in the <em>General Statistics</em> table.</p>
<p>To use, create a tab-separated file with two columns. The first column contains the
search strings and the second the replacement strings:</p>
<pre><code class="language-tsv">IDX102934   Sample_1
IDX102935   Sample_2
IDX102936   Sample_3</code></pre>
<p>Note that by default, <em>partial</em> matches are replaced. So if a log file gives a sample
name of <code>IDX102934_mytool</code> then the result will be <code>Sample_1_mytool</code>.
There are two config options to fine-tune this behaviour:</p>
<p>Setting <code>sample_names_replace_exact</code> to <code>True</code> in a MultiQC config file will tell
MultiQC to only change a sample name if the pattern <em>fully</em> matches the search string.
In the above example, <code>IDX102934_mytool</code> would remain unchanged.</p>
<p>Setting <code>sample_names_replace_complete</code> to <code>True</code>, the replacement string will be used
as a complete replacement if the search pattern matches at all.
In the above example, <code>IDX102934_mytool</code> would become <code>Sample_1</code>.</p>
<blockquote>
<p>NB: Use this method with caution! If aggressive cleaning of sample names results in
multiple samples with identical identifiers, they will be overwritten.</p>
</blockquote>
<p>To have more control over replacements, you can use regular expressions.
If you set <code>sample_names_replace_regex</code> to <code>True</code> in a MultiQC config file
and then create a file that contains regex search strings and even Python regex
group identifiers in the replace string. For example:</p>
<pre><code class="language-tsv">SAMPLE(\d)_([PS]E)_(\d) XXX_\1_\2_\3</code></pre>
<p>With this file, <code>SAMPLE1_PE_2</code> would be renamed to <code>XXX_1_PE_2</code>.
<code>SAMPLE3_SE_4</code> would be renamed to <code>XXX_3_SE_4</code>.</p>
<p>Setting <code>sample_names_replace_exact</code> to <code>True</code> also works for regular expression
searches. The code uses the <code>re.fullmatch</code> function, so no <code>^</code> or <code>$</code> anchors are needed.</p>
<p>Setting <code>sample_names_replace_complete</code> is ignored when using regexes.
If you want this behaviour then configure your regular expression to match the entire string.
For example, <code>*(\d)_([PS]E)_(\d) \1_\2_\3</code> would rename <code>SAMPLE1_PE_2</code> to <code>1_PE_2</code>.</p>
<p>Finally, if you prefer not to use <code>--replace-names</code> with a TSV file, you
can set the search patterns in a MultiQC config file directly.
For example:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sample_names_replace:</span>
<span class="hljs-attr">  IDX102934:</span> <span class="hljs-string">Sample_1</span>
<span class="hljs-attr">  IDX102935:</span> <span class="hljs-string">Sample_2</span>
<span class="hljs-attr">  IDX102936:</span> <span class="hljs-string">Sample_3</span></code></pre>
<p>Remember that backslashes must be escaped in YAML. So if using regular expressions
you will need to use double-backslashes. You may also need to quote strings:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sample_names_replace_regex:</span> <span class="hljs-literal">True</span>
<span class="hljs-attr">sample_names_replace:</span>
  <span class="hljs-string">"SAMPLE(\\d)_([PS]E)_(\\d)"</span><span class="hljs-string">:</span> <span class="hljs-string">"XXX_\\1_\\2_\\3"</span></code></pre>
<h2 id="bulk-sample-renaming-in-reports"><a href="#bulk-sample-renaming-in-reports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bulk sample renaming in reports</h2>
<p>Although it is possible to rename samples manually and in bulk using the
<a href="#renaming-samples">report toolbox</a>, it's often desirable to embed such renaming patterns
into the report so that they can be shared with others. For example, a typical case could be
for a sequencing centre that has internal sample IDs and also user-supplied sample names.
Or public sample identifiers such as SRA numbers as well as more meaningful names.</p>
<p>It's possible to supply a file with one or more sets of sample names using the <code>--sample-names</code>
command line option. This file should be a tab-delimited file with a header row (used for
the report button labels) and then any number of renamed sample identifiers. For example:</p>
<pre><code class="language-tsv">MultiQC Names   Proper Names    AWESOME NAMES
SRR1067503_1    Sample_1    MYBESTSAMP_1
SRR1067505_1    Sample_2    MYBESTSAMP_2
SRR1067510_1    Sample_3    MYBESTSAMP_3</code></pre>
<p>If supplied, buttons will be generated at the top of the report with your labels.
Clicking these will populate and apply the Toolbox renaming panel.</p>
<blockquote>
<p><strong>NB:</strong> Sample renaming works with partial substrings - these will be replaced!</p>
</blockquote>
<p>It's also possible to supply such renaming patterns within a config file (useful if you're
already generating a config file for a run). In this case, you need to set the variables
<code>sample_names_rename_buttons</code> and <code>sample_names_rename</code>. For example:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sample_names_rename_buttons:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">"MultiQC Names"</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">"Proper Names"</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">"AWESOME NAMES"</span>
<span class="hljs-attr">sample_names_rename:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">["SRR1067503_1",</span> <span class="hljs-string">"Sample_1"</span><span class="hljs-string">,</span> <span class="hljs-string">"MYBESTSAMP_1"</span><span class="hljs-string">]</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">["SRR1067505_1",</span> <span class="hljs-string">"Sample_2"</span><span class="hljs-string">,</span> <span class="hljs-string">"MYBESTSAMP_2"</span><span class="hljs-string">]</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">["SRR1067510_1",</span> <span class="hljs-string">"Sample_3"</span><span class="hljs-string">,</span> <span class="hljs-string">"MYBESTSAMP_3"</span><span class="hljs-string">]</span></code></pre>
<h2 id="show--hide-samples-buttons"><a href="#show--hide-samples-buttons" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Show / Hide samples buttons</h2>
<p>It is possible to filter which samples are visible through the <a href="#hiding-samples">report toolbox</a>,
but it can be desirable to embed such patterns into the report so that they can be shared
with others. One example can be to add filters for batches, to easily scan if certain
quality metrics overlap between these batches.</p>
<p>It's possible to supply a file with one or more patterns to filter samples on using the
<code>--sample-filters</code> command line option. This file should be a tab-delimited file with each
row containing the button name, whether the pattern should be hidden (<code>hide</code>) or shown (<code>show</code>)
and the patterns to be applied (all subsequent columns). If you want to use a regular expression
pattern, opposed to the default globbing, you can use <code>hide_re</code> and <code>show_re</code>.</p>
<p>For example, to filter on read pair groups, you could use the following file:</p>
<pre><code class="language-tsv">Read Group 1    show    _R1
Read Group 2    show    _R2</code></pre>
<p>Or in regex mode:</p>
<pre><code class="language-tsv">Read Group 1    show_re .*_R1
Read Group 2    show_re .*_R2</code></pre>
<p>To filter on controls and sample groups you could use:</p>
<pre><code class="language-tsv">Controls    show    input_
Conditions  show    group_1_    group_2_    group_3_</code></pre>
<p>MultiQC automatically adds an <code>Show all</code> button at the start, which reverts back to showing all samples.</p>
<p>If you prefer, you can also add these buttons using a MultiQC config file:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">show_hide_buttons:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">Read</span> <span class="hljs-string">Group</span> <span class="hljs-number">1</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">Read</span> <span class="hljs-string">Group</span> <span class="hljs-number">2</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">Controls</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">Conditions</span>
<span class="hljs-attr">show_hide_mode:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">show</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">show</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">show</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">show</span>
<span class="hljs-attr">show_hide_patterns:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">_R1</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">_R2</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">input_</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">["group_1_",</span> <span class="hljs-string">"group_2_"</span><span class="hljs-string">,</span> <span class="hljs-string">"group_3_"</span><span class="hljs-string">]</span></code></pre>
<h2 id="module-and-section-comments"><a href="#module-and-section-comments" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Module and section comments</h2>
<p>Sometimes you may want to add a custom comment above specific sections in the report. You can
do this with the config option <code>section_comments</code> as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">section_comments:</span>
<span class="hljs-attr">  featurecounts:</span> <span class="hljs-string">"This comment is for a module header, but should still work"</span>
<span class="hljs-attr">  star_alignments:</span> <span class="hljs-string">"This new way of commenting above sections is **awesome**!"</span></code></pre>
<p>Comments can be written in Markdown. The <code>section_comments</code> keys should correspond to the HTML IDs
of the report section. You can find these by clicking on a navigation link in the report and seeing
the <code>#section_id</code> at the end of the browser URL.</p>
<p>To add a section comment for General Statistics, use the ID <code>general_stats</code>.</p>
<h2 id="removing-modules-or-sections"><a href="#removing-modules-or-sections" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Removing modules or sections</h2>
<p>If you don't want an entire module to be used in a MultiQC report, use the <code>-e</code>/<code>--exclude</code>
command line flags to skip running that tool. You can also use the config option <code>exclude_modules</code>:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">exclude_modules:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">fastqc</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">cutadapt</span></code></pre>
<p>If you want to run <em>only</em> specific modules, you can do that with <code>-m</code>/<code>--module</code> or the
config option <code>run_modules</code>:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">run_modules:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">fastqc</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">cutadapt</span></code></pre>
<p>If you would like to remove just one section of a module report, you can do so with the
<code>remove_sections</code> config option as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">remove_sections:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">section-id-one</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">second-section-id</span></code></pre>
<p>The section ID is the string appended to the URL when clicking a report section in the navigation.
For example, the GATK module has a section with the title <em>"Compare Overlap"</em>. When clicking that
in the report's left hand side navigation, the web browser URL has <code>#gatk-compare-overlap</code>
appended. Here, you would add <code>gatk-compare-overlap</code> to the <code>remove_sections</code> config.</p>
<p>Finally, you can prevent MultiQC from finding the files for a module or submodule by customising
its search pattern. For example, to skip Picard Base Calling metrics, you could use the following:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sp:</span>
  <span class="hljs-string">picard/collectilluminabasecallingmetrics:</span>
<span class="hljs-attr">    skip:</span> <span class="hljs-literal">true</span></code></pre>
<p>The search pattern identifiers can be found in the documentation below for each module.</p>
<h4 id="removing-general-statistics"><a href="#removing-general-statistics" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Removing General Statistics</h4>
<p>The General Statistics is a bit of a special case in MultiQC, but there is added code to make it
behave well with the above mechanism. On the command line, you can specify <code>-e general_stats</code>.
Alternatively, you can set the following config flag in your MultiQC config:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">skip_generalstats:</span> <span class="hljs-literal">true</span></code></pre>
<h2 id="order-of-modules"><a href="#order-of-modules" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Order of modules</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#order-of-module-and-module-subsection-output">Order of module and module subsection output</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>By default, modules are included in the report as in the order specified in <code>config.module_order</code>.
Any modules found which aren't in this list are appended at the top of the report.</p>
<h4 id="top-modules"><a href="#top-modules" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Top modules</h4>
<p>To specify certain modules that should always come at the top of the report, you can configure
<code>config.top_modules</code> in your MultiQC configuration file. For example, to always have the FastQC
module at the top of reports, add the following to your <code>~/.multiqc_config.yaml</code> file:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">top_modules:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">"fastqc"</span></code></pre>
<h4 id="running-modules-multiple-times"><a href="#running-modules-multiple-times" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Running modules multiple times</h4>
<p>A module can be specified multiple times in either <code>config.module_order</code> or <code>config.top_modules</code>,
causing it to be run multiple times. By itself you'll just get two identical report sections.
However, you can also supply configuration options to the modules as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">top_modules:</span>
<span class="hljs-attr">  - moduleName:</span>
<span class="hljs-attr">      name:</span> <span class="hljs-string">"Module (filtered)"</span>
<span class="hljs-attr">      info:</span> <span class="hljs-string">"This section shows the module with different files"</span>
<span class="hljs-attr">      path_filters:</span>
<span class="hljs-bullet">        -</span> <span class="hljs-string">"*_special.txt"</span>
<span class="hljs-bullet">        -</span> <span class="hljs-string">"*_others.txt"</span>
<span class="hljs-attr">  - moduleName:</span>
<span class="hljs-attr">      name:</span> <span class="hljs-string">"Module (not-special)"</span>
<span class="hljs-attr">      path_filters_exclude:</span>
<span class="hljs-bullet">        -</span> <span class="hljs-string">"*_special.txt"</span></code></pre>
<p>These overwrite the defaults that are hardcoded in the module code. <code>path_filters</code> and <code>path_filters_exclude</code> being the exception. These filter the file searches for a given list of glob filename patterns:</p>
<table>
<thead>
<tr>
<th>Pattern</th>
<th>Meaning</th>
</tr>
</thead>
<tbody>
<tr>
<td>*</td>
<td>matches everything</td>
</tr>
<tr>
<td>?</td>
<td>matches any single character</td>
</tr>
<tr>
<td>[seq]</td>
<td>matches any character in seq</td>
</tr>
<tr>
<td>[!seq]</td>
<td>matches any character not in seq</td>
</tr>
</tbody>
</table>
<p>Note that exclusion superseeds inclusion for the path filters.</p>
<p>The other available configuration options are:</p>
<ul>
<li><code>name</code>: Section name</li>
<li><code>anchor</code>: Section report ID</li>
<li><code>target</code>: Intro link text</li>
<li><code>href</code>: Intro link URL</li>
<li><code>info</code>: Intro text</li>
<li><code>extra</code>: Additional HTML after intro.</li>
<li><code>custom_config</code>: Custom module-level settings. Translated into <code>config.moduleName</code>, but specifically for this section.</li>
</ul>
<p>For example, to run the FastQC module twice, before and after adapter trimming, you could
use the following config:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">module_order:</span>
<span class="hljs-attr">  - fastqc:</span>
<span class="hljs-attr">      name:</span> <span class="hljs-string">"FastQC (trimmed)"</span>
<span class="hljs-attr">      anchor:</span> <span class="hljs-string">"fastqc_trimmed"</span>
<span class="hljs-attr">      info:</span> <span class="hljs-string">"This section of the report shows FastQC results after adapter trimming."</span>
<span class="hljs-attr">      target:</span> <span class="hljs-string">""</span>
<span class="hljs-attr">      path_filters:</span>
<span class="hljs-bullet">        -</span> <span class="hljs-string">"*_1_trimmed_fastqc.zip"</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">cutadapt</span>
<span class="hljs-attr">  - fastqc:</span>
<span class="hljs-attr">      name:</span> <span class="hljs-string">"FastQC (raw)"</span>
<span class="hljs-attr">      anchor:</span> <span class="hljs-string">"fastqc_raw"</span>
<span class="hljs-attr">      path_filters:</span>
<span class="hljs-bullet">        -</span> <span class="hljs-string">"*_1_fastqc.zip"</span></code></pre>
<p>Note that if you change the <code>name</code> then you will get multiples of columns in the
<em>General Statistics</em> table. If unchanged, the topmost module may overwrite output from
the first iteration.</p>
<p>If you set a custom <code>anchor</code>, then this can be used for other configuration options.
For example, using the anchors above and the <code>report_section_order</code> described below:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">report_section_order:</span>
<span class="hljs-attr">  fastqc_trimmed:</span>
<span class="hljs-attr">    before:</span> <span class="hljs-string">fastqc_raw</span></code></pre>
<blockquote>
<p>NB: Currently, you can not list a module name in both <code>top_modules</code> and <code>module_order</code>.
Let me know if this is a problem..</p>
</blockquote>
<h3 id="order-of-module-and-module-subsection-output"><a href="#order-of-module-and-module-subsection-output" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Order of module and module subsection output</h3>
<p>The <code>module_order</code> config changes the order in which each MultiQC module is executed.
However, sometimes it's desirable to customise the order of specific sections in a report,
independent of the order of module execution. For example, the <code>custom_content</code> module can
generate multiple sections from different input files.
Also, <code>module_order</code> does not allow you to change the sequence of sections within a MultiQC module.</p>
<p>To change the order of MultiQC outputs, follow a link in a report navigation to skip to the section
you want to move (either a major section header or a subheading). Find the ID of that section by looking at the URL.
For example, clicking on <em>FastQC</em> changes the URL to <code>multiqc_report.html#fastqc</code> - the ID is
the text after (not including) the <code>#</code> symbol: <code>fastqc</code>.
The FastQC <em>Status Checks</em> subsection is <code>multiqc_report.html#fastqc_status_checks</code> and has the id <code>fastqc_status_checks</code>.</p>
<p>Next, specify the <code>report_section_order</code> option in your MultiQC config file. Modules and sections in
the report are given a number ranging from 10 (section at bottom of report), incrementing by +10
for each section. You can change this number (eg. a very low number to always get at the bottom
of the report or very high to always be at the top), or you can move a section to before or after
another existing section (has no effect if the other named ID is not in the report).</p>
<blockquote>
<p>Note that module sub-sections can only be move <em>within</em> their module. So you can't have the
FastQC <em>Adapter Content</em> section shown under the GATK module header.</p>
</blockquote>
<p>You can also use this config option to completely remove module sub-sections.
To do this, just set the subsection ID to <code>remove</code> (NB: no <code>:</code> or <code>-</code>).
This only works for module subsections. To remove an entire module, use the <code>-e</code>/<code>--exclude</code> flag.</p>
<p>For example, you could add the following to your MultiQC config file:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">report_section_order:</span>
<span class="hljs-attr">  module_output_1:</span>
<span class="hljs-attr">    order:</span> <span class="hljs-bullet">-1000</span>
<span class="hljs-attr">  module_output_2:</span>
<span class="hljs-attr">    after:</span> <span class="hljs-string">"diffsection"</span>
<span class="hljs-attr">  mod_section_1:</span>
<span class="hljs-attr">    before:</span> <span class="hljs-string">"othersection"</span>
<span class="hljs-attr">  mod_section_2:</span> <span class="hljs-string">remove</span></code></pre>
<h2 id="customising-plots"><a href="#customising-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Customising plots</h2>
<p>Almost every plot in all MultiQC reports are created using standard plotting functions
and a plot config. You can override any plot config variable you like for any plot to
customise how these are generated.</p>
<p>To do this, first find the plot that you would like to customise and copy it's unique ID.
You can find this by clicking export - the name next to the checkbox is the ID.</p>
<p>Next, you need to find the plot config key(s) that you would like to change.
You can find these by reading the MultiQC documentation below.</p>
<p>For example, to set a new limit for the Picard InsertSizeMetrics x-axis, you can use the following:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_plot_config:</span>
<span class="hljs-attr">  picard_insert_size:</span>
<span class="hljs-attr">    xmax:</span> <span class="hljs-number">300</span></code></pre>
<p>You can customise multiple variables for multiple plots:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_plot_config:</span>
  <span class="hljs-comment"># Show the percentages tab by default for the FastQC sequence counts plot</span>
<span class="hljs-attr">  fastqc_sequence_counts_plot:</span>
<span class="hljs-attr">    cpswitch_c_active:</span> <span class="hljs-literal">False</span>

  <span class="hljs-comment"># Only show up to 20bp on the x axis for cutadapt, change the title</span>
<span class="hljs-attr">  cutadapt_plot:</span>
<span class="hljs-attr">    xmax:</span> <span class="hljs-number">20</span>
<span class="hljs-attr">    title:</span> <span class="hljs-string">"How many base pairs have been removed from the data"</span>

  <span class="hljs-comment"># Add a coloured band in the background to show what is a good result</span>
  <span class="hljs-comment"># Yes I know this doesn't make sense for this plot, it's just an example ;)</span>
<span class="hljs-attr">  bismark_mbias:</span>
<span class="hljs-attr">    yPlotBands:</span>
<span class="hljs-attr">      - from:</span> <span class="hljs-number">0</span>
<span class="hljs-attr">        to:</span> <span class="hljs-number">40</span>
<span class="hljs-attr">        color:</span> <span class="hljs-string">"#e6c3c3"</span>
<span class="hljs-attr">      - from:</span> <span class="hljs-number">40</span>
<span class="hljs-attr">        to:</span> <span class="hljs-number">80</span>
<span class="hljs-attr">        color:</span> <span class="hljs-string">"#e6dcc3"</span>
<span class="hljs-attr">      - from:</span> <span class="hljs-number">80</span>
<span class="hljs-attr">        to:</span> <span class="hljs-number">100</span>
<span class="hljs-attr">        color:</span> <span class="hljs-string">"#c3e6c3"</span></code></pre>
<p>As of version 1.8, this also works for customising the config of bargraph categories:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_plot_config:</span>
<span class="hljs-attr">  bowtie1_alignment:</span>
<span class="hljs-attr">    reads_aligned:</span>
<span class="hljs-attr">      color:</span> <span class="hljs-string">"#d84e2f"</span>
<span class="hljs-attr">    multimapped:</span>
<span class="hljs-attr">      color:</span> <span class="hljs-string">"#f2e63f"</span>
<span class="hljs-attr">    not_aligned:</span>
<span class="hljs-attr">      color:</span> <span class="hljs-string">"#8bbc21"</span></code></pre>
<h2 id="customising-tables"><a href="#customising-tables" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Customising tables</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#hiding-columns">Hiding columns</a></li>
<li><a href="#column-order">Column order</a></li>
<li><a href="#column-titles">Column titles</a></li>
<li><a href="#conditional-formatting">Conditional formatting</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Much like with the custom plot config above, you can override almost any configuration options for tables.
To see what's available, read the documentation about <a href="#creating-a-table">Creating a table</a> below.</p>
<p>Tables have configuration at two levels. Table-wide configs are the same as plot configs and can
be overridden with <code>custom_plot_config</code> as described above.</p>
<p>Headers have their own configuration which can be overriden with <code>custom_table_header_config</code>.</p>
<p>Examples are often more useful for this kind of thing than words, so here are a few:</p>
<p>For the Picard HSMetrics table, we can use a custom table header for the first column
and change the default minimum value for the colour scale for all columns:</p>
<blockquote>
<p>Here <code>min</code> is a <em>header</em> config but we're setting it at table config level.
This means it will be used as a default for all columns in the table if the module
doesn't itself define anything specific for that column.
If it does, you need to overwrite that specific column using <code>custom_table_header_config</code></p>
</blockquote>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_plot_config:</span>
<span class="hljs-attr">  picard_hsmetrics_table:</span>
<span class="hljs-attr">    col1_header:</span> <span class="hljs-string">"Identifiers"</span>
<span class="hljs-attr">    min:</span> <span class="hljs-number">1000</span></code></pre>
<p>Now for header-specific changes.
To change the number of decimals used in the General Statistics table for the Qualimap <em>Mean Coverage</em> column:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_table_header_config:</span>
<span class="hljs-attr">  general_stats_table:</span>
<span class="hljs-attr">    mean_coverage:</span>
<span class="hljs-attr">      format:</span> <span class="hljs-string">"{:,.20f}"</span></code></pre>
<h3 id="hiding-columns"><a href="#hiding-columns" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Hiding columns</h3>
<p>Report tables such as the General Statistics table can get quite wide. To help with this,
columns in the report can be hidden. Some MultiQC modules include columns which are hidden
by default, others may be uninteresting to some users.</p>
<p>To allow customisation of this behaviour, the defaults can be changed by adding to your
MultiQC config file. This is done with the <code>table_columns_visible</code> value. Open a MultiQC
report and click <em>Configure Columns</em> above a table. Make a note of the <em>Group</em> and <em>ID</em>
for the column that you'd like to alter. For example, to make the <code>% Duplicate Reads</code>
column from FastQC hidden by default, the <em>Group</em> is <code>FastQC</code> and the <em>ID</em> is
<code>percent_duplicates</code>. These are then added to the config as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">table_columns_visible:</span>
<span class="hljs-attr">  FastQC:</span>
<span class="hljs-attr">    percent_duplicates:</span> <span class="hljs-literal">False</span></code></pre>
<p>You can also specify a value for an entire module / table namespace.
This will then show or hide all columns for that module. For example:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">table_columns_visible:</span>
<span class="hljs-attr">  FastQC:</span> <span class="hljs-literal">False</span></code></pre>
<p>Note that you can set these values to <code>True</code> to show columns that would otherwise be hidden
by default.</p>
<h3 id="column-order"><a href="#column-order" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Column order</h3>
<p>In the same way, you can force a column to appear at the start or end of the table, or
indeed impose a custom ordering on all the columns, by setting the <code>table_columns_placement</code>.
High values push columns to the right hand side of the table and low to the left. The default
value is 1000. For example:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">table_columns_placement:</span>
<span class="hljs-attr">  Samtools:</span>
<span class="hljs-attr">    reads_mapped:</span> <span class="hljs-number">900</span>
<span class="hljs-attr">    properly_paired:</span> <span class="hljs-number">1010</span>
<span class="hljs-attr">    secondary:</span> <span class="hljs-number">1020</span></code></pre>
<p>In this case, since the default placement weighting is <code>1000</code>, the <code>reads_mapped</code> will end up as the
leftmost column and the other two will and up as the final columns on the right of the table.</p>
<p>The columns are organised by either <em>namespace</em> or table ID, then column ID.
In the above example, <code>Samtools</code> is the namespace in the General Statistics table -
the text that is at the start of the tooltip. For custom tables, the ID may be easier to use.</p>
<h3 id="column-titles"><a href="#column-titles" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Column titles</h3>
<p>Sometimes it may be helpful to adjust the default table column header to display a different title.
For example when running a module multiple times, or when different modules have columns with similar names.</p>
<p>To do this, use the approach described above to find the column <em>Group</em> and <em>ID</em> and combine with the <code>table_columns_name</code> config option.
For example, the following config will change the General Statistics column for FastQC from <em>% GC</em> to <em>Percent of bases that are GC</em></p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">table_columns_name:</span>
<span class="hljs-attr">  FastQC:</span>
<span class="hljs-attr">    percent_gc:</span> <span class="hljs-string">"Percent of bases that are GC"</span></code></pre>
<h3 id="conditional-formatting"><a href="#conditional-formatting" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Conditional formatting</h3>
<p>It's possible to highlight values in tables based on their value. This is done using the <code>table_cond_formatting_rules</code> config setting.
Rules can be applied to every table in the report (<code>all_columns</code>), specific tables (table ID), or specific columns (column ID).</p>
<p>The default rules are as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">table_cond_formatting_rules:</span>
<span class="hljs-attr">  all_columns:</span>
<span class="hljs-attr">    pass:</span>
<span class="hljs-attr">      - s_eq:</span> <span class="hljs-string">"pass"</span>
<span class="hljs-attr">      - s_eq:</span> <span class="hljs-string">"true"</span>
<span class="hljs-attr">    warn:</span>
<span class="hljs-attr">      - s_eq:</span> <span class="hljs-string">"warn"</span>
<span class="hljs-attr">      - s_eq:</span> <span class="hljs-string">"unknown"</span>
<span class="hljs-attr">    fail:</span>
<span class="hljs-attr">      - s_eq:</span> <span class="hljs-string">"fail"</span>
<span class="hljs-attr">      - s_eq:</span> <span class="hljs-string">"false"</span></code></pre>
<p>These make any table cells in the report that match the string <code>pass</code> or <code>true</code> have text with a green background, orange for <code>warn</code>, red for <code>fail</code> and so on.
There can be multiple tests for each style of formatting - if there is a match for any, it will be applied.
The following comparison operators are available:</p>
<ul>
<li><code>s_eq</code> - String exactly equals (case insensitive)</li>
<li><code>s_contains</code> - String contains (case insensitive)</li>
<li><code>s_ne</code> - String does not equal (case insensitive)</li>
<li><code>eq</code> - Value equals</li>
<li><code>ne</code> - Value does not equal</li>
<li><code>gt</code> - Value is greater than</li>
<li><code>lt</code> - Value is less than</li>
</ul>
<p>To have matches for a specific table or column, use that ID instead of <code>all_columns</code>.</p>
<p>For example, for the entire General Stats table:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">table_cond_formatting_rules:</span>
<span class="hljs-attr">  general_stats_table:</span>
<span class="hljs-attr">    pass:</span>
<span class="hljs-attr">      - gt:</span> <span class="hljs-number">80</span>
<span class="hljs-attr">    warn:</span>
<span class="hljs-attr">      - lt:</span> <span class="hljs-number">80</span>
<span class="hljs-attr">    fail:</span>
<span class="hljs-attr">      - lt:</span> <span class="hljs-number">70</span></code></pre>
<p>Or for one column in the General Stats table:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">table_cond_formatting_rules:</span>
<span class="hljs-attr">  mqc-generalstats-uniquely_mapped_percent:</span>
<span class="hljs-attr">    pass:</span>
<span class="hljs-attr">      - gt:</span> <span class="hljs-number">80</span>
<span class="hljs-attr">    warn:</span>
<span class="hljs-attr">      - lt:</span> <span class="hljs-number">80</span>
<span class="hljs-attr">    fail:</span>
<span class="hljs-attr">      - lt:</span> <span class="hljs-number">70</span></code></pre>
<p>Note that the formatting is done in a specific order - <code>pass</code>/<code>warn</code>/<code>fail</code> by default, so that anything matching both <code>warn</code> and <code>fail</code> will be formatted as <code>fail</code> for example. This can be customised with <code>table_cond_formatting_colours</code> (see below).</p>
<p>To find the unique ID for your table / column, right click it in a report and inspect it's HTML (<em>Inpsect</em> in Chrome).</p>
<ul>
<li>Tables should look something like <code>&lt;table id="general_stats_table" class="table table-condensed mqc_table" data-title="General Statistics"&gt;</code>, where <code>general_stats_table</code> is the ID.</li>
<li>Table cells should look something like <code>&lt;td class="data-coloured mqc-generalstats-Assigned"&gt;</code>, where the <code>mqc-generalstats-Assigned</code> bit is the unique ID.</li>
</ul>
<blockquote>
<p>I know this isn't the same method of IDs as above and isn't super easy to do. Sorry!</p>
</blockquote>
<p>It's possible to highlight matches in any number of colours. MultiQC comes with the following defaults:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">table_cond_formatting_colours:</span>
<span class="hljs-attr">  - blue:</span> <span class="hljs-string">"#337ab7"</span>
<span class="hljs-attr">  - lbue:</span> <span class="hljs-string">"#5bc0de"</span>
<span class="hljs-attr">  - pass:</span> <span class="hljs-string">"#5cb85c"</span>
<span class="hljs-attr">  - warn:</span> <span class="hljs-string">"#f0ad4e"</span>
<span class="hljs-attr">  - fail:</span> <span class="hljs-string">"#d9534f"</span></code></pre>
<p>These can be overridden or added to with any string / CSS hex colour 
combinations you like. You can generate hex colour codes with lots of 
tools, for example <a href="http://htmlcolorcodes.com/">http://htmlcolorcodes.com/</a></p>
<p>Note that the different sets of rules are formatted in order. So if a value matches both <code>pass</code> and <code>fail</code> then it will be formatted as a <code>fail</code></p>
<h2 id="number-base-multiplier"><a href="#number-base-multiplier" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Number base (multiplier)</h2>
<p>To make numbers in the General Statistics table easier to read and compare quickly,
MultiQC sometimes divides them by one million (typically read counts). If your
samples have very low read counts then this can result in the table showing
counts of <code>0.0</code>, which isn't very helpful.</p>
<p>To change this behaviour, you can customise three config variables in your MultiQC
config. The defaults are as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">read_count_multiplier:</span> <span class="hljs-number">0.000001</span>
<span class="hljs-attr">read_count_prefix:</span> <span class="hljs-string">"M"</span>
<span class="hljs-attr">read_count_desc:</span> <span class="hljs-string">"millions"</span></code></pre>
<p>So, to show thousands of reads instead of millions, change these to:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">read_count_multiplier:</span> <span class="hljs-number">0.001</span>
<span class="hljs-attr">read_count_prefix:</span> <span class="hljs-string">"K"</span>
<span class="hljs-attr">read_count_desc:</span> <span class="hljs-string">"thousands"</span></code></pre>
<p>The same options are also available for numbers of base pairs:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">base_count_multiplier:</span> <span class="hljs-number">0.000001</span>
<span class="hljs-attr">base_count_prefix:</span> <span class="hljs-string">"Mb"</span>
<span class="hljs-attr">base_count_desc:</span> <span class="hljs-string">"millions"</span></code></pre>
<p>And for long reads:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">long_read_count_multiplier:</span> <span class="hljs-number">0.001</span>
<span class="hljs-attr">long_read_count_prefix:</span> <span class="hljs-string">"K"</span>
<span class="hljs-attr">long_read_count_desc:</span> <span class="hljs-string">"thousands"</span></code></pre>
<h2 id="number-formatting"><a href="#number-formatting" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Number formatting</h2>
<p>By default, the interactive HighCharts plots in MultiQC reports use spaces for thousand
separators and points for decimal places (<em>e.g.</em> <code>1 234 567.89</code>). Different countries
have different preferences for this, so you can customise the two using a couple of
configuration parameters - <code>decimalPoint_format</code> and <code>thousandsSep_format</code>.</p>
<p>For example, the following config would result in the following alternative
number formatting: <code>1234567,89</code>.</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">decimalPoint_format:</span> <span class="hljs-string">","</span>
<span class="hljs-attr">thousandsSep_format:</span> <span class="hljs-string">""</span></code></pre>
<p>This formatting currently only applies to the interactive charts. It may be extended
to apply elsewhere in the future (submit a new issue if you spot somewhere where you'd like it).</p>
<h2 id="troubleshooting"><a href="#troubleshooting" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Troubleshooting</h2>
<p>One tricky bit that caught me out whilst writing this is the different type casting
between Python, YAML and Jinja2 templates. This is especially true when using an
empty variable:</p>
<pre><code class="language-python hljs"><span class="hljs-comment"># Python</span>
my_var = <span class="hljs-literal">None</span></code></pre>
<pre><code class="language-yaml hljs"><span class="hljs-comment"># YAML</span>
<span class="hljs-attr">my_var:</span> <span class="hljs-literal">null</span></code></pre>
<pre><code class="language-python hljs"><span class="hljs-comment"># Jinja2</span>
<span class="hljs-keyword">if</span> myvar <span class="hljs-keyword">is</span> none <span class="hljs-comment"># Note - Lower case!</span></code></pre></div><div class="docs_block" id="pipelines.md"><h1 id="using-multiqc-in-pipelines"><a href="#using-multiqc-in-pipelines" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using MultiQC in pipelines</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#nextflow">Nextflow</a>
<ul class="nav nav-stacked">
<li><a href="#clashing-input-filenames">Clashing input filenames</a></li>
<li><a href="#custom-run-name">Custom run name</a></li>
<li><a href="#multiqc-config-file">MultiQC config file</a></li>
</ul>
</li>
<li><a href="#snakemake">Snakemake</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>MultiQC has been designed to be placed at the end of bioinformatics workflows
and works well when it's a final step in a pipeline. Here you can find a few
tips about integration with different workflow tools.</p>
<p>I'll use FastQC as an example input in all of these examples as it's a common
use case. However, the concepts should apply for any of the tools that MultiQC supports.</p>
<blockquote>
<p>Remember that you can use <a href="https://multiqc.info/docs/#custom-content">Custom Content</a>
feature to easily collect pipeline-specific metadata (software version numbers,
pipeline run-time data, links to documentation) in to a format that can be inserted
in to your report.</p>
</blockquote>
<p>If you know exactly which modules will be used by MultiQC, you can use the
<code>-m</code>/<code>--modules</code> flag to specify just these. This will speed up MultiQC a little.
This will probably only make a noticeable impact if your pipeline has thousands
of input files for MultiQC.</p>
<p>In the following, we show examples for <a href="#nextflow">Nextflow</a> and <a href="#snakemake">Snakemake</a> integration.</p>
<h2 id="nextflow"><a href="#nextflow" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Nextflow</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#clashing-input-filenames">Clashing input filenames</a></li>
<li><a href="#custom-run-name">Custom run name</a></li>
<li><a href="#multiqc-config-file">MultiQC config file</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>An example mini-pipeline for <a href="https://www.nextflow.io/">nextflow</a> which runs FastQC
and MultiQC is below.
See the <a href="https://nf-co.re/">nf-core</a> pipelines for lots of examples of full nextflow
pipelines that use MultiQC.</p>
<pre><code class="language-groovy hljs"><span class="hljs-meta">#!/usr/bin/env nextflow</span>

params.reads = <span class="hljs-string">"data/*{1,2}.fastq.gz"</span>
Channel.fromFilePairs( params.reads ).into { read_files_fastqc }

process fastqc {
<span class="hljs-symbol">    input:</span>
    set val(name), file(reads) from read_files_fastqc
<span class="hljs-symbol">
    output:</span>
    file <span class="hljs-string">"*_fastqc.{zip,html}"</span> into fastqc_results
<span class="hljs-symbol">
    script:</span>
    <span class="hljs-string">"""
    fastqc -q $reads
    """</span>
}

process multiqc {
<span class="hljs-symbol">    input:</span>
    file (<span class="hljs-string">'fastqc/*'</span>) from fastqc_results.collect().ifEmpty([])
<span class="hljs-symbol">
    output:</span>
    file <span class="hljs-string">"multiqc_report.html"</span> into multiqc_report
    file <span class="hljs-string">"multiqc_data"</span>
<span class="hljs-symbol">
    script:</span>
    <span class="hljs-string">"""
    multiqc .
    """</span>
}</code></pre>
<p>You will need to specify channels for all files that MultiQC needs to run, so that nextflow
stages them correctly in the MultiQC work directory.</p>
<p>Note that <code>.collect()</code> is needed to make MultiQC run once for all upstream outputs.</p>
<p>The <code>.ifEmpty([])</code> add on isn't really needed here, but is helpful in larger pipelines where
some processes may be optional. Without this, if any channels are not processed then MultiQC
won't run.</p>
<h3 id="clashing-input-filenames"><a href="#clashing-input-filenames" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Clashing input filenames</h3>
<p>If a nextflow process tries to stage more than one input file with an identical filename,
it will throw an error. Putting inputs into their own subfolder (<code>file ('fastqc/*')</code>) in
the above examples is not really needed, but reduces the chance of input filename clashes.</p>
<p>If you're using a tool that gives the same filename to each file that MultiQC uses, you'll
need to tell nextflow to rename the inputs to prevent clashes.</p>
<p>For example, <a href="https://ccb.jhu.edu/software/stringtie/">StringTie</a> prints statistics to
STDOUT that MultiQC uses to generate reports. We can easily collect this in nextflow by
using the <code>.command.log</code> file that nextflow saves as an output file from the process
(<code>file ".command.log" into stringtie_log</code>). However, now every sample has the same filename
for MultiQC.</p>
<p>We get around this by using <a href="https://www.nextflow.io/docs/latest/process.html#dynamic-input-file-names">dynamic input file names</a>
with nextflow:</p>
<pre><code class="language-groovy hljs">file (<span class="hljs-string">'stringtie/stringtie_log*'</span>) from stringtie_log.collect().ifEmpty([])</code></pre>
<p>This <code>file</code> pattern renames each stringtie log file to <code>stringtie_log1</code>,
<code>stringtie_log2</code> and so on, ensuring that we avoid any filename clashes.</p>
<p>Note that MultiQC finds output from some tools based on their filename, so use with caution
(you may need to define some custom <a href="https://multiqc.info/docs/#module-search-patterns">module search patterns</a>).</p>
<h3 id="custom-run-name"><a href="#custom-run-name" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Custom run name</h3>
<p>When you launch nextflow, you can use the <code>-name</code> command line flag (single hyphen) to give
a name to that specific pipeline run. You can configure your pipeline to pass this on to
MultiQC. It can then be used as the report title and filename.</p>
<p>Here's an example snippet for this:</p>
<pre><code class="language-groovy hljs">custom_runName = params.name
<span class="hljs-keyword">if</span>( !(workflow.runName ==~ <span class="hljs-regexp">/[a-z]+_[a-z]+/</span>) ){
  custom_runName = workflow.runName
}

process multiqc {
<span class="hljs-symbol">    input:</span>
    file (<span class="hljs-string">'fastqc/*'</span>) from ch_fastqc_results_for_multiqc.collect().ifEmpty([])
<span class="hljs-symbol">
    output:</span>
    file <span class="hljs-string">"*_report.html"</span> into ch_multiqc_report
    file <span class="hljs-string">"*_data"</span>
<span class="hljs-symbol">
    script:</span>
    rtitle = custom_runName ? "--title \"$custom_runName\"" : <span class="hljs-string">''</span>
    rfilename = custom_runName ? "--filename " + custom_runName.replaceAll('\\W','_').replaceAll('_+','_') + "_multiqc_report" : <span class="hljs-string">''</span>
    <span class="hljs-string">"""
    multiqc $rtitle $rfilename .
    """</span>
}</code></pre>
<p>Note that we use <code>params.name</code> as a placeholder, this gives the benefit that both <code>-name</code>
and the common typo <code>--name</code> work for this case.</p>
<p>There isn't an easy way to know if a custom value for <code>-name</code> has been given to nextflow,
but all default names are two lowercase words with a single underscore so if the name
matches this pattern then we ignore it.</p>
<h3 id="multiqc-config-file"><a href="#multiqc-config-file" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiQC config file</h3>
<p>It can be nice to use a config file for MultiQC to add in some static content to reports
about your pipeline (eg. <code>report_comment</code>).
Remember that even this config file should also be in a nextflow channel,
so that nextflow correctly stages it (especially important when running on the cloud).</p>
<p>This snippet works with a <code>params</code> variable again, so that pipeline users can replace
the config file with one of their own if they wish.</p>
<pre><code class="language-groovy hljs">params.multiqc_config = <span class="hljs-string">"$baseDir/assets/multiqc_config.yaml"</span>
Channel.fromPath(params.multiqc_config, <span class="hljs-string">checkIfExists:</span> <span class="hljs-literal">true</span>).set { ch_config_for_multiqc }

process multiqc {
<span class="hljs-symbol">    input:</span>
    file multiqc_config from ch_config_for_multiqc
    file (<span class="hljs-string">'fastqc/*'</span>) from fastqc_results.collect().ifEmpty([])
<span class="hljs-symbol">
    output:</span>
    file <span class="hljs-string">"multiqc_report.html"</span> into multiqc_report
    file <span class="hljs-string">"multiqc_data"</span>
<span class="hljs-symbol">
    script:</span>
    <span class="hljs-string">"""
    multiqc --config $multiqc_config .
    """</span>
}</code></pre>
<h2 id="snakemake"><a href="#snakemake" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Snakemake</h2>
<p>The best-practice for using MultiQC in <a href="https://snakemake.readthedocs.io/">Snakemake</a> pipelines is to use a predefined <a href="https://snakemake-wrappers.readthedocs.io/en/stable/wrappers/multiqc.html">Snakemake wrapper</a>. For example, see the following mini-pipeline:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">configfile:</span> <span class="hljs-string">"config.yaml"</span>

<span class="hljs-string">rule</span> <span class="hljs-attr">all:</span>
<span class="hljs-attr">  input:</span> <span class="hljs-string">"multiqc_report.html"</span>

<span class="hljs-string">rule</span> <span class="hljs-attr">fastqc:</span>
<span class="hljs-attr">  input:</span> <span class="hljs-string">"data/{sample}.fastq.gz"</span>
<span class="hljs-attr">  output:</span> <span class="hljs-string">html="fastqc/{sample}.html",</span>
    <span class="hljs-string">zip="fastqc/{sample}.zip"</span>
<span class="hljs-attr">  wrapper:</span> <span class="hljs-string">"0.31.1/bio/fastqc"</span>

<span class="hljs-string">rule</span> <span class="hljs-attr">multiqc:</span>
<span class="hljs-attr">  input:</span> <span class="hljs-string">expand("fastqc/{sample}.html",</span> <span class="hljs-string">sample=config["samples"])</span>
<span class="hljs-attr">  output:</span> <span class="hljs-string">"multiqc_report.html"</span>
<span class="hljs-attr">  wrapper:</span> <span class="hljs-string">"0.31.1/bio/multiqc"</span></code></pre>
<p>Snakemake wrappers not only deliver predefined and unit tested code 
for generating the requested output with the respective tool, but also 
define the required software stack in terms of conda packages.</p>
<p>You can run the above workflow as follows:</p>
<pre><code class="language-bash hljs">snakemake --use-conda</code></pre>
<p>This first installs all the required tools into isolated conda 
environments, and then executes all necessary steps to create the target
 that is given in the top rule. In other words, it will execute FastQC 
twice (creating <code>fastqc/1.html</code> and <code>fastqc/2.html</code>) and MultiQC once, creating <code>multiqc_report.html</code>.</p>
<p>Of course, using conda is optional, but it greatly increases reproducibility.
Snakemake is not limited to wrappers (although its <a href="https://snakemake-wrappers.readthedocs.io/">wrapper repository</a> provides many in the field of bioinformatics), but also supports direct execution of <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#rules">shell commands</a> and integration of <a href="https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#external-scripts">custom scripts</a> (e.g., for plotting).</p>
<p>If you prefer to use MultiQC without a snakemake wrapper, you can see a minimal example on GitHub: <a href="https://github.com/jakevc/snakemake_multiqc">jakevc/snakemake_multiqc</a>. This has an example script and some test data for you to play with.</p></div><div class="docs_block" id="downstream.md"><h1 id="downstream-analysis"><a href="#downstream-analysis" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Downstream analysis</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#tidymultiqc">TidyMultiqc</a></li>
<li><a href="#megaqc">MegaQC</a></li>
<li><a href="#chronqc">ChronQC</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Whilst MultiQC is typically used as a final reporting step in an 
analysis, it can also be used as an intermediate in your analysis.</p>
<p>MultiQC saves a directory of machine-readable outputs called <code>multiqc_data/</code>. In here there are files from each module and table, as well as a verbose <code>multiqc.log</code> file and a <code>multiqc_data.json</code> file that contains just about everything.</p>
<p>Most of these files are tab-separated <code>.tsv</code> files by default, but you can choose to have them as JSON, YAML if you prefer with the <code>-k</code>/<code>--data-format</code> flag or the <code>data_format</code> option in a config file.</p>
<p>These files can be useful as MultiQC essentially standardises the 
outputs from a lot of different tools.
Typical usage of MultiQC outputs could be filtering of large datasets 
(eg. single-cell analysis) or trend-monitoring of repeated runs.</p>
<p>Below are a few tools that are specifically designed to work with MultiQC.
They are not created by or endorsed by the MultiQC author but may be helpful for your research.</p>
<h2 id="tidymultiqc"><a href="#tidymultiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>TidyMultiqc</h2>
<ul>
<li>Homepage: <a href="https://cran.r-project.org/package=TidyMultiqc">https://CRAN.R-project.org/package=TidyMultiqc</a></li>
<li>Source: <a href="https://github.com/TMiguelT/TidyMultiqc">https://github.com/TMiguelT/TidyMultiqc</a></li>
</ul>
<p>Provides the means to convert <code>multiqc_data.json</code> files into <code>tidy</code> data frames for downstream analysis in R.</p>
<p>This analysis might involve cohort analysis, quality control 
visualisation, change-point detection, statistical process control, 
clustering, or any other type of quality analysis.</p>
<h2 id="megaqc"><a href="#megaqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MegaQC</h2>
<ul>
<li>Homepage: <a href="https://megaqc.info/">https://megaqc.info</a></li>
<li>Source: <a href="https://github.com/ewels/MegaQC">https://github.com/ewels/MegaQC</a></li>
</ul>
<p>Started off by MultiQC author <a href="https://github.com/ewels/">@ewels</a>
 this project is now under active development by a team of several 
contributors. It is functional but still has several parts of its 
codebase in in development. Additional contributors and maintainors very
 welcome!</p>
<p>MegaQC imports data from multiple MultiQC runs and provides an 
interface to explore this with an interactive web server using a 
database backend.
It can plot data over time, across runs and even has an interactive 
dashboard builder.
It's useful for anyone who wants to monitor MultiQC statistics (eg. 
clinical labs) or work interactively with large datasets (eg. single 
cell analysis).</p>
<h2 id="chronqc"><a href="#chronqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>ChronQC</h2>
<ul>
<li>Docs: <a href="https://chronqc.readthedocs.io/">https://chronqc.readthedocs.io</a></li>
<li>Source: <a href="https://github.com/nilesh-tawari/ChronQC">https://github.com/nilesh-tawari/ChronQC</a></li>
</ul>
<p>ChronQC is a quality control (QC) tracking system for clinical 
implementation of next-generation sequencing (NGS). ChronQC generates 
time series plots for various QC metrics, which allows comparison of the
 current run to historical runs. ChronQC has multiple features for 
tracking QC data including Westgard rules for clinical validity, 
laboratory-defined thresholds, and historical observations within a 
specified period. Users can record their notes and corrective actions 
directly onto the plots for long-term recordkeeping.</p></div><div class="docs_block" id="troubleshooting.md"><h1 id="troubleshooting-1"><a href="#troubleshooting-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Troubleshooting</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#not-enough-samples-found">Not enough samples found</a>
<ul class="nav nav-stacked">
<li><a href="#big-log-files">Big log files</a></li>
</ul>
</li>
<li><a href="#no-logs-found-for-a-tool">No logs found for a tool</a></li>
<li><a href="#error-messages-about-mkl-trial-mode--licences">Error messages about mkl trial mode / licences</a></li>
<li><a href="#locale-error-messages">Locale Error Messages</a></li>
<li><a href="#no-space-left-on-device">No space left on device</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Hopefully MultiQC will be easy to use and run without any hitches. If you have
any problems, please do get in touch with the developer
(<a href="http://phil.ewels.co.uk/">Phil Ewels</a>) by e-mail or by
<a href="https://github.com/ewels/MultiQC/issues/new">submitting an issue</a> on github.
Before that, here are a few things previously encountered that may help...</p>
<h2 id="not-enough-samples-found"><a href="#not-enough-samples-found" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Not enough samples found</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#big-log-files">Big log files</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>In this scenario, MultiQC finds <em>some</em> logs for the bioinformatics tool
in question, but not all of your samples appear in the report. This is
the most common question I get regarding MultiQC operation.</p>
<p>Usually, this happens because sample names collide. This happens innocently
a lot - MultiQC overwrites previous results of the same name and you get
the last one seen in the report. You can see warnings about this by running
MultiQC in verbose mode with the <code>-v</code> flag, or looking at the generated log
file in <code>multiqc_data/multiqc.log</code>. If you are unsure about what log file
ended up in the report, look at <code>multiqc_data/multiqc_sources.txt</code> which
lists each source file used.</p>
<p>To solve this, try running MultiQC with the <code>-d</code> and <code>-s</code> flags.
The <a href="http://multiqc.info/docs/#clashing-sample-names">Clashing sample names</a>
section of the docs explains this in more detail.</p>
<h3 id="big-log-files"><a href="#big-log-files" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Big log files</h3>
<p>Another reason that log files can be skipped is if the log filesize is very
large. For example, this could happen with very long concatenated standard out
files. By default, MultiQC skips any file that is larger than 10MB to keep
execution fast. The verbose log output (<code>-v</code> or <code>multiqc_data/multiqc.log</code>) will
show you if files are being skipped with messages such as these:</p>
<pre><code class="language-txt">[DEBUG  ]  Ignoring file as too large: filename.txt</code></pre>
<p>You can configure the threshold and parse your files by changing the
<code>log_filesize_limit</code> config option. For example, to parse files up to 2GB in
size, add the following to your MultiQC config file:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">log_filesize_limit:</span> <span class="hljs-number">2000000000</span></code></pre>
<h2 id="no-logs-found-for-a-tool"><a href="#no-logs-found-for-a-tool" class="header-link"><span class="glyphicon glyphicon-link"></span></a>No logs found for a tool</h2>
<p>In this case, you have run a bioinformatics tool and have some log files in
a directory. When you run MultiQC with that directory, it finds nothing
for the tool in question.</p>
<p>There are a couple of things you can check here:</p>
<ol>
<li>Is the tool definitely
<a href="https://github.com/ewels/MultiQC">supported by MultiQC</a>? If not, why
not <a href="https://github.com/ewels/MultiQC/issues/new">open an issue</a> to
request it!</li>
<li>Did your bioinformatics tool definitely run properly? I've spent quite
a bit of time debugging MultiQC modules only to realise that the output
files from the tool were empty or incomplete. If your data is missing,
take a look and the raw files and make sure that there's something to see!</li>
</ol>
<p>If everything looks fine, then MultiQC probably needs extending to support
your data. Tools have different versions, different parameters and different
output formats that can confuse the parsing code.
Please <a href="https://github.com/ewels/MultiQC/issues/new">open an issue</a> with
your log files and we can get it fixed.</p>
<h2 id="error-messages-about-mkl-trial-mode--licences"><a href="#error-messages-about-mkl-trial-mode--licences" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Error messages about mkl trial mode / licences</h2>
<p>In this case you run MultiQC and get something like this:</p>
<pre><code class="language-console">$ multiqc .

Vendor:  Continuum Analytics, Inc.
Package: mkl
Message: trial mode EXPIRED 2 days ago

    You cannot run mkl without a license any longer.
    A license can be purchased it at: http://continuum.io
    We are sorry for any inconveniences.

    SHUTTING DOWN PYTHON INTERPRETER</code></pre>
<p>The <code>mkl</code> library provides optimisations for <code>numpy</code>, a requirement of
<code>MatPlotLib</code>. Recent versions of Conda have a bundled version which should
come with a licence and remove the warning. See
<a href="https://docs.continuum.io/mkl-optimizations/index#dismissing-mkl-trial-warnings">this page</a>
for more info. If you already have Conda installed you can get the updated
version by running:</p>
<pre><code class="language-bash hljs">conda remove mkl-rt
conda install -f mkl</code></pre>
<p>Another way around it is to uninstall <code>mkl</code>. It seems that <code>numpy</code> works
without it fine:</p>
<pre><code class="language-bash hljs">conda remove --features mkl</code></pre>
<p>Problem solved! See more
<a href="http://stackoverflow.com/questions/25204021/anaconda-running-python-cannot-run-mkl-without-a-license">here</a> and
<a href="https://www.continuum.io/blog/developer-blog/anaconda-25-release-now-mkl-optimizations">here</a>.</p>
<p>If you're not using Conda, try installing MultiQC with that instead. You
can find instructions <a href="http://multiqc.info/docs/#installing-with-conda">here</a>.</p>
<h2 id="locale-error-messages"><a href="#locale-error-messages" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Locale Error Messages</h2>
<p>Two MultiQC dependencies have been known to throw errors due to problems
with the Python locale settings, or rather the lack of those settings.</p>
<p>MatPlotLib can complain that some strings (such as <code>en_SE</code>) aren't allowed.
Running MultiQC gives the following error:</p>
<pre><code class="language-bash hljs">multiqc --version</code></pre>
<pre><code class="language-python hljs"><span class="hljs-comment"># ..long traceback.. #</span>
 File <span class="hljs-string">"/sw/comp/python/2.7.6_milou/lib/python2.7/locale.py"</span>, line <span class="hljs-number">443</span>, <span class="hljs-keyword">in</span> _parse_localename
   <span class="hljs-keyword">raise</span> ValueError, <span class="hljs-string">'unknown locale: %s'</span> % localename
ValueError: unknown locale: UTF<span class="hljs-number">-8</span></code></pre>
<p>Click can have a similar problem if the locale isn't set when using
Python 3. That generates an error that looks like this:</p>
<pre><code class="language-python hljs"><span class="hljs-comment"># ..truncated traceback.. #</span>
File <span class="hljs-string">"click/_unicodefun.py"</span>, line <span class="hljs-number">118</span>, <span class="hljs-keyword">in</span> _verify_python3_env <span class="hljs-string">'for mitigation steps.'</span> + extra)

RuntimeError: Click will abort further execution because Python <span class="hljs-number">3</span> was configured to use ASCII
<span class="hljs-keyword">as</span> encoding <span class="hljs-keyword">for</span> the environment.  Consult http://click.pocoo.org/python3/<span class="hljs-keyword">for</span> mitigation steps.</code></pre>
<p>You can fix both of these problems by changing your system locale
to something that will be recognised. One way to do this is by adding
these lines to your <code>.bashrc</code> in your home directory (or <code>.bash_profile</code>):</p>
<pre><code class="language-bash hljs"><span class="hljs-built_in">export</span> LC_ALL=en_US.UTF-8
<span class="hljs-built_in">export</span> LANG=en_US.UTF-8</code></pre>
<p>Other locale strings are also fine, as long as the variables are set and valid.</p>
<h2 id="no-space-left-on-device"><a href="#no-space-left-on-device" class="header-link"><span class="glyphicon glyphicon-link"></span></a>No space left on device</h2>
<p>If you're running with very large datasets or have an unusually small temporary file
storage location, you may run in to the following error:</p>
<pre><code class="language-python hljs">  File <span class="hljs-string">"/usr/local/lib/python3.8/site-packages/multiqc/utils/util_functions.py"</span>, line <span class="hljs-number">71</span>, <span class="hljs-keyword">in</span> write_data_file
    print( jsonstr.encode(<span class="hljs-string">'utf-8'</span>, <span class="hljs-string">'ignore'</span>).decode(<span class="hljs-string">'utf-8'</span>), file=f)
OSError: [Errno <span class="hljs-number">28</span>] No space left on device</code></pre>
<p>This happens because MultiQC writes all output files to a temporary directory before
moving them to their final location (depending on configuration, MultiQC can change its output filenames
based on the data that it parses).</p>
<p>To solve this problem you can manually set the temp folder to another folder that has more space.
This is best done with an environment variable which is understood by the base Python installation, <code>TMPDIR</code>.</p>
<p>For example:</p>
<pre><code class="language-bash hljs"><span class="hljs-built_in">export</span> TMPDIR=/path/to/my/tmp</code></pre>
<p>Note that you will have to do this in each session where you run MultiQC. To automatically apply
it at every login, add the above line to your <code>~/.bashrc</code> file.</p>
<p>Be aware that setting this could have unforeseen consequences as it could affect the behaviour of other tools.</p></div></div><div class="docs_section">
<h1 class="section-header" id="multiqc-modules"><a href="#multiqc-modules" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiQC Modules</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#pre-alignment">Pre-alignment</a>
<ul class="nav nav-stacked">
<li><a href="#adapter-removal">Adapter Removal</a></li>
<li><a href="#afterqc">AfterQC</a></li>
<li><a href="#bcl2fastq">bcl2fastq</a></li>
<li><a href="#bclconvert">bclconvert</a></li>
<li><a href="#biobloom-tools">BioBloom Tools</a></li>
<li><a href="#ccs">CCS</a></li>
<li><a href="#checkqc">CheckQC</a></li>
<li><a href="#clipandmerge">ClipAndMerge</a></li>
<li><a href="#cluster-flow">Cluster Flow</a></li>
<li><a href="#cutadapt">Cutadapt</a></li>
<li><a href="#fastp">Fastp</a></li>
<li><a href="#fastq-screen">FastQ Screen</a></li>
<li><a href="#fastqc">FastQC</a>
<ul class="nav nav-stacked">
<li><a href="#theoretical-gc-content">Theoretical GC Content</a></li>
<li><a href="#changing-the-order-of-sections">Changing the order of sections</a></li>
</ul>
</li>
<li><a href="#flash">FLASh</a></li>
<li><a href="#flexbar">Flexbar</a></li>
<li><a href="#interop">InterOp</a></li>
<li><a href="#ivar">iVar</a></li>
<li><a href="#jellyfish">Jellyfish</a></li>
<li><a href="#kat">KAT</a></li>
<li><a href="#leehom">leeHom</a></li>
<li><a href="#lima">Lima</a></li>
<li><a href="#minionqc">minionqc</a></li>
<li><a href="#pbmarkdup">pbmarkdup</a></li>
<li><a href="#pychopper">Pychopper</a></li>
<li><a href="#pycoqc">pycoQC</a></li>
<li><a href="#seqyclean">SeqyClean</a></li>
<li><a href="#sickle">Sickle</a></li>
<li><a href="#skewer">Skewer</a></li>
<li><a href="#sortmerna">SortMeRNA</a></li>
<li><a href="#trimmomatic">Trimmomatic</a></li>
</ul>
</li>
<li><a href="#aligners">Aligners</a>
<ul class="nav nav-stacked">
<li><a href="#bbmap">BBMap</a></li>
<li><a href="#biscuit">BISCUIT</a>
<ul class="nav nav-stacked">
<li><a href="#insert-size-distribution">Insert Size Distribution</a></li>
</ul>
</li>
<li><a href="#bismark">Bismark</a></li>
<li><a href="#bowtie-1">Bowtie 1</a></li>
<li><a href="#bowtie-2">Bowtie 2</a></li>
<li><a href="#dragen">DRAGEN</a></li>
<li><a href="#hic-pro">HiC-Pro</a></li>
<li><a href="#hicup">HiCUP</a></li>
<li><a href="#hisat2">HISAT2</a></li>
<li><a href="#kallisto">Kallisto</a></li>
<li><a href="#longranger">Longranger</a></li>
<li><a href="#malt">MALT</a></li>
<li><a href="#salmon">Salmon</a></li>
<li><a href="#star">STAR</a></li>
<li><a href="#tophat">TopHat</a></li>
</ul>
</li>
<li><a href="#post-alignment">Post-alignment</a>
<ul class="nav nav-stacked">
<li><a href="#bamtools">Bamtools</a></li>
<li><a href="#bcftools">Bcftools</a></li>
<li><a href="#biobambam2">biobambam2</a></li>
<li><a href="#busco">BUSCO</a></li>
<li><a href="#bustools">bustools</a></li>
<li><a href="#conpair">Conpair</a></li>
<li><a href="#damageprofiler">DamageProfiler</a></li>
<li><a href="#dedup">DeDup</a></li>
<li><a href="#deeptools">deepTools</a></li>
<li><a href="#disambiguate">Disambiguate</a></li>
<li><a href="#eigenstratdatabasetools">EigenStratDatabaseTools</a></li>
<li><a href="#featurecounts">featureCounts</a></li>
<li><a href="#fgbio">Fgbio</a></li>
<li><a href="#gatk">GATK</a>
<ul class="nav nav-stacked">
<li><a href="#baserecalibrator">BaseRecalibrator</a></li>
<li><a href="#varianteval">VariantEval</a></li>
</ul>
</li>
<li><a href="#gffcompare">GffCompare</a></li>
<li><a href="#goleft-indexcov">goleft indexcov</a></li>
<li><a href="#hap.py">Hap.py</a></li>
<li><a href="#hicexplorer">HiCExplorer</a></li>
<li><a href="#homer">HOMER</a></li>
<li><a href="#hops">HOPS</a></li>
<li><a href="#htseq">HTSeq</a></li>
<li><a href="#jcvi-genome-annotation">JCVI Genome Annotation</a></li>
<li><a href="#kaiju">Kaiju</a></li>
<li><a href="#kraken">Kraken</a></li>
<li><a href="#macs2">MACS2</a></li>
<li><a href="#methylqa">methylQA</a></li>
<li><a href="#mirtop">mirtop</a></li>
<li><a href="#mirtrace">miRTrace</a></li>
<li><a href="#mosdepth">mosdepth</a></li>
<li><a href="#mtnucratio">MTNucRatio</a></li>
<li><a href="#multivcfanalyzer">MultiVCFAnalyzer</a></li>
<li><a href="#nanostat">NanoStat</a></li>
<li><a href="#ngsderive">ngsderive</a></li>
<li><a href="#odgi">odgi</a></li>
<li><a href="#optitype">OptiType</a></li>
<li><a href="#pangolin">Pangolin</a></li>
<li><a href="#peddy">Peddy</a></li>
<li><a href="#phantompeakqualtools">phantompeakqualtools</a></li>
<li><a href="#picard">Picard</a></li>
<li><a href="#preseq">Preseq</a></li>
<li><a href="#prokka">Prokka</a></li>
<li><a href="#purple">PURPLE</a></li>
<li><a href="#qc3c">qc3C</a></li>
<li><a href="#qorts">QoRTs</a></li>
<li><a href="#qualimap">Qualimap</a></li>
<li><a href="#quast">QUAST</a></li>
<li><a href="#rna-seqc">RNA-SeQC</a></li>
<li><a href="#rockhopper">Rockhopper</a></li>
<li><a href="#rsem">RSEM</a></li>
<li><a href="#rseqc">RSeQC</a></li>
<li><a href="#sambamba">Sambamba</a>
<ul class="nav nav-stacked">
<li><a href="#markdup">markdup</a></li>
</ul>
</li>
<li><a href="#samblaster">Samblaster</a></li>
<li><a href="#samtools">Samtools</a>
<ul class="nav nav-stacked">
<li><a href="#idxstats">idxstats</a></li>
</ul>
</li>
<li><a href="#sargasso">Sargasso</a></li>
<li><a href="#sentieon">Sentieon</a></li>
<li><a href="#sex.deterrmine">Sex.DetErrMine</a></li>
<li><a href="#slamdunk">Slamdunk</a></li>
<li><a href="#snippy">Snippy</a></li>
<li><a href="#snpeff">SnpEff</a></li>
<li><a href="#snpsplit">SNPsplit</a></li>
<li><a href="#somalier">Somalier</a></li>
<li><a href="#stacks">Stacks</a></li>
<li><a href="#supernova">Supernova</a></li>
<li><a href="#theta2">THeTA2</a></li>
<li><a href="#varscan2">VarScan2</a></li>
<li><a href="#vcftools">VCFTools</a>
<ul class="nav nav-stacked">
<li><a href="#important-general-note">Important General Note</a></li>
<li><a href="#implemented">Implemented</a></li>
<li><a href="#to-do">To do</a></li>
</ul>
</li>
<li><a href="#vep">VEP</a></li>
<li><a href="#verifybamid">VerifyBAMID</a></li>
<li><a href="#whatshap">WhatsHap</a></li>
</ul>
</li>

</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<div class="docs_block modules_block" id="pre-alignment-section"><h1 id="pre-alignment"><a href="#pre-alignment" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Pre-alignment</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#adapter-removal">Adapter Removal</a></li>
<li><a href="#afterqc">AfterQC</a></li>
<li><a href="#bcl2fastq">bcl2fastq</a></li>
<li><a href="#bclconvert">bclconvert</a></li>
<li><a href="#biobloom-tools">BioBloom Tools</a></li>
<li><a href="#ccs">CCS</a></li>
<li><a href="#checkqc">CheckQC</a></li>
<li><a href="#clipandmerge">ClipAndMerge</a></li>
<li><a href="#cluster-flow">Cluster Flow</a></li>
<li><a href="#cutadapt">Cutadapt</a></li>
<li><a href="#fastp">Fastp</a></li>
<li><a href="#fastq-screen">FastQ Screen</a></li>
<li><a href="#fastqc">FastQC</a>
<ul class="nav nav-stacked">
<li><a href="#theoretical-gc-content">Theoretical GC Content</a></li>
<li><a href="#changing-the-order-of-sections">Changing the order of sections</a></li>
</ul>
</li>
<li><a href="#flash">FLASh</a></li>
<li><a href="#flexbar">Flexbar</a></li>
<li><a href="#interop">InterOp</a></li>
<li><a href="#ivar">iVar</a></li>
<li><a href="#jellyfish">Jellyfish</a></li>
<li><a href="#kat">KAT</a></li>
<li><a href="#leehom">leeHom</a></li>
<li><a href="#lima">Lima</a></li>
<li><a href="#minionqc">minionqc</a></li>
<li><a href="#pbmarkdup">pbmarkdup</a></li>
<li><a href="#pychopper">Pychopper</a></li>
<li><a href="#pycoqc">pycoQC</a></li>
<li><a href="#seqyclean">SeqyClean</a></li>
<li><a href="#sickle">Sickle</a></li>
<li><a href="#skewer">Skewer</a></li>
<li><a href="#sortmerna">SortMeRNA</a></li>
<li><a href="#trimmomatic">Trimmomatic</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div><h2 id="adapter-removal"><a href="#adapter-removal" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Adapter Removal</h2><p class="mod_ext_link"><a href="https://github.com/mikkelschubert/adapterremoval" target="_blank">https://github.com/mikkelschubert/adapterremoval</a></p><p>This program searches for and removes remnant adapter sequences from High-Throughput
Sequencing (HTS) data and (optionally) trims low quality bases from the 3' end of
reads following adapter removal. AdapterRemoval can analyze both single end and
paired end data, and can be used to merge overlapping paired-ended reads into
(longer) consensus sequences. Additionally, the AdapterRemoval may be used to
recover a consensus adapter sequence for paired-ended data, for which this
information is not available.</p>
<p>The adapterRemoval module parses <code>*.settings</code> logs generated by
<a href="https://github.com/mikkelschubert/adapterremoval">Adapter Removal</a>,
a tool for rapid adapter trimming, identification, and read merging.</p>
<p>supported setting file results:</p>
<ul>
<li><code>single end</code></li>
<li><code>paired end noncollapsed</code></li>
<li><code>paired end collapsed</code></li>
</ul>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Adapter Removal file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">adapterRemoval:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.settings'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">AdapterRemoval</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="afterqc"><a href="#afterqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>AfterQC</h2><p class="mod_ext_link"><a href="https://github.com/OpenGene/AfterQC" target="_blank">https://github.com/OpenGene/AfterQC</a></p><p>The AfterQC module parses results generated by
<a href="https://github.com/OpenGene/AfterQC">AfterQC</a>. AfterQC can 
simply go through all fastq files in a folder and then output three 
folders: good, bad and QC folders, which contains good reads, bad reads 
and the QC results of each fastq file/pair.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                AfterQC file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">afterqc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.json'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">allow_mismatch_in_poly</span>
</code></pre>
          </div><h2 id="bcl2fastq"><a href="#bcl2fastq" class="header-link"><span class="glyphicon glyphicon-link"></span></a>bcl2fastq</h2><p class="mod_ext_link"><a href="https://support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html" target="_blank">https://support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html</a></p><p>There are two versions of this software: bcl2fastq for MiSeq and HiSeq
sequencing systems running RTA versions earlier than 1.8, and bcl2fastq2 for
Illumina sequencing systems running RTA version 1.18.54 and above. This module
currently only covers output from the latter.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                bcl2fastq file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">bcl2fastq:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">Stats.json</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">DemuxResults</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">300</span>
</code></pre>
          </div><h2 id="bclconvert"><a href="#bclconvert" class="header-link"><span class="glyphicon glyphicon-link"></span></a>bclconvert</h2><p class="mod_ext_link"><a href="https://support.illumina.com/sequencing/sequencing_software/bcl-convert.html" target="_blank">https://support.illumina.com/sequencing/sequencing_software/bcl-convert.html</a></p><p>This BclConvert module is based on the bcl2fastq multiqc module. It can parse multiple
bclconvert run outputs as long as they are from the same sequencing run. When doing this,
the undetermined reads will be 'corrected' and re-calculated (as an unknown read from
some one bclcovnert run might not be truly unknown, but simply from another run).</p>
<h4 id="calculate-estimated-depth"><a href="#calculate-estimated-depth" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Calculate estimated depth</h4>
<p>You can specify a genome size in config</p>
<p>It's often useful to talk about sequencing yield in terms of estimated depth of coverage.
In order to make MultiQC show the estimated depth for each sample, specify the reference genome/target size in your <a href="http://multiqc.info/docs/#configuring-multiqc">MultiQC configuration</a>:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">bclconvert:</span>
<span class="hljs-attr">  genome_size:</span> <span class="hljs-number">3049315783</span></code></pre>
<p>The coverage depth will be estimated as the yield Q30 dvivided by the genome size.</p>
<p>MultiQC comes with effective genome size presets for Human and Mouse, so you can
provide the genome build name instead, like this: <code>genome_size: hg38_genome</code>. The
following values are supported: <code>hg19_genome</code>, <code>hg38_genome</code>, <code>mm10_genome</code>.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                bclconvert file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">bclconvert/runinfo:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">RunInfo.xml</span>
<span class="hljs-string">bclconvert/demux:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">Demultiplex_Stats.csv</span>
<span class="hljs-string">bclconvert/quality_metrics:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">Quality_Metrics.csv</span>
<span class="hljs-string">bclconvert/adaptermetrics:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">Adapter_Metrics.csv</span>
<span class="hljs-string">bclconvert/unknown_barcodes:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">Top_Unknown_Barcodes.csv</span>
</code></pre>
          </div><h2 id="biobloom-tools"><a href="#biobloom-tools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>BioBloom Tools</h2><p class="mod_ext_link"><a href="https://github.com/bcgsc/biobloom/" target="_blank">https://github.com/bcgsc/biobloom/</a></p><p><a href="https://github.com/bcgsc/biobloom/">BioBloom Tools</a> (BBT) provides the means
to create filters for a given reference and then to categorize sequences.
This methodology is faster than alignment but does not provide mapping locations.
BBT was initially intended to be used for pre-processing and QC applications
like contamination detection, but is flexible to accommodate other purposes.
This tool is intended to be a pipeline component to replace costly alignment steps.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                BioBloom Tools file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">biobloomtools:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    filter_id	hits	misses	shared	rate_hit	rate_miss	rate_shared
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
</code></pre>
          </div><h2 id="ccs"><a href="#ccs" class="header-link"><span class="glyphicon glyphicon-link"></span></a>CCS</h2><p class="mod_ext_link"><a href="https://github.com/PacificBiosciences/ccs" target="_blank">https://github.com/PacificBiosciences/ccs</a></p><p>The CCS module parses the report file generated by
<a href="https://github.com/PacificBiosciences/ccs">ccs</a>, a PacBio tool to generate
highly accurate single-molecule consensus reads (HiFi Reads).</p>
<p>CCS takes multiple subreads of the same SMRTbell molecule and combines them
using a statistical model to produce one highly accurate consensus sequence,
also called HiFi read, with base quality values. This tool powers the Circular
Consensus Sequencing workflow in SMRT Link.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                CCS file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">ccs/v4:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">ZMWs</span> <span class="hljs-string">generating</span> <span class="hljs-string">CCS</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">1024</span>
<span class="hljs-string">ccs/v5:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'"id": "ccs_processing"'</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.json'</span>
</code></pre>
          </div><h2 id="checkqc"><a href="#checkqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>CheckQC</h2><p class="mod_ext_link"><a href="https://github.com/Molmed/checkQC" target="_blank">https://github.com/Molmed/checkQC</a></p><p>The CheckQC module parses results generated by
<a href="https://checkqc.readthedocs.io/en/latest/">CheckQC</a>,
a program designed to check a set of quality criteria against
an Illumina runfolder.</p>
<p>It parses a CheckQC JSON file, so make sure to use CheckQC with the
<code>--json</code> flag and collect the stdout in a file.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                CheckQC file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">checkqc:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">instrument_and_reagent_type</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.json'</span>
</code></pre>
          </div><h2 id="clipandmerge"><a href="#clipandmerge" class="header-link"><span class="glyphicon glyphicon-link"></span></a>ClipAndMerge</h2><p class="mod_ext_link"><a href="http://www.github.com/apeltzer/ClipAndMerge" target="_blank">http://www.github.com/apeltzer/ClipAndMerge</a></p><p>An
 application to clip adapter sequences and merge reads in ancient DNA 
analysis. Note, that versions &lt; 1.7.8 use the basename of the file 
path to distinguish samples, whereas newer versions produce logfiles 
with a sample identifer that gets parsed by MultiQC.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                ClipAndMerge file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">clipandmerge:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">ClipAndMerge</span> <span class="hljs-string">(</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">5</span>
</code></pre>
          </div><h2 id="cluster-flow"><a href="#cluster-flow" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Cluster Flow</h2><p class="mod_ext_link"><a href="http://clusterflow.io/" target="_blank">http://clusterflow.io</a></p><p>Cluster Flow is a simple and flexible bioinformatics pipeline tool.
It's designed to be quick and easy to install, with flexible configuration
and simple customization.</p>
<p>Cluster Flow easy enough to set up and use for non-bioinformaticians
(given a basic knowledge of the command line), and it's simplicity
makes it great for low to medium throughput analyses.</p>
<p>The MultiQC module for Cluster Flow parses <code>*_clusterflow.txt</code> logs
and finds consensus commands executed by modules in each pipeline run.</p>
<p>The Cluster Flow <code>*.run</code> files are also parsed and pipeline information
shown (some basic statistics plus the pipeline steps / params used).</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Cluster Flow file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">clusterflow/logs:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_clusterFlow.txt'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">clusterflow/runfiles:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.run'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Cluster</span> <span class="hljs-string">Flow</span> <span class="hljs-string">Run</span> <span class="hljs-string">File</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
</code></pre>
          </div><h2 id="cutadapt"><a href="#cutadapt" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Cutadapt</h2><p class="mod_ext_link"><a href="https://cutadapt.readthedocs.io/" target="_blank">https://cutadapt.readthedocs.io/</a></p><p>The Cutadapt module parses results generated by
<a href="https://cutadapt.readthedocs.io/">Cutadapt</a>,
a tool to find and remove adapter sequences, primers, poly-A
tails and other types of unwanted sequence from your high-throughput
sequencing reads.</p>
<p>This module should be able to parse logs from a wide range of versions of Cutadapt.
It's been tested with log files from v1.2.1, 1.6 and 1.8. Note that you will need
to change the search pattern for very old log files (such as v.1.2) with the following
MultiQC config:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sp:</span>
<span class="hljs-attr">  cutadapt:</span>
<span class="hljs-attr">    contents:</span> <span class="hljs-string">"cutadapt version"</span></code></pre>
<p>See the <a href="http://multiqc.info/docs/#module-search-patterns">module search patterns</a>
section of the MultiQC documentation for more information.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Cutadapt file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">cutadapt:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">This</span> <span class="hljs-string">is</span> <span class="hljs-string">cutadapt</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="fastp"><a href="#fastp" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Fastp</h2><p class="mod_ext_link"><a href="https://github.com/OpenGene/fastp" target="_blank">https://github.com/OpenGene/fastp</a></p><p>The Fastp module parses results generated by
<a href="https://github.com/OpenGene/fastp">Fastp</a>. Fastp can simply 
go through all fastq files in a folder and perform a series of quality 
control and filtering. Quality control and reporting are displayed both 
before and after filtering, allowing for a clear depiction of the 
consequences of the filtering process. Notably, the latter can be 
conducted on a variety of paramaters including quality scores, length, 
as well as the presence of adapters, polyG, or polyX tailing.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Fastp file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">fastp:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*fastp.json'</span>
</code></pre>
          </div><h2 id="fastq-screen"><a href="#fastq-screen" class="header-link"><span class="glyphicon glyphicon-link"></span></a>FastQ Screen</h2><p class="mod_ext_link"><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/" target="_blank">http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/</a></p><p>The FastQ Screen module parses results generated by
<a href="http://www.bioinformatics.babraham.ac.uk/projects/fastq_screen/">FastQ Screen</a>,
a tool that allows you to screen a library of sequences in FastQ format
against a set of sequence databases so you can see if the composition of
the library matches with what you expect.</p>
<p>By default, the module creates a plot that emulates the FastQ Screen output
with blue and red stacked bars showing unique and multimapping read counts.
This plot only works for a handful of samples however, so if
<code># samples * # organisms &gt;= 160</code>, a simpler stacked barplot is shown. This
is also shown when generating flat-image plots.</p>
<p>To always show this style of plot, add the following line to a MultiQC config file:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">fastqscreen_simpleplot:</span> <span class="hljs-literal">true</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                FastQ Screen file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">fastq_screen:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_screen.txt'</span>
</code></pre>
          </div><h2 id="fastqc"><a href="#fastqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>FastQC</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#theoretical-gc-content">Theoretical GC Content</a></li>
<li><a href="#changing-the-order-of-sections">Changing the order of sections</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div><p class="mod_ext_link"><a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/" target="_blank">http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></p><p>The FastQC module parses results generated by
<a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>,
a quality control tool for high throughput sequence data written
by Simon Andrews at the Babraham Institute.</p>
<p>FastQC generates a HTML report which is what most people use when
they run the program. However, it also helpfully generates a file
called <code>fastqc_data.txt</code> which is relatively easy to parse.</p>
<p>A typical run will produce the following files:</p>
<pre><code class="language-txt">mysample_fastqc.html
mysample_fastqc/
  Icons/
  Images/
  fastqc.fo
  fastqc_data.txt
  fastqc_report.html
  summary.txt</code></pre>
<p>Sometimes the directory is zipped, with just <code>mysample_fastqc.zip</code>.</p>
<p>The FastQC MultiQC module looks for files called <code>fastqc_data.txt</code>
or ending in <code>_fastqc.zip</code>. If the zip files are found, they are
read in memory and <code>fastqc_data.txt</code> parsed.</p>
<blockquote>
<p><strong>Note:</strong> The directory and zip file are often both present. To speed
up MultiQC execution, zip files will be skipped if the file name suggests
that they will share a sample name with data that has already been parsed.</p>
</blockquote>
<p>You can customise the patterns used for finding these files in your
MultiQC config (see <a href="#module-search-patterns">Module search patterns</a>).
The below code shows the default file patterns:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sp:</span>
  <span class="hljs-string">fastqc/data:</span>
<span class="hljs-attr">    fn:</span> <span class="hljs-string">"fastqc_data.txt"</span>
  <span class="hljs-string">fastqc/zip:</span>
<span class="hljs-attr">    fn:</span> <span class="hljs-string">"*_fastqc.zip"</span></code></pre>
<blockquote>
<p><strong>Note:</strong> Sample names are discovered by parsing the line beginning
<code>Filename</code> in <code>fastqc_data.txt</code>, <em>not</em> based on the FastQC report names.</p>
</blockquote>
<h3 id="theoretical-gc-content"><a href="#theoretical-gc-content" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Theoretical GC Content</h3>
<p>It is possible to plot a dashed line showing the theoretical GC content for a
reference genome. MultiQC comes with genome and transcriptome guides for Human
and Mouse. You can use these in your reports by adding the following MultiQC
config keys (see <a href="http://multiqc.info/docs/#configuring-multiqc">Configuring MultiQC</a>):</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">fastqc_config:</span>
<span class="hljs-attr">  fastqc_theoretical_gc:</span> <span class="hljs-string">"hg38_genome"</span></code></pre>
<p>Only one theoretical distribution can be plotted.
The following guides are available: <em>(txome = transcriptome)</em></p>
<ul>
<li><code>hg38_genome</code></li>
<li><code>hg38_txome</code></li>
<li><code>mm10_genome</code></li>
<li><code>mm10_txome</code></li>
</ul>
<p>Alternatively, a custom theoretical guide can be used in reports. To do this,
create a file with <code>fastqc_theoretical_gc</code> in the filename and place it with your
analysis files. It should be tab delimited with the following format (column 1 = %GC,
column 2 = % of genome):</p>
<pre><code class="language-bash hljs"><span class="hljs-comment"># FastQC theoretical GC content curve: YOUR REFERENCE NAME</span>
0   0.005311768
1   0.004108502
2   0.004060371
3   0.005066476
[...]</code></pre>
<p>You can generate these files using an R package called
<a href="https://github.com/mikelove/fastqcTheoreticalGC">fastqcTheoreticalGC</a>
written by <a href="https://github.com/mikelove">Mike Love</a>.
Please see the <a href="https://github.com/mikelove/fastqcTheoreticalGC">package readme</a>
for more details.</p>
<p>Result files from this package are searched for with the following search pattern
(can be customised as described above):</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sp:</span>
  <span class="hljs-string">fastqc/theoretical_gc:</span>
<span class="hljs-attr">    fn:</span> <span class="hljs-string">"*fastqc_theoretical_gc*"</span></code></pre>
<p>If you want to always use a specific custom file for MultiQC reports without having to
add it to the analysis directory, add the full file path to the same MultiQC config
variable described above:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">fastqc_config:</span>
<span class="hljs-attr">  fastqc_theoretical_gc:</span> <span class="hljs-string">"/path/to/your/custom_fastqc_theoretical_gc.txt"</span></code></pre>
<h3 id="changing-the-order-of-sections"><a href="#changing-the-order-of-sections" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Changing the order of sections</h3>
<p>Remember that it is possible to customise the order in which the different module sections appear
in the report if you wish.
See <a href="https://multiqc.info/docs/#order-of-module-and-module-subsection-output">the docs</a> for more information.</p>
<p>For example, to show the <em>Status Checks</em> section at the top, use the following config:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">report_section_order:</span>
<span class="hljs-attr">  fastqc_status_checks:</span>
<span class="hljs-attr">    order:</span> <span class="hljs-bullet">-1000</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                FastQC file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">fastqc/data:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">fastqc_data.txt</span>
<span class="hljs-string">fastqc/zip:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_fastqc.zip'</span>
<span class="hljs-string">fastqc/theoretical_gc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*fastqc_theoretical_gc*'</span>
</code></pre>
          </div><h2 id="flash"><a href="#flash" class="header-link"><span class="glyphicon glyphicon-link"></span></a>FLASh</h2><p class="mod_ext_link"><a href="https://ccb.jhu.edu/software/FLASH/" target="_blank">https://ccb.jhu.edu/software/FLASH/</a></p><p>The FLASh module parses the log messages generated by the <a href="https://ccb.jhu.edu/software/FLASH/">FLASh</a> read merger. To create a log file, you can use <code>tee</code>. From the FLASh help:</p>
<pre><code class="language-bash hljs">flash reads_1.fq reads_2.fq 2&gt;&amp;1 | tee logfilename.log</code></pre>
<p>The sample name is set by the first input filename listed in the log.
 However, this can be changed to using the first output filename (i.e. 
if you used FLASh's <code>--output-prefix=PREFIX</code> option) by using the following config:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">flash:</span>
<span class="hljs-attr">  use_output_name:</span> <span class="hljs-literal">true</span></code></pre>
<p>The module can also parse the <code>.hist</code> numeric histograms output by FLASh.</p>
<p>Note that the histogram's file format and extension are too generic 
by themselves which could result in the accidental parsing a file output
 by another tool. To get around this, the MultiQC module only parses 
files with the filename pattern <code>*flash*.hist</code>.</p>
<p>To customise this (for example, enabling for any file ending in <code>*.hist</code>), use the following config change:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sp:</span>
  <span class="hljs-string">flash/hist:</span>
<span class="hljs-attr">    fn:</span> <span class="hljs-string">"*.hist"</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                FLASh file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">flash/log:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'[FLASH]'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">flash/hist:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*flash*.hist'</span>
</code></pre>
          </div><h2 id="flexbar"><a href="#flexbar" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Flexbar</h2><p class="mod_ext_link"><a href="https://github.com/seqan/flexbar" target="_blank">https://github.com/seqan/flexbar</a></p><p><a href="https://github.com/seqan/flexbar">Flexbar</a> preprocesses high-throughput sequencing data
efficiently. It demultiplexes barcoded runs and removes adapter sequences.
Moreover, trimming and filtering features are provided.
Flexbar increases read mapping rates and improves genome as well as transcriptome assemblies.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Flexbar file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">flexbar:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'Flexbar - flexible barcode and adapter removal'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="interop"><a href="#interop" class="header-link"><span class="glyphicon glyphicon-link"></span></a>InterOp</h2><p class="mod_ext_link"><a href="http://illumina.github.io/interop/index.html" target="_blank">http://illumina.github.io/interop/index.html</a></p><p>This module parses the output from the InterOp Summary executable and creates a table view. The aim is to replicate the <code>Run &amp; Lane Metrics</code> table from the <a href="https://basespace.illumina.com/">Illumina Basespace</a> interface. The executable used can easily be installed from the BioConda channel using <code>conda install -c bioconda illumina-interop</code>.</p>
<p>The MultiQC interop module can parse the outputs of the <code>interop_summary</code> and <code>interop_index-summary</code> executables. Note that these must be run with the <code>--csv=1</code> option.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                InterOp file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">interop/summary:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'Level,Yield,Projected Yield,Aligned,Error Rate,Intensity C1,%&gt;=Q30'</span>
<span class="hljs-string">interop/index-summary:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'Total Reads,PF Reads,% Read Identified (PF),CV,Min,Max'</span>
</code></pre>
          </div><h2 id="ivar"><a href="#ivar" class="header-link"><span class="glyphicon glyphicon-link"></span></a>iVar</h2><p class="mod_ext_link"><a href="https://github.com/andersen-lab/ivar" target="_blank">https://github.com/andersen-lab/ivar</a></p><p>iVar is a computational package that contains functions broadly useful for viral amplicon-based sequencing.</p>
<p>This module parses the output from the <code>ivar trim</code> command and creates a table view.
Both output from V1 and V2 of the tool are supported and parsed accordingly.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                iVar file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">ivar/trim:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Number</span> <span class="hljs-string">of</span> <span class="hljs-string">references</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">8</span>
</code></pre>
          </div><h2 id="jellyfish"><a href="#jellyfish" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Jellyfish</h2><p class="mod_ext_link"><a href="http://www.cbcb.umd.edu/software/jellyfish/" target="_blank">http://www.cbcb.umd.edu/software/jellyfish/</a></p><p>JELLYFISH
 is a tool for fast, memory-efficient counting of k-mers in DNA. A k-mer
 is a substring of length k, and counting the occurrences of all such 
substrings is a central step in many analyses of DNA sequence. JELLYFISH
 can count k-mers using an order of magnitude less memory and an order 
of magnitude faster than other k-mer counting packages by using an 
efficient encoding of a hash table and by exploiting the 
"compare-and-swap" CPU instruction to increase parallelism.</p>
<p>The MultiQC module for Jellyfish parses <em>only</em> <code>*_jf.hist</code> files. The general usage of jellyfish to be parsed by MultiQC module needs to be:</p>
<ul>
<li><code>gunzip -c file.fastq.gz | jellyfish count -o file.jf -m ...</code></li>
<li><code>jellyfish histo -o file_jf.hist -f file.jf</code></li>
</ul>
<p>In case a user wants to customise the matching pattern for jellyfish, then multiqc can be run with the option <code>--cl_config "sp: { jellyfish: { fn: 'PATTERN' } }"</code> where <code>PATTERN</code> is the pattern to be matched. For example:</p>
<pre><code class="language-bash hljs">multiqc . --cl_config <span class="hljs-string">"sp: { jellyfish: { fn: '*.hist' } }"</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Jellyfish file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">jellyfish:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_jf.hist'</span>
</code></pre>
          </div><h2 id="kat"><a href="#kat" class="header-link"><span class="glyphicon glyphicon-link"></span></a>KAT</h2><p class="mod_ext_link"><a href="https://github.com/TGAC/KAT" target="_blank">https://github.com/TGAC/KAT</a></p><p>The
 KAT multiqc module interprets output from KAT distribution analysis 
json files, which typically contain information such as estimated genome
 size and heterozygosity rates from your k-mer spectra.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                KAT file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">kat:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.dist_analysis.json'</span>
</code></pre>
          </div><h2 id="leehom"><a href="#leehom" class="header-link"><span class="glyphicon glyphicon-link"></span></a>leeHom</h2><p class="mod_ext_link"><a href="https://github.com/grenaud/leeHom" target="_blank">https://github.com/grenaud/leeHom</a></p><p>leeHom is a Bayesian maximum <em>a posteriori</em> algorithm for stripping
sequencing adapters and merging overlapping portions of reads.
The algorithm is mostly aimed at ancient DNA and Illumina data but
can be used for any dataset.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                leeHom file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">leehom:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Adapter</span> <span class="hljs-string">dimers/chimeras</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="lima"><a href="#lima" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Lima</h2><p class="mod_ext_link"><a href="https://github.com/PacificBiosciences/barcoding" target="_blank">https://github.com/PacificBiosciences/barcoding</a></p><p>The Lima module parses the report and count files generated by
<a href="https://github.com/PacificBiosciences/barcoding">Lima</a>, a PacBio tool to
demultiplex PacBio single-molecule sequencing data.</p>
<p>By default, Lima will use <code>barcode1--barcode2</code> as the sample names. To prevent
these barcodes showing up in the General Statistics table of MultiQC, the Lima
results are added to their own section.</p>
<p>If you want to include the Lima results in the General Statistics table, you
can rename the <code>barcode1--barcode2</code> filenames to their apropriate samples using
the <a href="https://multiqc.info/docs/#sample-name-replacement">--replace-names</a>
option. Each sample that is specified in this way will be moved from the Lima
section to the General Statistics table.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Lima file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">lima/summary:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">ZMWs</span> <span class="hljs-string">above</span> <span class="hljs-string">all</span> <span class="hljs-string">thresholds</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">1024</span>
<span class="hljs-string">lima/counts:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    IdxFirst	IdxCombined	IdxFirstNamed	IdxCombinedNamed	Counts	MeanScore
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="minionqc"><a href="#minionqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>minionqc</h2><p class="mod_ext_link"><a href="https://github.com/roblanf/minion_qc" target="_blank">https://github.com/roblanf/minion_qc</a></p><p>The MinIONQC module parses results generated by <a href="https://github.com/roblanf/minion_qc">MinIONQC</a>.
It uses the <code>sequencing_summary.txt</code> files produced by ONT (Oxford Nanopore Technologies) long-read
base-callers to perform QC on the reads.
It allows quick-and-easy comparison of data from multiple flowcells</p>
<p>The MultiQC module parses data in the <code>summary.yaml</code> MinIONQC output files.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                minionqc file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">minionqc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">summary.yaml</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">total.gigabases</span>
</code></pre>
          </div><h2 id="pbmarkdup"><a href="#pbmarkdup" class="header-link"><span class="glyphicon glyphicon-link"></span></a>pbmarkdup</h2><p class="mod_ext_link"><a href="https://github.com/PacificBiosciences/pbmarkdup" target="_blank">https://github.com/PacificBiosciences/pbmarkdup</a></p><p>The pbmarkdup module parses results generated by
<a href="https://github.com/PacificBiosciences/pbmarkdup">pbmarkdup</a>, and adds the
<strong>% Unique Molecules</strong> and <strong>%Duplicate Reads</strong> (hidden) to the General Statistics
table.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                pbmarkdup file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">pbmarkdup:</span>
<span class="hljs-attr">  contents_re:</span> <span class="hljs-string">&gt;
    LIBRARY +READS +UNIQUE MOLECULES
    +DUPLICATE READS
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">5</span>
</code></pre>
          </div><h2 id="pychopper"><a href="#pychopper" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Pychopper</h2><p class="mod_ext_link"><a href="https://github.com/nanoporetech/pychopper" target="_blank">https://github.com/nanoporetech/pychopper</a></p><p>The MultiQC module parses the pychopper stats file. Pychopper needs to be run with the <code>-S stats_output</code> option to create the file. The name of the output file defines the sample name.</p>
<p>The stats file is a three column <code>tsv</code> file with the format <code>category name value</code>.</p>
<p>Currently only two stats are displayed in MultiQC. Two bargraphs are 
created for the read classication and the strand orientation of the 
identified full length transcripts. Additional stats could be included 
on further request.</p>
<p>The general stats table contains a value that displays the percentage
 of full length transcripts. This value is calculated from the 
cumulative length of reads where Pychopper found primers at both ends.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Pychopper file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">pychopper:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Classification</span>	<span class="hljs-string">Rescue</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">6</span>
</code></pre>
          </div><h2 id="pycoqc"><a href="#pycoqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>pycoQC</h2><p class="mod_ext_link"><a href="https://a-slide.github.io/pycoQC/" target="_blank">https://a-slide.github.io/pycoQC/</a></p><p>PycoQC relies on the <code>sequencing_summary.txt</code> file generated by Albacore and Guppy,
but if needed it can also generate a summary file from basecalled <code>fast5</code> files.</p>
<p>The package supports 1D and 1D2 runs generated with MinION, GridION and PromethION
devices and basecalled with Albacore 1.2.1+ or Guppy 2.1.3+.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                pycoQC file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">pycoqc:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'"pycoqc":'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
</code></pre>
          </div><h2 id="seqyclean"><a href="#seqyclean" class="header-link"><span class="glyphicon glyphicon-link"></span></a>SeqyClean</h2><p class="mod_ext_link"><a href="https://github.com/ibest/seqyclean" target="_blank">https://github.com/ibest/seqyclean</a></p><p>The SeqyClean module will visualize the results from a <a href="https://github.com/ibest/seqyclean">SeqyClean</a>,
 a comprehensive preprocessing software pipeline. SeqyClean removes 
noise from Fastq files to improve de-novo genome assembly and genome 
mapping.</p>
<p>The module parses the <code>*SummaryStatistics.tsv</code> files that results from a SeqyClean cleaning.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                SeqyClean file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">seqyclean:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_SummaryStatistics.tsv'</span>
</code></pre>
          </div><h2 id="sickle"><a href="#sickle" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Sickle</h2><p class="mod_ext_link"><a href="https://github.com/najoshi/sickle" target="_blank">https://github.com/najoshi/sickle</a></p><p>The Sickle module parses standard error generated by
<a href="https://github.com/najoshi/sickle">Sickle</a>, a
windowed Adaptive Trimming for fastq files using quality. StdOut can be captured by
directing it to a file e.g. <code>sickle command 2&gt; sickle_out.log</code></p>
<p>The module generates the sample names based on the filenames.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Sickle file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">sickle:</span>
<span class="hljs-attr">  contents_re:</span> <span class="hljs-string">'FastQ \w*\s?records kept: .*'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
</code></pre>
          </div><h2 id="skewer"><a href="#skewer" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Skewer</h2><p class="mod_ext_link"><a href="https://github.com/relipmoc/skewer" target="_blank">https://github.com/relipmoc/skewer</a></p><p>The Skewer module parses results generated by
<a href="https://github.com/relipmoc/skewer">Skewer</a>,
an adapter trimming tool specially designed for processing
next-generation sequencing (NGS) paired-end sequences.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Skewer file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">skewer:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'maximum error ratio allowed (-r):'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="sortmerna"><a href="#sortmerna" class="header-link"><span class="glyphicon glyphicon-link"></span></a>SortMeRNA</h2><p class="mod_ext_link"><a href="http://bioinfo.lifl.fr/RNA/sortmerna/" target="_blank">http://bioinfo.lifl.fr/RNA/sortmerna/</a></p><p>SortMeRNA
 is a program tool for filtering, mapping and OTU-picking NGS reads in 
metatranscriptomic and metagenomic data. The core algorithm is based on 
approximate seeds and allows for fast and sensitive analyses of 
nucleotide sequences. The main application of SortMeRNA is filtering 
ribosomal RNA from metatranscriptomic data.</p>
<p>The MultiQC module parses the log files, which are created when <code>SortMeRNA</code> is run with the <code>--log</code> option.</p>
<p>The default header in the 'General Statistics' table is '% rRNA'. Users can override this using the configuration option:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sortmerna:</span>
<span class="hljs-attr">  tab_header:</span> <span class="hljs-string">"My database hits"</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                SortMeRNA file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">sortmerna:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Minimal</span> <span class="hljs-string">SW</span> <span class="hljs-string">score</span> <span class="hljs-string">based</span> <span class="hljs-string">on</span> <span class="hljs-string">E-value</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="trimmomatic"><a href="#trimmomatic" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Trimmomatic</h2><p class="mod_ext_link"><a href="http://www.usadellab.org/cms/?page=trimmomatic" target="_blank">http://www.usadellab.org/cms/?page=trimmomatic</a></p><p>The Trimmomatic module parses standard error generated by
<a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a>,
a flexible read trimming tool for Illumina NGS data. StdErr can be captured by
directing it to a file e.g. <code>trimmomatic command 2&gt; trim_out.log</code></p>
<p>By default, the module generates the sample names based on the command line used
by Trimmomatic. If you prefer, you can tell the module to use the filenames as
sample names instead. To do so, use the following config option:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">trimmomatic:</span>
<span class="hljs-attr">  s_name_filenames:</span> <span class="hljs-literal">true</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Trimmomatic file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">trimmomatic:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Trimmomatic</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div></div><div class="docs_block modules_block" id="aligners-section"><h1 id="aligners"><a href="#aligners" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Aligners</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#bbmap">BBMap</a></li>
<li><a href="#biscuit">BISCUIT</a>
<ul class="nav nav-stacked">
<li><a href="#insert-size-distribution">Insert Size Distribution</a></li>
</ul>
</li>
<li><a href="#bismark">Bismark</a></li>
<li><a href="#bowtie-1">Bowtie 1</a></li>
<li><a href="#bowtie-2">Bowtie 2</a></li>
<li><a href="#dragen">DRAGEN</a></li>
<li><a href="#hic-pro">HiC-Pro</a></li>
<li><a href="#hicup">HiCUP</a></li>
<li><a href="#hisat2">HISAT2</a></li>
<li><a href="#kallisto">Kallisto</a></li>
<li><a href="#longranger">Longranger</a></li>
<li><a href="#malt">MALT</a></li>
<li><a href="#salmon">Salmon</a></li>
<li><a href="#star">STAR</a></li>
<li><a href="#tophat">TopHat</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div><h2 id="bbmap"><a href="#bbmap" class="header-link"><span class="glyphicon glyphicon-link"></span></a>BBMap</h2><p class="mod_ext_link"><a href="http://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/" target="_blank">http://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/</a></p><p>The BBMap module produces summary statistics from the
<a href="http://jgi.doe.gov/data-and-tools/bbtools/bb-tools-user-guide/">BBMap</a> suite of tools.
The module can summarise data from the following BBMap output files
(descriptions from command line help output):</p>
<ul>
<li><code>stats</code>
<ul>
<li>BBDuk filtering statistics.</li>
</ul></li>
<li><code>covstats</code> <em>(not yet implemented)</em>
<ul>
<li>Per-scaffold coverage info.</li>
</ul></li>
<li><code>rpkm</code> <em>(not yet implemented)</em>
<ul>
<li>Per-scaffold RPKM/FPKM counts.</li>
</ul></li>
<li><code>covhist</code>
<ul>
<li>Histogram of # occurrences of each depth level.</li>
</ul></li>
<li><code>basecov</code> <em>(not yet implemented)</em>
<ul>
<li>Coverage per base location.</li>
</ul></li>
<li><code>bincov</code> <em>(not yet implemented)</em>
<ul>
<li>Print binned coverage per location (one line per X bases).</li>
</ul></li>
<li><code>scafstats</code> <em>(not yet implemented)</em>
<ul>
<li>Statistics on how many reads mapped to which scaffold.</li>
</ul></li>
<li><code>refstats</code>
<ul>
<li>Statistics on how many reads mapped to which reference file; only for BBSplit.</li>
</ul></li>
<li><code>bhist</code>
<ul>
<li>Base composition histogram by position.</li>
</ul></li>
<li><code>qhist</code>
<ul>
<li>Quality histogram by position.</li>
</ul></li>
<li><code>qchist</code>
<ul>
<li>Count of bases with each quality value.</li>
</ul></li>
<li><code>aqhist</code>
<ul>
<li>Histogram of average read quality.</li>
</ul></li>
<li><code>bqhist</code>
<ul>
<li>Quality histogram designed for box plots.</li>
</ul></li>
<li><code>lhist</code>
<ul>
<li>Read length histogram.</li>
</ul></li>
<li><code>gchist</code>
<ul>
<li>Read GC content histogram.</li>
</ul></li>
<li><code>indelhist</code>
<ul>
<li>Indel length histogram.</li>
</ul></li>
<li><code>mhist</code>
<ul>
<li>Histogram of match, sub, del, and ins rates by read location.</li>
</ul></li>
<li><code>statsfile</code> <em>(not yet implemented)</em>
<ul>
<li>Mapping statistics are printed here.</li>
</ul></li>
</ul>
<p>Additional information on the BBMap tools is available on
<a href="http://seqanswers.com/forums/showthread.php?t=41057">SeqAnswers</a>.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                BBMap file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">bbmap/stats:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#Name	Reads	ReadsPct'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">4</span>
<span class="hljs-string">bbmap/aqhist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#Quality	count1	fraction1	count2	fraction2'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/bhist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#Pos	A	C	G	T	N'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/bincov:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#RefName	Cov	Pos	RunningPos'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
<span class="hljs-string">bbmap/bqhist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#BaseNum	count_1	min_1	max_1	mean_1	Q1_1	med_1	Q3_1	LW_1	RW_1	count_2	min_2	max_2	mean_2	Q1_2	med_2	Q3_2	LW_2	RW_2'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/covhist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#Coverage	numBases'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/covstats:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#ID	Avg_fold'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/ehist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#Errors	Count'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/gchist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#GC	Count'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">5</span>
<span class="hljs-string">bbmap/idhist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#Mean_reads'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/ihist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#InsertSize	Count'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">6</span>
<span class="hljs-string">bbmap/indelhist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#Length	Deletions	Insertions'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/lhist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#Length	Count'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/mhist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#BaseNum	Match1	Sub1	Del1	Ins1	N1	Other1	Match2	Sub2	Del2	Ins2	N2	Other2'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/qahist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#Deviation'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/qchist:</span>
<span class="hljs-attr">  contents_re:</span> <span class="hljs-string">|
    #Quality	count1	fraction1
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/qhist:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#BaseNum	Read1_linear	Read1_log	Read1_measured	Read2_linear	Read2_log	Read2_measured'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/rpkm:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#File	'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/statsfile_machine:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Reads</span> <span class="hljs-string">Used=</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">bbmap/statsfile:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'Reads Used:'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="biscuit"><a href="#biscuit" class="header-link"><span class="glyphicon glyphicon-link"></span></a>BISCUIT</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#insert-size-distribution">Insert Size Distribution</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div><p class="mod_ext_link"><a href="https://github.com/huishenlab/biscuit" target="_blank">https://github.com/huishenlab/biscuit</a></p><p>The BISCUIT module parses logs generated by
<a href="https://github.com/huishenlab/biscuit">BISCUIT</a> and the quality control script,
QC.sh, included with the BISCUIT software.</p>
<p>Note - as of MultiQC v1.9, the module supports only BISCUIT version v0.3.16 and onwards.
If you have BISCUIT data from before this, please use MultiQC v1.8.</p>
<h3 id="insert-size-distribution"><a href="#insert-size-distribution" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Insert Size Distribution</h3>
<p>The second tab of this plot uses the config option <code>read_count_multiplier</code>,
so if millions of reads is not useful for your data you can customise this.</p>
<p>See <a href="https://multiqc.info/docs/#number-base-multiplier">Number base (multiplier)</a>
in the documentation.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                BISCUIT file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">biscuit/align_mapq:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_mapq_table.txt'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">BISCUITqc</span> <span class="hljs-string">Mapping</span> <span class="hljs-string">Quality</span> <span class="hljs-string">Table</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
<span class="hljs-string">biscuit/align_strand:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_strand_table.txt'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">BISCUITqc</span> <span class="hljs-string">Strand</span> <span class="hljs-string">Table</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
<span class="hljs-string">biscuit/align_isize:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_isize_table.txt'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">BISCUITqc</span> <span class="hljs-string">Insert</span> <span class="hljs-string">Size</span> <span class="hljs-string">Table</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
<span class="hljs-string">biscuit/dup_report:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_dup_report.txt'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">BISCUITqc</span> <span class="hljs-string">Read</span> <span class="hljs-string">Duplication</span> <span class="hljs-string">Table</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
<span class="hljs-string">biscuit/qc_cv:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_cv_table.txt'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">BISCUITqc</span> <span class="hljs-string">Uniformity</span> <span class="hljs-string">Table</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
<span class="hljs-string">biscuit/covdist_all_base_botgc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_all_base_botgc_table.txt'</span>
<span class="hljs-string">biscuit/covdist_all_base:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_all_base_table.txt'</span>
<span class="hljs-string">biscuit/covdist_all_base_topgc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_all_base_topgc_table.txt'</span>
<span class="hljs-string">biscuit/covdist_q40_base_botgc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_q40_base_botgc_table.txt'</span>
<span class="hljs-string">biscuit/covdist_q40_base:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_q40_base_table.txt'</span>
<span class="hljs-string">biscuit/covdist_q40_base_topgc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_q40_base_topgc_table.txt'</span>
<span class="hljs-string">biscuit/covdist_all_cpg_botgc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_all_cpg_botgc_table.txt'</span>
<span class="hljs-string">biscuit/covdist_all_cpg:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_all_cpg_table.txt'</span>
<span class="hljs-string">biscuit/covdist_all_cpg_topgc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_all_cpg_topgc_table.txt'</span>
<span class="hljs-string">biscuit/covdist_q40_cpg_botgc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_q40_cpg_botgc_table.txt'</span>
<span class="hljs-string">biscuit/covdist_q40_cpg:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_q40_cpg_table.txt'</span>
<span class="hljs-string">biscuit/covdist_q40_cpg_topgc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_covdist_q40_cpg_topgc_table.txt'</span>
<span class="hljs-string">biscuit/cpg_retention_readpos:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_CpGRetentionByReadPos.txt'</span>
<span class="hljs-string">biscuit/cph_retention_readpos:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_CpHRetentionByReadPos.txt'</span>
<span class="hljs-string">biscuit/base_avg_retention_rate:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_totalBaseConversionRate.txt'</span>
<span class="hljs-string">biscuit/read_avg_retention_rate:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_totalReadConversionRate.txt'</span>
</code></pre>
          </div><h2 id="bismark"><a href="#bismark" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bismark</h2><p class="mod_ext_link"><a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/" target="_blank">http://www.bioinformatics.babraham.ac.uk/projects/bismark/</a></p><p>The Bismark module parses logs generated by
<a href="http://www.bioinformatics.babraham.ac.uk/projects/bismark/">Bismark</a>,
a tool to map bisulfite converted sequence reads and determine
cytosine methylation states.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Bismark file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">bismark/align:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_[SP]E_report.txt'</span>
<span class="hljs-string">bismark/dedup:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.deduplication_report.txt'</span>
<span class="hljs-string">bismark/meth_extract:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_splitting_report.txt'</span>
<span class="hljs-string">bismark/m_bias:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*M-bias.txt'</span>
<span class="hljs-string">bismark/bam2nuc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.nucleotide_stats.txt'</span>
</code></pre>
          </div><h2 id="bowtie-1"><a href="#bowtie-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bowtie 1</h2><p class="mod_ext_link"><a href="http://bowtie-bio.sourceforge.net/" target="_blank">http://bowtie-bio.sourceforge.net/</a></p><p>The Bowtie 1 module parses results generated by
<a href="http://bowtie-bio.sourceforge.net/">Bowtie</a>,
an ultrafast, memory-efficient short read aligner.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Bowtie 1 file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">bowtie1:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'# reads processed:'</span>
<span class="hljs-attr">  exclude_fn:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">bowtie.left_kept_reads.log</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">bowtie.left_kept_reads.m2g_um.log</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">bowtie.left_kept_reads.m2g_um_seg1.log</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">bowtie.left_kept_reads.m2g_um_seg2.log</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">bowtie.right_kept_reads.log</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">bowtie.right_kept_reads.m2g_um.log</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">bowtie.right_kept_reads.m2g_um_seg1.log</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">bowtie.right_kept_reads.m2g_um_seg2.log</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="bowtie-2"><a href="#bowtie-2" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bowtie 2</h2><p class="mod_ext_link"><a href="http://bowtie-bio.sourceforge.net/bowtie2/" target="_blank">http://bowtie-bio.sourceforge.net/bowtie2/</a></p><p>The Bowtie 2 module parses results generated by
<a href="http://bowtie-bio.sourceforge.net/bowtie2/">Bowtie 2</a>,
an ultrafast and memory-efficient tool for aligning sequencing
reads to long reference sequences.</p>
<p>Please note that the Bowtie 2 logs are difficult to parse as they don't contain
much extra information (such as what the input data was). A typical log looks like this:</p>
<pre><code class="language-txt">314537 reads; of these:
  314537 (100.00%) were paired; of these:
    111016 (35.30%) aligned concordantly 0 times
    193300 (61.46%) aligned concordantly exactly 1 time
    10221 (3.25%) aligned concordantly &gt;1 times
    ----
    111016 pairs aligned concordantly 0 times; of these:
      11377 (10.25%) aligned discordantly 1 time
    ----
    99639 pairs aligned 0 times concordantly or discordantly; of these:
      199278 mates make up the pairs; of these:
        112779 (56.59%) aligned 0 times
        85802 (43.06%) aligned exactly 1 time
        697 (0.35%) aligned &gt;1 times
82.07% overall alignment rate</code></pre>
<p>Bowtie 2 logs are from <code>STDERR</code> - some pipelines (such as <a href="http://clusterflow.io/">Cluster Flow</a>)
print the Bowtie 2 command before this, so MultiQC looks to see if this can be recognised in the same
file. If not, it takes the filename as the sample name.</p>
<p>Bowtie 2 is used by other tools too, so if your log file contains the word <code>bisulfite</code>, MultiQC will
assume that this is actually Bismark and ignore the Bowtie 2 logs.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Bowtie 2 file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">bowtie2:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'reads; of these:'</span>
<span class="hljs-attr">  exclude_contents:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">bisulfite</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">HiC-Pro</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="dragen"><a href="#dragen" class="header-link"><span class="glyphicon glyphicon-link"></span></a>DRAGEN</h2><p class="mod_ext_link"><a href="https://www.illumina.com/products/by-type/informatics-products/dragen-bio-it-platform.html" target="_blank">https://www.illumina.com/products/by-type/informatics-products/dragen-bio-it-platform.html</a></p><p><a href="https://www.illumina.com/products/by-type/informatics-products/dragen-bio-it-platform.html">Illumina DRAGEN</a>
is a Bio-IT Platform that provides ultra-rapid secondary analysis of sequencing data using field-programmable
gate array technology (FPGA).</p>
<p>DRAGEN has a number of different pipelines and outputs, including base calling, DNA and RNA alignment,
post-alignment processing and variant calling, covering virtually all stages of typical NGS data processing.
For each stage, it generates QC files with metrics resembling those of samtools-stats, mosdepth, bcftools-stats
and alike. This MultiQC module supports some of the output but not all.</p>
<ul>
<li><code>&lt;output prefix&gt;.wgs_fine_hist_&lt;tumor|normal&gt;.csv</code>
<ul>
<li>Coverage distribution and cumulative coverage plots</li>
</ul></li>
<li><code>&lt;output prefix&gt;.mapping_metrics.csv</code>
<ul>
<li>General stats table, a dedicated table, and a few barplots</li>
</ul></li>
<li><code>&lt;output prefix&gt;.wgs_coverage_metrics_&lt;tumor|normal&gt;.csv</code>
<ul>
<li>General stats table and a dedicated table</li>
</ul></li>
<li><code>&lt;output prefix&gt;.wgs_contig_mean_cov_&lt;tumor|normal&gt;.csv</code>
<ul>
<li>A histogram like in mosdepth, with each chrom as a category on X axis, plus a category for autosomal chromosomes average</li>
</ul></li>
<li><code>&lt;output prefix&gt;.fragment_length_hist.csv</code>
<ul>
<li>A histogram plot</li>
</ul></li>
<li><code>&lt;output prefix&gt;.ploidy_estimation_metrics.csv</code>
<ul>
<li>Add just Ploidy estimation into the general stats table</li>
</ul></li>
<li><code>&lt;output prefix&gt;.vc_metrics.csv</code>
<ul>
<li>A dedicated table and the total number of Variants into the general stats table</li>
</ul></li>
</ul>
<p>Each QC output adds a section into the report if a corresponding QC file is found.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                DRAGEN file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">dragen/vc_metrics:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.vc_metrics.csv'</span>
<span class="hljs-string">dragen/ploidy_estimation_metrics:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.ploidy_estimation_metrics.csv'</span>
<span class="hljs-string">dragen/contig_mean_cov:</span>
<span class="hljs-attr">  fn_re:</span> <span class="hljs-string">'.*\.(wgs|target_bed)_contig_mean_cov_?(tumor|normal)?\.csv'</span>
<span class="hljs-string">dragen/coverage_metrics:</span>
<span class="hljs-attr">  fn_re:</span> <span class="hljs-string">'.*\.(wgs|target_bed)_coverage_metrics_?(tumor|normal)?\.csv'</span>
<span class="hljs-string">dragen/fine_hist:</span>
<span class="hljs-attr">  fn_re:</span> <span class="hljs-string">'.*\.(wgs|target_bed)_fine_hist_?(tumor|normal)?\.csv'</span>
<span class="hljs-string">dragen/fragment_length_hist:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.fragment_length_hist.csv'</span>
<span class="hljs-string">dragen/mapping_metrics:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.mapping_metrics.csv'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    Number of unique reads (excl. duplicate
    marked reads)
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">50</span>
</code></pre>
          </div><h2 id="hic-pro"><a href="#hic-pro" class="header-link"><span class="glyphicon glyphicon-link"></span></a>HiC-Pro</h2><p class="mod_ext_link"><a href="https://github.com/nservant/HiC-Pro" target="_blank">https://github.com/nservant/HiC-Pro</a></p><p>The HiC-Pro module parses results generated by
<a href="https://github.com/nservant/HiC-Pro">HiC-Pro</a>, a tool for efficient processing and quality control of Hi-C data.</p>
<p>Please note - because this module shares sample identifiers across multiple files,
the <code>--fn_as_s_name</code> / <code>config.use_filename_as_sample_name</code> functionality has been disabled and has no effect.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                HiC-Pro file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">hicpro/mmapstat:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.mmapstat'</span>
<span class="hljs-string">hicpro/mpairstat:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.mpairstat'</span>
<span class="hljs-string">hicpro/mergestat:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.mergestat'</span>
<span class="hljs-string">hicpro/mRSstat:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.mRSstat'</span>
<span class="hljs-string">hicpro/assplit:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.assplit.stat'</span>
</code></pre>
          </div><h2 id="hicup"><a href="#hicup" class="header-link"><span class="glyphicon glyphicon-link"></span></a>HiCUP</h2><p class="mod_ext_link"><a href="http://www.bioinformatics.babraham.ac.uk/projects/hicup/" target="_blank">http://www.bioinformatics.babraham.ac.uk/projects/hicup/</a></p><p>The HiCUP module parses results generated by
<a href="http://www.bioinformatics.babraham.ac.uk/projects/hicup/">HiCUP</a>,
(Hi-C User Pipeline), a tool for mapping and performing quality
control on Hi-C data.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                HiCUP file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">hicup:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'HiCUP_summary_report*'</span>
</code></pre>
          </div><h2 id="hisat2"><a href="#hisat2" class="header-link"><span class="glyphicon glyphicon-link"></span></a>HISAT2</h2><p class="mod_ext_link"><a href="https://ccb.jhu.edu/software/hisat2/" target="_blank">https://ccb.jhu.edu/software/hisat2/</a></p><p><a href="https://ccb.jhu.edu/software/hisat2/">HISAT2</a>
is a fast and sensitive alignment program for mapping NGS reads
(both DNA and RNA) to a population of human genomes (as well as to a
single reference genome).</p>
<p>The HISAT2 MultiQC module parses summary statistics generated by
versions &gt;= v2.1.0 where the command line option <code>--new-summary</code>
has been specified.</p>
<p>Note that running HISAT2 without this option (and older versions)
gives log output identical to Bowtie2. These logs are indistinguishable
and summary statistics will appear in MultiQC reports labelled as Bowtie2.
See GitHub issues on the <a href="https://github.com/infphilo/hisat2/issues/48">HISAT2 repository</a>
and the <a href="https://github.com/ewels/MultiQC/issues/221">MultiQC repository</a>
for more information.</p>
<p>HISAT2 does not report the input file names in the log, so MultiQC
takes the filename as the sample. Note that if you specify
<code>--summary-file</code> when running HISAT2 the same summary output
appears both there and in the <code>stdout</code>. So if you save both with
different names you may end up with duplicate samples in your
MultiQC report.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                HISAT2 file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">hisat2:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'HISAT2 summary stats:'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="kallisto"><a href="#kallisto" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Kallisto</h2><p class="mod_ext_link"><a href="http://pachterlab.github.io/kallisto/" target="_blank">http://pachterlab.github.io/kallisto/</a></p><p>The Kallisto module parses logs generated by
<a href="http://pachterlab.github.io/kallisto/">Kallisto</a>,
a program for quantifying abundances of transcripts from RNA-Seq data, or more generally
of target sequences using high-throughput sequencing reads.</p>
<p><strong>Note</strong> - MultiQC parses the standard out from Kallisto, <em>not</em> any of its output files
(<code>abundance.h5</code>, <code>abundance.tsv</code>, and <code>run_info.json</code>). As such, you must capture the
Kallisto stdout to a file when running to use the MultiQC module.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Kallisto file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">kallisto:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'[quant] finding pseudoalignments for the reads'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="longranger"><a href="#longranger" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Longranger</h2><p class="mod_ext_link"><a href="https://www.10xgenomics.com/" target="_blank">https://www.10xgenomics.com/</a></p><p>Currently supported Longranger pipelines:</p>
<ul>
<li><code>wgs</code></li>
<li><code>targeted</code></li>
<li><code>align</code></li>
</ul>
<p>Usage:</p>
<pre><code class="language-bash hljs">longranger wgs --fastqs=/path/to/fastq --id=NA12878
multiqc /path/to/NA12878</code></pre>
<p>This module will look for the files <code>_invocation</code> and <code>summary.csv</code> in the the <code>NA12878</code> folder, i.e. the output folder of Longranger in this example. The file <code>summary.csv</code> is required. If the file <code>_invocation</code> is not found the sample will receive a generic name in the MultiQC report (<code>longranger#1</code>), instead of <code>NA12878</code> or whatever was given by the <code>--id</code> parameter.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Longranger file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">longranger/summary:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*summary.csv'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    longranger_version,instrument_ids,gems_detected,mean_dna_per_gem,bc_on_whitelist,bc_mean_qscore,n50_linked_reads_per_molecule
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
<span class="hljs-string">longranger/invocation:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">_invocation</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">call</span> <span class="hljs-string">PHASER_SVCALLER_CS(</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">2048</span>
</code></pre>
          </div><h2 id="malt"><a href="#malt" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MALT</h2><p class="mod_ext_link"><a href="https://uni-tuebingen.de/fakultaeten/mathematisch-naturwissenschaftliche-fakultaet/fachbereiche/informatik/lehrstuehle/algorithms-in-bioinformatics/software/malt/" target="_blank">https://uni-tuebingen.de/fakultaeten/mathematisch-naturwissenschaftliche-fakultaet/fachbereiche/informatik/lehrstuehle/algorithms-in-bioinformatics/software/malt/</a></p><p>MALT
 performs alignment of metagenomic reads against a database of reference
 sequences (such as NR, GenBank or Silva) and produces a MEGAN RMA file 
as output.</p>
<p>The MALT MultiQC module reads the header of the MALT log files
and procudes three MultiQC sections:</p>
<ul>
<li>A MALT summary statistics table</li>
<li>A Mappability bargraph</li>
<li>A Taxonomic assignment success bargraph</li>
</ul>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                MALT file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">malt:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'MaltRun - Aligns sequences using MALT (MEGAN alignment tool)'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
</code></pre>
          </div><h2 id="salmon"><a href="#salmon" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Salmon</h2><p class="mod_ext_link"><a href="http://combine-lab.github.io/salmon/" target="_blank">http://combine-lab.github.io/salmon/</a></p><p>The Salmon module parses results generated by
<a href="http://combine-lab.github.io/salmon/">Salmon</a>,
a tool for quantifying the expression of transcripts using RNA-seq data.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Salmon file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">salmon/meta:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">meta_info.json</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">salmon_version</span>
<span class="hljs-string">salmon/fld:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">flenDist.txt</span>
</code></pre>
          </div><h2 id="star"><a href="#star" class="header-link"><span class="glyphicon glyphicon-link"></span></a>STAR</h2><p class="mod_ext_link"><a href="https://github.com/alexdobin/STAR" target="_blank">https://github.com/alexdobin/STAR</a></p><p><a href="https://github.com/alexdobin/STAR">STAR</a> is an ultrafast universal RNA-seq aligner.</p>
<p>This MultiQC module parses summary statistics from the <code>Log.final.out</code> log files.
Sample names are taken either from the filename prefix (<code>sampleNameLog.final.out</code>)
when set with <code>--outFileNamePrefix</code> in STAR. If there is no filename prefix,
the sample name is set as the name of the directory containing the file.</p>
<p>In addition to this summary log file, the module parses <code>ReadsPerGene.out.tab</code>
files generated with <code>--quantMode GeneCounts</code>, if found.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                STAR file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">star:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*Log.final.out'</span>
<span class="hljs-string">star/genecounts:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*ReadsPerGene.out.tab'</span>
</code></pre>
          </div><h2 id="tophat"><a href="#tophat" class="header-link"><span class="glyphicon glyphicon-link"></span></a>TopHat</h2><p class="mod_ext_link"><a href="https://ccb.jhu.edu/software/tophat/" target="_blank">https://ccb.jhu.edu/software/tophat/</a></p><p>The TopHat module parses results generated by
<a href="https://ccb.jhu.edu/software/tophat/">TopHat</a>,
a fast splice junction mapper for RNA-Seq reads that aligns
RNA-Seq reads to mammalian-sized genomes.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                TopHat file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">tophat:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*align_summary.txt'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div></div><div class="docs_block modules_block" id="post-alignment-section"><h1 id="post-alignment"><a href="#post-alignment" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Post-alignment</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#bamtools">Bamtools</a></li>
<li><a href="#bcftools">Bcftools</a></li>
<li><a href="#biobambam2">biobambam2</a></li>
<li><a href="#busco">BUSCO</a></li>
<li><a href="#bustools">bustools</a></li>
<li><a href="#conpair">Conpair</a></li>
<li><a href="#damageprofiler">DamageProfiler</a></li>
<li><a href="#dedup">DeDup</a></li>
<li><a href="#deeptools">deepTools</a></li>
<li><a href="#disambiguate">Disambiguate</a></li>
<li><a href="#eigenstratdatabasetools">EigenStratDatabaseTools</a></li>
<li><a href="#featurecounts">featureCounts</a></li>
<li><a href="#fgbio">Fgbio</a></li>
<li><a href="#gatk">GATK</a>
<ul class="nav nav-stacked">
<li><a href="#baserecalibrator">BaseRecalibrator</a></li>
<li><a href="#varianteval">VariantEval</a></li>
</ul>
</li>
<li><a href="#gffcompare">GffCompare</a></li>
<li><a href="#goleft-indexcov">goleft indexcov</a></li>
<li><a href="#hap.py">Hap.py</a></li>
<li><a href="#hicexplorer">HiCExplorer</a></li>
<li><a href="#homer">HOMER</a></li>
<li><a href="#hops">HOPS</a></li>
<li><a href="#htseq">HTSeq</a></li>
<li><a href="#jcvi-genome-annotation">JCVI Genome Annotation</a></li>
<li><a href="#kaiju">Kaiju</a></li>
<li><a href="#kraken">Kraken</a></li>
<li><a href="#macs2">MACS2</a></li>
<li><a href="#methylqa">methylQA</a></li>
<li><a href="#mirtop">mirtop</a></li>
<li><a href="#mirtrace">miRTrace</a></li>
<li><a href="#mosdepth">mosdepth</a></li>
<li><a href="#mtnucratio">MTNucRatio</a></li>
<li><a href="#multivcfanalyzer">MultiVCFAnalyzer</a></li>
<li><a href="#nanostat">NanoStat</a></li>
<li><a href="#ngsderive">ngsderive</a></li>
<li><a href="#odgi">odgi</a></li>
<li><a href="#optitype">OptiType</a></li>
<li><a href="#pangolin">Pangolin</a></li>
<li><a href="#peddy">Peddy</a></li>
<li><a href="#phantompeakqualtools">phantompeakqualtools</a></li>
<li><a href="#picard">Picard</a></li>
<li><a href="#preseq">Preseq</a></li>
<li><a href="#prokka">Prokka</a></li>
<li><a href="#purple">PURPLE</a></li>
<li><a href="#qc3c">qc3C</a></li>
<li><a href="#qorts">QoRTs</a></li>
<li><a href="#qualimap">Qualimap</a></li>
<li><a href="#quast">QUAST</a></li>
<li><a href="#rna-seqc">RNA-SeQC</a></li>
<li><a href="#rockhopper">Rockhopper</a></li>
<li><a href="#rsem">RSEM</a></li>
<li><a href="#rseqc">RSeQC</a></li>
<li><a href="#sambamba">Sambamba</a>
<ul class="nav nav-stacked">
<li><a href="#markdup">markdup</a></li>
</ul>
</li>
<li><a href="#samblaster">Samblaster</a></li>
<li><a href="#samtools">Samtools</a>
<ul class="nav nav-stacked">
<li><a href="#idxstats">idxstats</a></li>
</ul>
</li>
<li><a href="#sargasso">Sargasso</a></li>
<li><a href="#sentieon">Sentieon</a></li>
<li><a href="#sex.deterrmine">Sex.DetErrMine</a></li>
<li><a href="#slamdunk">Slamdunk</a></li>
<li><a href="#snippy">Snippy</a></li>
<li><a href="#snpeff">SnpEff</a></li>
<li><a href="#snpsplit">SNPsplit</a></li>
<li><a href="#somalier">Somalier</a></li>
<li><a href="#stacks">Stacks</a></li>
<li><a href="#supernova">Supernova</a></li>
<li><a href="#theta2">THeTA2</a></li>
<li><a href="#varscan2">VarScan2</a></li>
<li><a href="#vcftools">VCFTools</a>
<ul class="nav nav-stacked">
<li><a href="#important-general-note">Important General Note</a></li>
<li><a href="#implemented">Implemented</a></li>
<li><a href="#to-do">To do</a></li>
</ul>
</li>
<li><a href="#vep">VEP</a></li>
<li><a href="#verifybamid">VerifyBAMID</a></li>
<li><a href="#whatshap">WhatsHap</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div><h2 id="bamtools"><a href="#bamtools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bamtools</h2><p class="mod_ext_link"><a href="https://github.com/pezmaster31/bamtools" target="_blank">https://github.com/pezmaster31/bamtools</a></p><p>The Bamtools module parses <code>bamtools stats</code> logs generated by
<a href="https://github.com/pezmaster31/bamtools">Bamtools</a>,
a programmer's API and an end-user's toolkit for handling BAM files.</p>
<p>Supported commands: <code>stats</code></p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Bamtools file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">bamtools/stats:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'Stats for BAM file(s):'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">10</span>
</code></pre>
          </div><h2 id="bcftools"><a href="#bcftools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bcftools</h2><p class="mod_ext_link"><a href="https://samtools.github.io/bcftools/" target="_blank">https://samtools.github.io/bcftools/</a></p><p>The Bcftools module parses results generated by
<a href="https://samtools.github.io/bcftools/">Bcftools</a>,
a suite of programs for interacting with variant call data.</p>
<p>Supported commands: <code>stats</code></p>
<h4 id="collapse-complementary-substitutions"><a href="#collapse-complementary-substitutions" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Collapse complementary substitutions</h4>
<p>In non-strand-specific data, reporting the total numbers of occurences for both changes
in a comlementary pair - like <code>A&gt;C</code> and <code>T&gt;G</code> - might not bring any additional information.
To collapse such statistics in the substitutions plot, you can add the following section into
<a href="http://multiqc.info/docs/#configuring-multiqc">your configuration</a>:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">bcftools:</span>
<span class="hljs-attr">  collapse_complementary_changes:</span> <span class="hljs-literal">true</span></code></pre>
<p>MultiQC will sum up all complementary changes and show only <code>A&gt;*</code> and <code>C&gt;*</code> substitutions
in the resulting plot.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Bcftools file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">bcftools/stats:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">This</span> <span class="hljs-string">file</span> <span class="hljs-string">was</span> <span class="hljs-string">produced</span> <span class="hljs-string">by</span> <span class="hljs-string">bcftools</span> <span class="hljs-string">stats</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="biobambam2"><a href="#biobambam2" class="header-link"><span class="glyphicon glyphicon-link"></span></a>biobambam2</h2><p class="mod_ext_link"><a href="https://github.com/gt1/biobambam2" target="_blank">https://github.com/gt1/biobambam2</a></p><p>Currently, the biobambam2 module only processes output from the <code>bamsormadup</code> command.
Not only that, but it cheats by using the module code from Picard/MarkDuplicates.
The output is so similar that the code simply sets up a module with unique name and
filename search pattern and then uses the parsing code from the Picard module.</p>
<p>Apart from behind the scenes coding, this module should work in exactly the same way
as all other MultiQC modules.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                biobambam2 file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">biobambam2/bamsormadup:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'# bamsormadup'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
</code></pre>
          </div><h2 id="busco"><a href="#busco" class="header-link"><span class="glyphicon glyphicon-link"></span></a>BUSCO</h2><p class="mod_ext_link"><a href="http://busco.ezlab.org/" target="_blank">http://busco.ezlab.org/</a></p><p>BUSCO v2 provides quantitative measures for the assessment of genome
assembly, gene set, and transcriptome completeness, based on
evolutionarily-informed expectations of gene content from near-universal
single-copy orthologs selected from OrthoDB v9.</p>
<p>The MultiQC module parses the <code>short_summary_[samplename].txt</code> files and
plots the proportion of BUSCO types found. MultiQC has been tested with
output from BUSCO v1.22 - v2.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                BUSCO file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">busco:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'short_summary*'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'BUSCO version is:'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="bustools"><a href="#bustools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>bustools</h2><p class="mod_ext_link"><a href="https://bustools.github.io/" target="_blank">https://bustools.github.io/</a></p><p>This MultiQC modules parses the report generated by Bustools inspect, and expects the file to be named <code>inspect.json</code>. This is the default naming pattern when you make use of the <a href="https://www.kallistobus.tools/">kallisto-bustools wrapper</a>.
 The sample name is set as the name of the directory containing the 
file. For the rest, this module should work in exactly the same way as 
all other MultiQC modules.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                bustools file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">bustools:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*inspect.json'</span>
</code></pre>
          </div><h2 id="conpair"><a href="#conpair" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Conpair</h2><p class="mod_ext_link"><a href="https://github.com/nygenome/Conpair" target="_blank">https://github.com/nygenome/Conpair</a></p><p><a href="https://github.com/nygenome/Conpair">Conpair</a>
 is a fast and robust method dedicated for human tumour-normal studies 
to perform concordance verification (= samples coming from the same 
individual), as well as cross-individual contamination level estimation 
in whole-genome and whole-exome sequencing experiments.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Conpair file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">conpair/concordance:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'markers (coverage per marker threshold : '</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
<span class="hljs-string">conpair/contamination:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'Tumor sample contamination level: '</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
</code></pre>
          </div><h2 id="damageprofiler"><a href="#damageprofiler" class="header-link"><span class="glyphicon glyphicon-link"></span></a>DamageProfiler</h2><p class="mod_ext_link"><a href="https://github.com/Integrative-Transcriptomics/DamageProfiler" target="_blank">https://github.com/Integrative-Transcriptomics/DamageProfiler</a></p><p>A tool for DNA damage pattern retrieval for ancient DNA analysis and verification.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                DamageProfiler file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">damageprofiler:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*dmgprof.json'</span>
</code></pre>
          </div><h2 id="dedup"><a href="#dedup" class="header-link"><span class="glyphicon glyphicon-link"></span></a>DeDup</h2><p class="mod_ext_link"><a href="http://www.github.com/apeltzer/DeDup" target="_blank">http://www.github.com/apeltzer/DeDup</a></p><p>Improved Duplicate Removal for merged/collapsed reads in ancient DNA analysis</p>
<p>By default, tables show read counts in thousands.
To customise this, you can set the following MultiQC config variables:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">ancient_read_count_prefix:</span> <span class="hljs-string">"K"</span>
<span class="hljs-attr">ancient_read_count_desc:</span> <span class="hljs-string">"thousands"</span>
<span class="hljs-attr">ancient_read_count_multiplier:</span> <span class="hljs-number">0.001</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                DeDup file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">dedup:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*dedup.json'</span>
</code></pre>
          </div><h2 id="deeptools"><a href="#deeptools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>deepTools</h2><p class="mod_ext_link"><a href="http://deeptools.readthedocs.io/" target="_blank">http://deeptools.readthedocs.io</a></p><p>deepTools
 addresses the challenge of handling the large amounts of data that are 
now routinely generated from DNA sequencing centers. deepTools contains 
useful modules to process the mapped reads data for multiple quality 
checks, creating <strong>normalized coverage files</strong> in standard 
bedGraph and bigWig file formats, that allow comparison between 
different files (for example, treatment and control). Finally, using 
such normalized and standardized files, deepTools can create many 
publication-ready <strong>visualizations</strong> to identify enrichments and for functional annotations of the genome.</p>
<p>The MultiQC module for deepTools parses a number of the text files 
that deepTools can produce. In particular, the following are supported:</p>
<ul>
<li><code>bamPEFragmentSize --table</code></li>
<li><code>bamPEFragmentSize --outRawFragmentLengths</code></li>
<li><code>estimateReadFiltering</code></li>
<li><code>plotCoverage ---outRawCounts</code> (as well as the content written normally to the console)</li>
<li><code>plotEnrichment --outRawCounts</code></li>
<li><code>plotFingerprint --outQualityMetrics --outRawCounts</code></li>
<li><code>plotPCA --outFileNameData</code></li>
<li><code>plotCorrelation --outFileCorMatrix</code></li>
<li><code>plotProfile --outFileNameData</code></li>
</ul>
<p>Please be aware that some tools (namely, <code>plotFingerprint --outRawCounts</code> and <code>plotCoverage --outRawCounts</code>) are only supported as of deepTools version 2.6. For earlier output from <code>plotCoverage --outRawCounts</code>, you can use <code>#'chr' 'start' 'end'</code> in <code>utils/search_patterns.yaml</code> (see <a href="http://multiqc.info/docs/#module-search-patterns">here</a> for more details). Also for these types of files, you may need to increase the maximum file size supported by MultiQC (<code>log_filesize_limit</code> in the MultiQC configuration file). You can find details regarding the configuration file location <a href="http://multiqc.info/docs/#configuring-multiqc">here</a>.</p>
<p>Note that sample names are parsed from the text files themselves, they are not derived from file names.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                deepTools file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">deeptools/bamPEFragmentSizeTable:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    	Frag. Sampled	Frag. Len. Min.	Frag.
    Len. 1st. Qu.	Frag. Len. Mean	Frag. Len.
    Median	Frag. Len. 3rd Qu.
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">deeptools/bamPEFragmentSizeDistribution:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#bamPEFragmentSize'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">deeptools/estimateReadFiltering:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    Sample	Total Reads	Mapped
    Reads	Alignments in blacklisted
    regions	Estimated mapped reads
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">deeptools/plotCorrelationData:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#plotCorrelation --outFileCorMatrix'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">deeptools/plotCoverageStdout:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'sample	mean	std	min	25%	50%	75%	max'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">deeptools/plotCoverageOutRawCounts:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#plotCoverage --outRawCounts'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">deeptools/plotEnrichment:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    file	featureType	percent	featureReadCount	totalReadCount
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">deeptools/plotFingerprintOutRawCounts:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#plotFingerprint --outRawCounts'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">deeptools/plotFingerprintOutQualityMetrics:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    Sample	AUC	Synthetic
    AUC	X-intercept	Synthetic
    X-intercept	Elbow Point	Synthetic Elbow
    Point
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">deeptools/plotPCAData:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#plotPCA --outFileNameData'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">deeptools/plotProfile:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">bin</span> <span class="hljs-string">labels</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="disambiguate"><a href="#disambiguate" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Disambiguate</h2><p class="mod_ext_link"><a href="https://github.com/AstraZeneca-NGS/disambiguate" target="_blank">https://github.com/AstraZeneca-NGS/disambiguate</a></p><p>Disambiguation algorithm for reads aligned to two species (e.g. human and mouse
genomes) from Tophat, Hisat2, STAR or BWA mem. Both a Python and C++
implementation are offered.</p>
<p>The MultiQC module for Disambiguate parses the summary files generated by
Disambiguate.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Disambiguate file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">disambiguate:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">unique</span> <span class="hljs-string">species</span> <span class="hljs-string">A</span> <span class="hljs-string">pairs</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
</code></pre>
          </div><h2 id="eigenstratdatabasetools"><a href="#eigenstratdatabasetools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>EigenStratDatabaseTools</h2><p class="mod_ext_link"><a href="https://github.com/TCLamnidis/EigenStratDatabaseTools" target="_blank">https://github.com/TCLamnidis/EigenStratDatabaseTools</a></p><p>A
 set of tools to compare and manipulate the contents of EingenStrat 
databases, and to calculate SNP coverage statistics in such databases.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                EigenStratDatabaseTools file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">eigenstratdatabasetools:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_eigenstrat_coverage.json'</span>
</code></pre>
          </div><h2 id="featurecounts"><a href="#featurecounts" class="header-link"><span class="glyphicon glyphicon-link"></span></a>featureCounts</h2><p class="mod_ext_link"><a href="http://bioinf.wehi.edu.au/featureCounts/" target="_blank">http://bioinf.wehi.edu.au/featureCounts/</a></p><p>The featureCounts module parses results generated by
<a href="http://bioinf.wehi.edu.au/featureCounts/">featureCounts</a>,
a highly efficient general-purpose read summarization
program that counts mapped reads for genomic features such as genes,
exons, promoter, gene bodies, genomic bins and chromosomal locations.</p>
<p>As of MultiQC v1.10, the module should also work with output from
<a href="https://bioconductor.org/packages/release/bioc/html/Rsubread.html">Rsubread</a>.
Note that your filenames must end in <code>.summary</code> to be discovered.
See <a href="#module-search-patterns">Module search patterns</a> for how to customise this.</p>
<p>Please note that if files are in "Rsubread mode" then lines will be split by any
whitespace, instead of tab characters. As such, filenames with spaces in will
cause the parsing to fail.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                featureCounts file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">featurecounts:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.summary'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="fgbio"><a href="#fgbio" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Fgbio</h2><p class="mod_ext_link"><a href="http://fulcrumgenomics.github.io/fgbio/tools/latest/" target="_blank">http://fulcrumgenomics.github.io/fgbio/tools/latest/</a></p><p>The fgbio MultiQC module currently supports tool the following outputs:</p>
<ul>
<li><a href="http://fulcrumgenomics.github.io/fgbio/tools/latest/GroupReadsByUmi.html/">GroupReadsByUmi</a></li>
<li><a href="http://fulcrumgenomics.github.io/fgbio/tools/latest/ErrorRateByReadPosition.html/">ErrorRateByReadPosition</a></li>
</ul>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Fgbio file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">fgbio/groupreadsbyumi:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">fraction_gt_or_eq_family_size</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
<span class="hljs-string">fgbio/errorratebyreadposition:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    read_number	position	bases_total	errors	error_rate	a_to_c_error_rate	a_to_g_error_rate	a_to_t_error_rate	c_to_a_error_rate	c_to_g_error_rate	c_to_t_error_rate
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
</code></pre>
          </div><h2 id="gatk"><a href="#gatk" class="header-link"><span class="glyphicon glyphicon-link"></span></a>GATK</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#baserecalibrator">BaseRecalibrator</a></li>
<li><a href="#varianteval">VariantEval</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div><p class="mod_ext_link"><a href="https://www.broadinstitute.org/gatk/" target="_blank">https://www.broadinstitute.org/gatk/</a></p><p>Developed by the <a href="http://www.broadinstitute.org/dsde">Data Science and Data Engineering</a>
group at the <a href="http://www.broadinstitute.org/">Broad Institute</a>, the GATK toolkit offers
a wide variety of tools with a primary focus on variant discovery and genotyping.
Its powerful processing engine and high-performance computing features make it capable
of taking on projects of any size.</p>
<p>Supported tools:</p>
<ul>
<li><code>BaseRecalibrator</code></li>
<li><code>VariantEval</code></li>
</ul>
<h3 id="baserecalibrator"><a href="#baserecalibrator" class="header-link"><span class="glyphicon glyphicon-link"></span></a>BaseRecalibrator</h3>
<p><a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/current/org_broadinstitute_gatk_tools_walkers_bqsr_BaseRecalibrator.php">BaseRecalibrator</a>
is a tool for detecting systematic errors in read base quality scores of aligned high-throughput
sequencing reads. It outputs a base quality score recalibration table that can be used in
conjunction with the
<a href="https://software.broadinstitute.org/gatk/documentation/tooldocs/current/org_broadinstitute_gatk_tools_walkers_readutils_PrintReads.php">PrintReads</a>
tool to recalibrate base quality scores.</p>
<h3 id="varianteval"><a href="#varianteval" class="header-link"><span class="glyphicon glyphicon-link"></span></a>VariantEval</h3>
<p><a href="https://software.broadinstitute.org/gatk/gatkdocs/current/org_broadinstitute_gatk_tools_walkers_varianteval_VariantEval.php">VariantEval</a>
is a general-purpose tool for variant evaluation. It gives information about percentage of
variants in dbSNP, genotype concordance, Ti/Tv ratios and a lot more.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                GATK file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">gatk/varianteval:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#:GATKTable:TiTvVariantEvaluator'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">gatk/base_recalibrator:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#:GATKTable:Arguments:Recalibration'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
</code></pre>
          </div><h2 id="gffcompare"><a href="#gffcompare" class="header-link"><span class="glyphicon glyphicon-link"></span></a>GffCompare</h2><p class="mod_ext_link"><a href="https://ccb.jhu.edu/software/stringtie/gffcompare.shtml" target="_blank">https://ccb.jhu.edu/software/stringtie/gffcompare.shtml</a></p><p>The program <code>gffcompare</code> can be used to compare, merge, annotate and estimate accuracy
of one or more GFF files (the "query" files), when compared with a reference annotation (also provided as GFF).</p>
<p>The <em>Sensitivity / Precision</em> values are displayed in a single plot,
different loci levels can be switched by choosing a different dataset.</p>
<blockquote>
<p><strong>NB:</strong> Please use <code>gffcompare</code> only with single samples.
Multi-Sample comparisons are not correctly rendered by this MultiQC module.</p>
</blockquote>
<p>Note that exported data in <code>multiqc_data/multiqc_gffcompare.{tsv,yaml,json}</code> only works when
exporting with YAML or JSON - the default <code>.tsv</code> output will not contain any data.
Please use <code>-k yaml</code> or <code>-k json</code> to export in a structured format.
It is hoped to refactor this code in a future release - please submit a PR if you are interested.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                GffCompare file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">gffcompare:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.stats'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'# gffcompare'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
</code></pre>
          </div><h2 id="goleft-indexcov"><a href="#goleft-indexcov" class="header-link"><span class="glyphicon glyphicon-link"></span></a>goleft indexcov</h2><p class="mod_ext_link"><a href="https://github.com/brentp/goleft/tree/master/indexcov" target="_blank">https://github.com/brentp/goleft/tree/master/indexcov</a></p><p>The goleft indexcov module parses results generated by
<a href="https://github.com/brentp/goleft/tree/master/indexcov">goleft indexcov</a>. It
uses the PED and ROC data files to create diagnostic plots of coverage per
sample, helping to identify sample gender and coverage issues.</p>
<p>By default, we attempt to only plot chromosomes using standard human-like naming
(chr1, chr2... chrX or 1, 2 ... X) but you can specify chromosomes for detailed
ROC plots for alternative naming schemes in your configuration with:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">goleft_indexcov_config:</span>
<span class="hljs-attr">  chromosomes:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">I</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">II</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">III</span></code></pre>
<p>The number of plotted chromosomes is limited to 50 by default, you can customise this with the following:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">goleft_indexcov_config:</span>
<span class="hljs-attr">  max_chroms:</span> <span class="hljs-number">80</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                goleft indexcov file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">goleft_indexcov/roc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*-indexcov.roc'</span>
<span class="hljs-string">goleft_indexcov/ped:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*-indexcov.ped'</span>
</code></pre>
          </div><h2 id="hap.py"><a href="#hap.py" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Hap.py</h2><p class="mod_ext_link"><a href="https://github.com/Illumina/hap.py" target="_blank">https://github.com/Illumina/hap.py</a></p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Hap.py file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">happy:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.summary.csv'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Type,Filter,TRUTH</span>
</code></pre>
          </div><h2 id="hicexplorer"><a href="#hicexplorer" class="header-link"><span class="glyphicon glyphicon-link"></span></a>HiCExplorer</h2><p class="mod_ext_link"><a href="https://hicexplorer.readthedocs.io/" target="_blank">https://hicexplorer.readthedocs.io</a></p><p>The
 HiCExplorer module parses results generated by HiCExplorere's 
hicBuildMatrix, a tool to create an interaction matrix out of mapped 
Hi-C reads.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                HiCExplorer file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">hicexplorer:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Min</span> <span class="hljs-string">rest.</span> <span class="hljs-string">site</span> <span class="hljs-string">distance</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">4096</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">26</span>
</code></pre>
          </div><h2 id="homer"><a href="#homer" class="header-link"><span class="glyphicon glyphicon-link"></span></a>HOMER</h2><p class="mod_ext_link"><a href="http://homer.ucsd.edu/homer/" target="_blank">http://homer.ucsd.edu/homer/</a></p><p>HOMER <em>(Hypergeometric Optimization of Motif EnRichment)</em> is a suite of tools for
Motif Discovery and next-gen sequencing analysis. HOMER contains many useful tools
for analyzing ChIP-Seq, GRO-Seq, RNA-Seq, DNase-Seq, Hi-C and numerous other types
of functional genomics sequencing data sets.</p>
<p>The HOMER MultiQC module currently only parses output from the <code>findPeaks</code> tool.
If you would like support to be added for other HOMER tools, please open a
<a href="https://github.com/ewels/MultiQC/issues/new">new issue</a> on the MultiQC GitHub page.</p>
<h4 id="findpeaks"><a href="#findpeaks" class="header-link"><span class="glyphicon glyphicon-link"></span></a>FindPeaks</h4>
<p>The HOMER findPeaks MultiQC module parses the summary statistics found at the top
of HOMER peak files. Three key statistics are shown in the General Statistics table,
all others are saved to <code>multiqc_data/multiqc_homer_findpeaks.txt</code>.</p>
<h4 id="tagdirectory"><a href="#tagdirectory" class="header-link"><span class="glyphicon glyphicon-link"></span></a>TagDirectory</h4>
<p>The HOMER tag directory submodule parses output from files
<a href="http://homer.ucsd.edu/homer/ngs/tagDir.html">tag directory</a> output files, generating
a number of diagnostic plots.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                HOMER file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">homer/findpeaks:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'# HOMER Peaks'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
<span class="hljs-string">homer/GCcontent:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">tagGCcontent.txt</span>
<span class="hljs-string">homer/genomeGCcontent:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">genomeGCcontent.txt</span>
<span class="hljs-string">homer/RestrictionDistribution:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'petagRestrictionDistribution.*.txt'</span>
<span class="hljs-string">homer/LengthDistribution:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">tagLengthDistribution.txt</span>
<span class="hljs-string">homer/tagInfo:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">tagInfo.txt</span>
<span class="hljs-string">homer/FreqDistribution:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">petag.FreqDistribution_1000.txt</span>
</code></pre>
          </div><h2 id="hops"><a href="#hops" class="header-link"><span class="glyphicon glyphicon-link"></span></a>HOPS</h2><p class="mod_ext_link"><a href="https://www.github.com/rhubler/HOPS" target="_blank">https://www.github.com/rhubler/HOPS</a></p><p>This module takes the JSON output of the HOPS postprocessing R script (Version
&gt;= 0.34). to recreate the possible positives heatmap, with the heat intensity
representing the number of 'ancient DNA characteristics' categories (small
edit distance, damage, both edit distance and aDNA damage) that a particular
taxon has.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                HOPS file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">hops:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">heatmap_overview_Wevid.json</span>
</code></pre>
          </div><h2 id="htseq"><a href="#htseq" class="header-link"><span class="glyphicon glyphicon-link"></span></a>HTSeq</h2><p class="mod_ext_link"><a href="http://www-huber.embl.de/HTSeq/doc/overview.html" target="_blank">http://www-huber.embl.de/HTSeq/doc/overview.html</a></p><p>HTSeq is a general purpose Python package that provides infrastructure to
process data from high-throughput sequencing assays. <code>htseq-count</code> is a tool
that is part of the main HTSeq package - it takes a file with aligned sequencing
reads, plus a list of genomic features and counts how many reads map to each feature.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                HTSeq file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">htseq:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">__too_low_aQual</span>
</code></pre>
          </div><h2 id="jcvi-genome-annotation"><a href="#jcvi-genome-annotation" class="header-link"><span class="glyphicon glyphicon-link"></span></a>JCVI Genome Annotation</h2><p class="mod_ext_link"><a href="https://github.com/tanghaibao/jcvi" target="_blank">https://github.com/tanghaibao/jcvi</a></p><p>JCVI Genome Annotation provides quantitative measures for the assessment of genome
annotation.</p>
<p>The JCVI module parses the output of <code>python -m jcvi.annotation.stats genestats &lt;input.gff&gt;</code>.
The file name is used as the sample name.
If the output from the <code>python -m jcvi.annotation.stats stats &lt;input.gff&gt;</code> is present in the same directory,
it is used to draw complementary plots.</p>
<p>A typical result directory will contain:</p>
<pre><code class="language-txt"> Exon_Count
&nbsp;&nbsp;  sample1.txt
&nbsp;&nbsp;  sample2.txt
 Exon_Length
&nbsp;&nbsp;  sample1.txt
&nbsp;&nbsp;  sample2.txt
 Gene_Length
&nbsp;&nbsp;  sample1.txt
&nbsp;&nbsp;  sample2.txt
 Intron_Length
&nbsp;&nbsp;  sample1.txt
&nbsp;&nbsp;  sample2.txt
 sample1_genestats.txt
 sample2_genestats.txt
</code></pre>
<p>The JCVI module has been tested with output from JCVI v1.0.9.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                JCVI Genome Annotation file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">jcvi:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'     o    % GC    % of genome    Average size (bp)    Median size (bp)    Number    Total length (Mb)'</span>
</code></pre>
          </div><h2 id="kaiju"><a href="#kaiju" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Kaiju</h2><p class="mod_ext_link"><a href="http://kaiju.binf.ku.dk/" target="_blank">http://kaiju.binf.ku.dk/</a></p><p>The Kaiju module parses output generated by kaiju2table from <a href="http://kaiju.binf.ku.dk/">Kaiju</a>,
a fast and sensitive taxonomic classification for metagenomics.
e.g:</p>
<pre><code class="language-bash hljs">kaiju -i R1.fq.gz -j R2.fq.gz -o output_kaiju.txt
kaiju2table -t nodes.dmp -n names.dmp -r species -o kaiju2table_species.txt output_kaiju.txt
kaiju2table -t nodes.dmp -n names.dmp -r phylum -o kaiju2table_phylum.txt output_kaiju.txt</code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Kaiju file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">kaiju:</span>
<span class="hljs-attr">  contents_re:</span> <span class="hljs-string">&gt;
    file\tpercent\treads\ttaxon_id\ttaxon_name
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="kraken"><a href="#kraken" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Kraken</h2><p class="mod_ext_link"><a href="https://ccb.jhu.edu/software/kraken2/" target="_blank">https://ccb.jhu.edu/software/kraken2/</a></p><p>The MultiQC module supports outputs from both Kraken and Kraken 2.</p>
<p>It works with report files generated using the <code>--report</code> flag, that look like the following:</p>
<pre><code class="language-ts">11.66   98148   98148   U   0   unclassified
88.34   743870  996 -   1   root
88.22   742867  0   -   131567    cellular organisms
88.22   742866  2071    D   2       Bacteria
87.95   740514  2914    P   1239          Firmicutes</code></pre>
<p>A bar graph is generated that shows the number of fragments for each sample that
fall into the top categories for each taxa rank. The top categories are calculated
by summing the library percentages across all samples.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Kraken file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">kraken:</span>
<span class="hljs-attr">  contents_re:</span> <span class="hljs-string">'^\s{0,2}(\d{1,3}\.\d{1,2})\t(\d+)\t(\d+)\t([\dUDKRPCOFGS-]{1,3})\t(\d+)\s+(.+)'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
</code></pre>
          </div><h2 id="macs2"><a href="#macs2" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MACS2</h2><p class="mod_ext_link"><a href="https://github.com/taoliu/MACS" target="_blank">https://github.com/taoliu/MACS</a></p><p>MACS2 <em>(Model-based Analysis of ChIP-Seq)</em> is a tool for identifying transcript
factor binding sites. MACS captures the influence of genome complexity to
evaluate the significance of enriched ChIP regions.</p>
<p>The MACS2 MultiQC module reads the <code>*_peaks.xls</code> results files and prints the
redundancy rates and number of peaks found in the General Statistics table.
Numerous additional values are parsed and saved to
<code>multiqc_data/multiqc_macs2.txt</code>.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                MACS2 file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">macs2:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_peaks.xls'</span>
</code></pre>
          </div><h2 id="methylqa"><a href="#methylqa" class="header-link"><span class="glyphicon glyphicon-link"></span></a>methylQA</h2><p class="mod_ext_link"><a href="http://methylqa.sourceforge.net/" target="_blank">http://methylqa.sourceforge.net/</a></p><p>The methylQA module parses results generated by
<a href="http://methylqa.sourceforge.net/">methylQA</a>,
a methylation sequencing data quality assessment tool.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                methylQA file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">methylQA:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.report'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="mirtop"><a href="#mirtop" class="header-link"><span class="glyphicon glyphicon-link"></span></a>mirtop</h2><p class="mod_ext_link"><a href="https://github.com/miRTop/mirtop" target="_blank">https://github.com/miRTop/mirtop</a></p><p>This tool is dedicated to the creation and management of miRNA alignment output using the standardized GFF3 format (<a href="https://github.com/miRTop/mirGFF3">https://github.com/miRTop/mirGFF3</a>).
A unified miRNA alignment format allows to easily compare the output of different alignement tools.</p>
<p>Currently, mirtop can convert into mirGFF3 the outputs of commonly 
used pipelines, such as seqbuster, isomiR-SEA, sRNAbench, Prost! as well
 as BAM files.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                mirtop file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">mirtop:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*_mirtop_stats.log'</span>
</code></pre>
          </div><h2 id="mirtrace"><a href="#mirtrace" class="header-link"><span class="glyphicon glyphicon-link"></span></a>miRTrace</h2><p class="mod_ext_link"><a href="https://github.com/friedlanderlab/mirtrace" target="_blank">https://github.com/friedlanderlab/mirtrace</a></p><p>The miRTrace module parses results generated by
<a href="https://github.com/friedlanderlab/mirtrace">miRTrace</a>, a quality
control software for small RNA sequencing data.</p>
<p>miRTrace performs adapter trimming and discards the reads that fail to pass
the QC filters. miRTrace specifically addresses sequencing quality, read length,
sequencing depth and miRNA complexity and also identifies the presence of both
miRNAs and undesirable sequences derived from tRNAs, rRNAs, or Illumina artifact
sequences.</p>
<p>miRTrace also profiles clade-specific miRNAs based on a comprehensive catalog
of clade-specific miRNA families identified previously. With this information,
miRTrace can detect exogenous miRNAs, which could be contamination derived,
e.g. index mis-assignment on sample demultiplexing, or biologically derived,
e.g. parasitic RNAs.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                miRTrace file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">mirtrace/summary:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">mirtrace-results.json</span>
<span class="hljs-string">mirtrace/length:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">mirtrace-stats-length.tsv</span>
<span class="hljs-string">mirtrace/contaminationbasic:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">mirtrace-stats-contamination_basic.tsv</span>
<span class="hljs-string">mirtrace/mirnacomplexity:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">mirtrace-stats-mirna-complexity.tsv</span>
</code></pre>
          </div><h2 id="mosdepth"><a href="#mosdepth" class="header-link"><span class="glyphicon glyphicon-link"></span></a>mosdepth</h2><p class="mod_ext_link"><a href="https://github.com/brentp/mosdepth" target="_blank">https://github.com/brentp/mosdepth</a></p><p><a href="https://github.com/brentp/mosdepth/">Mosdepth</a> performs fast BAM/CRAM depth calculation for WGS, exome, or targeted sequencing.</p>
<p>It can generate several output files all with a common prefix and different endings:</p>
<ul>
<li>per-base depth (<code>{prefix}.per-base.bed.gz</code>),</li>
<li>mean per-window depth given a window size (<code>{prefix}.regions.bed.gz</code>, if a BED file provided with <code>--by</code>),</li>
<li>mean per-region given a BED file of regions (<code>{prefix}.regions.bed.gz</code>, if a window size provided with <code>--by</code>),</li>
<li>a distribution of proportion of bases covered at or above a given threshhold for each chromosome and genome-wide (<code>{prefix}.mosdepth.global.dist.txt</code> and <code>{prefix}.mosdepth.region.dist.txt</code>),</li>
<li>quantized output that merges adjacent bases as long as they fall in the same coverage bins (<code>{prefix}.quantized.bed.gz</code>),</li>
<li>threshold output to indicate how many bases in each region are covered at the given thresholds (<code>{prefix}.thresholds.bed.gz</code>)</li>
</ul>
<p>The MultiQC module plots coverage distributions from 2 kinds of outputs:</p>
<ul>
<li><code>{prefix}.mosdepth.region.dist.txt</code></li>
<li><code>{prefix}.mosdepth.global.dist.txt</code></li>
</ul>
<p>Using "region" if exists, otherwise "global". Plotting 3 figures:</p>
<ul>
<li>Proportion of bases in the reference genome with, at least, a given depth of coverage (cumulative coverage distribution).</li>
<li>Proportion of bases in the reference genome with a given depth of coverage (absolute coverage distribution).</li>
<li>Average coverage per contig/chromosome.</li>
</ul>
<p>Also plotting the percentage of the genome covered at a threshold in the General Stats section.
The default thresholds are 1, 5, 10, 30, 50, which can be customised in the config as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">mosdepth_config:</span>
<span class="hljs-attr">  general_stats_coverage:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">10</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">20</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">40</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">200</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">30000</span></code></pre>
<p>You can also specify which columns would be hidden when the report loads (by default, all values are hidden except 30X):</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">general_stats_coverage_hidden:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-number">10</span>
<span class="hljs-bullet">  -</span> <span class="hljs-number">20</span>
<span class="hljs-bullet">  -</span> <span class="hljs-number">200</span></code></pre>
<p>For the per-contig coverage plot, you can include and exclude contigs based on name or pattern.</p>
<p>For example, you could add the following to your MultiQC config file:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">mosdepth_config:</span>
<span class="hljs-attr">  include_contigs:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">"chr*"</span>
<span class="hljs-attr">  exclude_contigs:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">"*_alt"</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">"*_decoy"</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">"*_random"</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">"chrUn*"</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">"HLA*"</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">"chrM"</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">"chrEBV"</span></code></pre>
<p>Note that exclusion superseeds inclusion for the contig filters.</p>
<p>If you want to see what is being excluded, you can set <code>show_excluded_debug_logs</code> to <code>True</code>:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">mosdepth_config:</span>
<span class="hljs-attr">  show_excluded_debug_logs:</span> <span class="hljs-literal">True</span></code></pre>
<p>This will then print a debug log message (use <code>multiqc -v</code>) for each excluded contig.
This is disabled by default as there can be very many in some cases.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                mosdepth file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">mosdepth/summary:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.mosdepth.summary.txt'</span>
<span class="hljs-string">mosdepth/global_dist:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.mosdepth.global.dist.txt'</span>
<span class="hljs-string">mosdepth/region_dist:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.mosdepth.region.dist.txt'</span>
</code></pre>
          </div><h2 id="mtnucratio"><a href="#mtnucratio" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MTNucRatio</h2><p class="mod_ext_link"><a href="http://www.github.com/apeltzer/MTNucRatioCalculator" target="_blank">http://www.github.com/apeltzer/MTNucRatioCalculator</a></p><p>A method to compute mitochondrial to nuclear reads ratios for NGS data.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                MTNucRatio file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">mtnucratio:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*mtnuc.json'</span>
</code></pre>
          </div><h2 id="multivcfanalyzer"><a href="#multivcfanalyzer" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiVCFAnalyzer</h2><p class="mod_ext_link"><a href="http://www.github.com/alexherbig/MultiVCFAnalyzer" target="_blank">http://www.github.com/alexherbig/MultiVCFAnalyzer</a></p><p>MultiVCFanalyzer
 reads multiple VCF files as produced by GATK UnifiedGenotyper, performs
 filtering based on a number of criteria, and provides the combined 
genotype calls in a number of formats that are suitable for follow-up 
analyses such as phylogenetic reconstruction, SNP effect analyses, 
population genetic analyses etc.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                MultiVCFAnalyzer file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">multivcfanalyzer:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">MultiVCFAnalyzer.json</span>
</code></pre>
          </div><h2 id="nanostat"><a href="#nanostat" class="header-link"><span class="glyphicon glyphicon-link"></span></a>NanoStat</h2><p class="mod_ext_link"><a href="https://github.com/wdecoster/nanostat" target="_blank">https://github.com/wdecoster/nanostat</a></p><p>The nanostat module parses text output generated by
<a href="https://github.com/wdecoster/nanostat/">NanoStat</a>, a program for summarising results of sequencing
on Oxford Nanopore methods (MinION, PromethION etc.)</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                NanoStat file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">nanostat:</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">4096</span>
<span class="hljs-attr">  contents_re:</span> <span class="hljs-string">'General summary:\\s*'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="ngsderive"><a href="#ngsderive" class="header-link"><span class="glyphicon glyphicon-link"></span></a>ngsderive</h2><p class="mod_ext_link"><a href="https://github.com/stjudecloud/ngsderive" target="_blank">https://github.com/stjudecloud/ngsderive</a></p><p><code>ngsderive</code>
 is a forensic analysis tool useful in backwards computing information 
from next-generation sequencing data. Notably, results are provided as a
 'best guess'  the tool does not claim 100% accuracy and results should
 be considered with that understanding.
Please see the <a href="https://stjudecloud.github.io/ngsderive">documentation</a> for more information.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                ngsderive file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">ngsderive/strandedness:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    File	TotalReads	ForwardPct	ReversePct	Predicted
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">ngsderive/instrument:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">File</span>	<span class="hljs-string">Instrument</span>	<span class="hljs-string">Confidence</span>	<span class="hljs-string">Basis</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">ngsderive/readlen:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    File	Evidence	MajorityPctDetected	ConsensusReadLength
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">ngsderive/encoding:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">File</span>	<span class="hljs-string">Evidence</span>	<span class="hljs-string">ProbableEncoding</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">ngsderive/junction_annotation:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    File	total_junctions	total_splice_events	known_junctions	partial_novel_junctions	complete_novel_junctions	known_spliced_reads	partial_novel_spliced_reads	complete_novel_spliced_reads
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="odgi"><a href="#odgi" class="header-link"><span class="glyphicon glyphicon-link"></span></a>odgi</h2><p class="mod_ext_link"><a href="https://github.com/pangenome/odgi" target="_blank">https://github.com/pangenome/odgi</a></p><p>The odgi module parses <a href="https://odgi.readthedocs.io/en/latest/rst/commands/odgi_stats.html">odgi stats</a> reports. It is specifically designed to parse the output of</p>
<pre><code class="language-sh hljs bash">odgi stats -i input_graph.og -y</code></pre>
<p>It is not guaranteed that output created using any other parameter combination can be parsed using this module.
It solely works with report files generated by <a href="https://pangenome.github.io/odgi/odgi_docs.html#_odgi_stats1">odgi stats</a> in <code>.yaml</code> format, which look like the following:</p>
<pre><code class="language-yaml hljs"><span class="hljs-meta">---</span>
<span class="hljs-attr">length:</span> <span class="hljs-number">206263</span>
<span class="hljs-attr">nodes:</span> <span class="hljs-number">3751</span>
<span class="hljs-attr">edges:</span> <span class="hljs-number">5195</span>
<span class="hljs-attr">paths:</span> <span class="hljs-number">13</span>
<span class="hljs-attr">num_weakly_connected_components:</span> <span class="hljs-number">1</span>
<span class="hljs-attr">weakly_connected_components:</span>
<span class="hljs-attr">  - component:</span>
<span class="hljs-attr">      id:</span> <span class="hljs-number">0</span>
<span class="hljs-attr">      nodes:</span> <span class="hljs-number">3751</span>
<span class="hljs-attr">      is_acyclic:</span> <span class="hljs-string">"no"</span>
<span class="hljs-attr">num_nodes_self_loops:</span>
<span class="hljs-attr">  total:</span> <span class="hljs-number">0</span>
<span class="hljs-attr">  unique:</span> <span class="hljs-number">0</span>
<span class="hljs-attr">A:</span> <span class="hljs-number">57554</span>
<span class="hljs-attr">C:</span> <span class="hljs-number">43275</span>
<span class="hljs-attr">G:</span> <span class="hljs-number">41944</span>
<span class="hljs-attr">T:</span> <span class="hljs-number">63490</span>
<span class="hljs-attr">mean_links_length:</span>
<span class="hljs-attr">  - length:</span>
<span class="hljs-attr">      path:</span> <span class="hljs-string">all_paths</span>
<span class="hljs-attr">      in_node_space:</span> <span class="hljs-number">1.64973</span>
<span class="hljs-attr">      in_nucleotide_space:</span> <span class="hljs-number">7.34035</span>
<span class="hljs-attr">      num_links_considered:</span> <span class="hljs-number">202793</span>
<span class="hljs-attr">      num_gap_links_not_penalized:</span> <span class="hljs-number">147940</span>
<span class="hljs-attr">sum_of_path_node_distances:</span>
<span class="hljs-attr">  - distance:</span>
<span class="hljs-attr">    path:</span> <span class="hljs-string">all_paths</span>
<span class="hljs-attr">    in_node_space:</span> <span class="hljs-number">5.53383</span>
<span class="hljs-attr">    in_nucleotide_space:</span> <span class="hljs-number">2.1454</span>
<span class="hljs-attr">    nodes:</span> <span class="hljs-number">202806</span>
<span class="hljs-attr">    nucleotides:</span> <span class="hljs-number">3757597</span>
<span class="hljs-attr">    num_penalties:</span> <span class="hljs-number">231</span>
<span class="hljs-attr">    num_penalties_different_orientation:</span> <span class="hljs-number">0</span></code></pre>
<p>For the odgi module to discover the <a href="https://odgi.readthedocs.io/en/latest/rst/commands/odgi_stats.html">odgi stats</a> reports, the file must match one of the following patterns:</p>
<ul>
<li>"*.og.stats.yaml"</li>
<li>"*.og.stats.yml"</li>
<li>"*.odgi.stats.yaml"</li>
<li>"*.odgi.stats.yml"</li>
</ul>
<p>A bar graph is generated, which shows the length, number of nodes, edges and paths for each sample.
Additionally, a second bar graph is generated visualizing the <code>in_node_space</code> and <code>in_nucleotide_space</code> for every sample.</p>
<p>It is possible to add custom content to your MultiQC report including
<a href="https://odgi.readthedocs.io/en/latest/rst/commands/odgi_viz.html">odgi viz</a> or <a href="https://odgi.readthedocs.io/en/latest/rst/commands/odgi_draw.html">odgi draw</a> PNGs.
Ensure that the names of the PNGs match <code>*_odgi_viz_mqc.png</code>.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                odgi file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">odgi:</span>
<span class="hljs-bullet">  -</span> 
<span class="hljs-attr">    fn:</span> <span class="hljs-string">'*.og.stats.yaml'</span>
<span class="hljs-bullet">  -</span> 
<span class="hljs-attr">    fn:</span> <span class="hljs-string">'*.og.stats.yml'</span>
<span class="hljs-bullet">  -</span> 
<span class="hljs-attr">    fn:</span> <span class="hljs-string">'*.odgi.stats.yaml'</span>
<span class="hljs-bullet">  -</span> 
<span class="hljs-attr">    fn:</span> <span class="hljs-string">'*.odgi.stats.yml'</span>
</code></pre>
          </div><h2 id="optitype"><a href="#optitype" class="header-link"><span class="glyphicon glyphicon-link"></span></a>OptiType</h2><p class="mod_ext_link"><a href="https://github.com/FRED-2/OptiType" target="_blank">https://github.com/FRED-2/OptiType</a></p><p>OptiType
 is a novel HLA genotyping algorithm based on integer linear 
programming, capable of producing accurate 4-digit HLA genotyping 
predictions from NGS data by simultaneously selecting all major and 
minor HLA Class I alleles.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                OptiType file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">optitype:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'	A1	A2	B1	B2	C1	C2	Reads	Objective'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="pangolin"><a href="#pangolin" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Pangolin</h2><p class="mod_ext_link"><a href="https://github.com/cov-lineages/pangolin" target="_blank">https://github.com/cov-lineages/pangolin</a></p><p>Pangolin <em>(Phylogenetic Assignment of Named Global Outbreak LINeages)</em> was developed to implement
the dynamic nomenclature of SARS-CoV-2 lineages, known as the Pango nomenclature.</p>
<p>It allows a user to assign a SARS-CoV-2 genome sequence the most likely lineage (Pango lineage) to SARS-CoV-2 query sequences.</p>
<p>You can find the Panogolin documentation here: <a href="https://cov-lineages.org/pangolin.html">https://cov-lineages.org/pangolin.html</a></p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Pangolin file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">pangolin:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">pangolin_version</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="peddy"><a href="#peddy" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Peddy</h2><p class="mod_ext_link"><a href="https://github.com/brentp/peddy" target="_blank">https://github.com/brentp/peddy</a></p><p>Peddy compares familial-relationships and sexes as reported in a
PED file with those inferred from a VCF.</p>
<p>It samples the VCF at about 25000 sites (plus chrX) to accurately
estimate relatedness, IBS0, heterozygosity, sex and ancestry.
It uses 2504 thousand genome samples as backgrounds to calibrate
the relatedness calculation and to make ancestry predictions.</p>
<p>It does this very quickly by sampling, by using C for computationally
intensive parts, and by parallelization.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Peddy file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">peddy/summary_table:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.peddy.ped'</span>
<span class="hljs-string">peddy/het_check:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.het_check.csv'</span>
<span class="hljs-string">peddy/ped_check:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.ped_check.csv'</span>
<span class="hljs-string">peddy/sex_check:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.sex_check.csv'</span>
<span class="hljs-string">peddy/background_pca:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.background_pca.json'</span>
</code></pre>
          </div><h2 id="phantompeakqualtools"><a href="#phantompeakqualtools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>phantompeakqualtools</h2><p class="mod_ext_link"><a href="https://www.encodeproject.org/software/phantompeakqualtools/" target="_blank">https://www.encodeproject.org/software/phantompeakqualtools/</a></p><p>Used
 to generate three quality metrics: NSC, RSC, and PBC. The NSC 
(Normalized strand cross-correlation) and RSC (relative strand 
cross-correlation) metrics use cross-correlation of stranded read 
density profiles to measure enrichment independently of peak calling. 
The PBC (PCR bottleneck coefficient) is an approximate measure of 
library complexity. PBC is the ratio of (non-redundant, uniquely 
mappable reads)/(uniquely mappable reads).</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                phantompeakqualtools file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">phantompeakqualtools/out:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.spp.out'</span>
</code></pre>
          </div><h2 id="picard"><a href="#picard" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Picard</h2><p class="mod_ext_link"><a href="http://broadinstitute.github.io/picard/" target="_blank">http://broadinstitute.github.io/picard/</a></p><p>The Picard module parses results generated by
<a href="http://broadinstitute.github.io/picard/">Picard</a>,
a set of Java command line tools for manipulating high-throughput
sequencing data.</p>
<p>Supported commands:</p>
<ul>
<li><code>AlignmentSummaryMetrics</code></li>
<li><code>BaseDistributionByCycle</code></li>
<li><code>CollectIlluminaBasecallingMetrics</code></li>
<li><code>CollectIlluminaLaneMetrics</code></li>
<li><code>CrosscheckFingerprints</code></li>
<li><code>ExtractIlluminaBarcodes</code></li>
<li><code>GcBiasMetrics</code></li>
<li><code>HsMetrics</code></li>
<li><code>InsertSizeMetrics</code></li>
<li><code>MarkDuplicates</code></li>
<li><code>MarkIlluminaAdapters</code></li>
<li><code>OxoGMetrics</code></li>
<li><code>QualityByCycleMetrics</code></li>
<li><code>QualityScoreDistributionMetrics</code></li>
<li><code>QualityYieldMetrics</code></li>
<li><code>RnaSeqMetrics</code></li>
<li><code>RrbsSummaryMetrics</code></li>
<li><code>ValidateSamFile</code></li>
<li><code>VariantCallingMetrics</code></li>
<li><code>WgsMetrics</code></li>
</ul>
<h4 id="coverage-levels"><a href="#coverage-levels" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Coverage Levels</h4>
<p>It's possible to customise the HsMetrics <em>"Target Bases 30X"</em> coverage and
WgsMetrics <em>"Fraction of Bases over 30X"</em> that are
shown in the general statistics table. This must correspond to field names in the
picard report, such as <code>PCT_TARGET_BASES_2X</code> / <code>PCT_10X</code>. Any numbers not found in the
reports will be ignored.</p>
<p>The coverage levels available for HsMetrics are
<a href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#HsMetrics">typically</a>
1, 2, 10, 20, 30, 40, 50 and 100X.</p>
<p>The coverage levels available for WgsMetrics are
<a href="http://broadinstitute.github.io/picard/picard-metric-definitions.html#CollectWgsMetrics.WgsMetrics">typically</a>
1, 5, 10, 15, 20, 25, 30, 40, 50, 60, 70, 80, 90 and 100X.</p>
<p>To customise this, add the following to your MultiQC config:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">picard_config:</span>
<span class="hljs-attr">  general_stats_target_coverage:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">10</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">50</span></code></pre>
<h4 id="crosscheckfingerprints"><a href="#crosscheckfingerprints" class="header-link"><span class="glyphicon glyphicon-link"></span></a>CrosscheckFingerprints</h4>
<p>In addition to adding a table of results, a <code>Crosschecks All Expected</code> column will be added to the General Statistics. If all comparisons for a sample were <code>Expected</code>, then the value of the field will be <code>True</code> and green. If not it will be <code>False</code> and Red.</p>
<p>You can customize the columns show in the CrosscheckFingerprints table with the config keys <code>CrosscheckFingerprints_table_cols</code> and <code>CrosscheckFingerprints_table_cols_hidden</code>. For example:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">picard_config:</span>
<span class="hljs-attr">  CrosscheckFingerprints_table_cols:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">RESULT</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">LOD_SCORE</span>
<span class="hljs-attr">  CrosscheckFingerprints_table_cols_hidden:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">LEFT_LANE</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">RIGHT_LANE</span></code></pre>
<p>The column names will be normalized, ex <code>LOD_SCORE -&gt; Lod score</code>.</p>
<p>Note that if <code>CALCULATE_TUMOR_AWARE_RESULTS</code> was set to true on the CLI for any of the CrosscheckFingerprints result files, then the <code>LOD_SCORE_TUMOR_NORMAL</code> and <code>LOD_SCORE_NORMAL_TUMOR</code> will be displayed.</p>
<h4 id="hsmetrics"><a href="#hsmetrics" class="header-link"><span class="glyphicon glyphicon-link"></span></a>HsMetrics</h4>
<p>Note that the <em>Target Region Coverage</em> plot is generated using the <code>PCT_TARGET_BASES_</code> table columns from the HsMetrics output (not immediately obvious when looking at the log files).</p>
<p>You can customize the columns shown in the HsMetrics table with the config keys <code>HsMetrics_table_cols</code> and <code>HsMetrics_table_cols_hidden</code>. For example:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">picard_config:</span>
<span class="hljs-attr">  HsMetrics_table_cols:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">NEAR_BAIT_BASES</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">OFF_BAIT_BASES</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">ON_BAIT_BASES</span>
<span class="hljs-attr">  HsMetrics_table_cols_hidden:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">MAX_TARGET_COVERAGE</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">MEAN_BAIT_COVERAGE</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">MEAN_TARGET_COVERAGE</span></code></pre>
<p>Only values listed in <code>HsMetrics_table_cols</code> will be included in the table.
Anything listed in <code>HsMetrics_table_cols_hidden</code> will be hidden by default.</p>
<h4 id="insertsizemetrics"><a href="#insertsizemetrics" class="header-link"><span class="glyphicon glyphicon-link"></span></a>InsertSizeMetrics</h4>
<p>By default, the insert size plot is smoothed to contain a maximum of 500 data points per sample.
This is to prevent the MultiQC report from being very large with big datasets.
If you would like to customise this value to get a better resolution you can set the following
MultiQC config values, with the new maximum number of points:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">picard_config:</span>
<span class="hljs-attr">  insertsize_smooth_points:</span> <span class="hljs-number">10000</span></code></pre>
<p>The plotted maximum insert size can be set with:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">picard_config:</span>
<span class="hljs-attr">  insertsize_xmax:</span> <span class="hljs-number">10000</span></code></pre>
<h4 id="markduplicates"><a href="#markduplicates" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MarkDuplicates</h4>
<p>If a <code>BAM</code> file contains multiple read groups, Picard MarkDuplicates generates a report
with multiple metric lines, one for each "library".</p>
<p>By default, MultiQC will sum the values for every library it finds and recompute the
<code>PERCENT_DUPLICATION</code> and <code>ESTIMATED_LIBRARY_SIZE</code> fields, giving a single set of results
for each <code>BAM</code> file.</p>
<p>If instead you would prefer each library to be treated as a separate sample, you can do so
by setting the following MultiQC config:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">picard_config:</span>
<span class="hljs-attr">  markdups_merge_multiple_libraries:</span> <span class="hljs-literal">False</span></code></pre>
<p>This prevents the merge and recalculation and appends the library name to the sample name.</p>
<p>This behaviour is present in MultiQC since version 1.9. Before this, only the metrics from the
first library were taken and all others were ignored.</p>
<h4 id="validatesamfile-search-pattern"><a href="#validatesamfile-search-pattern" class="header-link"><span class="glyphicon glyphicon-link"></span></a>ValidateSamFile Search Pattern</h4>
<p>Generally, Picard adds identifiable content to the output of function
 calls. This is not the case for ValidateSamFile. In order to identify 
logs the MultiQC Picard submodule <code>ValidateSamFile</code> will 
search for filenames that contain 'validatesamfile' or 
'ValidateSamFile'. One can customise the used search pattern by 
overwriting the <code>picard/sam_file_validation</code> pattern in your MultiQC config. For example:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">sp:</span>
  <span class="hljs-string">picard/sam_file_validation:</span>
<span class="hljs-attr">    fn:</span> <span class="hljs-string">"*[Vv]alidate[Ss]am[Ff]ile*"</span></code></pre>
<h4 id="wgsmetrics"><a href="#wgsmetrics" class="header-link"><span class="glyphicon glyphicon-link"></span></a>WgsMetrics</h4>
<p>The coverage histogram from Picard typically shows a normal distribution with a very long tail.
To make the plot easier to view, by default the module plots the line up to 99% of the data.
This typically removes the long tail and gives a more useful graph.</p>
<p>If you would like, you can set a specific value for the maximum coverage to cut the graph at.
By setting this to a very large value, you will disable the cutting (the graph will automatically
limit the axis at the maximum data point). You can do this as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">picard_config:</span>
<span class="hljs-attr">  wgsmetrics_histogram_max_cov:</span> <span class="hljs-number">500</span></code></pre>
<p>If running with very high coverage samples or using the Picard <code>CAP_COVERAGE</code> option,
the coverage histogram can become very large indeed. For eaxmple, if reporting coverages of 1 million,
it will have 1 million data points per sample. That can crash the browser and take a long time to run.</p>
<p>There are two customisation MultiQC options to help with this.
Firstly, MultiQC will automatically "smooth" the histogram to a maximum of <code>1000</code> data points by binning.
This should stop the browser from crashing. You can tweak how many bins are used with the following:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">picard_config:</span>
<span class="hljs-attr">  wgsmetrics_histogram_smooth:</span> <span class="hljs-number">1000</span></code></pre>
<p>Change <code>1000</code> to whatever number you want. If you don't want any smoothing, set it to a very high number
bigger than the number of data points you have.</p>
<p>Secondly, if you would prefer to instead simply skip the histogram, you can set the following:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">picard_config:</span>
<span class="hljs-attr">  wgsmetrics_skip_histogram:</span> <span class="hljs-literal">True</span></code></pre>
<p>This will omit that section from the report entirely, and also skip parsing the histogram data.
By specifying this option you may speed up the run time for MultiQC with these types of files
significantly.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Picard file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">picard/alignment_metrics:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">AlignmentSummaryMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/basedistributionbycycle:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">BaseDistributionByCycleMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/crosscheckfingerprints:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">CrosscheckFingerprints</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/gcbias:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">GcBias</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/hsmetrics:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">HsMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/insertsize:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">InsertSizeMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/markdups:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">DuplicationMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/oxogmetrics:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">OxoGMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/pcr_metrics:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">TargetedPcrMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/quality_by_cycle:</span>
<span class="hljs-attr">  contents_re:</span> <span class="hljs-string">'[Qq]uality[Bb]y[Cc]ycle'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">MEAN_QUALITY</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/quality_score_distribution:</span>
<span class="hljs-attr">  contents_re:</span> <span class="hljs-string">'[Qq]uality[Ss]core[Dd]istribution'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">COUNT_OF_Q</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/quality_yield_metrics:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">QualityYieldMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/rnaseqmetrics:</span>
<span class="hljs-attr">  contents_re:</span> <span class="hljs-string">'[Rr]na[Ss]eq[Mm]etrics'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'## METRICS CLASS'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/rrbs_metrics:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">RrbsSummaryMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/sam_file_validation:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*[Vv]alidate[Ss]am[Ff]ile*'</span>
<span class="hljs-string">picard/variant_calling_metrics:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.variant_calling_detail_metrics'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">CollectVariantCallingMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/wgs_metrics:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">CollectWgsMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/collectilluminabasecallingmetrics:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">CollectIlluminaBasecallingMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/collectilluminalanemetrics:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">CollectIlluminaLaneMetrics</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/extractilluminabarcodes:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">ExtractIlluminaBarcodes</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">picard/markilluminaadapters:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">MarkIlluminaAdapters</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="preseq"><a href="#preseq" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Preseq</h2><p class="mod_ext_link"><a href="http://smithlabresearch.org/software/preseq/" target="_blank">http://smithlabresearch.org/software/preseq/</a></p><p>The Preseq module parses results generated by
<a href="http://smithlabresearch.org/software/preseq/">Preseq</a>,
a tool that estimates the complexity of a library, showing how many additional
unique reads are sequenced for increasing total read count.</p>
<p>When <code>preseq lc_extrap</code> is run with the default parameters, the extrapolation points
reach 10 billion molecules making the plot difficult to interpret in most scenarios.
It also includes a lot of data in the reports, which can unnecessarily inflate report
file sizes. To avoid this, MultiQC trims back the x axis until each dataset
shows 80% of its maximum y-value (unique molecules).</p>
<p>To disable this feature and show all of the data, add the following to your
<a href="http://multiqc.info/docs/#configuring-multiqc">MultiQC configuration</a>:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">preseq:</span>
<span class="hljs-attr">  notrim:</span> <span class="hljs-literal">true</span></code></pre>
<h4 id="using-coverage-instead-of-read-counts"><a href="#using-coverage-instead-of-read-counts" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Using coverage instead of read counts</h4>
<p>Preseq reports its numbers as "Molecule counts". This isn't always very intuitive,
and it's often easier to talk about sequencing depth in terms of coverage.
You can plot the estimated coverage instead by specifying the reference genome or target size,
and the read length in your <a href="http://multiqc.info/docs/#configuring-multiqc">MultiQC configuration</a>:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">preseq:</span>
<span class="hljs-attr">  genome_size:</span> <span class="hljs-number">3049315783</span>
<span class="hljs-attr">  read_length:</span> <span class="hljs-number">300</span></code></pre>
<p>These parameters make the script take every molecule count and divide it by
(genome_size / read_length).</p>
<p>MultiQC comes with effective genome size presets for Human and Mouse, so you can
provide the genome build name instead, like this: <code>genome_size: hg38_genome</code>. The
following values are supported: <code>hg19_genome</code>, <code>hg38_genome</code>, <code>mm10_genome</code>.</p>
<p>When the genome and read sizes are provided, MultiQC will plot the molecule counts
on the X axis ("total" data) and coverages on the Y axis ("unique" data).
However, you can customize what to plot on each axis (counts or coverage), e.g.:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">preseq:</span>
<span class="hljs-attr">  x_axis:</span> <span class="hljs-string">counts</span>
<span class="hljs-attr">  y_axis:</span> <span class="hljs-string">coverage</span></code></pre>
<h4 id="plotting-externally-calculated-read-counts"><a href="#plotting-externally-calculated-read-counts" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Plotting externally calculated read counts</h4>
<p>To mark on the plot the read counts calculated externally from BAM or fastq files,
create a file with <code>preseq_real_counts</code> in the filename and place it with your analysis files.
It should be space or tab delimited with 2 or 3 columns (column 1 = preseq file name,
column 2 = real read count, optional column 3 = real unique read count). For example:</p>
<pre><code class="language-tsv">Sample_1.preseq.txt 3638261 3638011
Sample_2.preseq.txt 1592394 1592133
[...]</code></pre>
<p>You can generate a line for such a file using samtools:</p>
<pre><code class="language-bash hljs"><span class="hljs-built_in">echo</span> <span class="hljs-string">"Sample_1.preseq.txt "</span>$(samtools view -c -F 4 Sample_1.bam)<span class="hljs-string">" "</span>$(samtools view -c -F 1028 Sample_1.bam)</code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Preseq file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">preseq:</span>
<span class="hljs-bullet">  -</span> 
<span class="hljs-attr">    contents:</span> <span class="hljs-string">EXPECTED_DISTINCT</span>
<span class="hljs-attr">    num_lines:</span> <span class="hljs-number">2</span>
<span class="hljs-bullet">  -</span> 
<span class="hljs-attr">    contents:</span> <span class="hljs-string">distinct_reads</span>
<span class="hljs-attr">    num_lines:</span> <span class="hljs-number">2</span>
<span class="hljs-string">preseq/real_counts:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*preseq_real_counts*'</span>
</code></pre>
          </div><h2 id="prokka"><a href="#prokka" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Prokka</h2><p class="mod_ext_link"><a href="http://www.vicbioinformatics.com/software.prokka.shtml" target="_blank">http://www.vicbioinformatics.com/software.prokka.shtml</a></p><p>The Prokka module analyses summary results from the
<a href="http://www.vicbioinformatics.com/software.prokka.shtml">Prokka</a> annotation
pipeline for prokaryotic genomes.
The Prokka module accepts two configuration options:</p>
<ul>
<li><code>prokka_table</code>: default <code>False</code>. Show a table in the report.</li>
<li><code>prokka_barplot</code>: default <code>True</code>. Show a barplot in the report.</li>
<li><code>prokka_fn_snames</code>: default <code>False</code>. Use filenames for sample names (see below).</li>
</ul>
<p>Sample names are generated using the first line in the prokka reports:</p>
<pre><code class="language-txt">organism: Helicobacter pylori Sample1</code></pre>
<p>The module assumes that the first two words are the organism name and
the third is the sample name. So the above will give a sample name of
<code>Sample1</code>.</p>
<p>If you prefer, you can set <code>config.prokka_fn_snames</code> to <code>True</code> and MultiQC
will instead use the log filename as the sample name.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Prokka file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">prokka:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'contigs:'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
</code></pre>
          </div><h2 id="purple"><a href="#purple" class="header-link"><span class="glyphicon glyphicon-link"></span></a>PURPLE</h2><p class="mod_ext_link"><a href="https://github.com/hartwigmedical/hmftools/tree/master/purity-ploidy-estimator" target="_blank">https://github.com/hartwigmedical/hmftools/tree/master/purity-ploidy-estimator</a></p><p>Purple combines B-allele frequency (BAF), read depth ratios, somatic variants and
structural variant breakpoints to estimate the purity and copy number profile
of a tumor sample, and also predicts gender, the MSI status, tumor mutational
load and burden, clonality and the whole genome duplication status.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                PURPLE file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">purple/qc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.purple.qc'</span>
<span class="hljs-string">purple/purity:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.purple.purity.tsv'</span>
</code></pre>
          </div><h2 id="qc3c"><a href="#qc3c" class="header-link"><span class="glyphicon glyphicon-link"></span></a>qc3C</h2><p class="mod_ext_link"><a href="http://github.com/cerebis/qc3C" target="_blank">http://github.com/cerebis/qc3C</a></p><p>qc3C
 allows researchers to assess the fraction of read-pairs within a Hi-C 
library that are a product of proximity ligation -- in effect the Hi-C 
signal strength.
Using a k-mer based approach, signal strength is inferred directly from 
reads and therefore no reference is required.
Reference based assessment is also available and can provide further 
details.</p>
<p>With this information in hand, researchers are able to decide how 
much sequencing will be needed to achieve their experimental aims.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                qc3C file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">qc3C:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.qc3C.json'</span>
</code></pre>
          </div><h2 id="qorts"><a href="#qorts" class="header-link"><span class="glyphicon glyphicon-link"></span></a>QoRTs</h2><p class="mod_ext_link"><a href="http://hartleys.github.io/QoRTs/" target="_blank">http://hartleys.github.io/QoRTs/</a></p><p>The QoRTs software package is a fast, efficient, and portable multifunction
toolkit designed to assist in the analysis, quality control, and data
management of RNA-Seq datasets. Its primary function is to aid in the
detection and identification of errors, biases, and artifacts produced
by paired-end high-throughput RNA-Seq technology. In addition, it can produce
count data designed for use with differential expression and differential
exon usage tools, as well as individual-sample and/or group-summary genome
track files suitable for use with the UCSC genome browser.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                QoRTs file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">qorts:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">BENCHMARK_MinutesOnSamIteration</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">100</span>
</code></pre>
          </div><h2 id="qualimap"><a href="#qualimap" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Qualimap</h2><p class="mod_ext_link"><a href="http://qualimap.bioinfo.cipf.es/" target="_blank">http://qualimap.bioinfo.cipf.es/</a></p><p>The Qualimap module parses results generated by
<a href="http://qualimap.bioinfo.cipf.es/">Qualimap</a>,
a platform-independent application to facilitate the quality
control of alignment sequencing data and its derivatives like feature
counts.</p>
<p>The MultiQC module supports the Qualimap commands <code>BamQC</code> and <code>RNASeq</code>.
Note that Qualimap must be run with the <code>-outdir</code> option as well as
<code>-outformat HTML</code> (which is on by default). MultiQC uses files
found within the <code>raw_data_qualimapReport</code> folder (as well as <code>genome_results.txt</code>).</p>
<p>Qualimap adds lots of columns to the General Statistics table. To avoid making the table
too wide and bloated, some of these are hidden by default (<code>Error Rate</code>, <code>M Aligned</code>, <code>M Total reads</code>).
You can override these defaults in your MultiQC config file - for example, to show
<code>Error Rate</code> by default and hide <code>Ins. size</code> by default, add the following:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">table_columns_visible:</span>
<span class="hljs-attr">  QualiMap:</span>
<span class="hljs-attr">    general_error_rate:</span> <span class="hljs-literal">True</span>
<span class="hljs-attr">    median_insert_size:</span> <span class="hljs-literal">False</span></code></pre>
<p>See the <a href="http://multiqc.info/docs/#hiding-columns">relevant section of the documentation</a> for more detail.</p>
<p>In addition to this, it's possible to customise which coverage thresholds calculated
by the Qualimap BamQC module <em>(default: 1, 5, 10, 30, 50)</em> and which of these are hidden in the
General Statistics tablewhen the report loads <em>(default: all hidden except 30X)</em>.</p>
<p>To do this, add something like the following to your MultiQC config file:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">qualimap_config:</span>
<span class="hljs-attr">  general_stats_coverage:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">10</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">20</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">40</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">200</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">30000</span>
<span class="hljs-attr">  general_stats_coverage_hidden:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">10</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">20</span>
<span class="hljs-bullet">    -</span> <span class="hljs-number">200</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Qualimap file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">qualimap/bamqc/genome_results:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">genome_results.txt</span>
<span class="hljs-string">qualimap/bamqc/coverage:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">coverage_histogram.txt</span>
<span class="hljs-string">qualimap/bamqc/insert_size:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">insert_size_histogram.txt</span>
<span class="hljs-string">qualimap/bamqc/genome_fraction:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">genome_fraction_coverage.txt</span>
<span class="hljs-string">qualimap/bamqc/gc_dist:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">mapped_reads_gc-content_distribution.txt</span>
<span class="hljs-string">qualimap/rnaseq/rnaseq_results:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">rnaseq_qc_results.txt</span>
<span class="hljs-string">qualimap/rnaseq/coverage:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">coverage_profile_along_genes_(total).txt</span>
</code></pre>
          </div><h2 id="quast"><a href="#quast" class="header-link"><span class="glyphicon glyphicon-link"></span></a>QUAST</h2><p class="mod_ext_link"><a href="http://quast.bioinf.spbau.ru/" target="_blank">http://quast.bioinf.spbau.ru/</a></p><p><a href="http://quast.bioinf.spbau.ru/">QUAST</a> evaluates genome assemblies by
computing various metrics, including</p>
<ul>
<li>N50, length for which the collection of all contigs of that length or
longer covers at least 50% of assembly length</li>
<li>NG50, where length of the reference genome is being covered</li>
<li>NA50 and NGA50, where aligned blocks instead of contigs are taken</li>
<li>Misassemblies, misassembled and unaligned contigs or contigs bases</li>
<li>Genes and operons covered</li>
</ul>
<p>The QUAST MultiQC module parses the <code>report.tsv</code> files generated by
QUAST and adds key metrics to the report General Statistics table.
All statistics for all samples are saved to <code>multiqc_data/multiqc_quast.txt</code>.</p>
<h4 id="configuration"><a href="#configuration" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Configuration</h4>
<p>By default, the QUAST module is configured to work with large <em>de-novo</em> genomes,
showing thousands of contigs, mega-base pairs and other sensible defaults.</p>
<p>If these aren't appropriate for your genomes, you can configure them as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">quast_config:</span>
<span class="hljs-attr">  contig_length_multiplier:</span> <span class="hljs-number">0.001</span>
<span class="hljs-attr">  contig_length_suffix:</span> <span class="hljs-string">"Kbp"</span>
<span class="hljs-attr">  total_length_multiplier:</span> <span class="hljs-number">0.000001</span>
<span class="hljs-attr">  total_length_suffix:</span> <span class="hljs-string">"Mbp"</span>
<span class="hljs-attr">  total_number_contigs_multiplier:</span> <span class="hljs-number">0.001</span>
<span class="hljs-attr">  total_number_contigs_suffix:</span> <span class="hljs-string">"K"</span></code></pre>
<p>The default module values are shown above. See the
<a href="http://multiqc.info/docs/#configuring-multiqc">main MultiQC documentation</a>
for more information about how to configure MultiQC.</p>
<h4 id="metaquast"><a href="#metaquast" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MetaQUAST</h4>
<p>The QUAST module will also parse output from
<a href="http://quast.sourceforge.net/metaquast">MetaQUAST</a> runs (<code>metaquast.py</code>).</p>
<p>The <code>combined_reference/report.tsv</code> file is parsed, and folders
<code>runs_per_reference</code> and <code>not_aligned</code> are ignored.</p>
<p>If you want to run MultiQC against auxiliary MetaQUAST runs, you must
explicitly pass these files to MultiQC:</p>
<pre><code class="language-bash hljs">multiqc runs_per_reference/reference_1/report.tsv</code></pre>
<p>Note that you can pass as many file paths to MultiQC as you like
and use glob expansion (eg. <code>runs_per_reference/*/report.tsv</code>).</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                QUAST file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">quast:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">report.tsv</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="rna-seqc"><a href="#rna-seqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>RNA-SeQC</h2><p class="mod_ext_link"><a href="https://github.com/broadinstitute/rnaseqc" target="_blank">https://github.com/broadinstitute/rnaseqc</a></p><p>The RSeQC module parses results generated by
<a href="https://github.com/broadinstitute/rnaseqc">RNA-SeQC</a>,
(not to be confused with <a href="http://rseqc.sourceforge.net/">RSeQC</a>,
which MultiQC also supports).</p>
<p>RNA-SeQC is a java program which computes a series of quality
control metrics for RNA-seq data.</p>
<p>This module shows the Spearman correlation heatmap if both
Spearman and Pearson's are found. To plot Pearson's by default instead,
add the following to your MultiQC config file:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">rna_seqc:</span>
<span class="hljs-attr">  default_correlation:</span> <span class="hljs-string">pearson</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                RNA-SeQC file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">rna_seqc/metrics_v1:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*metrics.tsv'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'Sample	Note	'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">rna_seqc/metrics_v2:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*metrics.tsv'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">High</span> <span class="hljs-string">Quality</span> <span class="hljs-string">Ambiguous</span> <span class="hljs-string">Alignment</span> <span class="hljs-string">Rate</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">rna_seqc/coverage:</span>
<span class="hljs-attr">  fn_re:</span> <span class="hljs-string">meanCoverageNorm_(high|medium|low)\.txt</span>
<span class="hljs-string">rna_seqc/correlation:</span>
<span class="hljs-attr">  fn_re:</span> <span class="hljs-string">corrMatrix(Pearson|Spearman)\.txt</span>
</code></pre>
          </div><h2 id="rockhopper"><a href="#rockhopper" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Rockhopper</h2><p class="mod_ext_link"><a href="https://cs.wellesley.edu/~btjaden/Rockhopper/index.html" target="_blank">https://cs.wellesley.edu/~btjaden/Rockhopper/index.html</a></p><p>Rockhopper
 aligns reads to coding sequences, rRNAs, tRNAs, and miscellaneous RNAs 
on both the sense and anti-sense strand. These statistics are summarized
 in the Rockhopper bar plot in this module.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Rockhopper file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">rockhopper:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">summary.txt</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    Number of gene-pairs predicted to be
    part of the same operon
</span><span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">500000</span>
</code></pre>
          </div><h2 id="rsem"><a href="#rsem" class="header-link"><span class="glyphicon glyphicon-link"></span></a>RSEM</h2><p class="mod_ext_link"><a href="https://deweylab.github.io/RSEM/" target="_blank">https://deweylab.github.io/RSEM/</a></p><p>The rsem module parses results generated by
<a href="https://deweylab.github.io/RSEM/">RSEM</a>,
a software package for estimating gene and isoform expression levels
from RNA-Seq data</p>
<p>Supported scripts:</p>
<ul>
<li><code>rsem-calculate-expression</code></li>
</ul>
<p>This module search for the file <code>.cnt</code> created by RSEM into directory named <code>PREFIX.stat</code></p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                RSEM file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">rsem:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.cnt'</span>
</code></pre>
          </div><h2 id="rseqc"><a href="#rseqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>RSeQC</h2><p class="mod_ext_link"><a href="http://rseqc.sourceforge.net/" target="_blank">http://rseqc.sourceforge.net/</a></p><p>The RSeQC module parses results generated by
<a href="http://rseqc.sourceforge.net/">RSeQC</a>,
a package that provides a number of useful modules that can
comprehensively evaluate high throughput RNA-seq data.</p>
<p>Supported scripts:</p>
<ul>
<li><code>bam_stat</code></li>
<li><code>gene_body_coverage</code></li>
<li><code>infer_experiment</code></li>
<li><code>inner_distance</code></li>
<li><code>junction_annotation</code></li>
<li><code>junction_saturation</code></li>
<li><code>read_distribution</code></li>
<li><code>read_duplication</code></li>
<li><code>read_gc</code></li>
<li><code>tin</code></li>
</ul>
<p>You can choose to hide sections of RSeQC output and customise their order.
To do this, add and customise the following to your MultiQC config file:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">rseqc_sections:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">read_distribution</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">tin</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">gene_body_coverage</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">inner_distance</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">read_gc</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">read_duplication</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">junction_annotation</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">junction_saturation</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">infer_experiment</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">bam_stat</span></code></pre>
<p>Change the order to rearrage sections or remove to hide them from the report.</p>
<p>Note that some scripts (for example, <code>junction_annotation.py</code>) produce the results used by MultiQC as standard-error.
To use with MultiQC, make sure that you redirect this to a file using <code>2&gt; mysample.log</code>.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                RSeQC file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">rseqc/bam_stat:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'Proper-paired reads map to different chrom:'</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">500000</span>
<span class="hljs-string">rseqc/gene_body_coverage:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.geneBodyCoverage.txt'</span>
<span class="hljs-string">rseqc/inner_distance:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.inner_distance_freq.txt'</span>
<span class="hljs-string">rseqc/junction_annotation:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'Partial Novel Splicing Junctions:'</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">500000</span>
<span class="hljs-string">rseqc/junction_saturation:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.junctionSaturation_plot.r'</span>
<span class="hljs-string">rseqc/read_gc:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.GC.xls'</span>
<span class="hljs-string">rseqc/read_distribution:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'Group               Total_bases         Tag_count           Tags/Kb'</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">500000</span>
<span class="hljs-string">rseqc/read_duplication_pos:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.pos.DupRate.xls'</span>
<span class="hljs-string">rseqc/infer_experiment:</span>
<span class="hljs-bullet">  -</span> 
<span class="hljs-attr">    fn:</span> <span class="hljs-string">'*infer_experiment.txt'</span>
<span class="hljs-bullet">  -</span> 
<span class="hljs-attr">    contents:</span> <span class="hljs-string">Fraction</span> <span class="hljs-string">of</span> <span class="hljs-string">reads</span> <span class="hljs-string">explained</span> <span class="hljs-string">by</span>
<span class="hljs-attr">    max_filesize:</span> <span class="hljs-number">500000</span>
<span class="hljs-string">rseqc/tin:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.summary.txt'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">TIN(median)</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="sambamba"><a href="#sambamba" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Sambamba</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#markdup">markdup</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div><p class="mod_ext_link"><a href="https://lomereiter.github.io/sambamba/" target="_blank">https://lomereiter.github.io/sambamba/</a></p><p><a href="https://lomereiter.github.io/sambamba/">Sambamba</a> is a suite of programs for
users to quickly and efficiently process their high-throughput sequencing data.
It is functionally similar to Samtools, but the source code is written in the
D Language; it allows for faster performance while still being easy to use.</p>
<p>Supported commands:</p>
<ul>
<li><code>markdup</code></li>
</ul>
<h3 id="markdup"><a href="#markdup" class="header-link"><span class="glyphicon glyphicon-link"></span></a>markdup</h3>
<p>This module parses key phrases in the output log files to find duplicate +
unique reads and then calculates duplicate rate per sample. It will
will work for both single and paired-end data.
The absolute number of reads by type are displayed in a stacked bar plot,
and duplicate rates are in the general statistics table.</p>
<p>Duplicate rates are calculated as follows:</p>
<h4 id="paired-end"><a href="#paired-end" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Paired end</h4>
<blockquote>
<p><code>duplicate_rate = duplicateReads / (sortedEndPairs * 2 + singleEnds - singleUnmatchedPairs) * 100</code></p>
</blockquote>
<h4 id="single-end"><a href="#single-end" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Single end</h4>
<blockquote>
<p><code>duplicate_rate = duplicateReads / singleEnds * 100</code></p>
</blockquote>
<p>If Sambamba Markdup is invoked using Snakemake, the following bare-bones
rule should work fine:</p>
<pre><code class="language-python hljs">rule markdup:
  input:
    <span class="hljs-string">"data/align/{sample}.bam"</span>
  output:
    <span class="hljs-string">"data/markdup/{sample}.markdup.bam"</span>
  log:
    <span class="hljs-string">"data/logs/{sample}.log"</span>
  shell:
    <span class="hljs-string">"sambamba markdup {input} {output} &gt; {log} 2&gt;&amp;1"</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Sambamba file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">sambamba/markdup:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    finding positions of the duplicate reads
    in the file
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">50</span>
</code></pre>
          </div><h2 id="samblaster"><a href="#samblaster" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Samblaster</h2><p class="mod_ext_link"><a href="https://github.com/GregoryFaust/samblaster" target="_blank">https://github.com/GregoryFaust/samblaster</a></p><p>The Samblaster module parses results generated by
<a href="https://github.com/GregoryFaust/samblaster">Samblaster</a>,
a tool to mark duplicates and extract discordant and split
reads from sam files.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Samblaster file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">samblaster:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'samblaster: Version'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="samtools"><a href="#samtools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Samtools</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#idxstats">idxstats</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div><p class="mod_ext_link"><a href="http://www.htslib.org/" target="_blank">http://www.htslib.org</a></p><p>The Samtools module parses results generated by
<a href="http://www.htslib.org/">Samtools</a>,
a suite of programs for interacting with high-throughput
sequencing data.</p>
<p>Supported commands:</p>
<ul>
<li><code>stats</code></li>
<li><code>flagstats</code></li>
<li><code>idxstats</code></li>
<li><code>rmdup</code></li>
</ul>
<h3 id="idxstats"><a href="#idxstats" class="header-link"><span class="glyphicon glyphicon-link"></span></a>idxstats</h3>
<p>The <code>samtools idxstats</code> prints its results to standard
out (no consistent file name) and has no header lines
(no way to recognise from content of file). As such, <code>idxstats</code>
result files must have the string <code>idxstat</code> somewhere in the filename.</p>
<p>There are a few MultiQC config options that you can add to
customise how the idxstats module works. A typical configuration
could look as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-comment"># Always include these chromosomes in the plot</span>
<span class="hljs-attr">samtools_idxstats_always:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">X</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">Y</span>

<span class="hljs-comment"># Never include these chromosomes in the plot</span>
<span class="hljs-attr">samtools_idxstats_ignore:</span>
<span class="hljs-bullet">  -</span> <span class="hljs-string">MT</span>

<span class="hljs-comment"># Threshold where chromosomes are ignored in the plot.</span>
<span class="hljs-comment"># Should be a fraction, default is 0.001 (0.1% of total)</span>
<span class="hljs-attr">samtools_idxstats_fraction_cutoff:</span> <span class="hljs-number">0.001</span>

<span class="hljs-comment"># Name of the X and Y chromosomes.</span>
<span class="hljs-comment"># If not specified, MultiQC will search for any chromosome</span>
<span class="hljs-comment"># names that look like x, y, chrx or chry (case insensitive search)</span>
<span class="hljs-attr">samtools_idxstats_xchr:</span> <span class="hljs-string">myXchr</span>
<span class="hljs-attr">samtools_idxstats_ychr:</span> <span class="hljs-string">myYchr</span></code></pre>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Samtools file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">samtools/stats:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">This</span> <span class="hljs-string">file</span> <span class="hljs-string">was</span> <span class="hljs-string">produced</span> <span class="hljs-string">by</span> <span class="hljs-string">samtools</span> <span class="hljs-string">stats</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">samtools/flagstat:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">&gt;
    in total (QC-passed reads + QC-failed
    reads)
</span><span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">samtools/idxstats:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*idxstat*'</span>
<span class="hljs-string">samtools/rmdup:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'[bam_rmdup'</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="sargasso"><a href="#sargasso" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Sargasso</h2><p class="mod_ext_link"><a href="http://statbio.github.io/Sargasso/" target="_blank">http://statbio.github.io/Sargasso/</a></p><p>The sargasso module parses results generated by
<a href="http://statbio.github.io/Sargasso/">Sargasso</a>,
a tool for separating mixed-species RNA-seq reads according to
their species of origin.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Sargasso file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">sargasso:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">overall_filtering_summary.txt</span>
</code></pre>
          </div><h2 id="sentieon"><a href="#sentieon" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Sentieon</h2><p class="mod_ext_link"><a href="https://www.sentieon.com/products/" target="_blank">https://www.sentieon.com/products/</a></p><p>The Sentieon module parses output from the Sentieon dna-seq suite of tools,
which themselves are implementations of certain Picard metrics
<a href="http://broadinstitute.github.io/picard/">Picard</a>,
a set of Java command line tools for manipulating high-throughput
sequencing data.</p>
<p>Supported commands:</p>
<ul>
<li><code>InsertSizeMetrics</code></li>
<li><code>GcBiasMetrics</code></li>
<li><code>AlignmentSummaryMetrics</code></li>
</ul>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Sentieon file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">sentieon/alignment_metrics:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-bullet">--algo</span> <span class="hljs-string">AlignmentStat</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">sentieon/insertsize:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-bullet">--algo</span> <span class="hljs-string">InsertSizeMetricAlgo</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
<span class="hljs-string">sentieon/gcbias:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-bullet">--algo</span> <span class="hljs-string">GCBias</span>
<span class="hljs-attr">  shared:</span> <span class="hljs-literal">true</span>
</code></pre>
          </div><h2 id="sex.deterrmine"><a href="#sex.deterrmine" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Sex.DetErrMine</h2><p class="mod_ext_link"><a href="https://github.com/TCLamnidis/Sex.DetERRmine" target="_blank">https://github.com/TCLamnidis/Sex.DetERRmine</a></p><p>A
 python script to calculate the relative coverage of X and Y 
chromosomes, and their associated error bars, from the depth of coverage
 at specified SNPs.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Sex.DetErrMine file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">sexdeterrmine:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">sexdeterrmine.json</span>
</code></pre>
          </div><h2 id="slamdunk"><a href="#slamdunk" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Slamdunk</h2><p class="mod_ext_link"><a href="http://t-neumann.github.io/slamdunk/" target="_blank">http://t-neumann.github.io/slamdunk/</a></p><p><a href="http://t-neumann.github.io/slamdunk/">Slamdunk</a> is a tool to analyze data from the SLAM-Seq sequencing protocol.</p>
<p>This module should be able to parse logs from v0.2.2-dev onwards.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Slamdunk file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">slamdunk/summary:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'# slamdunk summary'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">slamdunk/PCA:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'# slamdunk PCA'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">slamdunk/rates:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'# slamdunk rates'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">slamdunk/utrrates:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'# slamdunk utrrates'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">slamdunk/tcperreadpos:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'# slamdunk tcperreadpos'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-string">slamdunk/tcperutrpos:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'# slamdunk tcperutr'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="snippy"><a href="#snippy" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Snippy</h2><p class="mod_ext_link"><a href="https://github.com/tseemann/snippy" target="_blank">https://github.com/tseemann/snippy</a></p><p>The Snippy module parses output generated by
<a href="https://github.com/tseemann/snippy">snippy</a>. The following commands are implemented:</p>
<ul>
<li><code>snippy</code>
<ul>
<li>Variant type descriptive statistics.</li>
<li>Parses summary <code>prefix.txt</code> file that is generated.</li>
</ul></li>
<li><code>snippy-core</code>
<ul>
<li>Core genome alignment descriptive statistics.</li>
<li>Parses summary <code>prefix.txt</code> file that is generated.</li>
</ul></li>
</ul>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Snippy file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">snippy/snippy:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">snippy</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">20</span>
<span class="hljs-string">snippy/snippy-core:</span>
<span class="hljs-attr">  contents_re:</span> <span class="hljs-string">&gt;
    ID\tLENGTH\tALIGNED\tUNALIGNED\tVARIANT\tHET\tMASKED\tLOWCOV
</span><span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div><h2 id="snpeff"><a href="#snpeff" class="header-link"><span class="glyphicon glyphicon-link"></span></a>SnpEff</h2><p class="mod_ext_link"><a href="http://snpeff.sourceforge.net/" target="_blank">http://snpeff.sourceforge.net/</a></p><p>The SnpEff module parses results generated by
<a href="http://snpeff.sourceforge.net/">SnpEff</a>,
a genetic variant annotation and effect prediction toolbox.
It annotates and predicts the effects of variants on genes (such as amino
acid changes).</p>
<p>MultiQC parses the summary <code>.csv</code> file that is generated by SnpEff. Note that
you must run SnpEff with <code>-csvStats &lt;filename&gt;</code> for this to be generated.
See the <a href="http://snpeff.sourceforge.net/SnpEff_manual.html#outputSummary">SnpEff</a>
documentation for more information.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                SnpEff file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">snpeff:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">SnpEff_version</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">5000000</span>
</code></pre>
          </div><h2 id="snpsplit"><a href="#snpsplit" class="header-link"><span class="glyphicon glyphicon-link"></span></a>SNPsplit</h2><p class="mod_ext_link"><a href="https://www.bioinformatics.babraham.ac.uk/projects/SNPsplit/" target="_blank">https://www.bioinformatics.babraham.ac.uk/projects/SNPsplit/</a></p><p>Currently
 only the "Allele-tagging" and "Allele-sorting" reports are supported.
The log files from the genome creation steps are not parsed and there 
are no plots/tables produced from the "SNP coverage" report.</p>
<p>Differences between the numbers in the tagging and sorting reports 
are due to paired-end reads.
For these, if only a single mate in a pair is assigned to a genome then 
it will "rescue" its mate and both will be "sorted" into that genome 
(even though only one of them was tagged).
Conversely, if the mates in a pair are tagged as arising from different 
genomes, then the pair as a whole is unassignable.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                SNPsplit file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">snpsplit/old:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'Writing allele-flagged output file to:'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">2</span>
<span class="hljs-string">snpsplit/new:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*SNPsplit_report.yaml'</span>
</code></pre>
          </div><h2 id="somalier"><a href="#somalier" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Somalier</h2><p class="mod_ext_link"><a href="https://github.com/brentp/somalier" target="_blank">https://github.com/brentp/somalier</a></p><p>Somalier can be used to find sample swaps or duplicates in cancer
projects, where there is often no jointly-called VCF across samples.</p>
<p>It is also extremely efficient and so can be used to find all-vs-all
relatedness estimates for thousands of samples.</p>
<p>It also outputs information on sex, depth, heterozgyosity, and ancestry
to be use for general QC.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Somalier file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">somalier/somalier-ancestry:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.somalier-ancestry.tsv'</span>
<span class="hljs-string">somalier/samples:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.samples.tsv'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#family_id'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">5</span>
<span class="hljs-string">somalier/pairs:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.pairs.tsv'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">hom_concordance</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">5</span>
</code></pre>
          </div><h2 id="stacks"><a href="#stacks" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Stacks</h2><p class="mod_ext_link"><a href="http://catchenlab.life.illinois.edu/stacks/" target="_blank">http://catchenlab.life.illinois.edu/stacks/</a></p><h4 id="very-important-note"><a href="#very-important-note" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Very important note</h4>
<p>This module will only work with Stacks version 2.1 or greater.
Furthermore, this module is designed to only parse some of the output from the <code>denovo_map</code> pipeline.
If you are missing some functionality, please submit an issue on the <a href="https://github.com/ewels/MultiQC">MultiQC github page</a></p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Stacks file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">stacks/gstacks:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">gstacks.log.distribs</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">BEGIN</span> <span class="hljs-string">effective_coverages_per_sample</span>
<span class="hljs-string">stacks/populations:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">populations.log.distribs</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">BEGIN</span> <span class="hljs-string">missing_samples_per_loc_prefilters</span>
<span class="hljs-string">stacks/sumstats:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.sumstats_summary.tsv'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'# Pop ID	Private	Num_Indv	Var	StdErr	P	Var'</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">1000000</span>
</code></pre>
          </div><h2 id="supernova"><a href="#supernova" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Supernova</h2><p class="mod_ext_link"><a href="https://www.10xgenomics.com/" target="_blank">https://www.10xgenomics.com/</a></p><h4 id="important-notes"><a href="#important-notes" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Important notes</h4>
<p>Due to the size of the <code>histogram_kmer_count.json</code> files, 
MultiQC is likely to skip these files. To be able to display these you 
will need to change the MultiQC configuration to allow for larger 
logfiles, see the MultiQC <a href="http://multiqc.info/docs/#troubleshooting-1">documentation</a>. For instance, if you run MultiQC as part of an analysis pipeline, you can create a <code>multiqc_config.yaml</code> file in the working directory, containing the following line:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">log_filesize_limit:</span> <span class="hljs-number">100000000</span></code></pre>
<h4 id="general-notes"><a href="#general-notes" class="header-link"><span class="glyphicon glyphicon-link"></span></a>General Notes</h4>
<p>The Supernova module parses the reports from an assembly run. As a bare minimum it requires the file <code>report.txt</code>, found in the folder <code>sampleID/outs/</code>, to function. Note! If you are anything like the author (@remiolsen), you might only have files (often renamed to, e.g. <code>sampleID-report.txt</code>) lying around due to disk space limitations and for ease of sharing with your colleagues. This module will search for <code>*report*.txt</code>. If available the stats in the report file will be superseded by the higher precision numbers found in the file <code>sampleID/outs/assembly/stats/summary.json</code>. In the same folder, this module will search for the following plots and render them:</p>
<ul>
<li><code>histogram_molecules.json</code> -- Inferred molecule lengths</li>
<li><code>histogram_kmer_count.json</code> -- Kmer multiplicity</li>
</ul>
<p>This module has been tested using Supernova versions <code>1.1.4</code> and <code>1.2.0</code></p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                Supernova file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">supernova/report:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*report*.txt'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">100</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'- assembly checksum ='</span>
<span class="hljs-string">supernova/summary:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">summary.json</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">120</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'"lw_mean_mol_len":'</span>
<span class="hljs-string">supernova/molecules:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">histogram_molecules.json</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">10</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'"description": "molecules",'</span>
<span class="hljs-string">supernova/kmers:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">histogram_kmer_count.json</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">10</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'"description": "kmer_count",'</span>
</code></pre>
          </div><h2 id="theta2"><a href="#theta2" class="header-link"><span class="glyphicon glyphicon-link"></span></a>THeTA2</h2><p class="mod_ext_link"><a href="http://compbio.cs.brown.edu/projects/theta/" target="_blank">http://compbio.cs.brown.edu/projects/theta/</a></p><p><a href="http://compbio.cs.brown.edu/projects/theta/">THeTA2</a> (Tumor Heterogeneity Analysis)
is an algorithm that estimates the tumour purity and clonal / subclonal copy number
aberrations directly from high-throughput DNA sequencing data.</p>
<p>The THeTA2 MultiQC module plots the % germline and % tumour subclone for each sample.
Note that each sample can have multiple maximum likelihood solutions - the MultiQC
module plots proportions for the first one in the results file (<code>*.BEST.results</code>).
Also note that if there are more than 5 tumour subclones, their percentages are summed.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                THeTA2 file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-attr">theta2:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.BEST.results'</span>
</code></pre>
          </div><h2 id="varscan2"><a href="#varscan2" class="header-link"><span class="glyphicon glyphicon-link"></span></a>VarScan2</h2><p class="mod_ext_link"><a href="http://dkoboldt.github.io/varscan/" target="_blank">http://dkoboldt.github.io/varscan/</a></p><p>VarScan is a platform-independent mutation caller for targeted, exome, and whole-genome
resequencing data generated on Illumina, SOLiD, Life/PGM, Roche/454, and similar instruments.
The newest version, VarScan 2, is written in Java, so it runs on most operating systems.</p>
<p>VarScan can be used to detect different types of variation:</p>
<ul>
<li>Germline variants (SNPs an dindels) in individual samples or pools of samples.</li>
<li>Multi-sample variants (shared or private) in multi-sample datasets (with mpileup).</li>
<li>Somatic mutations, LOH events, and germline variants in tumor-normal pairs.</li>
<li>Somatic copy number alterations (CNAs) in tumor-normal exome data.</li>
</ul>
<p>The MultiQC module can read output from <code>mpileup2cns</code>, <code>mpileup2snp</code> and <code>mpileup2indel</code> logfiles.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                VarScan2 file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">varscan2/mpileup2snp:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Only</span> <span class="hljs-string">SNPs</span> <span class="hljs-string">will</span> <span class="hljs-string">be</span> <span class="hljs-string">reported</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
<span class="hljs-string">varscan2/mpileup2indel:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Only</span> <span class="hljs-string">indels</span> <span class="hljs-string">will</span> <span class="hljs-string">be</span> <span class="hljs-string">reported</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
<span class="hljs-string">varscan2/mpileup2cns:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">Only</span> <span class="hljs-string">variants</span> <span class="hljs-string">will</span> <span class="hljs-string">be</span> <span class="hljs-string">reported</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">3</span>
</code></pre>
          </div><h2 id="vcftools"><a href="#vcftools" class="header-link"><span class="glyphicon glyphicon-link"></span></a>VCFTools</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#important-general-note">Important General Note</a></li>
<li><a href="#implemented">Implemented</a></li>
<li><a href="#to-do">To do</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div><p class="mod_ext_link"><a href="https://vcftools.github.io/" target="_blank">https://vcftools.github.io</a></p><h3 id="important-general-note"><a href="#important-general-note" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Important General Note</h3>
<ul>
<li>Depending on the size and density of the variant data (vcf),
some of the stat files generated by vcftools can be very large.
If you find that some of your input files are missing, increase
the <a href="http://multiqc.info/docs/#big-log-files">config.log_filesize_limit</a>
so that the large file(s) will not be skipped by MultiQC. Note, however,
that this might make MultiQC very slow!</li>
</ul>
<p>This module parses the outputs from VCFTools' various commands:</p>
<h3 id="implemented"><a href="#implemented" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Implemented</h3>
<ul>
<li><code>relatedness2</code>
<ul>
<li>Plots a heatmap of pairwise sample relatedness.</li>
<li>Not to be confused with the similarly-named command <code>relatedness</code></li>
</ul></li>
<li><code>TsTv-by-count</code>
<ul>
<li>Plots the transition to transversion ratio as a function of
alternative allele count (using only bi-allelic SNPs).</li>
</ul></li>
<li><code>TsTv-by-qual</code>
<ul>
<li>Plots the transition to transversion ratio as a function of SNP
quality threshold (using only bi-allelic SNPs).</li>
</ul></li>
<li><code>TsTv-summary</code>
<ul>
<li>Plots a bargraph of the summary counts of each type of transition and
transversion SNPs.</li>
</ul></li>
</ul>
<h3 id="to-do"><a href="#to-do" class="header-link"><span class="glyphicon glyphicon-link"></span></a>To do</h3>
<p>VCFTools has a number of outputs not yet supported in MultiQC which
would be good to add. Please check GitHub if you'd like these added
or (better still), would like to contribute!</p>
<ul>
<li>basic stats</li>
<li>relatedness</li>
<li>freq</li>
<li>depth</li>
<li><a href="https://vcftools.github.io/man_latest.html">Everything else</a></li>
</ul>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                VCFTools file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">vcftools/relatedness2:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.relatedness2'</span>
<span class="hljs-string">vcftools/tstv_by_count:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.TsTv.count'</span>
<span class="hljs-string">vcftools/tstv_by_qual:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.TsTv.qual'</span>
<span class="hljs-string">vcftools/tstv_summary:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.TsTv.summary'</span>
</code></pre>
          </div><h2 id="vep"><a href="#vep" class="header-link"><span class="glyphicon glyphicon-link"></span></a>VEP</h2><p class="mod_ext_link"><a href="https://www.ensembl.org/info/docs/tools/vep/index.html" target="_blank">https://www.ensembl.org/info/docs/tools/vep/index.html</a></p><p>The VEP module parses the summary statistics generated by
<a href="https://www.ensembl.org/info/docs/tools/vep/index.html">Ensembl VEP</a>.</p>
<p>MultiQC parses the VEP summary statistics stored in either HTML or plain text format.
Beside VEP's default naming convention, you can run VEP with one of the options below to use this module:</p>
<ul>
<li><code>--stats_file [OUTPUT_FILENAME]_summary.html</code> <em>(VEP's default naming convention)</em></li>
<li><code>--stats_file [SAMPLE_NAME].vep.html</code> <em>(without the <code>vep</code> or <code>summary</code> suffix, MultiQC will ignore the HTML files)</em></li>
<li><code>--stats_file [SAMPLE_NAME]_vep.html</code></li>
<li><code>--stats_text --stats_file [SAMPLE_NAME].vep.txt</code></li>
<li><code>--stats_text --stats_file [SAMPLE_NAME]_vep.txt</code></li>
</ul>
<p>See the <a href="https://www.ensembl.org/info/docs/tools/vep/vep_formats.html#stats">VEP</a>
documentation for more information.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                VEP file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">vep/vep_html:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.html'</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">VEP</span> <span class="hljs-string">summary</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">10</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">1000000</span>
<span class="hljs-string">vep/vep_txt:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'[VEP run statistics]'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
<span class="hljs-attr">  max_filesize:</span> <span class="hljs-number">100000</span>
</code></pre>
          </div><h2 id="verifybamid"><a href="#verifybamid" class="header-link"><span class="glyphicon glyphicon-link"></span></a>VerifyBAMID</h2><p class="mod_ext_link"><a href="https://genome.sph.umich.edu/wiki/VerifyBamID" target="_blank">https://genome.sph.umich.edu/wiki/VerifyBamID</a></p><p>A
 key step in any genetic analysis is to verify whether data being 
generated matches expectations. verifyBamID checks whether reads in a 
BAM file match previous genotypes for a specific sample. In addition, it
 detects possible sample mixture from population allele frequency only, 
which can be particularly useful when the genotype data is not 
available.</p>
<p>Using a mathematical model that relates observed sequence reads to an
 hypothetical true genotype, verifyBamID tries to decide whether 
sequence reads match a particular individual or are more likely to be 
contaminated (including a small proportion of foreign DNA), derived from
 a closely related individual, or derived from a completely different 
individual.</p>
<p>This module currently only imports data from the <code>.selfSM</code> output.
The chipmix and freemix columns are imported into the general statistics table.
A verifyBAMID section is then added, with a table containing the entire selfSM file.</p>
<p>If no chip data was parsed, these columns will not be added to the MultiQC report.</p>
<p>Should you wish to remove one of these columns from the general 
statistics table add the below lines to the table_columns_visible 
section of your config file</p>
<pre><code class="hljs">table_columns_visible:
    verifyBAMID:
        CHIPMIX: False
        FREEMIX: False</code></pre>
<p>This was designed to work with verifyBamID 1.1.3 January 2018</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                VerifyBAMID file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">verifybamid/selfsm:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">'*.selfSM'</span>
</code></pre>
          </div><h2 id="whatshap"><a href="#whatshap" class="header-link"><span class="glyphicon glyphicon-link"></span></a>WhatsHap</h2><p class="mod_ext_link"><a href="https://whatshap.readthedocs.io/" target="_blank">https://whatshap.readthedocs.io</a></p><hr>
<p>The WhatsHap module parses results generated by
<a href="https://whatshap.readthedocs.io/">WhatsHap</a>, and is currently restricted to the
output from <code>whatshap stats --tsv</code>.</p>
          <div class="panel panel-default search-patterns-panel">
            <div class="panel-heading">
                WhatsHap file search patterns
                <a href="#module-search-patterns">See docs <i class="fa fa-chevron-up"></i></a>
            </div>
            <pre class="panel-body"><code class="language-yaml hljs"><span class="hljs-string">whatshap/stats:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">'#sample	chromosome	file_name	variants	phased	unphased	singletons'</span>
<span class="hljs-attr">  num_lines:</span> <span class="hljs-number">1</span>
</code></pre>
          </div></div></div><div class="docs_section">
<h1 class="section-header" id="custom-content"><a href="#custom-content" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Custom Content</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#introduction-1">Introduction</a>
<ul class="nav nav-stacked">
<li><a href="#data-from-a-released-tool">Data from a released tool</a></li>
<li><a href="#images">Images</a></li>
<li><a href="#multiqc-specific-data-file">MultiQC-specific data file</a></li>
<li><a href="#data-as-part-of-multiqc-config">Data as part of MultiQC config</a></li>
<li><a href="#separate-configuration-and-data-files">Separate configuration and data files</a></li>
</ul>
</li>
<li><a href="#configuration-1">Configuration</a>
<ul class="nav nav-stacked">
<li><a href="#grouping-sections-and-subsections">Grouping sections and subsections</a></li>
<li><a href="#order-of-sections">Order of sections</a></li>
<li><a href="#section-configuration">Section configuration</a></li>
<li><a href="#plot-configuration">Plot configuration</a></li>
<li><a href="#tricky-extras">Tricky extras</a>
<ul class="nav nav-stacked">
<li><a href="#differences-between-tables-and-general-stats">Differences between Tables and General Stats</a></li>
<li><a href="#first-columns-in-tables-are-special">First columns in tables are special</a></li>
</ul>
</li>
<li><a href="#linting">Linting</a></li>
</ul>
</li>
<li><a href="#examples">Examples</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<div class="docs_block" id="custom_content.md"><h1 id="introduction-1"><a href="#introduction-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Introduction</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#data-from-a-released-tool">Data from a released tool</a></li>
<li><a href="#images">Images</a></li>
<li><a href="#multiqc-specific-data-file">MultiQC-specific data file</a></li>
<li><a href="#data-as-part-of-multiqc-config">Data as part of MultiQC config</a></li>
<li><a href="#separate-configuration-and-data-files">Separate configuration and data files</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Bioinformatics projects often include non-standardised analyses, with results from custom
scripts or in-house packages. It can be frustrating to have a MultiQC report describing
results from 90% of your pipeline but missing the final key plot. To help with this,
MultiQC has a special <em>"custom content"</em> module.</p>
<p>Custom content parsing is a little more restricted than standard modules. Specifically:</p>
<ul>
<li>Only one plot per section is possible</li>
<li>Plot customisation is more limited</li>
</ul>
<p>All plot types can be generated using custom content - see the
<a href="https://github.com/ewels/MultiQC_TestData/tree/master/data/custom_content">test files</a>
for examples of how data should be structured.</p>
<blockquote>
<p><strong>Note</strong>: Use the name <code>custom_content</code> to refer to this module within configuration
settings that require a module name, such as <a href="#order-of-modules"><code>module_order</code></a> or
<a href="#removing-modules-or-sections"><code>run_modules</code></a>.</p>
</blockquote>
<h2 id="data-from-a-released-tool"><a href="#data-from-a-released-tool" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Data from a released tool</h2>
<p>If your data comes from a released bioinformatics tool, you shouldn't be using this
feature of MultiQC! Sure, you can probably get it to work, but it's better if a
fully-fledged core MultiQC module is written instead. That way, other users of MultiQC
can also benefit from results parsing.</p>
<p>Note that proper MultiQC modules are more robust and powerful than this custom-content
feature. You can also <a href="http://multiqc.info/docs/#writing-new-modules">write modules</a>
in <a href="http://multiqc.info/docs/#multiqc-plugins">MultiQC plugins</a> if they're not suitable for
general release.</p>
<h2 id="images"><a href="#images" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Images</h2>
<p>As of MultiQC v1.7, you can import custom images into your MultiQC reports.
Simply add <code>_mqc</code> to the end of the filename for <code>.png</code>, <code>.jpg</code> or <code>.jpeg</code> files, for example:
<code>my_image_file_mqc.png</code> or <code>summmary_diagram.jpeg</code>.</p>
<p>Images will be embedded within the HTML file, so will be self contained.
Note that this means that it's very possible to make the HTML file very very large if abused!</p>
<p>The report section name and description will be automatically based on the filename.</p>
<p>Note that if you are using <code>sp:</code> to take in images with a custom filename you need to also set <code>ignore_images: false</code> in your config. For example:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_data:</span>
<span class="hljs-attr">  my_custom_content_image:</span>
<span class="hljs-attr">    section_name:</span> <span class="hljs-string">"My nice image"</span>
<span class="hljs-attr">sp:</span>
<span class="hljs-attr">  my_custom_content_image:</span>
<span class="hljs-attr">    fn:</span> <span class="hljs-string">"*.png"</span>
<span class="hljs-attr">ignore_images:</span> <span class="hljs-literal">false</span></code></pre>
<h2 id="multiqc-specific-data-file"><a href="#multiqc-specific-data-file" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiQC-specific data file</h2>
<p>If you can choose exactly how your data output looks, then the easiest way to parse it
is to use a MultiQC-specific format. If the filename ends in <code>*_mqc.(yaml|yml|json|txt|csv|tsv|log|out|png|jpg|jpeg|html)</code>
then it will be found by any standard MultiQC installation with no additional customisation
required (v0.9 onwards).</p>
<p>These files contain configuration information specifying how the data should be parsed,
alongside the data. If you want to use YAML, this is an example of how it should look:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">id:</span> <span class="hljs-string">"my_pca_section"</span>
<span class="hljs-attr">section_name:</span> <span class="hljs-string">"PCA Analysis"</span>
<span class="hljs-attr">description:</span> <span class="hljs-string">"This plot shows the first two components from a principal component analysis."</span>
<span class="hljs-attr">plot_type:</span> <span class="hljs-string">"scatter"</span>
<span class="hljs-attr">pconfig:</span>
<span class="hljs-attr">  id:</span> <span class="hljs-string">"pca_scatter_plot"</span>
<span class="hljs-attr">  title:</span> <span class="hljs-string">"PCA Plot"</span>
<span class="hljs-attr">  xlab:</span> <span class="hljs-string">"PC1"</span>
<span class="hljs-attr">  ylab:</span> <span class="hljs-string">"PC2"</span>
<span class="hljs-attr">data:</span>
<span class="hljs-attr">  sample_1:</span> <span class="hljs-string">{</span> <span class="hljs-attr">x:</span> <span class="hljs-number">12</span><span class="hljs-string">,</span> <span class="hljs-attr">y:</span> <span class="hljs-number">14</span> <span class="hljs-string">}</span>
<span class="hljs-attr">  sample_2:</span> <span class="hljs-string">{</span> <span class="hljs-attr">x:</span> <span class="hljs-number">8</span><span class="hljs-string">,</span> <span class="hljs-attr">y:</span> <span class="hljs-number">6</span> <span class="hljs-string">}</span>
<span class="hljs-attr">  sample_3:</span> <span class="hljs-string">{</span> <span class="hljs-attr">x:</span> <span class="hljs-number">5</span><span class="hljs-string">,</span> <span class="hljs-attr">y:</span> <span class="hljs-number">11</span> <span class="hljs-string">}</span>
<span class="hljs-attr">  sample_4:</span> <span class="hljs-string">{</span> <span class="hljs-attr">x:</span> <span class="hljs-number">9</span><span class="hljs-string">,</span> <span class="hljs-attr">y:</span> <span class="hljs-number">12</span> <span class="hljs-string">}</span></code></pre>
<p>The file format can also be JSON:</p>
<pre><code class="language-json hljs">{
  <span class="hljs-attr">"id"</span>: <span class="hljs-string">"custom_data_lineplot"</span>,
  <span class="hljs-attr">"section_name"</span>: <span class="hljs-string">"Custom JSON File"</span>,
  <span class="hljs-attr">"description"</span>: <span class="hljs-string">"This plot is a self-contained JSON file."</span>,
  <span class="hljs-attr">"plot_type"</span>: <span class="hljs-string">"linegraph"</span>,
  <span class="hljs-attr">"pconfig"</span>: {
    <span class="hljs-attr">"id"</span>: <span class="hljs-string">"custom_data_linegraph"</span>,
    <span class="hljs-attr">"title"</span>: <span class="hljs-string">"Output from my JSON file"</span>,
    <span class="hljs-attr">"ylab"</span>: <span class="hljs-string">"Number of things"</span>,
    <span class="hljs-attr">"xDecimals"</span>: <span class="hljs-literal">false</span>
  },
  <span class="hljs-attr">"data"</span>: {
    <span class="hljs-attr">"sample_1"</span>: { <span class="hljs-attr">"1"</span>: <span class="hljs-number">12</span>, <span class="hljs-attr">"2"</span>: <span class="hljs-number">14</span>, <span class="hljs-attr">"3"</span>: <span class="hljs-number">10</span>, <span class="hljs-attr">"4"</span>: <span class="hljs-number">7</span>, <span class="hljs-attr">"5"</span>: <span class="hljs-number">16</span> },
    <span class="hljs-attr">"sample_2"</span>: { <span class="hljs-attr">"1"</span>: <span class="hljs-number">9</span>, <span class="hljs-attr">"2"</span>: <span class="hljs-number">11</span>, <span class="hljs-attr">"3"</span>: <span class="hljs-number">15</span>, <span class="hljs-attr">"4"</span>: <span class="hljs-number">18</span>, <span class="hljs-attr">"5"</span>: <span class="hljs-number">21</span> }
  }
}</code></pre>
<p>Note that if you're using <code>plot_type: html</code> then <code>data</code> just takes a string, with no sample keys.</p>
<p>For maximum compatibility with other tools, you can also use comma-separated or tab-separated files.
Include commented header lines with plot configuration in YAML format:</p>
<pre><code class="language-bash hljs"><span class="hljs-comment"># id: "Output from my script'</span>
<span class="hljs-comment"># section_name: 'Custom data file'</span>
<span class="hljs-comment"># description: 'This output is described in the file header. Any MultiQC installation will understand it without prior configuration.'</span>
<span class="hljs-comment"># format: 'tsv'</span>
<span class="hljs-comment"># plot_type: 'bargraph'</span>
<span class="hljs-comment"># pconfig:</span>
<span class="hljs-comment">#    id: 'custom_bargraph_w_header'</span>
<span class="hljs-comment">#    ylab: 'Number of things'</span>
Category_1    374
Category_2    229
Category_3    39
Category_4    253</code></pre>
<p>You can easily inject custom HTML snippets by ending the filename with <code>_mqc.html</code> - again the
embedded config works in a similar way, but with a HTML comment:</p>
<pre><code class="language-html hljs xml"><span class="hljs-comment">&lt;!--
id: 'custom-html'
section_name: 'Custom HTML'
description: 'This section is created using a custom HTML file'
--&gt;</span>
<span class="hljs-tag">&lt;<span class="hljs-name">p</span>&gt;</span>Some custom HTML content here.<span class="hljs-tag">&lt;/<span class="hljs-name">p</span>&gt;</span></code></pre>
<p>If no configuration is given, MultiQC will do its best to guess how to visualise your data appropriately.
To see examples of typical file structures which are understood, see the
<a href="https://github.com/ewels/MultiQC_TestData/tree/master/data/custom_content/no_config">test data</a>
used to develop this code. Something will be probably be shown, but it may produce unexpected results.</p>
<blockquote>
<p><strong>Note:</strong> Check <a href="#tricky-extras">Tricky extras</a> for certain caveats about formatting headers for custom
<code>tsv</code> or <code>csv</code> files, particularly for the first column.</p>
</blockquote>
<h2 id="data-as-part-of-multiqc-config"><a href="#data-as-part-of-multiqc-config" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Data as part of MultiQC config</h2>
<p>If you are already using a MultiQC config file to add data to your report (for example,
<a href="http://multiqc.info/docs/#customising-reports">titles / introductory text</a>), you can
give data within this file too. This can be in any MultiQC config file (for example,
passed on the command line with <code>-c my_yaml_file.yaml</code>). This is useful as you can
keep everything contained within a single file (including stuff unrelated to this
specific <em>custom content</em> feature of MultiQC).</p>
<p>To be understood by MultiQC, the <code>custom_data</code> key must be found.
This must contain a section with a unique id, specific to your new report section.
Finally, the contents of this second dictionary will look the same as the above
stand-alone <code>YAML</code> files. For example:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_data:</span>
<span class="hljs-attr">  my_data_type:</span>
<span class="hljs-attr">    id:</span> <span class="hljs-string">"mqc_config_file_section"</span>
<span class="hljs-attr">    section_name:</span> <span class="hljs-string">"My Custom Section"</span>
<span class="hljs-attr">    description:</span> <span class="hljs-string">"This data comes from a single multiqc_config.yaml file"</span>
<span class="hljs-attr">    plot_type:</span> <span class="hljs-string">"bargraph"</span>
<span class="hljs-attr">    pconfig:</span>
<span class="hljs-attr">      id:</span> <span class="hljs-string">"barplot_config_only"</span>
<span class="hljs-attr">      title:</span> <span class="hljs-string">"MultiQC Config Data Plot"</span>
<span class="hljs-attr">      ylab:</span> <span class="hljs-string">"Number of things"</span>
<span class="hljs-attr">    data:</span>
<span class="hljs-attr">      sample_a:</span>
<span class="hljs-attr">        first_thing:</span> <span class="hljs-number">12</span>
<span class="hljs-attr">        second_thing:</span> <span class="hljs-number">14</span>
<span class="hljs-attr">      sample_b:</span>
<span class="hljs-attr">        first_thing:</span> <span class="hljs-number">8</span>
<span class="hljs-attr">        second_thing:</span> <span class="hljs-number">6</span>
<span class="hljs-attr">      sample_c:</span>
<span class="hljs-attr">        first_thing:</span> <span class="hljs-number">11</span>
<span class="hljs-attr">        second_thing:</span> <span class="hljs-number">5</span>
<span class="hljs-attr">      sample_d:</span>
<span class="hljs-attr">        first_thing:</span> <span class="hljs-number">12</span>
<span class="hljs-attr">        second_thing:</span> <span class="hljs-number">9</span></code></pre>
<p>Or to add data to the General Statistics table:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_data:</span>
<span class="hljs-attr">  my_genstats:</span>
<span class="hljs-attr">    plot_type:</span> <span class="hljs-string">"generalstats"</span>
<span class="hljs-attr">    pconfig:</span>
<span class="hljs-attr">      - col_1:</span>
<span class="hljs-attr">          max:</span> <span class="hljs-number">100</span>
<span class="hljs-attr">          min:</span> <span class="hljs-number">0</span>
<span class="hljs-attr">          scale:</span> <span class="hljs-string">"RdYlGn"</span>
<span class="hljs-attr">          suffix:</span> <span class="hljs-string">"%"</span>
<span class="hljs-attr">      - col_2:</span>
<span class="hljs-attr">          min:</span> <span class="hljs-number">0</span>
<span class="hljs-attr">    data:</span>
<span class="hljs-attr">      sample_a:</span>
<span class="hljs-attr">        col_1:</span> <span class="hljs-number">14.32</span>
<span class="hljs-attr">        col_2:</span> <span class="hljs-number">1.2</span>
<span class="hljs-attr">      sample_b:</span>
<span class="hljs-attr">        col_1:</span> <span class="hljs-number">84.84</span>
<span class="hljs-attr">        col_2:</span> <span class="hljs-number">1.9</span></code></pre>
<blockquote>
<p><strong>Note:</strong> Use a <strong>list</strong> of headers in <code>pconfig</code> (keys prepended with <code>-</code>) to specify the order
of columns in the General Statistics table.</p>
</blockquote>
<p>See the <a href="http://multiqc.info/docs/#step-3-adding-to-the-general-statistics-table">general statistics docs</a>
for more information about configuring data for the General Statistics table.</p>
<h2 id="separate-configuration-and-data-files"><a href="#separate-configuration-and-data-files" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Separate configuration and data files</h2>
<p>It's not always possible or desirable to include MultiQC configuration within a data file.
If this is the case, you can add to the MultiQC configuration to specify how input files
should be parsed.</p>
<p>As described in the above <a href="#data-as-part-of-multiqc-config"><em>Data as part of MultiQC config</em></a> section,
this configuration should be held within a section called <code>custom_data</code> with a section-specific id.
The only difference is that no <code>data</code> subsection is given and a search pattern for the given id must
be supplied.</p>
<p>Search patterns are added <a href="http://multiqc.info/docs/#module-search-patterns">as with any other module</a>.
Ensure that the search pattern key is the same as your <code>custom_data</code> section ID.</p>
<p>For example, a MultiQC config file could look as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-comment"># Other MultiQC config stuff here</span>
<span class="hljs-attr">custom_data:</span>
<span class="hljs-attr">  example_files:</span>
<span class="hljs-attr">    file_format:</span> <span class="hljs-string">"tsv"</span>
<span class="hljs-attr">    section_name:</span> <span class="hljs-string">"Coverage Decay"</span>
<span class="hljs-attr">    description:</span> <span class="hljs-string">"This plot comes from files acommpanied by a mutliqc_config.yaml file for configuration"</span>
<span class="hljs-attr">    plot_type:</span> <span class="hljs-string">"linegraph"</span>
<span class="hljs-attr">    pconfig:</span>
<span class="hljs-attr">      id:</span> <span class="hljs-string">"example_coverage_lineplot"</span>
<span class="hljs-attr">      title:</span> <span class="hljs-string">"Coverage Decay"</span>
<span class="hljs-attr">      ylab:</span> <span class="hljs-string">"X Coverage"</span>
<span class="hljs-attr">      ymax:</span> <span class="hljs-number">100</span>
<span class="hljs-attr">      ymin:</span> <span class="hljs-number">0</span>
<span class="hljs-attr">sp:</span>
<span class="hljs-attr">  example_files:</span>
<span class="hljs-attr">    fn:</span> <span class="hljs-string">"example_files_*"</span></code></pre>
<p>And work with the following data file:
<code>example_files_Sample_1.txt</code>:</p>
<pre><code class="language-bash hljs">0   98.22076066
1   97.96764159
2   97.78227175
3   97.61262195
<span class="hljs-comment"># [...]</span></code></pre>
<p>This kind of customisation should work with most Custom Content types.
For example, using an image called <code>some_science_mqc.jpeg</code> gives us a report section <code>some_science</code>,
which we can then add a nicer name and description to:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_data:</span>
<span class="hljs-attr">  some_science:</span>
<span class="hljs-attr">    section_name:</span> <span class="hljs-string">"Some real science"</span>
<span class="hljs-attr">    description:</span> <span class="hljs-string">"This description comes from multiqc_config.yaml and helps to annotate the Custom Content image."</span></code></pre>
<p>As mentioned above - if no configuration is given, MultiQC will do its best to guess how to visualise
your data appropriately. To see examples of typical file structures which are understood, see the
<a href="https://github.com/ewels/MultiQC_TestData/tree/master/data/custom_content/no_config">test data</a>
used to develop this code.</p>
<h1 id="configuration-1"><a href="#configuration-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Configuration</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#grouping-sections-and-subsections">Grouping sections and subsections</a></li>
<li><a href="#order-of-sections">Order of sections</a></li>
<li><a href="#section-configuration">Section configuration</a></li>
<li><a href="#plot-configuration">Plot configuration</a></li>
<li><a href="#tricky-extras">Tricky extras</a>
<ul class="nav nav-stacked">
<li><a href="#differences-between-tables-and-general-stats">Differences between Tables and General Stats</a></li>
<li><a href="#first-columns-in-tables-are-special">First columns in tables are special</a></li>
</ul>
</li>
<li><a href="#linting">Linting</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<h2 id="grouping-sections-and-subsections"><a href="#grouping-sections-and-subsections" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Grouping sections and subsections</h2>
<p>If you have multiple content types that you would like to group together with MultiQC sub-sections,
you can do so using the following keys:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">parent_id:</span> <span class="hljs-string">custom_section</span>
<span class="hljs-attr">parent_name:</span> <span class="hljs-string">"Some grouped data"</span>
<span class="hljs-attr">parent_description:</span> <span class="hljs-string">"This parent section contains one or more sub-sections below it"</span></code></pre>
<p>Any custom-content files that share the same <code>parent_id</code> will be grouped.</p>
<p>Note that some things, such as <code>parent_name</code> are taken from the first file that MultiQC finds
with this <code>parent_id</code>. So it's a good idea to specify this in every file.
<code>parent_description</code> and <code>extra</code> is taken from the first file where it is set.</p>
<blockquote>
<p><code>parent_id</code> only works within Custom Content.
It is not currently possible to add custom content output into a report section
from a core MultiQC module.</p>
</blockquote>
<h2 id="order-of-sections"><a href="#order-of-sections" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Order of sections</h2>
<p>If you have multiple different Custom Content sections, their order will be random
and may vary between runs. To avoid this, you can specify an order in your MultiQC
config as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">custom_content:</span>
<span class="hljs-attr">  order:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">first_cc_section</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">second_cc_section</span></code></pre>
<p>Each section name should be the ID assigned to that section. You can explicitly set
this (see below), or the Custom Content module will automatically assign an ID.
To find out what your custom content section ID is, generate a report and click
the side navigation to your section. The browser URL should update and show something
that looks like this:</p>
<pre><code class="language-txt">multiqc_report.html#my_cc_section</code></pre>
<p>The section ID is the part after the <code>#</code> (<code>my_cc_section</code> in the above section).</p>
<p>Note that any Custom Content sections found that are <em>not</em> specified in the config
will be placed at the top of the report.</p>
<h2 id="section-configuration"><a href="#section-configuration" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Section configuration</h2>
<p>See below for how these config options can be specified (either within the data file
or in a MultiQC config file). All of these configuration parameters
are optional, and MultiQC will do its best to guess sensible defaults if they are
not specified.</p>
<p>All possible configuration keys and their default values are shown below:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">id:</span> <span class="hljs-literal">null</span> <span class="hljs-comment"># Unique ID for report section.</span>
<span class="hljs-attr">section_anchor:</span> <span class="hljs-string">&lt;id&gt;</span> <span class="hljs-comment"># Used in report section #soft-links</span>
<span class="hljs-attr">section_name:</span> <span class="hljs-string">&lt;id&gt;</span> <span class="hljs-comment"># Nice name used for the report section header</span>
<span class="hljs-attr">section_href:</span> <span class="hljs-literal">null</span> <span class="hljs-comment"># External URL for the data, to find more information</span>
<span class="hljs-attr">description:</span> <span class="hljs-literal">null</span> <span class="hljs-comment"># Introductory text to be printed under the section header</span>
<span class="hljs-attr">section_extra:</span> <span class="hljs-literal">null</span> <span class="hljs-comment"># Custom HTML to add after the section description</span>
<span class="hljs-attr">file_format:</span> <span class="hljs-literal">null</span> <span class="hljs-comment"># File format of the data (eg. csv / tsv)</span>
<span class="hljs-attr">plot_type:</span>
  <span class="hljs-literal">null</span> <span class="hljs-comment"># The plot type to visualise the data with.</span>
  <span class="hljs-comment"># generalstats | table | bargraph | linegraph | scatter | heatmap | beeswarm</span>
<span class="hljs-attr">pconfig:</span> <span class="hljs-string">{}</span> <span class="hljs-comment"># Configuration for the plot.</span></code></pre>
<blockquote>
<p>Data types <code>generalstats</code> and <code>beeswarm</code> are <em>only</em> possible by setting the above
configuration keys (these can't be guessed by data format).</p>
</blockquote>
<p>Note that any <em>custom content</em> data found with the same section <code>id</code> will be merged
into the same report section / plot. The other section configuration keys are merged
for each file, with identical keys overwriting what was previously parsed.</p>
<p>This approach means that it's possible to have a single file containing data for multiple
samples, but it's also possible to have one file per sample and still have all of them
summarised.</p>
<blockquote>
<p>If you're using <code>plot_type: 'generalstats'</code> then a report section will not be created and
most of the configuration keys above are ignored.</p>
</blockquote>
<h2 id="plot-configuration"><a href="#plot-configuration" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Plot configuration</h2>
<p>Configuration of specific plots follows the same syntax as used when writing modules.
To find out more, please see the later docs. Specifically, the plot config docs for
<a href="#bar-graphs">bar graphs</a>,
<a href="#line-graphs">line graphs</a>,
<a href="#scatter-plots">scatter plots</a>,
<a href="#creating-a-table">tables</a>,
<a href="#beeswarm-plots-dot-plots">beeswarm plots</a> and
<a href="#heatmaps">heatmaps</a>.</p>
<p>Wherever you see <code>pconfig</code>, any key can be used within the above syntax.</p>
<h2 id="tricky-extras"><a href="#tricky-extras" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Tricky extras</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#differences-between-tables-and-general-stats">Differences between Tables and General Stats</a></li>
<li><a href="#first-columns-in-tables-are-special">First columns in tables are special</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Because of the way this module works, there are a few specifics that can trip you up.
Most of these should probably be fixed one day. Feel free to complain on gitter or submit a pull request!
I'll try to keep a list here to help the wary...</p>
<h3 id="differences-between-tables-and-general-stats"><a href="#differences-between-tables-and-general-stats" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Differences between Tables and General Stats</h3>
<p>Although they're both tables, note that general stats configures columns with a list
in the <code>pconfig</code> scope (see above example). Files that are just tables use <code>headers</code> instead.</p>
<h3 id="first-columns-in-tables-are-special"><a href="#first-columns-in-tables-are-special" class="header-link"><span class="glyphicon glyphicon-link"></span></a>First columns in tables are special</h3>
<p>The first column in every table is reserved for the sample name. As such, it shouldn't contain data.
All header configuration will be ignored for the first column. The only exception is name:
this can be tweaked using the somewhat tricky <code>col1_header</code> field in the <code>pconfig</code> scope (see table docs).
Alternatively, you can customise the column name by including a 'header row' in the first line of the <code>tsv</code>
or <code>csv</code> itself specifying the column names, with the first column with the name of your choice, and
subsequent columns including the key(s) defined in the header.</p>
<h2 id="linting"><a href="#linting" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Linting</h2>
<p>MultiQC has been developed to be as forgiving as possible and will handle lots of
invalid or ignored configurations. This is useful for most users but can make life
difficult when getting MultiQC to work with a new custom content format.</p>
<p>To help with this, you can run with the <code>--lint</code> flag, which will give explicit
warnings about anything that is not optimally configured. For example:</p>
<pre><code class="language-bash hljs">multiqc --lint test_data</code></pre>
<h1 id="examples"><a href="#examples" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Examples</h1>
<p>Probably the best way to get to grips with Custom Content is to see some examples.
The MultiQC automated testing runs with a bunch of different files, and I try to add to
these all the time.</p>
<p>You can see these examples here: <a href="https://github.com/ewels/MultiQC_TestData/tree/master/data/custom_content">https://github.com/ewels/MultiQC_TestData/tree/master/data/custom_content</a></p>
<p>For example, to see a file which generates a table in a report by itself, you can
have a look at <code>embedded_config/table_headers_mqc.txt</code> (<a href="https://github.com/ewels/MultiQC_TestData/blob/master/data/custom_content/embedded_config/table_headers_mqc.txt">link</a>).</p></div></div><div class="docs_section">
<h1 class="section-header" id="coding-with-multiqc"><a href="#coding-with-multiqc" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Coding with MultiQC</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#writing-new-modules">Writing New Modules</a>
<ul class="nav nav-stacked">
<li><a href="#introduction-2">Introduction</a></li>
<li><a href="#core-modules--plugins">Core modules / plugins</a></li>
<li><a href="#multiqc-lint-tests">MultiQC Lint Tests</a></li>
<li><a href="#code-formatting">Code formatting</a>
<ul class="nav nav-stacked">
<li><a href="#use-version-control">Use version control</a></li>
<li><a href="#black">Black</a></li>
<li><a href="#prettier">Prettier</a></li>
<li><a href="#markdownlint">Markdownlint</a></li>
</ul>
</li>
<li><a href="#initial-setup">Initial setup</a>
<ul class="nav nav-stacked">
<li><a href="#multiqc-file-structure">MultiQC file structure</a></li>
<li><a href="#submodule">Submodule</a></li>
<li><a href="#entry-points">Entry points</a></li>
<li><a href="#multiqc-config">MultiQC config</a></li>
<li><a href="#documentation">Documentation</a></li>
<li><a href="#changelog">Changelog</a></li>
<li><a href="#multiqcmodule-class">MultiqcModule Class</a></li>
<li><a href="#logging">Logging</a></li>
</ul>
</li>
<li><a href="#step-1-find-log-files">Step 1 - Find log files</a></li>
<li><a href="#step-2-parse-data-from-the-input-files">Step 2 - Parse data from the input files</a>
<ul class="nav nav-stacked">
<li><a href="#filtering-by-parsed-sample-names">Filtering by parsed sample names</a></li>
<li><a href="#no-files-found">No files found</a></li>
<li><a href="#custom-sample-names">Custom sample names</a></li>
<li><a href="#identical-sample-names">Identical sample names</a></li>
<li><a href="#printing-to-the-sources-file">Printing to the sources file</a></li>
</ul>
</li>
<li><a href="#step-3-adding-to-the-general-statistics-table">Step 3 - Adding to the general statistics table</a>
<ul class="nav nav-stacked">
<li><a href="#table-colour-scales">Table colour scales</a></li>
</ul>
</li>
<li><a href="#step-4-writing-data-to-a-file">Step 4 - Writing data to a file</a></li>
<li><a href="#step-5-create-report-sections">Step 5 - Create report sections</a></li>
<li><a href="#step-6-plot-some-data">Step 6 - Plot some data</a></li>
<li><a href="#appendices">Appendices</a>
<ul class="nav nav-stacked">
<li><a href="#user-configuration">User configuration</a></li>
<li><a href="#profiling-performance">Profiling Performance</a></li>
<li><a href="#adding-custom-css--javascript">Adding Custom CSS / Javascript</a></li>
</ul>
</li>

</ul>
</li>
<li><a href="#plotting-functions">Plotting Functions</a>
<ul class="nav nav-stacked">
<li><a href="#common-options">Common options</a></li>
<li><a href="#bar-graphs">Bar graphs</a>
<ul class="nav nav-stacked">
<li><a href="#switching-datasets">Switching datasets</a></li>
<li><a href="#interactive--flat-image-plots">Interactive / Flat image plots</a></li>
</ul>
</li>
<li><a href="#line-graphs">Line graphs</a>
<ul class="nav nav-stacked">
<li><a href="#switching-datasets-1">Switching datasets</a></li>
<li><a href="#additional-data-series">Additional data series</a></li>
</ul>
</li>
<li><a href="#scatter-plots">Scatter Plots</a></li>
<li><a href="#creating-a-table">Creating a table</a>
<ul class="nav nav-stacked">
<li><a href="#table-decimal-places">Table decimal places</a></li>
<li><a href="#table-colour-scales-1">Table colour scales</a></li>
<li><a href="#custom-cell-background-colours">Custom cell background colours</a></li>
<li><a href="#zero-centrepoints">Zero centrepoints</a></li>
<li><a href="#conditional-formatting-of-data-values">Conditional formatting of data values</a></li>
</ul>
</li>
<li><a href="#beeswarm-plots-dot-plots">Beeswarm plots (dot plots)</a></li>
<li><a href="#heatmaps">Heatmaps</a></li>
<li><a href="#javascript-functions">Javascript Functions</a>
<ul class="nav nav-stacked">
<li><a href="#plotting-line-graphs">Plotting line graphs</a></li>
<li><a href="#plotting-bar-graphs">Plotting bar graphs</a></li>
<li><a href="#switching-counts-and-percentages">Switching counts and percentages</a></li>
<li><a href="#switching-plot-datasets">Switching plot datasets</a></li>
<li><a href="#custom-event-triggers">Custom event triggers</a></li>
</ul>
</li>

</ul>
</li>
<li><a href="#multiqc-plugins">MultiQC Plugins</a>
<ul class="nav nav-stacked">
<li><a href="#entry-points-1">Entry Points</a></li>
<li><a href="#modules">Modules</a></li>
<li><a href="#templates">Templates</a></li>
<li><a href="#command-line-options">Command line options</a></li>
<li><a href="#hooks">Hooks</a></li>
</ul>
</li>
<li><a href="#writing-new-templates">Writing New Templates</a>
<ul class="nav nav-stacked">
<li><a href="#core-or-plugin">Core or plugin</a></li>
<li><a href="#creating-a-template-skeleton">Creating a template skeleton</a></li>
<li><a href="#child-templates">Child templates</a></li>
<li><a href="#extra-init-variables">Extra init variables</a></li>
<li><a href="#jinja-template-variables">Jinja template variables</a></li>
<li><a href="#appendices-1">Appendices</a>
<ul class="nav nav-stacked">
<li><a href="#custom-plotting-functions">Custom plotting functions</a></li>
</ul>
</li>

</ul>
</li>
<li><a href="#updating-for-compatibility">Updating for compatibility</a>
<ul class="nav nav-stacked">
<li><a href="#v1.0-updates">v1.0 Updates</a></li></ul></li></ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<div class="docs_block" id="modules.md"><h1 id="writing-new-modules"><a href="#writing-new-modules" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Writing New Modules</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#introduction-2">Introduction</a></li>
<li><a href="#core-modules--plugins">Core modules / plugins</a></li>
<li><a href="#multiqc-lint-tests">MultiQC Lint Tests</a></li>
<li><a href="#code-formatting">Code formatting</a>
<ul class="nav nav-stacked">
<li><a href="#use-version-control">Use version control</a></li>
<li><a href="#black">Black</a></li>
<li><a href="#prettier">Prettier</a></li>
<li><a href="#markdownlint">Markdownlint</a></li>
</ul>
</li>
<li><a href="#initial-setup">Initial setup</a>
<ul class="nav nav-stacked">
<li><a href="#multiqc-file-structure">MultiQC file structure</a></li>
<li><a href="#submodule">Submodule</a></li>
<li><a href="#entry-points">Entry points</a></li>
<li><a href="#multiqc-config">MultiQC config</a></li>
<li><a href="#documentation">Documentation</a></li>
<li><a href="#changelog">Changelog</a></li>
<li><a href="#multiqcmodule-class">MultiqcModule Class</a></li>
<li><a href="#logging">Logging</a></li>
</ul>
</li>
<li><a href="#step-1-find-log-files">Step 1 - Find log files</a></li>
<li><a href="#step-2-parse-data-from-the-input-files">Step 2 - Parse data from the input files</a>
<ul class="nav nav-stacked">
<li><a href="#filtering-by-parsed-sample-names">Filtering by parsed sample names</a></li>
<li><a href="#no-files-found">No files found</a></li>
<li><a href="#custom-sample-names">Custom sample names</a></li>
<li><a href="#identical-sample-names">Identical sample names</a></li>
<li><a href="#printing-to-the-sources-file">Printing to the sources file</a></li>
</ul>
</li>
<li><a href="#step-3-adding-to-the-general-statistics-table">Step 3 - Adding to the general statistics table</a>
<ul class="nav nav-stacked">
<li><a href="#table-colour-scales">Table colour scales</a></li>
</ul>
</li>
<li><a href="#step-4-writing-data-to-a-file">Step 4 - Writing data to a file</a></li>
<li><a href="#step-5-create-report-sections">Step 5 - Create report sections</a></li>
<li><a href="#step-6-plot-some-data">Step 6 - Plot some data</a></li>
<li><a href="#appendices">Appendices</a>
<ul class="nav nav-stacked">
<li><a href="#user-configuration">User configuration</a></li>
<li><a href="#profiling-performance">Profiling Performance</a></li>
<li><a href="#adding-custom-css--javascript">Adding Custom CSS / Javascript</a></li>
</ul>
</li>

</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<h2 id="introduction-2"><a href="#introduction-2" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Introduction</h2>
<p>Writing a new module can at first seem a daunting task. However, MultiQC
has been written <em>(and refactored)</em> to provide a lot of functionality
as common functions.</p>
<p>Provided that you are familiar with writing Python and you have a read
through the guide below, you should be on your way in no time!</p>
<p>If you have any problems, feel free to contact the author - details
here: <a href="https://github.com/ewels">@ewels</a></p>
<h2 id="core-modules--plugins"><a href="#core-modules--plugins" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Core modules / plugins</h2>
<p>New modules can either be written as part of MultiQC or in a stand-alone
plugin. If your module is for a publicly available tool, please add it
to the main program and contribute your code back when complete via a
pull request.</p>
<p>If your module is for something <em>very</em> niche, which no-one else can use,
you can write it as part of a custom plugin. The process is almost identical,
though it keeps the code bases separate. For more information about this,
see the docs about <em>MultiQC Plugins</em> below.</p>
<h2 id="multiqc-lint-tests"><a href="#multiqc-lint-tests" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiQC Lint Tests</h2>
<p>MultiQC has been developed to be as forgiving as possible and will handle lots of
invalid or ignored code. This is useful most of the time but can be difficult when
writing new MultiQC modules (especially during pull-request reviews).</p>
<p>To help with this, you can run with the <code>--lint</code> flag, which will give explicit
warnings about anything that is not optimally configured. For example:</p>
<pre><code class="language-bash hljs">multiqc --lint test_data</code></pre>
<p>Note that the automated MultiQC continuous integration testing runs in this mode,
so you will need to pass all lint tests for those checks to pass. This is required
for any pull-requests.</p>
<h2 id="code-formatting"><a href="#code-formatting" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Code formatting</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#use-version-control">Use version control</a></li>
<li><a href="#black">Black</a></li>
<li><a href="#prettier">Prettier</a></li>
<li><a href="#markdownlint">Markdownlint</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>In addition to testing MultiQC functionality, the MultiQC code base is also checked for
consistency and formatting.</p>
<p>Everyone has their own preferences when it comes to writing any code, both in the methods
used but also with simple things like whitespace and whether to use <code>"</code> or <code>'</code>.
When reviewing code contributions in pull-requests, these variations in coding style
introduce an additional mental overhead. Inconsistent code style across the package
also makes it harder for newcomers to get into the code.</p>
<p>Code formatting / linting tools are able to assess files in many different languages
and check that a set of "soft" formatting rules are adhered to, to enforce code consistency.
Better still, many of these tools can automatically change the formatting so that developers
can write code in whatever style they prefer and defer this task to automation.</p>
<p>Much like source control, gloves in a lab, and wearing a seatbelt, code formatters and code linting
is an annoying inconvience at first for most people which in time becomes an indespesible
tool in the maintenance of high quality software.</p>
<p>When it comes to MultiQC, three tools are used to set and check the code base:</p>
<ul>
<li><a href="https://github.com/psf/black">Black</a> - Python Code</li>
<li><a href="https://prettier.io/">Prettier</a> - Everything else (almost)</li>
<li><a href="https://github.com/igorshubovych/markdownlint-cli">markdownlint-cli</a> - Stricter markdown rules</li>
</ul>
<p><strong>All developers must run these tools when submitting changes via Pull-Requests!</strong>
Automated continuous integration tests will run using GitHub Actions to check that all files pass the above tests.
If any files do not, that test will fail giving a red :x: next to the pull request.</p>
<blockquote>
<p>Make sure that your configuration is working properly and that you're not changing loads of files
that you haven't worked with. Pull-requests will not be merged with such changes.</p>
</blockquote>
<p>All three tools should be relatively easy to install and run, and have integration with the majority
of code editors. Once set up, they can run on save and you'll never need to think about them again.</p>
<p>For example, <a href="https://code.visualstudio.com/">Visual Studio Code</a> has
<a href="https://code.visualstudio.com/docs/python/editing#_formatting">built-in support for Black</a> and
plugins for <a href="https://github.com/prettier/prettier-vscode">Prettier</a> and
<a href="https://github.com/DavidAnson/vscode-markdownlint">Markdownlint</a>.</p>
<p>Atom also has plugins for <a href="https://atom.io/packages/python-black">Black</a>,
<a href="https://atom.io/packages/prettier-atom">Prettier</a> and
<a href="https://atom.io/packages/linter-node-markdownlint">Markdownlint</a>.</p>
<h3 id="use-version-control"><a href="#use-version-control" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Use version control</h3>
<p>These tools are set up to edit source code in place. If you're not used to working with them,
make sure that you have saved your changes <em>and made a commit</em> with those changes before running them!
Then if all goes wrong and they make loads of unexpected changes, you can run <code>git checkout .</code> to
reset everything before trying again.</p>
<h3 id="black"><a href="#black" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Black</h3>
<p>MultiQC is written in Python and so the majority of the codebase is Python.
<a href="https://github.com/psf/black">Black</a> describes itself as <em>"The uncompromising code formatter"</em>
and has very few configuration options on purpose.</p>
<p>Black is very easy to install and run:</p>
<pre><code class="language-bash hljs">pip install black
black .</code></pre>
<p>Black should automatically find the config file in the root of the directory called
<code>pyproject.toml</code> which configures it to use a line length of 120 characters.
This should be found automatically.</p>
<h3 id="prettier"><a href="#prettier" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Prettier</h3>
<p>In addition to Python, MultiQC also has a lot of markdown documentation,
YAML configuration files, custom JavaScript and more. To format these files
we use a tool called <a href="https://prettier.io/">Prettier</a>.</p>
<p>Prettier is available via the Node Package Manager (npm). If you've not used this before,
you'll need to install Node first: <a href="https://nodejs.org/en/download/">https://nodejs.org/en/download/</a></p>
<p>Once <code>npm</code> is working, you can <a href="https://prettier.io/docs/en/install.html">install Prettier</a> globally.
Then run with the <code>--write</code> flag to edit files in place.</p>
<pre><code class="language-bash hljs">npm install -g prettier
prettier --write .</code></pre>
<p>Prettier has two config files in the repository root: <code>.prettierrc.yaml</code> and <code>.prettierignore</code>.
These should be found automatically by Prettier.
One configures Prettier to use longer line lengths of 120 characters and the other tells it to skip
bundled source files and all HTML files (personally I'm not a fan of what it does with HTML).</p>
<h3 id="markdownlint"><a href="#markdownlint" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Markdownlint</h3>
<p>Prettier formats markdown using <a href="https://github.com/remarkjs/remark">remark</a>, which works great.
However, <a href="https://github.com/DavidAnson/markdownlint">Markdownlint</a> has a lot more strict rules
and checks for a load of additional nice things (eg. specifying code type for syntax highlighting of code blocks).</p>
<p>Markdownlint doesn't fix things in place like the two tools above, instead it just checks files
and reports problems that must be fixed manually.</p>
<p>MultiQC uses <a href="https://github.com/igorshubovych/markdownlint-cli">markdownlint-cli</a> to run tests.
Once <code>npm</code> is working (see above), it's a simple install and run:</p>
<pre><code class="language-bash hljs">npm install -g markdownlint-cli
markdownlint .</code></pre>
<p>There is a config file for markdownlint in the root of the repository called <code>.markdownlint.yaml</code>
which disables and tweaks quite a lot of rules. This should be found automatically.</p>
<p>Note that the editor plugins above run this as you edit files, giving little wiggly lines
to warn you about problems as you type.</p>
<h2 id="initial-setup"><a href="#initial-setup" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Initial setup</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#multiqc-file-structure">MultiQC file structure</a></li>
<li><a href="#submodule">Submodule</a></li>
<li><a href="#entry-points">Entry points</a></li>
<li><a href="#multiqc-config">MultiQC config</a></li>
<li><a href="#documentation">Documentation</a></li>
<li><a href="#changelog">Changelog</a></li>
<li><a href="#multiqcmodule-class">MultiqcModule Class</a></li>
<li><a href="#logging">Logging</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<h3 id="multiqc-file-structure"><a href="#multiqc-file-structure" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiQC file structure</h3>
<p>The source code for MultiQC is separated into different folders.
Most of the files you won't have to touch - the relevant files that
you will need to edit or create are as follows:</p>
<pre><code class="language-console"> docs
&nbsp;&nbsp;  README.md
&nbsp;&nbsp;  modules
&nbsp;&nbsp;  &nbsp;&nbsp;  &lt;your_module&gt;.md
 multiqc
&nbsp;&nbsp;  modules
&nbsp;&nbsp; |&nbsp;&nbsp;  &lt;your_module&gt;
&nbsp;&nbsp; &nbsp;&nbsp;  &nbsp;&nbsp;  __init__.py
&nbsp;&nbsp; &nbsp;&nbsp;  &nbsp;&nbsp;  &lt;your_module&gt;.py
&nbsp;&nbsp;  utils
&nbsp;&nbsp;  &nbsp;&nbsp;  search_patterns.yaml
 CHANGELOG.md
 setup.py</code></pre>
<p>These files are described in more detail below.</p>
<h3 id="submodule"><a href="#submodule" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Submodule</h3>
<p>MultiQC modules are Python submodules - as such, they need their own
directory in <code>/multiqc/</code> with an <code>__init__.py</code> file. The directory should
share its name with the module. To follow common practice, the module
code itself usually then goes in a separate python file (also with the same
name, i.e. <code>multiqc/modname/modname.py</code>) which is then imported by the
<code>__init__.py</code> file with:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> __future__ <span class="hljs-keyword">import</span> absolute_import
<span class="hljs-keyword">from</span> .modname <span class="hljs-keyword">import</span> MultiqcModule</code></pre>
<h3 id="entry-points"><a href="#entry-points" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Entry points</h3>
<p>Once your submodule files are in place, you need to tell MultiQC that they
are available as an analysis module. This is done within <code>setup.py</code> using
<a href="http://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins">entry points</a>.
In <code>setup.py</code> you will see some code that looks like this:</p>
<pre><code class="language-python hljs">entry_points = {
    <span class="hljs-string">'multiqc.modules.v1'</span>: [
        <span class="hljs-string">'bismark = multiqc.modules.bismark:MultiqcModule'</span>,
        [...]
    ]
}</code></pre>
<p>Copy one of the existing module lines and change it to use your module name.
The order is irrelevant, so stick to alphabetical if in doubt.
Once this is done, you will need to update your installation of MultiQC:</p>
<pre><code class="language-bash hljs">pip install -e .</code></pre>
<h3 id="multiqc-config"><a href="#multiqc-config" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiQC config</h3>
<p>So that MultiQC knows what order modules should be run in, you need to add
your module to the core config file.</p>
<p>In <code>multiqc/utils/config_defaults.yaml</code> you should see a list variable called <code>module_order</code>.
This contains the name of modules in order of precedence. Add your module here
in an appropriate position.</p>
<h3 id="documentation"><a href="#documentation" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Documentation</h3>
<p>Next up, you need to create a documentation file for your module. The reason
for this is twofold: firstly, docs are important to help people to use, debug
and extend MultiQC (you're reading this, aren't you?). Secondly,
having the file there with the appropriate YAML front matter will make the
module show up on the <a href="http://multiqc.info/">MultiQC homepage</a> so that everyone
knows it exists. This process is automated once the file is added to the core
repository.</p>
<p>This docs file should be placed in <code>docs/modules/&lt;your_module_name&gt;.md</code> and
should have the following structure:</p>
<pre><code class="language-markdown hljs">---
Name: Tool Name
URL: http://www.amazing-bfx-tool.com
Description: &gt;
  This amazing tool does some really cool stuff. You can describe it
<span class="hljs-section">  here and split onto multiple lines if you want. Not too long though!
---</span>

Your documentation goes here. Feel free to use markdown and write whatever
you think would be helpful. Please avoid using heading levels 1 to 3.</code></pre>
<p>Make a reference to this in the YAML <em>frontmatter</em> list at the top of
<code>docs/README.md</code> - this allows the website to find the file to build
the documentation.</p>
<h3 id="changelog"><a href="#changelog" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Changelog</h3>
<p>Last but not least, remember to add your new module to the <code>CHANGELOG.md</code>,
so that people know that it's there.</p>
<h3 id="multiqcmodule-class"><a href="#multiqcmodule-class" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiqcModule Class</h3>
<p>If you've copied one of the other entry point statements, it will have ended
in <code>:MultiqcModule</code> - this tells MultiQC to try to execute a class or function
called <code>MultiqcModule</code>.</p>
<p>To use the helper functions bundled with MultiQC, you should extend this
class from <code>multiqc.modules.base_module.BaseMultiqcModule</code> in your
module code file (i.e. <code>multiqc/modname/modname.py</code>). This will give
you access to a number of functions on the <code>self</code> namespace. For example:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc.modules.base_module <span class="hljs-keyword">import</span> BaseMultiqcModule

<span class="hljs-class"><span class="hljs-keyword">class</span> <span class="hljs-title">MultiqcModule</span><span class="hljs-params">(BaseMultiqcModule)</span>:</span>
    <span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">__init__</span><span class="hljs-params">(self)</span>:</span>
        <span class="hljs-comment"># Initialise the parent object</span>
        super(MultiqcModule, self).__init__(
          name=<span class="hljs-string">'My Module'</span>,
          anchor=<span class="hljs-string">'mymod'</span>,
          href=<span class="hljs-string">"http://www.awesome_bioinfo.com/my_module"</span>,
          info=<span class="hljs-string">"is an example analysis module used for writing documentation."</span>,
          doi=<span class="hljs-string">"01.2345/journal/abc123"</span>
        )</code></pre>
<p>The <code>__init__</code> variables are used to create the header, URL link,
analysis module credits and description in the report.</p>
<p>The available arguments when initialising a module as follows:</p>
<ul>
<li><code>name</code> - Name of your module</li>
<li><code>anchor</code> - A HTML-safe anchor that will be used after the <code>#</code> in the URL</li>
<li><code>href</code> - Link to the homepage for the tool</li>
<li><code>info</code> - Very short description text about the tool</li>
<li><code>doi</code> - One or more publication DOIs (<code>str</code>, or <code>list</code> of <code>str</code>s)</li>
<li><code>comment</code> - Additional comment text for module. Usually user-supplied in a config.</li>
<li><code>extra</code> - Additional custom HTML after description text.</li>
<li><code>target</code> - Name of the module in the description (default: <code>name</code>)</li>
<li><code>autoformat</code> - (default: <code>True</code>)</li>
<li><code>autoformat_type</code> - (default: <code>markdown</code>)</li>
</ul>
<p>Ok, that should be it! The <code>__init__()</code> function will now be executed every
time MultiQC runs. Try adding a <code>print("Hello World!")</code> statement and see
if it appears in the MultiQC logs at the appropriate time...</p>
<h3 id="logging"><a href="#logging" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Logging</h3>
<p>Last thing - MultiQC modules have a standardised way of producing output,
so you shouldn't really use <code>print()</code> statements for your <code>Hello World</code> in
your module code ;).</p>
<p>Instead, use the <code>logger</code> module as follows:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">import</span> logging
log = logging.getLogger(__name__)
<span class="hljs-comment"># Initialise your class and so on</span>
log.info(<span class="hljs-string">'Hello World!'</span>)</code></pre>
<p>Log messages can come in a range of formats:</p>
<ul>
<li><code>log.debug</code>
<ul>
<li>Thes only show if MultiQC is run in <code>-v</code>/<code>--verbose</code> mode</li>
</ul></li>
<li><code>log.info</code>
<ul>
<li>For more important status updates</li>
</ul></li>
<li><code>log.warning</code>
<ul>
<li>Alert user about problems that don't halt execution</li>
</ul></li>
<li><code>log.error</code> and <code>log.critical</code>
<ul>
<li>Not often used, these are for show-stopping problems</li>
</ul></li>
</ul>
<h2 id="step-1-find-log-files"><a href="#step-1-find-log-files" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 1 - Find log files</h2>
<p>The first thing that your module will need to do is to find analysis log
files. You can do this by searching for a filename fragment, or a string
within the file. It's possible to search for both (a match on either
will return the file) and also to have multiple strings possible.</p>
<p>First, add your default patterns to <code>multiqc/utils/search_patterns.yaml</code></p>
<p>Each search has a yaml key, with one or more search criteria.</p>
<p>The yaml key must begin with the name of your module. If you have multiple
search patterns for a single module, follow the module name with a forward
slash and then any string. For example, see the <code>fastqc</code> module search patterns:</p>
<pre><code class="language-yaml hljs"><span class="hljs-string">fastqc/data:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">"fastqc_data.txt"</span>
<span class="hljs-string">fastqc/zip:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">"_fastqc.zip"</span></code></pre>
<p>The following search criteria sub-keys can then be used:</p>
<ul>
<li><code>fn</code>
<ul>
<li>A glob filename pattern, used with the Python <a href="https://docs.python.org/2/library/fnmatch.html"><code>fnmatch</code></a> function</li>
</ul></li>
<li><code>fn_re</code>
<ul>
<li>A regex filename pattern</li>
</ul></li>
<li><code>contents</code>
<ul>
<li>A string to match within the file contents (checked line by line)</li>
</ul></li>
<li><code>contents_re</code>
<ul>
<li>A regex to match within the file contents (checked line by line)</li>
<li>NB: Regex must match entire line (add <code>.*</code> to start and end of pattern to avoid this)</li>
</ul></li>
<li><code>exclude_fn</code>
<ul>
<li>A glob filename pattern which will exclude a file if matched</li>
</ul></li>
<li><code>exclude_fn_re</code>
<ul>
<li>A regex filename pattern which will exclude a file if matched</li>
</ul></li>
<li><code>exclude_contents</code>
<ul>
<li>A string which will exclude the file if matched within the file contents (checked line by line)</li>
</ul></li>
<li><code>exclude_contents_re</code>
<ul>
<li>A regex which will exclude the file if matched within the file contents (checked line by line)</li>
</ul></li>
<li><code>num_lines</code>
<ul>
<li>The number of lines to search through for the <code>contents</code> string. Default: all lines.</li>
</ul></li>
<li><code>shared</code>
<ul>
<li>By default, once a file has been assigned to a module it is not searched again. Specify <code>shared: true</code> when your file can be shared between multiple tools (for example, part of a <code>stdout</code> stream).</li>
</ul></li>
<li><code>max_filesize</code>
<ul>
<li>Files larger than the <code>log_filesize_limit</code> config key 
(default: 10MB) are skipped. If you know your files will be smaller than
 this and need to search by contents, you can specify this value (in 
bytes) to skip any files smaller than this limit.</li>
</ul></li>
</ul>
<p>Please try to use <code>num_lines</code> and <code>max_filesize</code> where possible as they will speed up
MultiQC execution time.</p>
<p>Note that <code>exclude_</code> keys are tested after a file is detected with one or
more of the other patterns.</p>
<p>For example, two typical modules could specify search patterns as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">mymod:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">"_myprogram.txt"</span>
<span class="hljs-attr">myothermod:</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">"This is myprogram v1.3"</span></code></pre>
<p>You can also supply a list of different patterns for a single log file type if needed.
If any of the patterns are matched, the file will be returned:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">mymod:</span>
<span class="hljs-attr">  - fn:</span> <span class="hljs-string">"mylog.txt"</span>
<span class="hljs-attr">  - fn:</span> <span class="hljs-string">"different_fn.out"</span></code></pre>
<p>You can use <em>AND</em> logic by specifying keys within a single list item. For example:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">mymod:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">"mylog.txt"</span>
<span class="hljs-attr">  contents:</span> <span class="hljs-string">"mystring"</span>
<span class="hljs-attr">myothermod:</span>
<span class="hljs-attr">  - fn:</span> <span class="hljs-string">"different_fn.out"</span>
<span class="hljs-attr">    contents:</span> <span class="hljs-string">"This is myprogram v1.3"</span>
<span class="hljs-attr">  - fn:</span> <span class="hljs-string">"another.txt"</span>
<span class="hljs-attr">    contents:</span> <span class="hljs-string">"What are these files anyway?"</span></code></pre>
<p>Here, a file must have the filename <code>mylog.txt</code> <em>and</em> contain the string <code>mystring</code>.</p>
<p>You can match subsets of files by using <code>exclude_</code> keys as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">mymod:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">"*.myprog.txt"</span>
<span class="hljs-attr">  exclude_fn:</span> <span class="hljs-string">"not_these_*"</span>
<span class="hljs-attr">myothermod:</span>
<span class="hljs-attr">  fn:</span> <span class="hljs-string">"mylog.txt"</span>
<span class="hljs-attr">  exclude_contents:</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">"trimmed"</span>
<span class="hljs-bullet">    -</span> <span class="hljs-string">"sorted"</span></code></pre>
<p>Note that the <code>exclude_</code> patterns can have either a single value or a list of values.
They are always considered using OR logic - any matches will reject the file.</p>
<p>Remember that users can overwrite these defaults in their own config files.
This is helpful as people have weird and wonderful processing pipelines with
their own conventions.</p>
<p>Once your strings are added, you can find files in your module with the
base function <code>self.find_log_files()</code>, using the key you set in the YAML:</p>
<pre><code class="language-python hljs">self.find_log_files(<span class="hljs-string">'mymod'</span>)</code></pre>
<p>This function yields a dictionary with various information about each matching
file. The <code>f</code> key contains the contents of the matching file:</p>
<pre><code class="language-python hljs"><span class="hljs-comment"># Find all files for mymod</span>
<span class="hljs-keyword">for</span> myfile <span class="hljs-keyword">in</span> self.find_log_files(<span class="hljs-string">'mymod'</span>):
    print( myfile[<span class="hljs-string">'f'</span>] )       <span class="hljs-comment"># File contents</span>
    print( myfile[<span class="hljs-string">'s_name'</span>] )  <span class="hljs-comment"># Sample name (from cleaned filename)</span>
    print( myfile[<span class="hljs-string">'fn'</span>] )      <span class="hljs-comment"># Filename</span>
    print( myfile[<span class="hljs-string">'root'</span>] )    <span class="hljs-comment"># Directory file was in</span></code></pre>
<p>If <code>filehandles=True</code> is specified, the <code>f</code> key contains a file handle
instead:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">for</span> f <span class="hljs-keyword">in</span> self.find_log_files(<span class="hljs-string">'mymod'</span>, filehandles=<span class="hljs-literal">True</span>):
    <span class="hljs-comment"># f['f'] is now a filehandle instead of contents</span>
    <span class="hljs-keyword">for</span> l <span class="hljs-keyword">in</span> f[<span class="hljs-string">'f'</span>]:
        print( l )</code></pre>
<p>This is good if the file is large, as Python doesn't read the entire
file into memory in one go.</p>
<h2 id="step-2-parse-data-from-the-input-files"><a href="#step-2-parse-data-from-the-input-files" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 2 - Parse data from the input files</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#filtering-by-parsed-sample-names">Filtering by parsed sample names</a></li>
<li><a href="#no-files-found">No files found</a></li>
<li><a href="#custom-sample-names">Custom sample names</a></li>
<li><a href="#identical-sample-names">Identical sample names</a></li>
<li><a href="#printing-to-the-sources-file">Printing to the sources file</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>What most MultiQC modules do once they have found matching analysis files
is to pass the matched file contents to another function, responsible
for parsing the data from the file. How this parsing is done will depend
on the format of the log file and the type of data being read. See below
for a basic example, based loosely on the preseq module:</p>
<pre><code class="language-python hljs"><span class="hljs-class"><span class="hljs-keyword">class</span> <span class="hljs-title">MultiqcModule</span><span class="hljs-params">(BaseMultiqcModule)</span>:</span>
    <span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">__init__</span><span class="hljs-params">(self)</span>:</span>
        <span class="hljs-comment"># [...]</span>
        self.mod_data = dict()
        <span class="hljs-keyword">for</span> f <span class="hljs-keyword">in</span> self.find_log_files(<span class="hljs-string">'mymod'</span>):
            self.mod_data[f[<span class="hljs-string">'s_name'</span>]] = self.parse_logs(f[<span class="hljs-string">'f'</span>])

    <span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">parse_logs</span><span class="hljs-params">(self, f)</span>:</span>
        data = {}
        <span class="hljs-keyword">for</span> l <span class="hljs-keyword">in</span> f.splitlines():
            s = l.split()
            data[s[<span class="hljs-number">0</span>]] = s[<span class="hljs-number">1</span>]
        <span class="hljs-keyword">return</span> data</code></pre>
<h3 id="filtering-by-parsed-sample-names"><a href="#filtering-by-parsed-sample-names" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Filtering by parsed sample names</h3>
<p>MultiQC users can use the <code>--ignore-samples</code> flag to skip sample names
that match specific patterns. As sample names are generated in a different
way by every module, this filter has to be applied after log parsing.</p>
<p>There is a core function to do this task - assuming that your data is
in a dictionary with the first key as sample name, pass it through the
<code>self.ignore_samples</code> function as follows:</p>
<pre><code class="language-python hljs">self.yourdata = self.ignore_samples(self.yourdata)</code></pre>
<p>This will remove any dictionary keys where the sample name matches
a user pattern.</p>
<p>If your data structure is not in the <code>sample_name: data</code> format then
you can check each sample name individually using the
<code>self.is_ignore_sample()</code> function:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">if</span> self.is_ignore_sample(f[<span class="hljs-string">'s_name'</span>]):
    print(<span class="hljs-string">"We will not use this sample!"</span>)</code></pre>
<p>Note that this function should be used <em>after</em> cleaning the sample name with
<code>self.clean_s_name()</code>.</p>
<h3 id="no-files-found"><a href="#no-files-found" class="header-link"><span class="glyphicon glyphicon-link"></span></a>No files found</h3>
<p>If your module cannot find any matching files, it needs to raise an
exception of type <code>UserWarning</code>. This tells the core MultiQC program
that no modules were found. For example:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">if</span> len(self.mod_data) == <span class="hljs-number">0</span>:
    <span class="hljs-keyword">raise</span> UserWarning</code></pre>
<p>Note that this has to be raised as early as possible, so that it halts
the module progress. For example, if no logs are found then the module
should not create any files or try to do any computation.</p>
<h3 id="custom-sample-names"><a href="#custom-sample-names" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Custom sample names</h3>
<p>Typically, sample names are taken from cleaned log filenames (the default
<code>f['s_name']</code> value returned). However, if possible, it's better to use
the name of the input file (allowing for concatenated log files).
To do this, you should use the <code>self.clean_s_name()</code> function, as
this will prepend the directory name if requested on the command line:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">for</span> f <span class="hljs-keyword">in</span> self.find_log_files(<span class="hljs-string">'mymod'</span>):
    input_fname, data = self.my_custom_parsing(f) <span class="hljs-comment"># Or parsed however</span>
    s_name = self.clean_s_name(input_fname, f)</code></pre>
<p>This function has already been applied to the contents of <code>f['s_name']</code>,
so it is only required when using something different for the sample identifier.</p>
<blockquote>
<p><code>self.clean_s_name()</code> <strong>must</strong> be used on sample names parsed from the file
contents. Without it, features such as prepending directories (<code>--dirs</code>)
will not work.</p>
</blockquote>
<p>The second argument should be the dictionary returned by the <code>self.find_log_files()</code> function.
The root path is used for <code>--dirs</code> and the search pattern key is used
for fine-grained configuration of the config option <code>use_filename_as_sample_name</code>.</p>
<p>If you are using non-standard values for the logfile root, filename or search pattern
key, these can be specified. The function def looks like this:</p>
<pre><code class="language-python hljs"><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">clean_s_name</span><span class="hljs-params">(self, s_name, f=None, root=None, filename=None, seach_pattern_key=None)</span>:</span></code></pre>
<p>A tyical example is when the sample name is the log file directory.
In this case, the root should be the dirname of that directory.
This is non-standard, and would be specified as follows:</p>
<pre><code class="language-python hljs">s_name = self.clean_s_name(f[<span class="hljs-string">"root"</span>], f, root=os.path.dirname(f[<span class="hljs-string">"root"</span>]))</code></pre>
<h3 id="identical-sample-names"><a href="#identical-sample-names" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Identical sample names</h3>
<p>If modules find samples with identical names, then the previous sample
is overwritten. It's good to print a log statement when this happens,
for debugging. However, most of the time it makes sense - programs often
create log files <em>and</em> print to <code>stdout</code> for example.</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">if</span> f[<span class="hljs-string">'s_name'</span>] <span class="hljs-keyword">in</span> self.bowtie_data:
    log.debug(<span class="hljs-string">"Duplicate sample name found! Overwriting: {}"</span>.format(f[<span class="hljs-string">'s_name'</span>]))</code></pre>
<h3 id="printing-to-the-sources-file"><a href="#printing-to-the-sources-file" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Printing to the sources file</h3>
<p>Finally, once you've found your file we want to add this information to the
<code>multiqc_sources.txt</code> file in the MultiQC report data directory. This lists
every sample name and the file from which this data came from. This is especially
useful if sample names are being overwritten as it lists the source used. This code
is typically written immediately after the above warning.</p>
<p>If you've used the <code>self.find_log_files</code> function, writing to the sources file
is as simple as passing the log file variable to the <code>self.add_data_source</code> function:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">for</span> f <span class="hljs-keyword">in</span> self.find_log_files(<span class="hljs-string">'mymod'</span>):
    self.add_data_source(f)</code></pre>
<p>If you have different files for different sections of the module, or are
customising the sample name, you can tweak the fields. The default arguments
are as shown:</p>
<pre><code class="language-python hljs">self.add_data_source(f=<span class="hljs-literal">None</span>, s_name=<span class="hljs-literal">None</span>, source=<span class="hljs-literal">None</span>, module=<span class="hljs-literal">None</span>, section=<span class="hljs-literal">None</span>)</code></pre>
<h2 id="step-3-adding-to-the-general-statistics-table"><a href="#step-3-adding-to-the-general-statistics-table" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 3 - Adding to the general statistics table</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#table-colour-scales">Table colour scales</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Now that you have your parsed data, you can start inserting it into the
MultiQC report. At the top of every report is the 'General Statistics'
table. This contains metrics from all modules, allowing cross-module
comparison.</p>
<p>There is a helper function to add your data to this table. It can take
a lot of configuration options, but most have sensible defaults. At
it's simplest, it works as follows:</p>
<pre><code class="language-python hljs">data = {
    <span class="hljs-string">'sample_1'</span>: {
        <span class="hljs-string">'first_col'</span>: <span class="hljs-number">91.4</span>,
        <span class="hljs-string">'second_col'</span>: <span class="hljs-string">'78.2%'</span>
    },
    <span class="hljs-string">'sample_2'</span>: {
        <span class="hljs-string">'first_col'</span>: <span class="hljs-number">138.3</span>,
        <span class="hljs-string">'second_col'</span>: <span class="hljs-string">'66.3%'</span>
    }
}
self.general_stats_addcols(data)</code></pre>
<p>To give more informative table headers and configure things like
data scales and colour schemes, you can supply an extra dict:</p>
<pre><code class="language-python hljs">headers = OrderedDict()
headers[<span class="hljs-string">'first_col'</span>] = {
    <span class="hljs-string">'title'</span>: <span class="hljs-string">'First'</span>,
    <span class="hljs-string">'description'</span>: <span class="hljs-string">'My First Column'</span>,
    <span class="hljs-string">'scale'</span>: <span class="hljs-string">'RdYlGn-rev'</span>
}
headers[<span class="hljs-string">'second_col'</span>] = {
    <span class="hljs-string">'title'</span>: <span class="hljs-string">'Second'</span>,
    <span class="hljs-string">'description'</span>: <span class="hljs-string">'My Second Column'</span>,
    <span class="hljs-string">'max'</span>: <span class="hljs-number">100</span>,
    <span class="hljs-string">'min'</span>: <span class="hljs-number">0</span>,
    <span class="hljs-string">'scale'</span>: <span class="hljs-string">'Blues'</span>,
    <span class="hljs-string">'suffix'</span>: <span class="hljs-string">'%'</span>
}
self.general_stats_addcols(data, headers)</code></pre>
<p>Here are all options for headers, with defaults:</p>
<pre><code class="language-python hljs">headers[<span class="hljs-string">'name'</span>] = {
    <span class="hljs-string">'namespace'</span>: <span class="hljs-string">''</span>,                <span class="hljs-comment"># Module name. Auto-generated for core modules in General Statistics.</span>
    <span class="hljs-string">'title'</span>: <span class="hljs-string">'[ dict key ]'</span>,        <span class="hljs-comment"># Short title, table column title</span>
    <span class="hljs-string">'description'</span>: <span class="hljs-string">'[ dict key ]'</span>,  <span class="hljs-comment"># Longer description, goes in mouse hover text</span>
    <span class="hljs-string">'max'</span>: <span class="hljs-literal">None</span>,                    <span class="hljs-comment"># Minimum value in range, for bar / colour coding</span>
    <span class="hljs-string">'min'</span>: <span class="hljs-literal">None</span>,                    <span class="hljs-comment"># Maximum value in range, for bar / colour coding</span>
    <span class="hljs-string">'scale'</span>: <span class="hljs-string">'GnBu'</span>,                <span class="hljs-comment"># Colour scale for colour coding. Set to False to disable.</span>
    <span class="hljs-string">'suffix'</span>: <span class="hljs-literal">None</span>,                 <span class="hljs-comment"># Suffix for value (eg. '%')</span>
    <span class="hljs-string">'format'</span>: <span class="hljs-string">'{:,.1f}'</span>,            <span class="hljs-comment"># Output format() string</span>
    <span class="hljs-string">'shared_key'</span>: <span class="hljs-literal">None</span>              <span class="hljs-comment"># See below for description</span>
    <span class="hljs-string">'modify'</span>: <span class="hljs-literal">None</span>,                 <span class="hljs-comment"># Lambda function to modify values</span>
    <span class="hljs-string">'hidden'</span>: <span class="hljs-literal">False</span>,                <span class="hljs-comment"># Set to True to hide the column on page load</span>
    <span class="hljs-string">'placement'</span> : <span class="hljs-number">1000.0</span>,           <span class="hljs-comment"># Alter the default ordering of columns in the table</span>
}</code></pre>
<ul>
<li><code>namespace</code>
<ul>
<li>This prepends the column title in the mouse hover: <em>Namespace: Title</em>.</li>
<li>The 'Configure Columns' modal displays this under the 'Group' column.</li>
<li>It's automatically generated for core modules in the General Statistics table,
though this can be overwritten (useful for example with custom-content).</li>
</ul></li>
<li><code>scale</code>
<ul>
<li>Colour scales are the names of ColorBrewer palettes. See below for available scales.</li>
<li>Add <code>-rev</code> to the name of a colour scale to reverse it</li>
<li>Set to <code>False</code> to disable colouring and background bars</li>
</ul></li>
<li><code>shared_key</code>
<ul>
<li>Any string can be specified here, if other columns are found that share
the same key, a consistent colour scheme and data scale will be used in
the table. Typically this is set to things like <code>read_count</code>, so that
the read count in a sample can be seen varying across analysis modules.</li>
</ul></li>
<li><code>modify</code>
<ul>
<li>A python <code>lambda</code> function to change the data in some way when it is
inserted into the table.</li>
</ul></li>
<li><code>hidden</code>
<ul>
<li>Setting this to <code>True</code> will hide the column when the report loads. It can
then be shown through the <em>Configure Columns</em> modal in the report. This can
be useful when data could be sometimes useful. For example, some modules
show "percentage aligned" on page load but hide "number of reads aligned".</li>
</ul></li>
<li><code>placement</code>
<ul>
<li>If you feel that the results from your module should appear at the left side
of the table set this value less than 1000. Or to move the column right, set
it greater than 1000. This value can be any float.</li>
</ul></li>
</ul>
<p>The typical use for the <code>modify</code> string is to divide large numbers such as read counts,
to make them easier to interpret. If handling read counts, there are three config variables
that should be used to allow users to change the multiplier for read counts: <code>read_count_multiplier</code>,
<code>read_count_prefix</code> and <code>read_count_desc</code>. For example:</p>
<pre><code class="language-python hljs"><span class="hljs-string">'title'</span>: <span class="hljs-string">'{} Reads'</span>.format(config.read_count_prefix),
<span class="hljs-string">'description'</span>: <span class="hljs-string">'Number of reads ({})'</span>.format(config.read_count_desc),
<span class="hljs-string">'modify'</span>: <span class="hljs-keyword">lambda</span> x: x * config.read_count_multiplier,</code></pre>
<p>Similar config options apply for base pairs: <code>base_count_multiplier</code>, <code>base_count_prefix</code> and
<code>base_count_desc</code>.</p>
<p>And for the read count of long reads: <code>long_read_count_multiplier</code>, <code>long_read_count_prefix</code> and
<code>long_read_count_desc</code>.</p>
<p>A third parameter can be passed to this function, <code>namespace</code>. This is usually
not needed - MultiQC automatically takes the name of the module that is calling
the function and uses this. However, sometimes it can be useful to overwrite this.</p>
<h3 id="table-colour-scales"><a href="#table-colour-scales" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Table colour scales</h3>
<p>Colour scales are taken from <a href="http://colorbrewer2.org/">ColorBrewer2</a>.
Colour scales can be reversed by adding the suffix <code>-rev</code> to the name. For example, <code>RdYlGn-rev</code>.</p>
<p>The following scales are available:</p>
<p><img src="MultiQC%20installation_files/cbrewer_scales.png" alt="color brewer"></p>
<h2 id="step-4-writing-data-to-a-file"><a href="#step-4-writing-data-to-a-file" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 4 - Writing data to a file</h2>
<p>In addition to printing data to the General Stats, MultiQC modules typically
also write to text-files to allow people to easily use the data in downstream
applications. This also gives the opportunity to output additional data that
may not be appropriate for the General Statistics table.</p>
<p>Again, there is a base class function to help you with this - just supply it
with a dictionary and a filename:</p>
<pre><code class="language-python hljs">data = {
    <span class="hljs-string">'sample_1'</span>: {
        <span class="hljs-string">'first_col'</span>: <span class="hljs-number">91.4</span>,
        <span class="hljs-string">'second_col'</span>: <span class="hljs-string">'78.2%'</span>
    },
    <span class="hljs-string">'sample_2'</span>: {
        <span class="hljs-string">'first_col'</span>: <span class="hljs-number">138.3</span>,
        <span class="hljs-string">'second_col'</span>: <span class="hljs-string">'66.3%'</span>
    }
}
self.write_data_file(data, <span class="hljs-string">'multiqc_mymod'</span>)</code></pre>
<p>If your output has a lot of columns, you can supply the additional
argument <code>sort_cols = True</code> to have the columns alphabetically sorted.</p>
<p>This function will also pay attention to the default / command line
supplied data format and behave accordingly. So the written file could
be a tab-separated file (default), <code>JSON</code> or <code>YAML</code>.</p>
<p>Note that any keys with more than 2 levels of nesting will be ignored
when being written to tab-separated files.</p>
<h2 id="step-5-create-report-sections"><a href="#step-5-create-report-sections" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 5 - Create report sections</h2>
<p>Great! It's time to start creating sections of the report with more information.
To do this, use the <code>self.add_section()</code> helper function.
This supports the following arguments:</p>
<ul>
<li><code>name</code>: Name of the section, used for the title</li>
<li><code>anchor</code>: The URL anchor - must be unique, used when clicking the name in the side-nav</li>
<li><code>description</code>: A very short descriptive text to go above the plot (markdown).</li>
<li><code>comment</code>: A comment to add under the description. Big and blue text, mostly for users to customise the report (markdown).</li>
<li><code>helptext</code>: Longer help text explaining what users should look for (markdown).</li>
<li><code>plot</code>: Results from one of the MultiQC plotting functions</li>
<li><code>content</code>: Any custom HTML</li>
<li><code>autoformat</code>: Default <code>True</code>. Automatically format the <code>description</code>, <code>comment</code> and <code>helptext</code> strings.</li>
<li><code>autoformat_type</code>: Default <code>markdown</code>. Autoformat text type. Currently only <code>markdown</code> supported.</li>
</ul>
<p>For example:</p>
<pre><code class="language-python hljs">self.add_section (
    name = <span class="hljs-string">'Second Module Section'</span>,
    anchor = <span class="hljs-string">'mymod-second'</span>,
    plot = linegraph.plot(data2)
)
self.add_section (
    name = <span class="hljs-string">'First Module Section'</span>,
    anchor = <span class="hljs-string">'mymod-first'</span>,
    description = <span class="hljs-string">'My amazing module output, from the first section'</span>,
    helptext = <span class="hljs-string">"""
        If you're not sure _how_ to interpret the data, we can help!
        Most modules use multi-line strings for these text blocks,
        with triple quotation marks.

        * Markdown
        * Lists
        * Are
        * `Great`
    """</span>,
    plot = bargraph.plot(data)
)
self.add_section (
    content = <span class="hljs-string">'&lt;p&gt;Some custom HTML.&lt;/p&gt;'</span>
)</code></pre>
<p>If a module has more than one section, these will automatically be labelled and linked
in the left side-bar navigation (unless <code>name</code> is not specified).</p>
<h2 id="step-6-plot-some-data"><a href="#step-6-plot-some-data" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Step 6 - Plot some data</h2>
<p>Ok, you have some data, now the fun bit - visualising it! Each of the plot
types is described in the <em>Plotting Functions</em> section of the docs.</p>
<h2 id="appendices"><a href="#appendices" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Appendices</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#user-configuration">User configuration</a></li>
<li><a href="#profiling-performance">Profiling Performance</a></li>
<li><a href="#adding-custom-css--javascript">Adding Custom CSS / Javascript</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<h3 id="user-configuration"><a href="#user-configuration" class="header-link"><span class="glyphicon glyphicon-link"></span></a>User configuration</h3>
<p>Instead of hardcoding defaults, it's a great idea to allow users to configure
the behaviour of MultiQC module code.</p>
<p>It's pretty easy to use the built in MultiQC configuration settings to do this,
so that users can set up their config as described
<a href="http://multiqc.info/docs/#configuring-multiqc">above in the docs</a>.</p>
<p>To do this, just assume that your configuration variables are available in the
MultiQC <code>config</code> module and have sensible defaults. For example:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc <span class="hljs-keyword">import</span> config

mymod_config = getattr(config, <span class="hljs-string">'mymod_config'</span>, {})
my_custom_config_var = mymod_config.get(<span class="hljs-string">'my_custom_config_var'</span>, <span class="hljs-number">5</span>)</code></pre>
<p>You now have a variable <code>my_custom_config_var</code> with a default value of 5, but that
can be configured by a user as follows:</p>
<pre><code class="language-yaml hljs"><span class="hljs-attr">mymod_config:</span>
<span class="hljs-attr">  my_custom_config_var:</span> <span class="hljs-number">200</span></code></pre>
<p>Please be sure to use a unique top-level config name to avoid clashes - prefixing
with your module name is a good idea as in the example above. Keep all module config
options under the same top-level name for clarity.</p>
<p>Finally, don't forget to document the usage of your module-specific configuration
in <code>docs/modules/mymodule.md</code> so that people know how to use it.</p>
<h3 id="profiling-performance"><a href="#profiling-performance" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Profiling Performance</h3>
<p>It's important that MultiQC runs quickly and efficiently, especially on big
projects with large numbers of samples. The recommended method to check this is
by using <code>cProfile</code> to profile the code execution.</p>
<p>To do this, first find out where your copy of MultiQC is located:</p>
<pre><code class="language-console">$ which multiqc
/Users/you/anaconda/envs/myenv/bin/multiqc</code></pre>
<p>Then run MultiQC with this path and the <code>cProfile</code> module as follows
(the flags at the end can be any regular MultiQC flags):</p>
<pre><code class="language-bash hljs">python -m cProfile -o multiqc_profile.prof /Users/you/anaconda/envs/myenv/bin/multiqc -f .</code></pre>
<p>You can create a <code>.bashrc</code> alias to make this easier to run:</p>
<pre><code class="language-bash hljs"><span class="hljs-built_in">alias</span> profile_multiqc=<span class="hljs-string">'python -m cProfile -o multiqc_profile.prof /Users/you/anaconda/envs/myenv/bin/multiqc '</span>
profile_multiqc -f .</code></pre>
<p>MultiQC should run as normal, but produce the additional binary file <code>multiqc_profile.prof</code>.
This can then be visualised with software such as <a href="https://jiffyclub.github.io/snakeviz/">SnakeViz</a>.</p>
<p>To install SnakeViz and visualise the results, do the following:</p>
<pre><code class="language-bash hljs">pip install snakeviz
snakeviz multiqc_profile.prof</code></pre>
<p>A web page should open where you can explore the execution times of different nested functions.
It's a good idea to run MultiQC with a comparable number of results from other tools (eg. FastQC)
to have a reference to compare against for how long the code should take to run.</p>
<h3 id="adding-custom-css--javascript"><a href="#adding-custom-css--javascript" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Adding Custom CSS / Javascript</h3>
<p>If you would like module-specific CSS and / or JavaScript added to the template,
just add to the <code>self.css</code> and <code>self.js</code> dictionaries that come with the
<code>BaseMultiqcModule</code> class. The key should be the filename that you want your file to
have in the generated report folder <em>(this is ignored in the default template, which
includes the content file directly in the HTML)</em>. The dictionary value should be
the path to the desired file. For example, see how it's done in the FastQC module:</p>
<pre><code class="language-python hljs">self.css = {
    <span class="hljs-string">'assets/css/multiqc_fastqc.css'</span> :
        os.path.join(os.path.dirname(__file__), <span class="hljs-string">'assets'</span>, <span class="hljs-string">'css'</span>, <span class="hljs-string">'multiqc_fastqc.css'</span>)
}
self.js = {
    <span class="hljs-string">'assets/js/multiqc_fastqc.js'</span> :
        os.path.join(os.path.dirname(__file__), <span class="hljs-string">'assets'</span>, <span class="hljs-string">'js'</span>, <span class="hljs-string">'multiqc_fastqc.js'</span>)
}</code></pre></div><div class="docs_block" id="plots.md"><h1 id="plotting-functions"><a href="#plotting-functions" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Plotting Functions</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#common-options">Common options</a></li>
<li><a href="#bar-graphs">Bar graphs</a>
<ul class="nav nav-stacked">
<li><a href="#switching-datasets">Switching datasets</a></li>
<li><a href="#interactive--flat-image-plots">Interactive / Flat image plots</a></li>
</ul>
</li>
<li><a href="#line-graphs">Line graphs</a>
<ul class="nav nav-stacked">
<li><a href="#switching-datasets-1">Switching datasets</a></li>
<li><a href="#additional-data-series">Additional data series</a></li>
</ul>
</li>
<li><a href="#scatter-plots">Scatter Plots</a></li>
<li><a href="#creating-a-table">Creating a table</a>
<ul class="nav nav-stacked">
<li><a href="#table-decimal-places">Table decimal places</a></li>
<li><a href="#table-colour-scales-1">Table colour scales</a></li>
<li><a href="#custom-cell-background-colours">Custom cell background colours</a></li>
<li><a href="#zero-centrepoints">Zero centrepoints</a></li>
<li><a href="#conditional-formatting-of-data-values">Conditional formatting of data values</a></li>
</ul>
</li>
<li><a href="#beeswarm-plots-dot-plots">Beeswarm plots (dot plots)</a></li>
<li><a href="#heatmaps">Heatmaps</a></li>
<li><a href="#javascript-functions">Javascript Functions</a>
<ul class="nav nav-stacked">
<li><a href="#plotting-line-graphs">Plotting line graphs</a></li>
<li><a href="#plotting-bar-graphs">Plotting bar graphs</a></li>
<li><a href="#switching-counts-and-percentages">Switching counts and percentages</a></li>
<li><a href="#switching-plot-datasets">Switching plot datasets</a></li>
<li><a href="#custom-event-triggers">Custom event triggers</a></li>
</ul>
</li>

</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>MultiQC plotting functions are held within <code>multiqc.plots</code> submodules.
To use them, simply import the modules you want, eg.:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc.plots <span class="hljs-keyword">import</span> bargraph, linegraph</code></pre>
<p>Once you've done that, you will have access to the corresponding plotting
functions:</p>
<pre><code class="language-python hljs">bargraph.plot()
linegraph.plot()
scatter.plot()
table.plot()
beeswarm.plot()
heatmap.plot()</code></pre>
<p>These have been designed to work in a similar manner to each other - you
pass a data structure to them, along with optional extras such as categories
and configuration options, and they return a string of HTML to add to the
report. You can add this to the module introduction or sections as described
above. For example:</p>
<pre><code class="language-python hljs">self.add_section (
    name = <span class="hljs-string">'Module Section'</span>,
    anchor = <span class="hljs-string">'mymod_section'</span>,
    description = <span class="hljs-string">'This plot shows some really nice data.'</span>,
    helptext = <span class="hljs-string">'This longer string (can be **markdown**) helps explain how to interpret the plot'</span>,
    plot = bargraph.plot(self.parsed_data, categories, pconfig)
)</code></pre>
<h2 id="common-options"><a href="#common-options" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Common options</h2>
<p>All plots should as a minimum have a config with an <code>id</code> and a <code>title</code>.
MultiQC is written to work with sensible defaults, so won't complain if you
don't supply these, but it's good practice for usability (the ID is used as
a filename when exporting plots, and all plots should have a title when exported).</p>
<p>Plot titles should use the format <em>Module name: Plot name</em> (this is partly for
ease of use within MegaQC and other downstream tools).</p>
<h2 id="bar-graphs"><a href="#bar-graphs" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Bar graphs</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#switching-datasets">Switching datasets</a></li>
<li><a href="#interactive--flat-image-plots">Interactive / Flat image plots</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Simple data can be plotted in bar graphs. Many MultiQC modules make use
of stacked bar graphs. Here, the <code>bargraph.plot()</code> function comes to
the rescue. A basic example is as follows:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc.plots <span class="hljs-keyword">import</span> bargraph
data = {
    <span class="hljs-string">'sample 1'</span>: {
        <span class="hljs-string">'aligned'</span>: <span class="hljs-number">23542</span>,
        <span class="hljs-string">'not_aligned'</span>: <span class="hljs-number">343</span>,
    },
    <span class="hljs-string">'sample 2'</span>: {
        <span class="hljs-string">'not_aligned'</span>: <span class="hljs-number">7328</span>,
        <span class="hljs-string">'aligned'</span>: <span class="hljs-number">1275</span>,
    }
}
html_content = bargraph.plot(data)</code></pre>
<p>To specify the order of categories in the plot, you can supply a list of
dictionary keys. This can also be used to exclude a key from the plot.</p>
<pre><code class="language-python hljs">cats = [<span class="hljs-string">'aligned'</span>, <span class="hljs-string">'not_aligned'</span>]
html_content = bargraph.plot(data, cats)</code></pre>
<p>If <code>cats</code> is given as a dict instead of a list, you can specify a nice name
and a colour too. Make it an OrderedDict to specify the order:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> collections <span class="hljs-keyword">import</span> OrderedDict
cats = OrderedDict()
cats[<span class="hljs-string">'aligned'</span>] = {
    <span class="hljs-string">'name'</span>: <span class="hljs-string">'Aligned Reads'</span>,
    <span class="hljs-string">'color'</span>: <span class="hljs-string">'#8bbc21'</span>
}
cats[<span class="hljs-string">'not_aligned'</span>] = {
    <span class="hljs-string">'name'</span>: <span class="hljs-string">'Unaligned Reads'</span>,
    <span class="hljs-string">'color'</span>: <span class="hljs-string">'#f7a35c'</span>
}</code></pre>
<p>Finally, a third variable should be supplied with configuration variables for
the plot. The defaults are as follows:</p>
<pre><code class="language-python hljs">config = {
    <span class="hljs-comment"># Building the plot</span>
    <span class="hljs-string">'id'</span>: <span class="hljs-string">'&lt;random string&gt;'</span>,                <span class="hljs-comment"># HTML ID used for plot</span>
    <span class="hljs-string">'cpswitch'</span>: <span class="hljs-literal">True</span>,                       <span class="hljs-comment"># Show the 'Counts / Percentages' switch?</span>
    <span class="hljs-string">'cpswitch_c_active'</span>: <span class="hljs-literal">True</span>,              <span class="hljs-comment"># Initial display with 'Counts' specified? False for percentages.</span>
    <span class="hljs-string">'cpswitch_counts_label'</span>: <span class="hljs-string">'Counts'</span>,      <span class="hljs-comment"># Label for 'Counts' button</span>
    <span class="hljs-string">'cpswitch_percent_label'</span>: <span class="hljs-string">'Percentages'</span> <span class="hljs-comment"># Label for 'Percentages' button</span>
    <span class="hljs-string">'logswitch'</span>: <span class="hljs-literal">False</span>,                     <span class="hljs-comment"># Show the 'Log10' switch?</span>
    <span class="hljs-string">'logswitch_active'</span>: <span class="hljs-literal">False</span>,              <span class="hljs-comment"># Initial display with 'Log10' active?</span>
    <span class="hljs-string">'logswitch_label'</span>: <span class="hljs-string">'Log10'</span>,             <span class="hljs-comment"># Label for 'Log10' button</span>
    <span class="hljs-string">'hide_zero_cats'</span>: <span class="hljs-literal">True</span>,                 <span class="hljs-comment"># Hide categories where data for all samples is 0</span>
    <span class="hljs-comment"># Customising the plot</span>
    <span class="hljs-string">'title'</span>: <span class="hljs-literal">None</span>,                          <span class="hljs-comment"># Plot title - should be in format "Module Name: Plot Title"</span>
    <span class="hljs-string">'xlab'</span>: <span class="hljs-literal">None</span>,                           <span class="hljs-comment"># X axis label</span>
    <span class="hljs-string">'ylab'</span>: <span class="hljs-literal">None</span>,                           <span class="hljs-comment"># Y axis label</span>
    <span class="hljs-string">'labelSize'</span>: <span class="hljs-number">8</span>,                         <span class="hljs-comment"># Axis label font size</span>
    <span class="hljs-string">'ymax'</span>: <span class="hljs-literal">None</span>,                           <span class="hljs-comment"># Max y limit</span>
    <span class="hljs-string">'ymin'</span>: <span class="hljs-literal">None</span>,                           <span class="hljs-comment"># Min y limit</span>
    <span class="hljs-string">'yCeiling'</span>: <span class="hljs-literal">None</span>,                       <span class="hljs-comment"># Maximum value for automatic axis limit (good for percentages)</span>
    <span class="hljs-string">'yFloor'</span>: <span class="hljs-literal">None</span>,                         <span class="hljs-comment"># Minimum value for automatic axis limit</span>
    <span class="hljs-string">'yMinRange'</span>: <span class="hljs-literal">None</span>,                      <span class="hljs-comment"># Minimum range for axis</span>
    <span class="hljs-string">'yDecimals'</span>: <span class="hljs-literal">True</span>,                      <span class="hljs-comment"># Set to false to only show integer labels</span>
    <span class="hljs-string">'ylab_format'</span>: <span class="hljs-literal">None</span>,                    <span class="hljs-comment"># Format string for x axis labels. Defaults to {value}</span>
    <span class="hljs-string">'stacking'</span>: <span class="hljs-string">'normal'</span>,                   <span class="hljs-comment"># Set to None to have category bars side by side</span>
    <span class="hljs-string">'use_legend'</span>: <span class="hljs-literal">True</span>,                     <span class="hljs-comment"># Show / hide the legend</span>
    <span class="hljs-string">'click_func'</span>: <span class="hljs-literal">None</span>,                     <span class="hljs-comment"># Javascript function to be called when a point is clicked</span>
    <span class="hljs-string">'cursor'</span>: <span class="hljs-literal">None</span>,                         <span class="hljs-comment"># CSS mouse cursor type.</span>
    <span class="hljs-string">'tt_decimals'</span>: <span class="hljs-number">0</span>,                       <span class="hljs-comment"># Number of decimal places to use in the tooltip number</span>
    <span class="hljs-string">'tt_suffix'</span>: <span class="hljs-string">''</span>,                        <span class="hljs-comment"># Suffix to add after tooltip number</span>
    <span class="hljs-string">'tt_percentages'</span>: <span class="hljs-literal">True</span>,                 <span class="hljs-comment"># Show the percentages of each count in the tooltip</span>
    <span class="hljs-string">'height'</span>: <span class="hljs-number">512</span>                           <span class="hljs-comment"># The default height of the plot, in pixels</span>
}</code></pre>
<blockquote>
<p>The keys <code>id</code> and <code>title</code> should always be passed as a minimum. The <code>id</code> is used
for the plot name when exporting. If left unset, the Plot Export panel will call
the filename <code>mqc_hcplot_gtucwirdzx.png</code> (with some other random string).
Plots should always have titles, especially as they can stand by themselves
when exported. The title should have the format <code>Modulename: Plot Name</code></p>
</blockquote>
<h3 id="switching-datasets"><a href="#switching-datasets" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Switching datasets</h3>
<p>It's possible to have single plot with buttons to switch between different
datasets. To do this, give a list of data objects to the <code>plot</code> function
and specify the <code>data_labels</code> config option with the text to be used for the buttons:</p>
<pre><code class="language-python hljs">config = {
    <span class="hljs-string">'data_labels'</span>: [<span class="hljs-string">'Reads'</span>, <span class="hljs-string">'Bases'</span>]
}
html_content = bargraph.plot([data1, data2], pconfig=config)</code></pre>
<p>You can also customise the y-axis label and min/max values for each dataset:</p>
<pre><code class="language-python hljs">config = {
    <span class="hljs-string">'data_labels'</span>: [
        {<span class="hljs-string">'name'</span>: <span class="hljs-string">'Reads'</span>, <span class="hljs-string">'ylab'</span>: <span class="hljs-string">'Number of Reads'</span>},
        {<span class="hljs-string">'name'</span>: <span class="hljs-string">'Bases'</span>, <span class="hljs-string">'ylab'</span>: <span class="hljs-string">'Number of Base Pairs'</span>, <span class="hljs-string">'ymax'</span>:<span class="hljs-number">100</span>}
    ]
}</code></pre>
<p>If supplying multiple datasets, you can also supply a list of category
objects. Make sure that they are in the same order as the data.</p>
<p>Categories should contain data keys, so if you're supplying a list of two datasets,
you should supply a list of two sets of keys for the categories. MultiQC will try to
guess categories from the data keys if categories are missing.</p>
<p>For example, with two datasets supplied as above:</p>
<pre><code class="language-python hljs">cats = [
    [<span class="hljs-string">'aligned_reads'</span>,<span class="hljs-string">'unaligned_reads'</span>],
    [<span class="hljs-string">'aligned_base_pairs'</span>,<span class="hljs-string">'unaligned_base_pairs'</span>],
]</code></pre>
<p>Or with additional customisation such as name and colour:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> collections <span class="hljs-keyword">import</span> OrderedDict
cats = [OrderedDict(), OrderedDict()]
cats[<span class="hljs-number">0</span>][<span class="hljs-string">'aligned_reads'</span>] =        {<span class="hljs-string">'name'</span>: <span class="hljs-string">'Aligned Reads'</span>,        <span class="hljs-string">'color'</span>: <span class="hljs-string">'#8bbc21'</span>}
cats[<span class="hljs-number">0</span>][<span class="hljs-string">'unaligned_reads'</span>] =      {<span class="hljs-string">'name'</span>: <span class="hljs-string">'Unaligned Reads'</span>,      <span class="hljs-string">'color'</span>: <span class="hljs-string">'#f7a35c'</span>}
cats[<span class="hljs-number">1</span>][<span class="hljs-string">'aligned_base_pairs'</span>] =   {<span class="hljs-string">'name'</span>: <span class="hljs-string">'Aligned Base Pairs'</span>,   <span class="hljs-string">'color'</span>: <span class="hljs-string">'#8bbc21'</span>}
cats[<span class="hljs-number">1</span>][<span class="hljs-string">'unaligned_base_pairs'</span>] = {<span class="hljs-string">'name'</span>: <span class="hljs-string">'Unaligned Base Pairs'</span>, <span class="hljs-string">'color'</span>: <span class="hljs-string">'#f7a35c'</span>}
html_content = bargraph.plot([data, data], cats, config)</code></pre>
<p>Note that, as in this example, the plot data can be the same dictionary supplied twice.</p>
<h3 id="interactive--flat-image-plots"><a href="#interactive--flat-image-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Interactive / Flat image plots</h3>
<p>Note that the <code>bargraph.plot()</code> function can generate both interactive
JavaScript (HighCharts) powered report plots <em>and</em> flat image plots made using
MatPlotLib. This choice is made within the function based on config variables
such as number of dataseries and command line flags.</p>
<p>Note that both plot types should come out looking pretty much identical. If
you spot something that's missing in the flat image plots, let me know.</p>
<h2 id="line-graphs"><a href="#line-graphs" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Line graphs</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#switching-datasets-1">Switching datasets</a></li>
<li><a href="#additional-data-series">Additional data series</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>This base function works much like the above, but for two-dimensional
data, to produce line graphs. It expects a dictionary with sample identifiers,
each containing numeric <code>x:y</code> points. For example:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc.plots <span class="hljs-keyword">import</span> linegraph
data = {
    <span class="hljs-string">'sample 1'</span>: {
        <span class="hljs-string">'&lt;x val 1&gt;'</span>: <span class="hljs-string">'&lt;y val 1&gt;'</span>,
        <span class="hljs-string">'&lt;x val 2&gt;'</span>: <span class="hljs-string">'&lt;y val 2&gt;'</span>
    },
    <span class="hljs-string">'sample 2'</span>: {
        <span class="hljs-string">'&lt;x val 1&gt;'</span>: <span class="hljs-string">'&lt;y val 1&gt;'</span>,
        <span class="hljs-string">'&lt;x val 2&gt;'</span>: <span class="hljs-string">'&lt;y val 2&gt;'</span>
    }
}
html_content = linegraph.plot(data)</code></pre>
<p>Additionally, a configuration dict can be supplied. The defaults are as follows:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc.plots <span class="hljs-keyword">import</span> linegraph
config = {
    <span class="hljs-comment"># Building the plot</span>
    <span class="hljs-string">'id'</span>: <span class="hljs-string">'&lt;random string&gt;'</span>,     <span class="hljs-comment"># HTML ID used for plot</span>
    <span class="hljs-string">'categories'</span>: <span class="hljs-literal">False</span>,         <span class="hljs-comment"># Set to True to use x values as categories instead of numbers.</span>
    <span class="hljs-string">'colors'</span>: dict()             <span class="hljs-comment"># Provide dict with keys = sample names and values colours</span>
    <span class="hljs-string">'smooth_points'</span>: <span class="hljs-literal">None</span>,       <span class="hljs-comment"># Supply a number to limit number of points / smooth data</span>
    <span class="hljs-string">'smooth_points_sumcounts'</span>: <span class="hljs-literal">True</span>, <span class="hljs-comment"># Sum counts in bins, or average? Can supply list for multiple datasets</span>
    <span class="hljs-string">'logswitch'</span>: <span class="hljs-literal">False</span>,          <span class="hljs-comment"># Show the 'Log10' switch?</span>
    <span class="hljs-string">'logswitch_active'</span>: <span class="hljs-literal">False</span>,   <span class="hljs-comment"># Initial display with 'Log10' active?</span>
    <span class="hljs-string">'logswitch_label'</span>: <span class="hljs-string">'Log10'</span>,  <span class="hljs-comment"># Label for 'Log10' button</span>
    <span class="hljs-string">'extra_series'</span>: <span class="hljs-literal">None</span>,        <span class="hljs-comment"># See section below</span>
    <span class="hljs-comment"># Plot configuration</span>
    <span class="hljs-string">'title'</span>: <span class="hljs-literal">None</span>,               <span class="hljs-comment"># Plot title - should be in format "Module Name: Plot Title"</span>
    <span class="hljs-string">'xlab'</span>: <span class="hljs-literal">None</span>,                <span class="hljs-comment"># X axis label</span>
    <span class="hljs-string">'ylab'</span>: <span class="hljs-literal">None</span>,                <span class="hljs-comment"># Y axis label</span>
    <span class="hljs-string">'labelSize'</span>: <span class="hljs-number">8</span>,              <span class="hljs-comment"># Axis label font size</span>
    <span class="hljs-string">'xCeiling'</span>: <span class="hljs-literal">None</span>,            <span class="hljs-comment"># Maximum value for automatic axis limit (good for percentages)</span>
    <span class="hljs-string">'xFloor'</span>: <span class="hljs-literal">None</span>,              <span class="hljs-comment"># Minimum value for automatic axis limit</span>
    <span class="hljs-string">'xMinRange'</span>: <span class="hljs-literal">None</span>,           <span class="hljs-comment"># Minimum range for axis</span>
    <span class="hljs-string">'xmax'</span>: <span class="hljs-literal">None</span>,                <span class="hljs-comment"># Max x limit</span>
    <span class="hljs-string">'xmin'</span>: <span class="hljs-literal">None</span>,                <span class="hljs-comment"># Min x limit</span>
    <span class="hljs-string">'xLog'</span>: <span class="hljs-literal">False</span>,               <span class="hljs-comment"># Use log10 x axis?</span>
    <span class="hljs-string">'xDecimals'</span>: <span class="hljs-literal">True</span>,           <span class="hljs-comment"># Set to false to only show integer labels</span>
    <span class="hljs-string">'yCeiling'</span>: <span class="hljs-literal">None</span>,            <span class="hljs-comment"># Maximum value for automatic axis limit (good for percentages)</span>
    <span class="hljs-string">'yFloor'</span>: <span class="hljs-literal">None</span>,              <span class="hljs-comment"># Minimum value for automatic axis limit</span>
    <span class="hljs-string">'yMinRange'</span>: <span class="hljs-literal">None</span>,           <span class="hljs-comment"># Minimum range for axis</span>
    <span class="hljs-string">'ymax'</span>: <span class="hljs-literal">None</span>,                <span class="hljs-comment"># Max y limit</span>
    <span class="hljs-string">'ymin'</span>: <span class="hljs-literal">None</span>,                <span class="hljs-comment"># Min y limit</span>
    <span class="hljs-string">'yLog'</span>: <span class="hljs-literal">False</span>,               <span class="hljs-comment"># Use log10 y axis?</span>
    <span class="hljs-string">'yDecimals'</span>: <span class="hljs-literal">True</span>,           <span class="hljs-comment"># Set to false to only show integer labels</span>
    <span class="hljs-string">'yPlotBands'</span>: <span class="hljs-literal">None</span>,          <span class="hljs-comment"># Highlighted background bands. See http://api.highcharts.com/highcharts#yAxis.plotBands</span>
    <span class="hljs-string">'xPlotBands'</span>: <span class="hljs-literal">None</span>,          <span class="hljs-comment"># Highlighted background bands. See http://api.highcharts.com/highcharts#xAxis.plotBands</span>
    <span class="hljs-string">'yPlotLines'</span>: <span class="hljs-literal">None</span>,          <span class="hljs-comment"># Highlighted background lines. See http://api.highcharts.com/highcharts#yAxis.plotLines</span>
    <span class="hljs-string">'xPlotLines'</span>: <span class="hljs-literal">None</span>,          <span class="hljs-comment"># Highlighted background lines. See http://api.highcharts.com/highcharts#xAxis.plotLines</span>
    <span class="hljs-string">'xLabelFormat'</span>: <span class="hljs-string">'{value}'</span>,   <span class="hljs-comment"># Format string for the axis labels</span>
    <span class="hljs-string">'yLabelFormat'</span>: <span class="hljs-string">'{value}'</span>,   <span class="hljs-comment"># Format string for the axis labels</span>
    <span class="hljs-string">'tt_label'</span>: <span class="hljs-string">'{point.x}: {point.y:.2f}'</span>, <span class="hljs-comment"># Use to customise tooltip label, eg. '{point.x} base pairs'</span>
    <span class="hljs-string">'tt_decimals'</span>: <span class="hljs-literal">None</span>,         <span class="hljs-comment"># Tooltip decimals when categories = True (when false use tt_label)</span>
    <span class="hljs-string">'tt_suffix'</span>: <span class="hljs-literal">None</span>,           <span class="hljs-comment"># Tooltip suffix when categories = True (when false use tt_label)</span>
    <span class="hljs-string">'pointFormat'</span>: <span class="hljs-literal">None</span>,         <span class="hljs-comment"># Replace the default HTML for the entire tooltip label</span>
    <span class="hljs-string">'click_func'</span>: function(){},  <span class="hljs-comment"># Javascript function to be called when a point is clicked</span>
    <span class="hljs-string">'cursor'</span>: <span class="hljs-literal">None</span>               <span class="hljs-comment"># CSS mouse cursor type. Defaults to pointer when 'click_func' specified</span>
    <span class="hljs-string">'reversedStacks'</span>: <span class="hljs-literal">False</span>      <span class="hljs-comment"># Reverse the order of the category stacks. Defaults True for plots with Log10 option</span>
    <span class="hljs-string">'height'</span>: <span class="hljs-number">512</span>                <span class="hljs-comment"># The default height of the plot, in pixels</span>
}
html_content = linegraph.plot(data, config)</code></pre>
<blockquote>
<p>The keys <code>id</code> and <code>title</code> should always be passed as a minimum. The <code>id</code> is used
for the plot name when exporting. If left unset, the Plot Export panel will call
the filename <code>mqc_hcplot_gtucwirdzx.png</code> (with some other random string).
Plots should always have titles, especially as they can stand by themselves
when exported. The title should have the format <code>Modulename: Plot Name</code></p>
</blockquote>
<h3 id="switching-datasets-1"><a href="#switching-datasets-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Switching datasets</h3>
<p>You can also have a single plot with buttons to switch between different
datasets. To do this, just supply a list of data dicts instead (same
formats as described above). For example:</p>
<pre><code class="language-python hljs">data = [
    {
        <span class="hljs-string">'sample 1'</span>: { <span class="hljs-string">'&lt;x val 1&gt;'</span>: <span class="hljs-string">'&lt;y val 1&gt;'</span>, <span class="hljs-string">'&lt;x val 2&gt;'</span>: <span class="hljs-string">'&lt;y val 2&gt;'</span> },
        <span class="hljs-string">'sample 2'</span>: { <span class="hljs-string">'&lt;x val 1&gt;'</span>: <span class="hljs-string">'&lt;y val 1&gt;'</span>, <span class="hljs-string">'&lt;x val 2&gt;'</span>: <span class="hljs-string">'&lt;y val 2&gt;'</span> }
    },
    {
        <span class="hljs-string">'sample 1'</span>: { <span class="hljs-string">'&lt;x val 1&gt;'</span>: <span class="hljs-string">'&lt;y val 1&gt;'</span>, <span class="hljs-string">'&lt;x val 2&gt;'</span>: <span class="hljs-string">'&lt;y val 2&gt;'</span> },
        <span class="hljs-string">'sample 2'</span>: { <span class="hljs-string">'&lt;x val 1&gt;'</span>: <span class="hljs-string">'&lt;y val 1&gt;'</span>, <span class="hljs-string">'&lt;x val 2&gt;'</span>: <span class="hljs-string">'&lt;y val 2&gt;'</span> }
    }
]</code></pre>
<p>You'll also want to add the following configuration options to
give names to the buttons and graph labels:</p>
<pre><code class="language-python hljs">config = {
    <span class="hljs-string">'data_labels'</span>: [
        {<span class="hljs-string">'name'</span>: <span class="hljs-string">'DS 1'</span>, <span class="hljs-string">'ylab'</span>: <span class="hljs-string">'Dataset 1'</span>, <span class="hljs-string">'xlab'</span>: <span class="hljs-string">'x Axis 1'</span>},
        {<span class="hljs-string">'name'</span>: <span class="hljs-string">'DS 2'</span>, <span class="hljs-string">'ylab'</span>: <span class="hljs-string">'Dataset 2'</span>, <span class="hljs-string">'xlab'</span>: <span class="hljs-string">'x Axis 2'</span>}
    ]
}</code></pre>
<p>All of these config values are optional, the function will default
to sensible values if things are missing.</p>
<h3 id="additional-data-series"><a href="#additional-data-series" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Additional data series</h3>
<p>Sometimes, it's good to be able to specify specific data series manually.
To do this, use <code>config['extra_series']</code>. For a single extra line this can
be a dict (as below). For multiple lines, use a list of dicts. For multiple
dataset plots, use a list of list of dicts.</p>
<p>For example, to add a dotted <code>x = y</code> reference line:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc.plots <span class="hljs-keyword">import</span> linegraph
config = {
    <span class="hljs-string">'extra_series'</span>: {
        <span class="hljs-string">'name'</span>: <span class="hljs-string">'x = y'</span>,
        <span class="hljs-string">'data'</span>: [[<span class="hljs-number">0</span>, <span class="hljs-number">0</span>], [max_x_val, max_y_val]],
        <span class="hljs-string">'dashStyle'</span>: <span class="hljs-string">'Dash'</span>,
        <span class="hljs-string">'lineWidth'</span>: <span class="hljs-number">1</span>,
        <span class="hljs-string">'color'</span>: <span class="hljs-string">'#000000'</span>,
        <span class="hljs-string">'marker'</span>: { <span class="hljs-string">'enabled'</span>: <span class="hljs-literal">False</span> },
        <span class="hljs-string">'enableMouseTracking'</span>: <span class="hljs-literal">False</span>,
        <span class="hljs-string">'showInLegend'</span>: <span class="hljs-literal">False</span>,
    }
}
html_content = linegraph.plot(data, config)</code></pre>
<h2 id="scatter-plots"><a href="#scatter-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Scatter Plots</h2>
<p>Scatter plots work in almost exactly the same way as line plots. Most (if not all)
config options are shared between the two. The data structure is similar but not identical:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc.plots <span class="hljs-keyword">import</span> scatter
data = {
    <span class="hljs-string">'sample 1'</span>: {
        x: <span class="hljs-string">'&lt;x val&gt;'</span>,
        y: <span class="hljs-string">'&lt;y val&gt;'</span>
    },
    <span class="hljs-string">'sample 2'</span>: {
        x: <span class="hljs-string">'&lt;x val&gt;'</span>,
        y: <span class="hljs-string">'&lt;y val&gt;'</span>
    }
}
html_content = scatter.plot(data)</code></pre>
<p>Note that you must use the keys <code>x</code> and <code>y</code> for each data point.</p>
<p>If you want more than one data point per sample, you can supply a list of
dictionaries instead. You can also optionally specify point colours and
sample name suffixes (these are appended to the sample name):</p>
<pre><code class="language-python hljs">data = {
    <span class="hljs-string">'sample 1'</span>: [
        { x: <span class="hljs-string">'&lt;x val&gt;'</span>, y: <span class="hljs-string">'&lt;y val&gt;'</span>, color: <span class="hljs-string">'#a6cee3'</span>, name: <span class="hljs-string">'Type 1'</span> },
        { x: <span class="hljs-string">'&lt;x val&gt;'</span>, y: <span class="hljs-string">'&lt;y val&gt;'</span>, color: <span class="hljs-string">'#1f78b4'</span>, name: <span class="hljs-string">'Type 2'</span> }
    ],
    <span class="hljs-string">'sample 2'</span>: [
        { x: <span class="hljs-string">'&lt;x val&gt;'</span>, y: <span class="hljs-string">'&lt;y val&gt;'</span>, color: <span class="hljs-string">'#b2df8a'</span>, name: <span class="hljs-string">'Type 1'</span> },
        { x: <span class="hljs-string">'&lt;x val&gt;'</span>, y: <span class="hljs-string">'&lt;y val&gt;'</span>, color: <span class="hljs-string">'#33a02c'</span>, name: <span class="hljs-string">'Type 2'</span> }
    ]
}</code></pre>
<p>Remember that MultiQC reports can contain large numbers of samples, so this plot type
is <strong>not</strong> suitable for large quantities of data - 20,000 genes might look good
for one sample, but when someone runs MultiQC with 500 samples, it will crash
the browser and be impossible to interpret.</p>
<p>See the above docs about line plots for most config options. The scatter plot
has a handful of unique ones in addition:</p>
<pre><code class="language-python hljs">pconfig = {
    <span class="hljs-string">'marker_colour'</span>: <span class="hljs-string">'rgba(124, 181, 236, .5)'</span>, <span class="hljs-comment"># string, base colour of points (recommend rgba / semi-transparent)</span>
    <span class="hljs-string">'marker_size'</span>: <span class="hljs-number">5</span>,               <span class="hljs-comment"># int, size of points</span>
    <span class="hljs-string">'marker_line_colour'</span>: <span class="hljs-string">'#999'</span>,   <span class="hljs-comment"># string, colour of point border</span>
    <span class="hljs-string">'marker_line_width'</span>: <span class="hljs-number">1</span>,         <span class="hljs-comment"># int, width of point border</span>
    <span class="hljs-string">'square'</span>: <span class="hljs-literal">False</span>                 <span class="hljs-comment"># Force the plot to stay square? (Maintain aspect ratio)</span>
}</code></pre>
<h2 id="creating-a-table"><a href="#creating-a-table" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Creating a table</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#table-decimal-places">Table decimal places</a></li>
<li><a href="#table-colour-scales-1">Table colour scales</a></li>
<li><a href="#custom-cell-background-colours">Custom cell background colours</a></li>
<li><a href="#zero-centrepoints">Zero centrepoints</a></li>
<li><a href="#conditional-formatting-of-data-values">Conditional formatting of data values</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>Tables should work just like the functions above (most like the bar
graph function). As a minimum, the function takes a dictionary containing
data - the first keys will be sample names (row headers) and each key
contained within will be a table column header.</p>
<p>You can also supply a list of key names to restrict the data in the table
to certain keys / columns. This also specifies the order that columns
should be displayed in.</p>
<p>For more customisation, the headers can be supplied as a dictionary. Each
key should match the keys used in the data dictionary, but values can
customise the output. If you want to specify the order of the columns, you
must use an <code>OrderedDict</code>.</p>
<p>Finally, the function accepts a config dictionary as a third parameter.
This can set global options for the table (eg. a title) and can also hold
default values to customise the output of all table columns.</p>
<p>The default header keys are:</p>
<pre><code class="language-python hljs">single_header = {
    <span class="hljs-string">'namespace'</span>: <span class="hljs-string">''</span>,                <span class="hljs-comment"># Name for grouping. Prepends desc and is in Config Columns modal</span>
    <span class="hljs-string">'title'</span>: <span class="hljs-string">'[ dict key ]'</span>,        <span class="hljs-comment"># Short title, table column title</span>
    <span class="hljs-string">'description'</span>: <span class="hljs-string">'[ dict key ]'</span>,  <span class="hljs-comment"># Longer description, goes in mouse hover text</span>
    <span class="hljs-string">'max'</span>: <span class="hljs-literal">None</span>,                    <span class="hljs-comment"># Minimum value in range, for bar / colour coding</span>
    <span class="hljs-string">'min'</span>: <span class="hljs-literal">None</span>,                    <span class="hljs-comment"># Maximum value in range, for bar / colour coding</span>
    <span class="hljs-string">'ceiling'</span>: <span class="hljs-literal">None</span>,                <span class="hljs-comment"># Maximum value for automatic bar limit</span>
    <span class="hljs-string">'floor'</span>: <span class="hljs-literal">None</span>,                  <span class="hljs-comment"># Minimum value for automatic bar limit</span>
    <span class="hljs-string">'minRange'</span>: <span class="hljs-literal">None</span>,               <span class="hljs-comment"># Minimum range for automatic bar</span>
    <span class="hljs-string">'scale'</span>: <span class="hljs-string">'GnBu'</span>,                <span class="hljs-comment"># Colour scale for colour coding. False to disable.</span>
    <span class="hljs-string">'bgcols'</span>: <span class="hljs-literal">None</span>,                 <span class="hljs-comment"># Dict with values: background colours for categorical data.</span>
    <span class="hljs-string">'colour'</span>: <span class="hljs-string">'&lt;auto&gt;'</span>,             <span class="hljs-comment"># Colour for column grouping</span>
    <span class="hljs-string">'suffix'</span>: <span class="hljs-literal">None</span>,                 <span class="hljs-comment"># Suffix for value (eg. '%')</span>
    <span class="hljs-string">'format'</span>: <span class="hljs-string">'{:,.1f}'</span>,            <span class="hljs-comment"># Value format string - default 1 decimal place</span>
    <span class="hljs-string">'cond_formatting_rules'</span>: <span class="hljs-literal">None</span>,  <span class="hljs-comment"># Rules for conditional formatting table cell values - see docs below</span>
    <span class="hljs-string">'cond_formatting_colours'</span>: <span class="hljs-literal">None</span>, <span class="hljs-comment"># Styles for conditional formatting of table cell values</span>
    <span class="hljs-string">'shared_key'</span>: <span class="hljs-literal">None</span>              <span class="hljs-comment"># See below for description</span>
    <span class="hljs-string">'modify'</span>: <span class="hljs-literal">None</span>,                 <span class="hljs-comment"># Lambda function to modify values</span>
    <span class="hljs-string">'hidden'</span>: <span class="hljs-literal">False</span>                 <span class="hljs-comment"># Set to True to hide the column on page load</span>
}</code></pre>
<p>A third parameter can be specified with settings for the whole table:</p>
<pre><code class="language-python hljs">table_config = {
    <span class="hljs-string">'namespace'</span>: <span class="hljs-string">''</span>,                         <span class="hljs-comment"># Name for grouping. Prepends desc and is in Config Columns modal</span>
    <span class="hljs-string">'id'</span>: <span class="hljs-string">'&lt;random string&gt;'</span>,                 <span class="hljs-comment"># ID used for the table</span>
    <span class="hljs-string">'table_title'</span>: <span class="hljs-string">'&lt;table id&gt;'</span>,             <span class="hljs-comment"># Title of the table. Used in the column config modal</span>
    <span class="hljs-string">'save_file'</span>: <span class="hljs-literal">False</span>,                      <span class="hljs-comment"># Whether to save the table data to a file</span>
    <span class="hljs-string">'raw_data_fn'</span>:<span class="hljs-string">'multiqc_&lt;table_id&gt;_table'</span> <span class="hljs-comment"># File basename to use for raw data file</span>
    <span class="hljs-string">'sortRows'</span>: <span class="hljs-literal">True</span>                         <span class="hljs-comment"># Whether to sort rows alphabetically</span>
    <span class="hljs-string">'only_defined_headers'</span>: <span class="hljs-literal">True</span>             <span class="hljs-comment"># Only show columns that are defined in the headers config</span>
    <span class="hljs-string">'col1_header'</span>: <span class="hljs-string">'Sample Name'</span>             <span class="hljs-comment"># The header used for the first column</span>
    <span class="hljs-string">'no_beeswarm'</span>: <span class="hljs-literal">False</span>    <span class="hljs-comment"># Force a table to always be plotted (beeswarm by default if many rows)</span>
}</code></pre>
<p>Most of the header keys can also be specified in the table config
(<code>namespace</code>, <code>scale</code>, <code>format</code>, <code>colour</code>, <code>hidden</code>, <code>max</code>, <code>min</code>, <code>ceiling</code>, <code>floor</code>, <code>minRange</code>, <code>shared_key</code>, <code>modify</code>).
These will then be applied to all columns prior to applying column-specific heading config.</p>
<p>A very basic example of creating a table is shown below:</p>
<pre><code class="language-python hljs">data = {
    <span class="hljs-string">'sample 1'</span>: {
        <span class="hljs-string">'aligned'</span>: <span class="hljs-number">23542</span>,
        <span class="hljs-string">'not_aligned'</span>: <span class="hljs-number">343</span>,
    },
    <span class="hljs-string">'sample 2'</span>: {
        <span class="hljs-string">'aligned'</span>: <span class="hljs-number">1275</span>,
        <span class="hljs-string">'not_aligned'</span>: <span class="hljs-number">7328</span>,
    }
}
table_html = table.plot(data)</code></pre>
<p>A more complicated version with ordered columns, defaults and column-specific
settings (eg. no decimal places):</p>
<pre><code class="language-python hljs">data = {
    <span class="hljs-string">'sample 1'</span>: {
        <span class="hljs-string">'aligned'</span>: <span class="hljs-number">23542</span>,
        <span class="hljs-string">'not_aligned'</span>: <span class="hljs-number">343</span>,
        <span class="hljs-string">'aligned_percent'</span>: <span class="hljs-number">98.563952271</span>
    },
    <span class="hljs-string">'sample 2'</span>: {
        <span class="hljs-string">'aligned'</span>: <span class="hljs-number">1275</span>,
        <span class="hljs-string">'not_aligned'</span>: <span class="hljs-number">7328</span>,
        <span class="hljs-string">'aligned_percent'</span>: <span class="hljs-number">14.820411484</span>
    }
}
headers = OrderedDict()
headers[<span class="hljs-string">'aligned_percent'</span>] = {
    <span class="hljs-string">'title'</span>: <span class="hljs-string">'% Aligned'</span>,
    <span class="hljs-string">'description'</span>: <span class="hljs-string">'Percentage of reads that aligned'</span>,
    <span class="hljs-string">'suffix'</span>: <span class="hljs-string">'%'</span>,
    <span class="hljs-string">'max'</span>: <span class="hljs-number">100</span>,
    <span class="hljs-string">'format'</span>: <span class="hljs-string">'{:,.0f}'</span> <span class="hljs-comment"># No decimal places please</span>
}
headers[<span class="hljs-string">'aligned'</span>] = {
    <span class="hljs-string">'title'</span>: <span class="hljs-string">'{} Aligned'</span>.format(config.read_count_prefix),
    <span class="hljs-string">'description'</span>: <span class="hljs-string">'Aligned Reads ({})'</span>.format(config.read_count_desc),
    <span class="hljs-string">'shared_key'</span>: <span class="hljs-string">'read_count'</span>,
    <span class="hljs-string">'modify'</span>: <span class="hljs-keyword">lambda</span> x: x * config.read_count_multiplier
}
config = {
    <span class="hljs-string">'namespace'</span>: <span class="hljs-string">'My Module'</span>,
    <span class="hljs-string">'min'</span>: <span class="hljs-number">0</span>,
    <span class="hljs-string">'scale'</span>: <span class="hljs-string">'GnBu'</span>
}
table_html = table.plot(data, headers, config)</code></pre>
<h3 id="table-decimal-places"><a href="#table-decimal-places" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Table decimal places</h3>
<p>You can customise how many decimal places a number has by using the <code>format</code> config
key for that column. The default format string is <code>'{:,.1f}'</code>, which specifies a
float number with a single decimal place. To remove decimals use <code>'{:,.0f}'</code>.
To have two decimal places, use <code>'{:,.2f}'</code>.</p>
<h3 id="table-colour-scales-1"><a href="#table-colour-scales-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Table colour scales</h3>
<p>Colour scales are taken from <a href="http://colorbrewer2.org/">ColorBrewer2</a>.
Colour scales can be reversed by adding the suffix <code>-rev</code> to the name. For example, <code>RdYlGn-rev</code>.</p>
<p>The following scales are available:</p>
<p><img src="MultiQC%20installation_files/cbrewer_scales.png" alt="color brewer"></p>
<h3 id="custom-cell-background-colours"><a href="#custom-cell-background-colours" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Custom cell background colours</h3>
<p>You can specify custom background colours for specific values using the <code>bgcols</code>
header config. This takes precedence over <code>scale</code>.</p>
<p>For example, a header config for a column could look like this:</p>
<pre><code class="language-python hljs">headers[tablecol] = {
    <span class="hljs-string">"title"</span>: <span class="hljs-string">"My table column"</span>,
    <span class="hljs-string">"bgcols"</span>: {
        <span class="hljs-string">"bad data"</span>: <span class="hljs-string">"#f8d7da"</span>,
        <span class="hljs-string">"ok data"</span>: <span class="hljs-string">"#fff3cd"</span>,
        <span class="hljs-string">"good data"</span>: <span class="hljs-string">"#d1e7dd"</span>
    }
}</code></pre>
<h3 id="zero-centrepoints"><a href="#zero-centrepoints" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Zero centrepoints</h3>
<p>If you set the header config <code>bars_zero_centrepoint</code> to <code>True</code>, the background bars
will use the absolute values to calculate bar width. So a value of <code>0</code> will have a bar
width of <code>0</code>, <code>20</code> a width of <code>20</code> and <code>-30</code> a width of <code>30</code>.</p>
<p>This works well with a divergent colour-scheme as the bar width shows the magnitude
of the value properly, whilst the colour scheme shows the difference between positive
and negative values.</p>
<p>For example:</p>
<pre><code class="language-python hljs">headers[tablecol] = {
    <span class="hljs-string">"title"</span>: <span class="hljs-string">"My table column"</span>,
    <span class="hljs-string">"scale"</span>: <span class="hljs-string">"RdYlGn"</span>,
    <span class="hljs-string">"bars_zero_centrepoint"</span>: <span class="hljs-literal">True</span>,
}</code></pre>
<h3 id="conditional-formatting-of-data-values"><a href="#conditional-formatting-of-data-values" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Conditional formatting of data values</h3>
<p>MultiQC has configuration options to allow users to configure <em>"conditional formatting"</em>,
with highlighted values in table cells (<a href="##conditional-formatting">see docs</a>).</p>
<p>Developers can also make use of this functionality within the header config dictionaries
for formatting data values.</p>
<p>The functionality follows the same logic as for user configs with the parameters
<code>cond_formatting_rules</code> and <code>cond_formatting_colours</code>. These correspond to the
user config options <code>table_cond_formatting_rules</code> and <code>table_cond_formatting_colours</code>,
with the exception that no column ID is needed for <code>table_cond_formatting_rules</code>.</p>
<p>For example, a simple header config could look as follows:</p>
<pre><code class="language-python hljs">headers[instrument] = {
    <span class="hljs-string">"title"</span>: <span class="hljs-string">"My table column"</span>,
    <span class="hljs-string">"cond_formatting_rules"</span>: {
        <span class="hljs-string">"pass"</span>: [{<span class="hljs-string">"s_eq"</span>: <span class="hljs-string">"good data"</span>}],
        <span class="hljs-string">"warn"</span>: [{<span class="hljs-string">"s_eq"</span>: <span class="hljs-string">"ok data"</span>}],
        <span class="hljs-string">"fail"</span>: [{<span class="hljs-string">"s_eq"</span>: <span class="hljs-string">"bad data"</span>}],
    }
}</code></pre>
<p>A more complex version with multiple rules could be:</p>
<pre><code class="language-python hljs">headers[tablecol] = {
    <span class="hljs-string">"title"</span>: <span class="hljs-string">"My table column"</span>,
    <span class="hljs-string">"cond_formatting_rules"</span>: {
        <span class="hljs-string">"brightgreen"</span>: [
            {<span class="hljs-string">"s_contains"</span>: <span class="hljs-string">"amazing"</span>},
            {<span class="hljs-string">"s_contains"</span>: <span class="hljs-string">"incredible"</span>},
        ],
        <span class="hljs-string">"brown"</span>: [{<span class="hljs-string">"s_ne"</span>: <span class="hljs-string">"rubbish-data"</span>}],
        <span class="hljs-string">"turquoise"</span>: [
            {<span class="hljs-string">"gt"</span>: <span class="hljs-number">4</span>},
            {<span class="hljs-string">"lt"</span>: <span class="hljs-number">12</span>},
        ],
    },
    <span class="hljs-string">"cond_formatting_colours"</span>: [
        {<span class="hljs-string">"brightgreen"</span>: <span class="hljs-string">"#39FF14"</span>},
        {<span class="hljs-string">"brown"</span>: <span class="hljs-string">"#A52A2A"</span>},
        {<span class="hljs-string">"turquoise"</span>: <span class="hljs-string">"#30D5C8"</span>},
    ]
}</code></pre>
<h2 id="beeswarm-plots-dot-plots"><a href="#beeswarm-plots-dot-plots" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Beeswarm plots (dot plots)</h2>
<p>Beeswarm plots work from the exact same data structure as tables, so the
usage is just the same. Except instead of calling <code>table</code>, call <code>beeswarm</code>:</p>
<pre><code class="language-python hljs">data = {
    <span class="hljs-string">'sample 1'</span>: {
        <span class="hljs-string">'aligned'</span>: <span class="hljs-number">23542</span>,
        <span class="hljs-string">'not_aligned'</span>: <span class="hljs-number">343</span>,
    },
    <span class="hljs-string">'sample 2'</span>: {
        <span class="hljs-string">'not_aligned'</span>: <span class="hljs-number">7328</span>,
        <span class="hljs-string">'aligned'</span>: <span class="hljs-number">1275</span>,
    }
}
beeswarm_html = beeswarm.plot(data)</code></pre>
<p>The function also accepts the same headers and config parameters.</p>
<h2 id="heatmaps"><a href="#heatmaps" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Heatmaps</h2>
<p>Heatmaps expect data in the structure of a list of lists. Then, a list
of sample names for the x-axis, and optionally for the y-axis (defaults
to the same as the x-axis).</p>
<pre><code class="language-python hljs">heatmap.plot(data, xcats, ycats, pconfig)</code></pre>
<p>A simple example:</p>
<pre><code class="language-python hljs">hmdata = [
    [<span class="hljs-number">0.9</span>, <span class="hljs-number">0.87</span>, <span class="hljs-number">0.73</span>, <span class="hljs-number">0.6</span>, <span class="hljs-number">0.2</span>, <span class="hljs-number">0.3</span>],
    [<span class="hljs-number">0.87</span>, <span class="hljs-number">1</span>, <span class="hljs-number">0.7</span>, <span class="hljs-number">0.6</span>, <span class="hljs-number">0.9</span>, <span class="hljs-number">0.3</span>],
    [<span class="hljs-number">0.73</span>, <span class="hljs-number">0.8</span>, <span class="hljs-number">1</span>, <span class="hljs-number">0.6</span>, <span class="hljs-number">0.9</span>, <span class="hljs-number">0.3</span>],
    [<span class="hljs-number">0.6</span>, <span class="hljs-number">0.8</span>, <span class="hljs-number">0.7</span>, <span class="hljs-number">1</span>, <span class="hljs-number">0.9</span>, <span class="hljs-number">0.3</span>],
    [<span class="hljs-number">0.2</span>, <span class="hljs-number">0.8</span>, <span class="hljs-number">0.7</span>, <span class="hljs-number">0.6</span>, <span class="hljs-number">1</span>, <span class="hljs-number">0.3</span>],
    [<span class="hljs-number">0.3</span>, <span class="hljs-number">0.8</span>, <span class="hljs-number">0.7</span>, <span class="hljs-number">0.6</span>, <span class="hljs-number">0.9</span>, <span class="hljs-number">1</span>],
]
names = [ <span class="hljs-string">'one'</span>, <span class="hljs-string">'two'</span>, <span class="hljs-string">'three'</span>, <span class="hljs-string">'four'</span>, <span class="hljs-string">'five'</span>, <span class="hljs-string">'six'</span> ]
hm_html = heatmap.plot(hmdata, names)</code></pre>
<p>Much like the other plots, you can change the way that the heatmap looks
using a config dictionary:</p>
<pre><code class="language-python hljs">pconfig = {
    <span class="hljs-string">'title'</span>: <span class="hljs-literal">None</span>,                 <span class="hljs-comment"># Plot title - should be in format "Module Name: Plot Title"</span>
    <span class="hljs-string">'xTitle'</span>: <span class="hljs-literal">None</span>,                <span class="hljs-comment"># X-axis title</span>
    <span class="hljs-string">'yTitle'</span>: <span class="hljs-literal">None</span>,                <span class="hljs-comment"># Y-axis title</span>
    <span class="hljs-string">'min'</span>: <span class="hljs-literal">None</span>,                   <span class="hljs-comment"># Minimum value (default: auto)</span>
    <span class="hljs-string">'max'</span>: <span class="hljs-literal">None</span>,                   <span class="hljs-comment"># Maximum value (default: auto)</span>
    <span class="hljs-string">'square'</span>: <span class="hljs-literal">True</span>,                <span class="hljs-comment"># Force the plot to stay square? (Maintain aspect ratio)</span>
    <span class="hljs-string">'xcats_samples'</span>: <span class="hljs-literal">True</span>,         <span class="hljs-comment"># Is the x-axis sample names? Set to False to prevent report toolbox from affecting.</span>
    <span class="hljs-string">'ycats_samples'</span>: <span class="hljs-literal">True</span>,         <span class="hljs-comment"># Is the y-axis sample names? Set to False to prevent report toolbox from affecting.</span>
    <span class="hljs-string">'colstops'</span>: []                 <span class="hljs-comment"># Scale colour stops. See below.</span>
    <span class="hljs-string">'reverseColors'</span>: <span class="hljs-literal">False</span>,        <span class="hljs-comment"># Reverse the order of the colour axis</span>
    <span class="hljs-string">'decimalPlaces'</span>: <span class="hljs-number">2</span>,            <span class="hljs-comment"># Number of decimal places for tooltip</span>
    <span class="hljs-string">'legend'</span>: <span class="hljs-literal">True</span>,                <span class="hljs-comment"># Colour axis key enabled or not</span>
    <span class="hljs-string">'borderWidth'</span>: <span class="hljs-number">0</span>,              <span class="hljs-comment"># Border width between cells</span>
    <span class="hljs-string">'datalabels'</span>: <span class="hljs-literal">True</span>,            <span class="hljs-comment"># Show values in each cell. Defaults True when less than 20 samples.</span>
    <span class="hljs-string">'datalabel_colour'</span>: <span class="hljs-string">'&lt;auto&gt;'</span>,  <span class="hljs-comment"># Colour of text for values. Defaults to auto contrast.</span>
    <span class="hljs-string">'height'</span>: <span class="hljs-number">512</span>                  <span class="hljs-comment"># The default height of the interactive plot, in pixels</span>
}</code></pre>
<p>The colour stops are a bit special and can be used to define a custom colour
scheme. These should be defined as a list of lists, with a number between 0 and 1
and a HTML colour. The default is <code>RdYlBu</code> from <a href="http://colorbrewer2.org/">ColorBrewer</a>:</p>
<pre><code class="language-python hljs">pconfig = {
    <span class="hljs-string">'colstops'</span> = [
        [<span class="hljs-number">0</span>, <span class="hljs-string">'#313695'</span>],
        [<span class="hljs-number">0.1</span>, <span class="hljs-string">'#4575b4'</span>],
        [<span class="hljs-number">0.2</span>, <span class="hljs-string">'#74add1'</span>],
        [<span class="hljs-number">0.3</span>, <span class="hljs-string">'#abd9e9'</span>],
        [<span class="hljs-number">0.4</span>, <span class="hljs-string">'#e0f3f8'</span>],
        [<span class="hljs-number">0.5</span>, <span class="hljs-string">'#ffffbf'</span>],
        [<span class="hljs-number">0.6</span>, <span class="hljs-string">'#fee090'</span>],
        [<span class="hljs-number">0.7</span>, <span class="hljs-string">'#fdae61'</span>],
        [<span class="hljs-number">0.8</span>, <span class="hljs-string">'#f46d43'</span>],
        [<span class="hljs-number">0.9</span>, <span class="hljs-string">'#d73027'</span>],
        [<span class="hljs-number">1</span>, <span class="hljs-string">'#a50026'</span>],
    ]
}</code></pre>
<h2 id="javascript-functions"><a href="#javascript-functions" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Javascript Functions</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#plotting-line-graphs">Plotting line graphs</a></li>
<li><a href="#plotting-bar-graphs">Plotting bar graphs</a></li>
<li><a href="#switching-counts-and-percentages">Switching counts and percentages</a></li>
<li><a href="#switching-plot-datasets">Switching plot datasets</a></li>
<li><a href="#custom-event-triggers">Custom event triggers</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>The javascript bundled in the default MultiQC template has a number of
helper functions to make your life easier.</p>
<blockquote>
<p>NB: The MultiQC Python functions make use of these, so it's very unlikely
that you'll need to use any of this. But it's here for reference.</p>
</blockquote>
<h3 id="plotting-line-graphs"><a href="#plotting-line-graphs" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Plotting line graphs</h3>
<p><code>plot_xy_line_graph (target, ds)</code></p>
<p>Plots a line graph with multiple series of (x,y) data pairs. Used by
the <a href="http://multiqc.info/docs/#line-graphs">linegraph.plot()</a>
python function.</p>
<p>Data and configuration must be added to the document level
<code>mqc_plots</code> variable on page load, using the target as the key.
The variables used are as follows:</p>
<pre><code class="language-javascript hljs">mqc_plots[target][<span class="hljs-string">"plot_type"</span>] = <span class="hljs-string">"xy_line"</span>;
mqc_plots[target][<span class="hljs-string">"config"</span>];
mqc_plots[target][<span class="hljs-string">"datasets"</span>];</code></pre>
<p>Multiple datasets can be added in the <code>['datasets']</code> array. The supplied
variable <code>ds</code> specifies which is plotted (defaults to <code>0</code>).</p>
<p>Available config options with default vars:</p>
<pre><code class="language-javascript hljs">config = {
  <span class="hljs-attr">title</span>: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Plot title</span>
  xlab: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// X axis label</span>
  ylab: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Y axis label</span>
  xCeiling: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Maximum value for automatic axis limit (good for percentages)</span>
  xFloor: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Minimum value for automatic axis limit</span>
  xMinRange: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Minimum range for axis</span>
  xmax: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Max x limit</span>
  xmin: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Min x limit</span>
  xDecimals: <span class="hljs-literal">true</span>, <span class="hljs-comment">// Set to false to only show integer labels</span>
  yCeiling: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Maximum value for automatic axis limit (good for percentages)</span>
  yFloor: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Minimum value for automatic axis limit</span>
  yMinRange: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Minimum range for axis</span>
  ymax: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Max y limit</span>
  ymin: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Min y limit</span>
  yDecimals: <span class="hljs-literal">true</span>, <span class="hljs-comment">// Set to false to only show integer labels</span>
  yPlotBands: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Highlighted background bands. See http://api.highcharts.com/highcharts#yAxis.plotBands</span>
  xPlotBands: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Highlighted background bands. See http://api.highcharts.com/highcharts#xAxis.plotBands</span>
  tt_label: <span class="hljs-string">"{point.x}: {point.y:.2f}"</span>, <span class="hljs-comment">// Use to customise tooltip label, eg. '{point.x} base pairs'</span>
  pointFormat: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Replace the default HTML for the entire tooltip label</span>
  click_func: <span class="hljs-function"><span class="hljs-keyword">function</span> (<span class="hljs-params"></span>) </span>{}, <span class="hljs-comment">// Javascript function to be called when a point is clicked</span>
  cursor: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// CSS mouse cursor type. Defaults to pointer when 'click_func' specified</span>
};</code></pre>
<p>An example of the markup expected, with the function being called:</p>
<pre><code class="language-html hljs xml"><span class="hljs-tag">&lt;<span class="hljs-name">div</span> <span class="hljs-attr">id</span>=<span class="hljs-string">"my_awesome_line_graph"</span> <span class="hljs-attr">class</span>=<span class="hljs-string">"hc-plot"</span>&gt;</span><span class="hljs-tag">&lt;/<span class="hljs-name">div</span>&gt;</span>
<span class="hljs-tag">&lt;<span class="hljs-name">script</span> <span class="hljs-attr">type</span>=<span class="hljs-string">"text/javascript"</span>&gt;</span><span class="javascript">
  mqc_plots[<span class="hljs-string">"#my_awesome_bar_plot"</span>][<span class="hljs-string">"plot_type"</span>] = <span class="hljs-string">"xy_line"</span>;
  mqc_plots[<span class="hljs-string">"#my_awesome_line_graph"</span>][<span class="hljs-string">"datasets"</span>] = [
    {
      <span class="hljs-attr">name</span>: <span class="hljs-string">"Sample 1"</span>,
      <span class="hljs-attr">data</span>: [
        [<span class="hljs-number">1</span>, <span class="hljs-number">1.5</span>],
        [<span class="hljs-number">1.5</span>, <span class="hljs-number">3.1</span>],
        [<span class="hljs-number">2</span>, <span class="hljs-number">6.4</span>],
      ],
    },
    {
      <span class="hljs-attr">name</span>: <span class="hljs-string">"Sample 2"</span>,
      <span class="hljs-attr">data</span>: [
        [<span class="hljs-number">1</span>, <span class="hljs-number">1.7</span>],
        [<span class="hljs-number">1.5</span>, <span class="hljs-number">4.3</span>],
        [<span class="hljs-number">2</span>, <span class="hljs-number">8.4</span>],
      ],
    },
  ];
  mqc_plots[<span class="hljs-string">"#my_awesome_line_graph"</span>][<span class="hljs-string">"config"</span>] = {
    <span class="hljs-attr">title</span>: <span class="hljs-string">"Best Plot Ever"</span>,
    <span class="hljs-attr">ylab</span>: <span class="hljs-string">"Pings"</span>,
    <span class="hljs-attr">xlab</span>: <span class="hljs-string">"Pongs"</span>,
  };
  $(<span class="hljs-function"><span class="hljs-keyword">function</span> (<span class="hljs-params"></span>) </span>{
    plot_xy_line_graph(<span class="hljs-string">"#my_awesome_line_graph"</span>);
  });
</span><span class="hljs-tag">&lt;/<span class="hljs-name">script</span>&gt;</span></code></pre>
<h3 id="plotting-bar-graphs"><a href="#plotting-bar-graphs" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Plotting bar graphs</h3>
<p><code>plot_stacked_bar_graph (target, ds)</code></p>
<p>Plots a bar graph with multiple series containing multiple categories.
Used by the <a href="http://multiqc.info/docs/#bar-graphs">bargraph.plot()</a>
python function.</p>
<p>Data and configuration must be added to the document level
<code>mqc_plots</code> variable on page load, using the target as the key.
The variables used are as follows:</p>
<pre><code class="language-javascript hljs">mqc_plots[target][<span class="hljs-string">"plot_type"</span>] = <span class="hljs-string">"bar_graph"</span>;
mqc_plots[target][<span class="hljs-string">"config"</span>];
mqc_plots[target][<span class="hljs-string">"datasets"</span>];
mqc_plots[target][<span class="hljs-string">"samples"</span>];</code></pre>
<p>All available config options with default vars:</p>
<pre><code class="language-javascript hljs">config = {
  <span class="hljs-attr">title</span>: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Plot title</span>
  xlab: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// X axis label</span>
  ylab: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Y axis label</span>
  ymax: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Max y limit</span>
  ymin: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Min y limit</span>
  yDecimals: <span class="hljs-literal">true</span>, <span class="hljs-comment">// Set to false to only show integer labels</span>
  ylab_format: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Format string for x axis labels. Defaults to {value}</span>
  stacking: <span class="hljs-string">"normal"</span>, <span class="hljs-comment">// Set to null to have category bars side by side (None in python)</span>
  xtype: <span class="hljs-string">"linear"</span>, <span class="hljs-comment">// Axis type. 'linear' or 'logarithmic'</span>
  use_legend: <span class="hljs-literal">true</span>, <span class="hljs-comment">// Show / hide the legend</span>
  click_func: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// Javascript function to be called when a point is clicked</span>
  cursor: <span class="hljs-literal">undefined</span>, <span class="hljs-comment">// CSS mouse cursor type. Defaults to pointer when 'click_func' specified</span>
  tt_percentages: <span class="hljs-literal">true</span>, <span class="hljs-comment">// Show the percentages of each count in the tooltip</span>
  reversedStacks: <span class="hljs-literal">false</span>, <span class="hljs-comment">// Reverse the order of the categories in the stack.</span>
};</code></pre>
<p>An example of the markup expected, with the function being called:</p>
<pre><code class="language-html hljs xml"><span class="hljs-tag">&lt;<span class="hljs-name">div</span> <span class="hljs-attr">id</span>=<span class="hljs-string">"my_awesome_bar_plot"</span> <span class="hljs-attr">class</span>=<span class="hljs-string">"hc-plot"</span>&gt;</span><span class="hljs-tag">&lt;/<span class="hljs-name">div</span>&gt;</span>
<span class="hljs-tag">&lt;<span class="hljs-name">script</span> <span class="hljs-attr">type</span>=<span class="hljs-string">"text/javascript"</span>&gt;</span><span class="javascript">
  mqc_plots[<span class="hljs-string">"#my_awesome_bar_plot"</span>][<span class="hljs-string">"plot_type"</span>] = <span class="hljs-string">"bar_graph"</span>;
  mqc_plots[<span class="hljs-string">"#my_awesome_bar_plot"</span>][<span class="hljs-string">"samples"</span>] = [<span class="hljs-string">"Sample 1"</span>, <span class="hljs-string">"Sample 2"</span>];
  mqc_plots[<span class="hljs-string">"#my_awesome_bar_plot"</span>][<span class="hljs-string">"datasets"</span>] = [
    { <span class="hljs-attr">data</span>: [<span class="hljs-number">4</span>, <span class="hljs-number">7</span>], <span class="hljs-attr">name</span>: <span class="hljs-string">"Passed Test"</span> },
    { <span class="hljs-attr">data</span>: [<span class="hljs-number">2</span>, <span class="hljs-number">3</span>], <span class="hljs-attr">name</span>: <span class="hljs-string">"Failed Test"</span> },
  ];
  mqc_plots[<span class="hljs-string">"#my_awesome_bar_plot"</span>][<span class="hljs-string">"config"</span>] = {
    <span class="hljs-attr">title</span>: <span class="hljs-string">"My Awesome Plot"</span>,
    <span class="hljs-attr">ylab</span>: <span class="hljs-string">"# Observations"</span>,
    <span class="hljs-attr">ymin</span>: <span class="hljs-number">0</span>,
    <span class="hljs-attr">stacking</span>: <span class="hljs-string">"normal"</span>,
  };
  $(<span class="hljs-function"><span class="hljs-keyword">function</span> (<span class="hljs-params"></span>) </span>{
    plot_stacked_bar_graph(<span class="hljs-string">"#my_awesome_bar_plot"</span>);
  });
</span><span class="hljs-tag">&lt;/<span class="hljs-name">script</span>&gt;</span></code></pre>
<h3 id="switching-counts-and-percentages"><a href="#switching-counts-and-percentages" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Switching counts and percentages</h3>
<p>If you're using the plotting functions above, it's easy to add a button which
switches between percentages and counts. Just add the following HTML above
your plot:</p>
<pre><code class="language-html hljs xml"><span class="hljs-tag">&lt;<span class="hljs-name">div</span> <span class="hljs-attr">class</span>=<span class="hljs-string">"btn-group switch_group"</span>&gt;</span>
  <span class="hljs-tag">&lt;<span class="hljs-name">button</span> <span class="hljs-attr">class</span>=<span class="hljs-string">"btn btn-default btn-sm active"</span> <span class="hljs-attr">data-action</span>=<span class="hljs-string">"set_numbers"</span> <span class="hljs-attr">data-target</span>=<span class="hljs-string">"#my_plot"</span>&gt;</span>Counts<span class="hljs-tag">&lt;/<span class="hljs-name">button</span>&gt;</span>
  <span class="hljs-tag">&lt;<span class="hljs-name">button</span> <span class="hljs-attr">class</span>=<span class="hljs-string">"btn btn-default btn-sm"</span> <span class="hljs-attr">data-action</span>=<span class="hljs-string">"set_percent"</span> <span class="hljs-attr">data-target</span>=<span class="hljs-string">"#my_plot"</span>&gt;</span>Percentages<span class="hljs-tag">&lt;/<span class="hljs-name">button</span>&gt;</span>
<span class="hljs-tag">&lt;/<span class="hljs-name">div</span>&gt;</span></code></pre>
<p><em>NB:</em> This markup is generated automatically with the Python <code>self.plot_bargraph()</code> function.</p>
<h3 id="switching-plot-datasets"><a href="#switching-plot-datasets" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Switching plot datasets</h3>
<p>Much like the counts / percentages buttons above, you can add a button which
switches the data displayed in a single plot. Make sure that both datasets
are stored in named javascript variables, then add the following markup:</p>
<pre><code class="language-html hljs xml"><span class="hljs-tag">&lt;<span class="hljs-name">div</span> <span class="hljs-attr">class</span>=<span class="hljs-string">"btn-group switch_group"</span>&gt;</span>
  <span class="hljs-tag">&lt;<span class="hljs-name">button</span>
    <span class="hljs-attr">class</span>=<span class="hljs-string">"btn btn-default btn-sm active"</span>
    <span class="hljs-attr">data-action</span>=<span class="hljs-string">"set_data"</span>
    <span class="hljs-attr">data-ylab</span>=<span class="hljs-string">"First Data"</span>
    <span class="hljs-attr">data-newdata</span>=<span class="hljs-string">"data_var_1"</span>
    <span class="hljs-attr">data-target</span>=<span class="hljs-string">"#my_plot"</span>
  &gt;</span>
    Data 1
  <span class="hljs-tag">&lt;/<span class="hljs-name">button</span>&gt;</span>
  <span class="hljs-tag">&lt;<span class="hljs-name">button</span>
    <span class="hljs-attr">class</span>=<span class="hljs-string">"btn btn-default btn-sm"</span>
    <span class="hljs-attr">data-action</span>=<span class="hljs-string">"set_data"</span>
    <span class="hljs-attr">data-ylab</span>=<span class="hljs-string">"Second Data"</span>
    <span class="hljs-attr">data-newdata</span>=<span class="hljs-string">"data_var_2"</span>
    <span class="hljs-attr">data-target</span>=<span class="hljs-string">"#my_plot"</span>
  &gt;</span>
    Data 2
  <span class="hljs-tag">&lt;/<span class="hljs-name">button</span>&gt;</span>
<span class="hljs-tag">&lt;/<span class="hljs-name">div</span>&gt;</span></code></pre>
<p>Note the CSS class <code>active</code> which specifies which button is 'pressed' on page load.
<code>data-ylab</code> and <code>data-xlab</code> can be used to specify the new axes labels.
<code>data-newdata</code> should be the name of the javascript object with the new data
to be plotted and <code>data-target</code> should be the CSS selector of the plot to change.</p>
<h3 id="custom-event-triggers"><a href="#custom-event-triggers" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Custom event triggers</h3>
<p>Some of the events that take place in the general javascript
code trigger jQuery events which you can hook into from within your
module's code. This allows you to take advantage of events generated
by the global theme whilst keeping your code modular.</p>
<pre><code class="language-javascript hljs">$(<span class="hljs-built_in">document</span>).on(<span class="hljs-string">"mqc_highlights"</span>, <span class="hljs-function"><span class="hljs-keyword">function</span> (<span class="hljs-params">e, f_texts, f_cols, regex_mode</span>) </span>{
  <span class="hljs-comment">// This trigger is called when the highlight strings are</span>
  <span class="hljs-comment">// updated. Three variables are given - an array of search</span>
  <span class="hljs-comment">// strings (f_texts), an array of colours with corresponding</span>
  <span class="hljs-comment">// indexes (f_cols) and a boolean var saying whether the</span>
  <span class="hljs-comment">// search should be treated as a string or a regex (regex_mode)</span>
});

$(<span class="hljs-built_in">document</span>).on(<span class="hljs-string">"mqc_renamesamples"</span>, <span class="hljs-function"><span class="hljs-keyword">function</span> (<span class="hljs-params">e, f_texts, t_texts, regex_mode</span>) </span>{
  <span class="hljs-comment">// This trigger is called when samples are renamed</span>
  <span class="hljs-comment">// Three variables are given - an array of search</span>
  <span class="hljs-comment">// strings (f_texts), an array of replacements with corresponding</span>
  <span class="hljs-comment">// indexes (t_texts) and a boolean var saying whether the</span>
  <span class="hljs-comment">// search should be treated as a string or a regex (regex_mode)</span>
});

$(<span class="hljs-built_in">document</span>).on(<span class="hljs-string">"mqc_hidesamples"</span>, <span class="hljs-function"><span class="hljs-keyword">function</span> (<span class="hljs-params">e, f_texts, regex_mode</span>) </span>{
  <span class="hljs-comment">// This trigger is called when the Hide Samples filters change.</span>
  <span class="hljs-comment">// Two variables are given - an array of search strings</span>
  <span class="hljs-comment">// (f_texts) and a boolean saying whether the search should</span>
  <span class="hljs-comment">// be treated as a string or a regex (regex_mode)</span>
});

$(<span class="hljs-string">"#YOUR_PLOT_ID"</span>).on(<span class="hljs-string">"mqc_plotresize"</span>, <span class="hljs-function"><span class="hljs-keyword">function</span> (<span class="hljs-params"></span>) </span>{
  <span class="hljs-comment">// This trigger is called when a plot handle is pulled,</span>
  <span class="hljs-comment">// resizing the height</span>
});

$(<span class="hljs-string">"#YOUR_PLOT_ID"</span>).on(<span class="hljs-string">"mqc_original_series_click"</span>, <span class="hljs-function"><span class="hljs-keyword">function</span> (<span class="hljs-params">e, name</span>) </span>{
  <span class="hljs-comment">// A plot able to show original images has had a point clicked.</span>
  <span class="hljs-comment">// 'name' contains the name of the series that was clicked</span>
});

$(<span class="hljs-string">"#YOUR_PLOT_ID"</span>).on(<span class="hljs-string">"mqc_original_chg_source"</span>, <span class="hljs-function"><span class="hljs-keyword">function</span> (<span class="hljs-params">e, name</span>) </span>{
  <span class="hljs-comment">// A plot with original images has had a request to change the</span>
  <span class="hljs-comment">// original image source (eg. pressing Prev / Next)</span>
});

$(<span class="hljs-string">"#YOUR_PLOT_ID"</span>).on(<span class="hljs-string">"mqc_plotexport_image"</span>, <span class="hljs-function"><span class="hljs-keyword">function</span> (<span class="hljs-params">e, cfg</span>) </span>{
  <span class="hljs-comment">// A trigger to export an image of the plot. cfg contains</span>
  <span class="hljs-comment">// config variables for the requested image.</span>
});

$(<span class="hljs-string">"#YOUR_PLOT_ID"</span>).on(<span class="hljs-string">"mqc_plotexport_data"</span>, <span class="hljs-function"><span class="hljs-keyword">function</span> (<span class="hljs-params">e, cfg</span>) </span>{
  <span class="hljs-comment">// A trigger to export a data file of the plot. cfg contains</span>
  <span class="hljs-comment">// config variables for the requested data.</span>
});</code></pre></div><div class="docs_block" id="plugins.md"><h1 id="multiqc-plugins"><a href="#multiqc-plugins" class="header-link"><span class="glyphicon glyphicon-link"></span></a>MultiQC Plugins</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#entry-points-1">Entry Points</a></li>
<li><a href="#modules">Modules</a></li>
<li><a href="#templates">Templates</a></li>
<li><a href="#command-line-options">Command line options</a></li>
<li><a href="#hooks">Hooks</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>MultiQC is written around a system designed for extensibility and plugins.
These features allow custom code to be written without polluting the central
code base.</p>
<p>Please note that we want MultiQC to grow as a community tool! So if you're
writing a module or theme that can be used by others, please keep it within
the main MultiQC framework and submit a pull request.</p>
<h2 id="entry-points-1"><a href="#entry-points-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Entry Points</h2>
<p>The plugin system works using setuptools
<a href="http://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins">entry points</a>.
In <code>setup.py</code> you will see a section of code that looks like this <em>(truncated)</em>:</p>
<pre><code class="language-python hljs">entry_points = {
    <span class="hljs-string">'multiqc.modules.v1'</span>: [
        <span class="hljs-string">'qualimap = multiqc.modules.qualimap:MultiqcModule'</span>,
    ],
    <span class="hljs-string">'multiqc.templates.v1'</span>: [
        <span class="hljs-string">'default = multiqc.templates.default'</span>,
    ],
    <span class="hljs-comment"># 'multiqc.cli_options.v1': [</span>
        <span class="hljs-comment"># 'my-new-option = myplugin.cli:new_option'</span>
    <span class="hljs-comment"># ],</span>
    <span class="hljs-comment"># 'multiqc.hooks.v1': [</span>
        <span class="hljs-comment"># 'before_config = myplugin.hooks:before_config',</span>
        <span class="hljs-comment"># 'config_loaded = myplugin.hooks:config_loaded',</span>
        <span class="hljs-comment"># 'execution_start = myplugin.hooks:execution_start',</span>
        <span class="hljs-comment"># 'before_modules = myplugin.hooks:before_modules',</span>
        <span class="hljs-comment"># 'after_modules = myplugin.hooks:after_modules',</span>
        <span class="hljs-comment"># 'execution_finish = myplugin.hooks:execution_finish',</span>
    <span class="hljs-comment"># ]</span>
},</code></pre>
<p>These sets of entry points can each be extended to add functionality
to MultiQC:</p>
<ul>
<li><code>multiqc.modules.v1</code>
<ul>
<li>Defines the module classes. Used to add new modules.</li>
</ul></li>
<li><code>multiqc.templates.v1</code>
<ul>
<li>Defines the templates. Can be used for new templates.</li>
</ul></li>
<li><code>multiqc.cli_options.v1</code>
<ul>
<li>Allows plugins to add new custom command line options</li>
</ul></li>
<li><code>multiqc.hooks.v1</code>
<ul>
<li>Code hooks for plugins to add new functionality</li>
</ul></li>
</ul>
<p>Any python program can create entry points with the same name, once installed
MultiQC will find these and run them accordingly. For an example of this in
action, see the <a href="https://github.com/ewels/MultiQC_NGI/blob/master/setup.py">MultiQC_NGI</a>
setup file:</p>
<pre><code class="language-python hljs">entry_points = {
        <span class="hljs-string">'multiqc.templates.v1'</span>: [
            <span class="hljs-string">'ngi = multiqc_ngi.templates.ngi'</span>,
            <span class="hljs-string">'genstat = multiqc_ngi.templates.genstat'</span>,
        ],
        <span class="hljs-string">'multiqc.cli_options.v1'</span>: [
            <span class="hljs-string">'project = multiqc_ngi.cli:pid_option'</span>
        ],
        <span class="hljs-string">'multiqc.hooks.v1'</span>: [
            <span class="hljs-string">'after_modules = multiqc_ngi.hooks:ngi_metadata'</span>,
        ]
    },</code></pre>
<p>Here, two new templates are added, a new command line option and a new code hook.</p>
<h2 id="modules"><a href="#modules" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Modules</h2>
<p>List items added to <code>multiqc.modules.v1</code> specify new modules. They should
be described as follows:</p>
<pre><code class="language-python hljs">modname = python_mod.dirname.submodname:classname<span class="hljs-string">'</span></code></pre>
<p>Once this is done, everything else should be the same as described in the
<a href="http://multiqc.info/docs/#writing-new-modules">writing modules</a> documentation.</p>
<h2 id="templates"><a href="#templates" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Templates</h2>
<p>As above, though no need to specify a class name at the end. See the
<a href="http://multiqc.info/docs/#writing-new-templates">writing templates</a> documentation for further instructions.</p>
<h2 id="command-line-options"><a href="#command-line-options" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Command line options</h2>
<p>MultiQC handles command line interaction using the <a href="http://click.pocoo.org/">click</a>
framework. You can use the <code>multiqc.cli_options.v1</code> entry point to add new
click decorators for command line options. For example, the MultiQC_NGI
plugin uses the entry point above with the following code in <code>cli.py</code>:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">import</span> click
pid_option = click.option(<span class="hljs-string">'--project'</span>, type=str)</code></pre>
<p>The values given from additional command line arguments are parsed by
MultiQC and put into <code>config.kwargs</code>. The above plugin later reads
the value given by the user with the <code>--project</code> flag in a hook:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">if</span> config.kwargs[<span class="hljs-string">'project'</span>] <span class="hljs-keyword">is</span> <span class="hljs-keyword">not</span> <span class="hljs-literal">None</span>:
  <span class="hljs-comment"># do some stuff</span></code></pre>
<p>See the <a href="http://click.pocoo.org/">click documentation</a> or the main
MultiQC script for more information and examples of adding command line
options.</p>
<h2 id="hooks"><a href="#hooks" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Hooks</h2>
<p>Hooks are a little more complicated - these define points in the core
MultiQC code where you can run custom functions. This can be useful as
your code is able to access data generated by other parts of the program.
For example, you could tie into the <code>after_modules</code> hook to insert data
processed by MultiQC modules into a database automatically.</p>
<p>Here, the entry point names are the hook titles, described as commented out
lines in the core MultiQC <code>setup.py</code>: <code>execution_start</code>, <code>config_loaded</code>,
<code>before_modules</code>, <code>after_modules</code> and <code>execution_finish</code>.</p>
<p>These should point to a function in your code which will be executed when
that hook fires. Your custom code can import the core MultiQC modules to
access configuration and loggers. For example:</p>
<pre><code class="language-python hljs"><span class="hljs-comment">#!/usr/bin/env python</span>
<span class="hljs-string">""" MultiQC hook functions - we tie into the MultiQC
core here to add in extra functionality. """</span>

<span class="hljs-keyword">import</span> logging
<span class="hljs-keyword">from</span> multiqc.utils <span class="hljs-keyword">import</span> report, config

log = logging.getLogger(<span class="hljs-string">'multiqc'</span>)

<span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">after_modules</span><span class="hljs-params">()</span>:</span>
  <span class="hljs-string">""" Plugin code to run when MultiQC modules have completed  """</span>
  num_modules = len(report.modules_output)
  status_string = <span class="hljs-string">"MultiQC hook - {} modules reported!"</span>.format(num_modules)
  log.critical(status_string)</code></pre></div><div class="docs_block" id="templates.md"><h1 id="writing-new-templates"><a href="#writing-new-templates" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Writing New Templates</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#core-or-plugin">Core or plugin</a></li>
<li><a href="#creating-a-template-skeleton">Creating a template skeleton</a></li>
<li><a href="#child-templates">Child templates</a></li>
<li><a href="#extra-init-variables">Extra init variables</a></li>
<li><a href="#jinja-template-variables">Jinja template variables</a></li>
<li><a href="#appendices-1">Appendices</a>
<ul class="nav nav-stacked">
<li><a href="#custom-plotting-functions">Custom plotting functions</a></li>
</ul>
</li>

</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>MultiQC is built around a templating system that uses the
<a href="http://jinja.pocoo.org/">Jinja</a> python package. This makes it very
easy to create new report templates that fit your needs.</p>
<h2 id="core-or-plugin"><a href="#core-or-plugin" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Core or plugin</h2>
<p>If your template could be of use to others, it would be great if you
could add it to the main MultiQC package. You can do this by creating a
fork of the <a href="https://github.com/ewels/MultiQC">MultiQC GitHub repository</a>,
adding your template and then creating a pull request to merge your changes
back to the main repository.</p>
<p>If it's very specific template, you can create a new Python package which
acts as a plugin. For more information about this, see the
<a href="http://multiqc.info/docs/#multiqc-plugins">plugins documentation</a>.</p>
<h2 id="creating-a-template-skeleton"><a href="#creating-a-template-skeleton" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Creating a template skeleton</h2>
<p>For a new template to be recognised by MultiQC, it must be a python submodule
directory with a <code>__init__.py</code> file. This must be referenced in the <code>setup.py</code>
installation script as an
<a href="http://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins">entry point</a>.</p>
<p>You can see the bundled templates defined in this way:</p>
<pre><code class="language-python hljs">entry_points = {
    <span class="hljs-string">'multiqc.templates.v1'</span>: [
        <span class="hljs-string">'default = multiqc.templates.default'</span>,
        <span class="hljs-string">'default_dev = multiqc.templates.default_dev'</span>,
        <span class="hljs-string">'simple = multiqc.templates.simple'</span>,
        <span class="hljs-string">'geo = multiqc.templates.geo'</span>,
    ]
}</code></pre>
<p>Note that these entry points can point to any Python modules, so if you're
writing a plugin module you can specify your module name instead. Just make
sure that <code>multiqc.templates.v1</code> is the same.</p>
<p>Once you've added the entry point, remember to install the package again:</p>
<pre><code class="language-bash hljs">pip install -e .</code></pre>
<p>Using <code>-e</code> tells <code>pip</code> to softlink the plugin files instead of
copying, so changes made whilst editing files will be reflected when you
run MultiQC.</p>
<p>The <code>__init__.py</code> files must define two variables - the path to the template
directory and the main jinja template file:</p>
<pre><code class="language-python hljs">template_dir = os.path.dirname(__file__)
base_fn = <span class="hljs-string">'base.html'</span></code></pre>
<h2 id="child-templates"><a href="#child-templates" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Child templates</h2>
<p>The default MultiQC template contains a <em>lot</em> of code. Importantly, it includes
1448 lines of custom JavaScript (at time of writing) which powers the plotting
and dynamic functions in the report. You probably don't want to rewrite all of
this for your template, so to make your life easier you can create a
<em>child template</em>.</p>
<p>To do this, add an extra variable to your template's <code>__init__.py</code>:</p>
<pre><code class="language-python hljs">template_parent = <span class="hljs-string">'default'</span></code></pre>
<p>This tells MultiQC to use the template files from the <code>default</code> template unless
a file with the same name is found in your child template. For instance, if
you just want to add your own logo in the header of the reports, you can create
your own <code>header.html</code> which will overwrite the default header.</p>
<p>Files within the default template have comments at the top explaining what
part of the report they generate.</p>
<h2 id="extra-init-variables"><a href="#extra-init-variables" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Extra init variables</h2>
<p>There are a few extra variables that can be added to the <code>__init__.py</code> file
to change how the report is generated.</p>
<p>Setting <code>output_dir</code> instructs MultiQC to put the report and it's contents
into a subdirectory. Set the string to your desired name. Note that this will
be prefixed if <code>-p</code>/<code>--prefix</code> is set at run time.</p>
<p>Secondly, you can copy additional files with your report when it is generated.
This is usually used to copy required images or scripts with the report. These
should be a list of file or directory paths, relative to the <code>__init__.py</code> file.
Directory contents will be copied recursively.</p>
<p>You can also override config options in the template. For example, setting
the value of <code>config.plots_force_flat</code> can force the report to only have
static image plots.</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc.utils <span class="hljs-keyword">import</span> config

output_subdir = <span class="hljs-string">'multiqc_report'</span>
copy_files = [<span class="hljs-string">'assets'</span>]
config.plots_force_flat = <span class="hljs-literal">True</span></code></pre>
<h2 id="jinja-template-variables"><a href="#jinja-template-variables" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Jinja template variables</h2>
<p>There are a number of variables that you can use within your Jinja template.
Two namespaces are available - <code>report</code> and <code>config</code>. You can print these
using the Jinja curly brace syntax, <em>eg.</em> <code>{{ config.version }}</code>. See the
<a href="http://jinja.pocoo.org/docs/dev/templates/">Jinja2 documentation</a> for more
information.</p>
<p>The default MultiQC template includes dependencies in the HTML so that the
report is standalone. If you would like to do the same, use the <code>include_file</code>
function. For example:</p>
<pre><code class="language-html hljs xml"><span class="hljs-tag">&lt;<span class="hljs-name">script</span>&gt;</span><span class="javascript">
  {
    {
      include_file(<span class="hljs-string">"js/jquery.min.js"</span>);
    }
  }
</span><span class="hljs-tag">&lt;/<span class="hljs-name">script</span>&gt;</span>
<span class="hljs-tag">&lt;<span class="hljs-name">img</span> <span class="hljs-attr">src</span>=<span class="hljs-string">"data:image/png;base64,{{ include_file('img/logo.png', b64=True) }}"</span> /&gt;</span></code></pre>
<h2 id="appendices-1"><a href="#appendices-1" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Appendices</h2><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#custom-plotting-functions">Custom plotting functions</a></li>
</ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<h3 id="custom-plotting-functions"><a href="#custom-plotting-functions" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Custom plotting functions</h3>
<p>If you don't like the default plotting functions built into MultiQC, you
can write your own! If you create a callable variable in a template called
either <code>bargraph</code> or <code>linegraph</code>, MultiQC will use that instead. For example:</p>
<pre><code class="language-python hljs"><span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">custom_linegraph</span><span class="hljs-params">(plotdata, pconfig)</span>:</span>
    <span class="hljs-keyword">return</span> <span class="hljs-string">'&lt;h1&gt;Awesome line graph here&lt;/h1&gt;'</span>
linegraph = custom_linegraph

<span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">custom_bargraph</span><span class="hljs-params">(plotdata, plotseries, pconfig)</span>:</span>
    <span class="hljs-keyword">return</span> <span class="hljs-string">'&lt;h1&gt;Awesome bar graph here&lt;/h1&gt;'</span>
bargraph = custom_bargraph</code></pre>
<p>These particular examples don't do very much, but hopefully you get the idea.
Note that you have to set the variable <code>linegraph</code> or <code>bargraph</code> to your function.</p></div><div class="docs_block" id="compatibility.md"><h1 id="updating-for-compatibility"><a href="#updating-for-compatibility" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Updating for compatibility</h1><div class="toc toc_mobile_subnav"><ul class="nav nav-stacked">
<li><a href="#v1.0-updates">v1.0 Updates</a></li></ul><p class="backtotop"><a href="#">Back to top</a></p></div>
<p>When releasing new versions of MultiQC we aim to maintain compatibility so that your existing
modules and plugins will keep working. However, in some cases we have to make changes that
require code to be modified. This section summarises the changes by MultiQC release.</p>
<h2 id="v1.0-updates"><a href="#v1.0-updates" class="header-link"><span class="glyphicon glyphicon-link"></span></a>v1.0 Updates</h2>
<p>MultiQC v1.0 brings a few changes in the way that MultiQC modules and
 plugins are written. Most are backwards-compatible, but there are a 
couple that could break external plugins.</p>
<h4 id="module-imports"><a href="#module-imports" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Module imports</h4>
<p>New MultiQC module imports have been refactored to make them less 
inter-dependent and fragile. This has a bunch of advantages, notably 
allowing better, more modular, unit testing (and hopefully more reliable
 and maintainable code).</p>
<p>All MultiQC modules and plugins will need to change some of their import statements.</p>
<p>There are two things that you probably need to change in your plugin modules to
make them work with the updated version of MultiQC, both to do with imports.
Instead of this style of importing modules:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc <span class="hljs-keyword">import</span> config, BaseMultiqcModule, plots</code></pre>
<p>You now need this:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc <span class="hljs-keyword">import</span> config
<span class="hljs-keyword">from</span> multiqc.plots <span class="hljs-keyword">import</span> bargraph   <span class="hljs-comment"># Load specific plot types here</span>
<span class="hljs-keyword">from</span> multiqc.modules.base_module <span class="hljs-keyword">import</span> BaseMultiqcModule</code></pre>
<p>Modules that directly reference <code>multiqc.BaseMultiqcModule</code> instead need to reference
<code>multiqc.modules.base_module.BaseMultiqcModule</code>.</p>
<p>Secondly, modules that use <code>import plots</code> now need to import the specific plots needed.
You will also need to update any plotting functions, removing the <code>plot.</code> prefix.</p>
<p>For example, change this:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">import</span> plots
<span class="hljs-keyword">return</span> plots.bargraph.plot(data, keys, pconfig)</code></pre>
<p>to this:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> plots <span class="hljs-keyword">import</span> bargraph
<span class="hljs-keyword">return</span> bargraph.plot(data, keys, pconfig)</code></pre>
<p>These changes have been made to simplify the module imports within MultiQC,
allowing specific parts of the codebase to be imported into a Python script
on their own. This enables small, atomic, clean unit testing.</p>
<p>If you have any questions, please open an issue.</p>
<blockquote>
<p>Many thanks to <a href="https://github.com/tbooth">@tbooth</a> at <a href="https://github.com/EdinburghGenomics">@EdinburghGenomics</a> for his patient work with this.</p>
</blockquote>
<h4 id="searching-for-files"><a href="#searching-for-files" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Searching for files</h4>
<p>The core <code>find_log_files</code> function has been rewritten and 
now works a little differently. Instead of searching all analysis files 
each time it's called (by every module), all files are searched once at 
the start of the MultiQC execution. This makes MultiQC run much faster.</p>
<p>To use the new syntax, add your search pattern to <code>config.sp</code> using the new <code>before_config</code> plugin hook:</p>
<p><code>setup.py</code>:</p>
<pre><code class="language-python hljs"><span class="hljs-comment"># [..]</span>
  <span class="hljs-string">'multiqc.hooks.v1'</span>: [
    <span class="hljs-string">'before_config = myplugin.mymodule:load_config'</span>
  ]</code></pre>
<p><code>mymodule.py</code>:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">from</span> multiqc.utils <span class="hljs-keyword">import</span> config
<span class="hljs-function"><span class="hljs-keyword">def</span> <span class="hljs-title">load_config</span><span class="hljs-params">()</span>:</span>
    my_search_patterns = {
        <span class="hljs-string">'my_plugin/my_mod'</span>: {<span class="hljs-string">'fn'</span>: <span class="hljs-string">'*_somefile.txt'</span>},
        <span class="hljs-string">'my_plugin/my_other_mod'</span>: {<span class="hljs-string">'fn'</span>: <span class="hljs-string">'*other_file.txt'</span>},
    }
    config.update_dict(config.sp, my_search_patterns)</code></pre>
<p>This will add in your search patterns to the default MultiQC config, 
before user config files are loaded (allowing people to overwrite your 
defaults as with other modules).</p>
<p>Now, you can find your files much as before, using the string specified above:</p>
<pre><code class="language-python hljs"><span class="hljs-keyword">for</span> f <span class="hljs-keyword">in</span> self.find_log_files(<span class="hljs-string">'my_plugin/my_mod'</span>):
  <span class="hljs-comment"># do something</span></code></pre>
<p>The old syntax (supplying a <code>dict</code> instead of a string to 
the function without any previous config setup) will still work, but you
 will get a depreciation notice. This functionality may be removed in 
the future.</p>
<h4 id="adding-report-sections"><a href="#adding-report-sections" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Adding report sections</h4>
<p>Until now, report sections were added by creating a list called <code>self.sections</code> and adding to it. If you only had a single section, the routine was to instead append to the <code>self.intro</code> string.</p>
<p>These methods have been depreciated in favour of a new function called <code>self.add_section()</code>. For example, instead of the previous:</p>
<pre><code class="language-python hljs">self.sections = list()
self.sections.append({
  <span class="hljs-string">'name'</span>: <span class="hljs-string">'My Section'</span>,
  <span class="hljs-string">'anchor'</span>: <span class="hljs-string">'my-html-id'</span>,
  <span class="hljs-string">'content'</span>: <span class="hljs-string">'&lt;p&gt;Description of what this plot shows.&lt;/p&gt;'</span> +
             linegraph.plot(data, pconfig)
})</code></pre>
<p>the syntax is now:</p>
<pre><code class="language-python hljs">self.add_section(
  name = <span class="hljs-string">'My Section'</span>,
  anchor = <span class="hljs-string">'my-html-id'</span>,
  description = <span class="hljs-string">'Description of what this plot shows.'</span>,
  helptext = <span class="hljs-string">'More extensive help text can about how to interpret this.'</span>
  plot = linegraph.plot(data, pconfig)
)</code></pre>
<p>Note that content should now be split up into three new keys: <code>description</code>, <code>helptext</code> and <code>plot</code>. This will allow consistent formatting and future developments with improved module help text. Text is wrapped in <code>&lt;p&gt;</code> tags by the function, so these are no longer needed. Raw content can still be provided in a <code>content</code> string as before if required.</p>
<p>All fields are optional. If <code>name</code> is omitted then the end result will be the same as previously done with <code>self.intro += content</code>.</p>
<h4 id="updated-number-formatting"><a href="#updated-number-formatting" class="header-link"><span class="glyphicon glyphicon-link"></span></a>Updated number formatting</h4>
<p>A couple of minor updates to how numbers are handled in tables may affect your configs. Firstly, format strings looking like <code>{:.1f}</code> should now be <code>{:,.1f}</code> (note the extra comma). This enables customisable number formatting with separated thousand groups.</p>
<p>Secondly, any table columns reporting a read count should use new 
config options to allow user-configurable multipliers. For example, 
instead of this:</p>
<pre><code class="language-python hljs">headers[<span class="hljs-string">'read_counts'</span>] = {
  <span class="hljs-string">'title'</span>: <span class="hljs-string">'M Reads'</span>,
  <span class="hljs-string">'description'</span>: <span class="hljs-string">'Read counts (millions)'</span>,
  <span class="hljs-string">'modify'</span>: <span class="hljs-keyword">lambda</span> x: x / <span class="hljs-number">1000000</span>,
  <span class="hljs-string">'format'</span>: <span class="hljs-string">'{:.,2f} M'</span>,
  <span class="hljs-string">'shared_key'</span>: <span class="hljs-string">'read_count'</span>
}</code></pre>
<p>you should now use this:</p>
<pre><code class="language-python hljs">headers[<span class="hljs-string">'read_counts'</span>] = {
  <span class="hljs-string">'title'</span>: <span class="hljs-string">'{} Reads'</span>.format(config.read_count_prefix),
  <span class="hljs-string">'description'</span>: <span class="hljs-string">'Total raw sequences ({})'</span>.format(config.read_count_desc),
  <span class="hljs-string">'modify'</span>: <span class="hljs-keyword">lambda</span> x: x * config.read_count_multiplier,
  <span class="hljs-string">'format'</span>: <span class="hljs-string">'{:,.2f} '</span> + config.read_count_prefix,
  <span class="hljs-string">'shared_key'</span>: <span class="hljs-string">'read_count'</span>
}</code></pre>
<p>Not as pretty, but allows users to view low depth coverage.</p></div></div>        </div>
      </div>
    </div> <!-- /container -->

    <footer id="footer">
      <div class="container">
        <p>Created by Phil Ewels:
          <a target="_blank" href="https://github.com/ewels"><i class="fa fa-github"></i> ewels</a> |
          <a target="_blank" href="https://twitter.com/tallphil"><i class="fa fa-twitter"></i> tallphil</a> |
          <a target="_blank" href="https://se.linkedin.com/in/philewels">LinkedIn</a> |
          <a target="_blank" href="https://www.researchgate.net/profile/Philip_Ewels">ResearchGate</a>
        </p>
        <a target="_blank" href="http://www.scilifelab.se/" class="scilife-footer">
          <img src="MultiQC%20installation_files/SLL_logo_blue_white.png">
        </a>
      </div>
    </footer>


    <!-- Placed at the end of the document so the pages load faster -->
    <script async="" src="MultiQC%20installation_files/analytics.js"></script><script src="MultiQC%20installation_files/jquery-1.js"></script>
    <script src="MultiQC%20installation_files/bootstrap.js"></script>
    <script src="MultiQC%20installation_files/highlight.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="MultiQC%20installation_files/docs.js"></script>

  <!-- Google Analytics -->
  <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
    (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
    m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','//www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-68098153-1', 'auto');
    ga('send', 'pageview');

  </script>

  

</body></html>